---
get---
title: "Anogenital_p30"
author: "Athena Golfinos-Owens"
date: '2023-03-10'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(Seurat)
library(DoubletFinder)
library(dplyr)
library(tidyr)
library(ggplot2)

color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3","#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", "#AA8282", "#D4B7B7", "#8600BF", "#BA5CE3", "#808000","#AEAE5C", "#1E90FF", "#00BFFF", "#56FF0D", "#FFFF00")
```

# [HUMAN GENOME ONLY] Vulvar scRNA-seq

## • Load in/save objects

```{r}
saveRDS(seu, file = paste('./seurat_objs/SoupX_FilteredH5/vulvar_combined_3samples_', Sys.Date(), '.RDS', sep = ''))
seu <- readRDS('./seurat_objs/SoupX_FilteredH5/vulvar_combined_3samples_2023-05-12.RDS')
```

## • Pre-processing seurat objects (with SoupX)

### •• Generate SoupX seurat objects

```{r}
library(SoupX)
#tmpDir <- '/Volumes/hqdinh2/Projects/RawData_FromUWBiotech/Vulvar_scRNAseq_032023/02.CellRangerOutput/Vulvar_scRNA-seq/V1T/count/'
#outdir <- './outputs/'
#seuout <- './seurat_objs/'
#sample <- "Vulva_#1_Tumor"

soupX_run(tmpDir = '/Volumes/hqdinh2/Projects/RawData_FromUWBiotech/Vulvar_scRNAseq_032023/02.CellRangerOutput/Vulvar_scRNA-seq/V2T/count/', sample = 'Vulva_#2_Tumor') #upper limit set at 600 UMI

soupX_run(tmpDir = '/Volumes/hqdinh2/Projects/RawData_FromUWBiotech/Vulvar_scRNAseq_032023/02.CellRangerOutput/Vulvar_scRNA-seq/V4T/count/', sample = 'Vulva_#4_Tumor') #upper limit set at 600 UMI

soupX_run(tmpDir = '/Volumes/hqdinh2/Projects/RawData_FromUWBiotech/Vulvar_scRNAseq_032023/02.CellRangerOutput/Vulvar_scRNA-seq/V4AN/count/', sample = 'Vulva_#4_AdjNorm')
```

### •• Preprocessing soupx seurat objects

```{r}
v1t <- readRDS('./seurat_objs/Vulva_#1_Tumor_2023-04-18.RDS')
v2t <- readRDS('./seurat_objs/Vulva_#2_Tumor_2023-04-18.RDS')
v4t <- readRDS('./seurat_objs/Vulva_#4_Tumor_2023-04-18.RDS')
v4an <- readRDS('./seurat_objs/Vulva_#4_AdjNorm_2023-04-18.RDS')

ncol(v1t) #12805 cells ~ 12% doublets
ncol(v2t) #6564 cells ~ 5.2% doublets
ncol(v4t) #3603 cells ~ 2.8% doublets
ncol(v4an) #189 cells ~ 0.1% doublets

v1t <- seu_pp1(v1t, harmony = F)
v1t <- seu_pp2(v1t, npcs = 30, reduction = 'pca')
DimPlot(v1t)

v2t <- seu_pp1(v2t, harmony = F)
v2t <- seu_pp2(v2t, npcs = 30, reduction = 'pca')
DimPlot(v2t)

v4t <- seu_pp1(v4t, harmony = F)
v4t <- seu_pp2(v4t, npcs = 30, reduction = 'pca')
DimPlot(v4t)

v4an <- seu_pp1(v4an, harmony = F)
v4an <- seu_pp2(v4an, npcs = 20, reduction = 'pca')
DimPlot(v4an)

sweep.res <- DoubletFinder::paramSweep_v3(v1t) 
sweep.stats <-DoubletFinder::summarizeSweep(sweep.res, GT = FALSE) 
bcmvn <- DoubletFinder::find.pK(sweep.stats)
barplot(bcmvn$BCmetric, names.arg = bcmvn$pK, las=2)

nExp_v1t <- round(ncol(v1t) * 0.12)  # expect 12% doublets
v1t <- DoubletFinder::doubletFinder_v3(v1t, pN = 0.25, pK = 0.19, nExp = nExp_v1t, PCs = 1:30)
DimPlot(v1t, group.by = 'DF.classifications_0.25_0.19_1352')
v1t <- v1t[,v1t$DF.classifications_0.25_0.19_1352 == 'Singlet']

saveRDS(v1t, file = paste('./seurat_objs/', Sys.Date(), '_v1t.RDS', sep = ''))



sweep.res <- DoubletFinder::paramSweep_v3(v2t) 
sweep.stats <-DoubletFinder::summarizeSweep(sweep.res, GT = FALSE) 
bcmvn <- DoubletFinder::find.pK(sweep.stats)

nExp_v2t <- round(ncol(v2t) * 0.052)  # expect 5.2% doublets
v2t <- DoubletFinder::doubletFinder_v3(v2t, pN = 0.25, pK = 0.005, nExp = nExp_v2t, PCs = 1:30)
DimPlot(v2t, group.by = 'DF.classifications_0.25_0.005_321')
v2t <- v2t[,v2t$DF.classifications_0.25_0.005_321 == 'Singlet']
saveRDS(v2t, file = paste('./seurat_objs/', Sys.Date(), '_v2t.RDS', sep = ''))


sweep.res <- DoubletFinder::paramSweep_v3(v4t) 
sweep.stats <-DoubletFinder::summarizeSweep(sweep.res, GT = FALSE) 
bcmvn <- DoubletFinder::find.pK(sweep.stats)
nExp_v4t <- round(ncol(v4t) * 0.028)  # expect 2.8% doublets
v4t <- DoubletFinder::doubletFinder_v3(v4t, pN = 0.25, pK = 0.02, nExp = nExp_v4t, PCs = 1:30)
DimPlot(v4t, group.by = 'DF.classifications_0.25_0.02_101')
v4t <- v4t[,v4t$DF.classifications_0.25_0.02_101 == 'Singlet']
saveRDS(v4t, file = paste('./seurat_objs/', Sys.Date(), '_v4t.RDS', sep = ''))

saveRDS(v4an, file = paste('./seurat_objs/', Sys.Date(), '_v4an.RDS', sep = ''))

all <- merge(v1t, v2t)
all <- merge(all, v4t)
all2 <- merge(all, v4an)
saveRDS(all, file = paste('./seurat_objs/NoSoupX_UMIgt100/vulvar_combined_3samples.RDS', sep = ''))
saveRDS(all2, file = paste('./seurat_objs/NoSoupX_UMIgt100/vulvar_combined_4samples.RDS', sep = ''))
```

## • Pre-processing RAW matrix into seurat objects (no SoupX)

```{r}
v1t_counts <- Read10X_h5('/Volumes/hqdinh2/Projects/RawData_FromUWBiotech/Vulvar_scRNAseq_032023/02.CellRangerOutput/sample_raw_feature_bc_matrix_V1T.h5', use.names = TRUE, unique.features = TRUE)
v1t <- CreateSeuratObject(v1t_counts, project = "Vulva_#1_Tumor", assay = "RNA", names.field = 1, names.delim = "_",meta.data = NULL)

v2t_counts <- Read10X_h5('/Volumes/hqdinh2/Projects/RawData_FromUWBiotech/Vulvar_scRNAseq_032023/02.CellRangerOutput/sample_raw_feature_bc_matrix_V2T.h5', use.names = TRUE, unique.features = TRUE)
v2t <- CreateSeuratObject(v2t_counts, project = "Vulva_#2_Tumor",
              assay = "RNA", names.field = 1, names.delim = "_",meta.data = NULL)

v4t_counts <- Read10X_h5('/Volumes/hqdinh2/Projects/RawData_FromUWBiotech/Vulvar_scRNAseq_032023/02.CellRangerOutput/sample_raw_feature_bc_matrix_V4T.h5', use.names = TRUE, unique.features = TRUE)
v4t <- CreateSeuratObject(v4t_counts, project = "Vulva_#4_Tumor",
              assay = "RNA", names.field = 1, names.delim = "_",meta.data = NULL)

v4an_counts <- Read10X_h5('/Volumes/hqdinh2/Projects/RawData_FromUWBiotech/Vulvar_scRNAseq_032023/02.CellRangerOutput/sample_raw_feature_bc_matrix_V4AN.h5', use.names = TRUE, unique.features = TRUE)
v4an <- CreateSeuratObject(v4an_counts, project = "Vulva_#4_AdjNorm",
              assay = "RNA", names.field = 1, names.delim = "_",meta.data = NULL)

v1t <- v1t[,v1t$nCount_RNA > 100]
v2t <- v2t[,v2t$nCount_RNA > 100]
v4t <- v4t[,v4t$nCount_RNA > 100]
v4an <- v4an[,v4an$nCount_RNA > 100]

ncol(v1t) #65216 cells ~ 12% doublets
ncol(v2t) #13604 cells ~ 6% doublets
ncol(v4t) #8110 cells ~ 3.6% doublets
ncol(v4an) #530 cells ~ 0.1% doublets

# run in the console
v1t <- seu_pp1(v1t, out = './outputs/', pct_doublets = 0.12, feature_subset = T, batch_id = 'orig.ident', reduction = 'pca', harmony = F, seurat_integration = F, normalize = T, test_npcs = 100, assay = 'RNA') #npcs = 30
DimPlot(v1t)

# run in the console
v2t <- seu_pp1(v2t, out = './outputs/', pct_doublets = 0.06, feature_subset = T, batch_id = 'orig.ident', reduction = 'pca', harmony = F, seurat_integration = F, normalize = T, test_npcs = 100, assay = 'RNA') #npcs = 30
DimPlot(v2t)

# run in the console
v4t <- seu_pp1(v4t, out = './outputs/', pct_doublets = 0.036, feature_subset = T, batch_id = 'orig.ident', reduction = 'pca', harmony = F, seurat_integration = F, normalize = T, test_npcs = 100, assay = 'RNA') #npcs = 30
DimPlot(v4t)

# run in the console
v4an <- seu_pp1(v4an, out = './outputs/', pct_doublets = 0.1, feature_subset = T, batch_id = 'orig.ident', reduction = 'pca', harmony = F, seurat_integration = F, normalize = T, test_npcs = 100, assay = 'RNA') #npcs = 30
DimPlot(v4an)


all <- merge(v1t, v2t)
all <- merge(all, v4t)
all <- merge(all, v4an)
saveRDS(all, file = paste('./seurat_objs/NoSoupX_UMIgt100/vulvar_3samples_combined.RDS', sep = ''))
```

### •• Pre-process merged object

```{r}
# run in console
all <- seu_pp1(all, out = './outputs/', remove_doublets = F, feature_subset = F, batch_id = 'orig.ident', reduction = 'harmonyRNA', harmony = TRUE, seurat_integration = FALSE, normalize = TRUE, test_npcs = 200, assay = 'RNA')
```

## • Identifying global clusters

```{r}
seu <- readRDS('./seurat_objs/vulvar_combined_3samples.RDS')
seu <- seu_pp1(seu, harmony = T, npcs = 200)
seu <- seu_pp2(seu, npcs = 75, reduction = 'harmonyRNA')
DimPlot(seu, group.by = 'orig.ident')

old_seu <- readRDS('./seurat_objs/Archived_objects/Pre-SoupX/vulvar_combined_2023-03-27.RDS')
meta <- old_seu[[c('global.cluster', 'orig.ident')]]
meta <- meta[rownames(meta) %in% colnames(seu),]
seu <- AddMetaData(seu, meta, col.name = 'global.cluster')

DimPlot(seu, group.by = 'global.cluster', label = T)

#aveRDS(seu, file = './seurat_objs/vulvar_combined_3samples.RDS')

FeaturePlot(seu, features = 'CD8A')

de_dotplot(seu, metadata = 'global.cluster', n_marks = 10)
test <- readRDS('./figures/2023-04-19_de_dotplot_markers.RDS')
plot(test[[2]])

DefaultAssay(seu) <- 'RNA'

# ENDOTHELIAL MARKERS
DotPlot(seu, features = c('RAMP2', 'GNG11', 'ECSCR', 'EGFL7', 'CLDN5', 'PECAM1', 'HYAL2', 'ARHGAP29', 'TIE1'), col.min = 0)
VlnPlot(seu, features = c('RAMP2', 'GNG11', 'ECSCR', 'EGFL7', 'CLDN5', 'PECAM1', 'HYAL2', 'ARHGAP29', 'TIE1'), pt.size = 0)


# FIBROBLAST MARKERS
DotPlot(seu, features = c('DCN', 'LUM', 'COL1A2', 'COL1A1', 'COL3A1', 'COL6A2', 'COL6A1', 'FTH1', 'C1S', 'COL6A3', 'CEBPB', 'CTSK', 'C1R', 'SFRP2', 'AEBP1'), col.min = 0)
VlnPlot(seu, features = c('DCN', 'LUM', 'COL1A2', 'COL1A1', 'COL3A1', 'COL6A2', 'COL6A1', 'FTH1', 'C1S', 'COL6A3', 'CEBPB', 'CTSK', 'C1R', 'SFRP2', 'AEBP1'), pt.size = 0)

# PERICYTE MARKERS
DotPlot(seu, features = c('MYL9', 'RGS5', 'SOD3', 'TAGLN', 'ACTA2', 'PPP1R14A', 'ID4', 'NOTCH3', 'SPARCL1'), col.min = 0)
VlnPlot(seu, features = c('MYL9', 'RGS5', 'SOD3', 'TAGLN', 'ACTA2', 'PPP1R14A', 'ID4', 'NOTCH3', 'SPARCL1'), pt.size = 0)

marks <- de_dotplot(seu, metadata = 'RNA_snn_res.0.2')

# MAST CELL MARKERS
DotPlot(seu, features = c('KIT', 'CLU', 'CPA3', 'TPSAB1', 'TPSB2', 'HPGDS'))
VlnPlot(seu, features = c('KIT', 'CLU', 'CPA3', 'TPSAB1', 'TPSB2', 'HPGDS'), pt.size = 0)

DotPlot(seu, group.by = 'RNA_snn_res.0.2', features = c('EPCAM', 'DCN', 'COL1A1', 'CD3E', 'LYZ', 'CD79A', 'JCHAIN', 'TPSB2', 'CLDN5', 'C1QA', 'CD1C', 'CLEC9A', 'CD14', 'TAGLN', 'KLRG1', 'FOXP3', 'LILRA4', 'SDC1'))

seu$global.cluster <- as.character(seu$RNA_snn_res.0.2)
seu$global.cluster[seu$global.cluster == '0'] <- '0:Epithelial cells'
seu$global.cluster[seu$global.cluster == '1'] <- '1:Epithelial cells'
seu$global.cluster[seu$global.cluster == '2'] <- '2:Myeloid cells'
seu$global.cluster[seu$global.cluster == '3'] <- '3:T/NK cells'
seu$global.cluster[seu$global.cluster == '4'] <- '4:Fibroblasts/stromal cells'
seu$global.cluster[seu$global.cluster == '5'] <- '5:Plasma/B cells'
seu$global.cluster[seu$global.cluster == '6'] <- '6:Endothelial cells'
seu$global.cluster[seu$global.cluster == '7'] <- '7:T/NK cells'
seu$global.cluster[seu$global.cluster == '8'] <- '8:Plasma/B cells'
seu$global.cluster[seu$global.cluster == '9'] <- '9:Neutrophils'
seu$global.cluster[seu$global.cluster == '10'] <- '10:Dendritic cells'
seu$global.cluster[seu$global.cluster == '11'] <- '11:Plasma/B cells'
seu$global.cluster[seu$global.cluster == '12'] <- '12:Myeloid cells'

#seu$hpv_status <- with(seu, ifelse(seu$orig.ident %in% c('Vulva_#1_Tumor', 'Vulva_#4_Tumor', 'Vulva_#4_AdjNorm'), 'HPV-', 'HPV+'))

DimPlot(seu, group.by = 'global.cluster', cols = color_clusters, label = T, repel = T)
```

![](images/paste-0DD4DBF3.png)

## • Plotting representative markers

```{r}
DotPlot(seu, group.by = 'RNA_snn_res.0.2', features = c('EPCAM', 'DCN', 'COL1A1', 'CD3E', 'LYZ', 'CD79A', 'JCHAIN', 'TPSB2', 'CLDN5', 'C1QA', 'CD1C', 'CLEC9A', 'CD14', 'TAGLN', 'KLRG1', 'FOXP3', 'LILRA4', 'SDC1'))

FeaturePlot(seu, features = c('EPCAM', 'DCN', 'COL1A1', 'CD3E', 'LYZ', 'CD79A', 'JCHAIN', 'TPSB2', 'CLDN5', 'C1QA', 'CD1C', 'CLEC9A', 'CD14', 'TAGLN', 'KLRG1', 'FOXP3', 'LILRA4', 'SDC1'), ncol = 3)

# from Bedard et al 2023
seu[['HIDDEN_signature']] <- PercentageFeatureSet(seu, features = c('ELF3', 'TMC4', 'TMPRSS4', 'TMC5', 'CLDN7', 'LLGL2', 'ERBB3', 'CLDN4', 'MACC1', 'MYO5B', 'RAB11FIP1', 'CEACAM6', 'MUC1', 'PDE4C', 'ERBB2')) #'BCAS1', 'GABR' not in object

FeaturePlot(seu, features = c('HIDDEN_signature'))
DimPlot(seurat, group.by = 'hpv_status')

seu$hpv_status <- seu$orig.ident
seu$hpv_status[seu$hpv_status %like any% c('%#1%', '%#4%')] <- 'HPV-'
seu$hpv_status[seu$hpv_status %like any% c('%#2%')] <- 'HPV+'
```

## • Frequency plot

```{r}
seu_mini <- seu[,!seu$orig.ident == 'Vulva_#4_AdjNorm']

freqs <- freq_boxplot(seu, metadata = 'global.cluster', split.meta = 'hpv_status', out = './figures/global_cluster_all_', out_width = 19, out_height = 10)

freqs2 <- freq_boxplot(seu_mini, metadata = 'global.cluster', split.meta = 'hpv_status', out = './figures/global_cluster_noAdjNorm_', out_width = 19, out_height = 10)

seu$global.cluster <- factor(Idents(seu), levels = c("Epithelial/Tumor cells", 'Monocytes/Macrophages', 'Endothelial cells', 'T cells', 'Fibroblasts', 'NK', 'CD20_Bcells', 'DCs', 'Neutrophils', 'pDCs', 'B/Plasma cells', 'Mast cells'))

freqs <- freq_plot(seu_mini, metadata = 'global.cluster', split.meta = 'orig.ident', out = './figures/global_cluster_all_', graph_type = 'stacked_barplot', out_width = 5, out_height = 6)
```

## • rBLAST

```{r}
library(rBLAST)
seq <- readRNAStringSet('./Vulva_fastqs_from_BAM/V4T/bamtofastq_S1_L001_R1_001.fasta')

bl <- blast(db="/Volumes/hqdinh2/Projects/Anogenital_p30/inputs/BLAST/references/nt_viruses.00")
```

## • [ACTIVE] CopyKAT (on individual)

### • Run

```{r}
v1t <- readRDS("/Volumes/hqdinh2/Projects/Anogenital_p30/seurat_objs/SoupX_FilteredH5/2023-04-19_v1t.RDS")

v2t <- readRDS("/Volumes/hqdinh2/Projects/Anogenital_p30/seurat_objs/SoupX_FilteredH5/2023-04-19_v2t.RDS")

v4t <- readRDS("/Volumes/hqdinh2/Projects/Anogenital_p30/seurat_objs/SoupX_FilteredH5/2023-04-19_v4t.RDS")

copykat_run <- function(seu, sample_name){
  
  require(copykat)
  
  exp.rawdata <- as.matrix(seu@assays$RNA@counts)
  copykat.test <- copykat(rawmat = exp.rawdata, 
                  id.type='S', cell.line ='no', 
                  ngene.chr=5, win.size=25, 
                  KS.cut=0.2, sam.name = 'test', 
                  distance='euclidean', n.cores = 4)
  
  return(copykat.test)
}

# running for V1T
v1t_ck <- copykat_run(v1t, 'v1t')
saveRDS(v1t_ck, file = './outputs/copyKAT_individual/v1t_ck.RDS')

v1t_predictions <- v1t_ck$prediction

ind <- match(v1t_predictions$cell.names, colnames(v1t))
v1t$copykat <- 'Filler'
v1t$copykat[ind] <- v1t_predictions$copykat.pred

DimPlot(v1t, group.by = 'copykat')
saveRDS(v1t, file = "/Volumes/hqdinh2/Projects/Anogenital_p30/seurat_objs/SoupX_FilteredH5/2023-04-19_v1t.RDS")


# running for V2T
v2t_ck <- copykat_run(v2t, 'v2t')
saveRDS(v2t_ck, file = './outputs/copyKAT_individual/v2t_ck.RDS')

v2t_predictions <- v2t_ck$prediction

ind <- match(v2t_predictions$cell.names, colnames(v2t))
v2t$copykat <- 'Filler'
v2t$copykat[ind] <- v2t_predictions$copykat.pred

DimPlot(v2t, group.by = 'copykat')
FeaturePlot(v2t, features = 'KRT17')
saveRDS(v2t, file = "/Volumes/hqdinh2/Projects/Anogenital_p30/seurat_objs/SoupX_FilteredH5/2023-04-19_v2t.RDS")

# running for V4T
v4t_ck <- copykat_run(v4t, 'v4t')
saveRDS(v4t_ck, file = './outputs/copyKAT_individual/v4t_ck.RDS')

v4t_predictions <- v4t_ck$prediction

ind <- match(v4t_predictions$cell.names, colnames(v4t))
v4t$copykat <- 'Filler'
v4t$copykat[ind] <- v4t_predictions$copykat.pred

DimPlot(v4t, group.by = 'copykat')
FeaturePlot(v4t, features = 'KRT17')
saveRDS(v4t, file = "/Volumes/hqdinh2/Projects/Anogenital_p30/seurat_objs/SoupX_FilteredH5/2023-04-19_v4t.RDS")
```

### •• Remap to larger object

```{r}
v1t_ck <- readRDS(file = './outputs/copyKAT_individual/v1t_ck.RDS')
v2t_ck <- readRDS(file = './outputs/copyKAT_individual/v2t_ck.RDS')
v4t_ck <- readRDS(file = './outputs/copyKAT_individual/v4t_ck.RDS')

all <- rbind(v1t_ck$prediction, v2t_ck$prediction, v4t_ck$prediction)

ncol(seu) #19267
nrow(all) #19267

ind <- match(all$cell.names, colnames(seu))
seu$copykat <- 'Filler'
seu$copykat[ind] <- all$copykat.pred

DimPlot(seu, group.by = 'copykat')
```

### •• Simple Statistics

```{r}

# on new copykat on individual samples (metadata found in vulvar_combined_3samples_2023-05-12.RDS)
length(seu$copykat[seu$copykat == 'aneuploid'])
# 6894

library(DescTools)
length(seu$copykat[seu$global.cluster %like% '%Tumor%'])
# 10356

# pct abnormal epithelial cells
length(seu$copykat[seu$copykat == 'aneuploid']) / length(seu$copykat[seu$global.cluster %like% '%Tumor%'])
#  0.665701


# on old copykat version on merged atlas of only epithelial cells (metadata found in vulvar_combined_3samples_2023-05-05.RDS)

length(seu$copykat[seu$copykat == 'aneuploid'])
# 3323

library(DescTools)
length(seu$copykat[seu$global.cluster %like% '%Tumor%'])
# 10356

# pct abnormal epithelial cells
length(seu$copykat[seu$copykat == 'aneuploid']) / length(seu$copykat[seu$global.cluster %like% '%Tumor%'])
#  0.3208768
```

### •• Plotting comparisons

```{r}
seu[['epithelial.score']] <- PercentageFeatureSet(object = seu, features = c('EPCAM', 'KRT8', 'KRT19'))

FeaturePlot(seu, features = 'epithelial.score',cols = c('light blue', 'dark red'))

```

## • CopyKAT (on merged atlas)

```{r}
seu_mini <- seu[,seu$global.cluster == 'Epithelial/Tumor cells']
ncol(seu)

library(copykat)

exp.rawdata <- as.matrix(seu@assays$RNA@counts)
copykat.test <- copykat(rawmat = exp.rawdata, id.type='S', cell.line ='no', ngene.chr=5, win.size=25, KS.cut=0.2, sam.name = 'test', distance='euclidean', n.cores = 4)

old_copykat <- readRDS("/Volumes/hqdinh2/Projects/Anogenital_p30/outputs/copykat_all_output_2023-05-05.RDS")

ck1 <- copykat$prediction

seu$copykat <- 'diploid'

ind <- match(ck1$cell.names, colnames(seu))
seu$copykat[ind] <- ck1$copykat.pred
```

# Testing inferCNV

```{r}
library(infercnv)

test_raw_counts <- read.delim(system.file("extdata", "oligodendroglioma_expression_downsampled.counts.matrix.gz", package = "infercnv"))

test_annotations <- read.delim(system.file("extdata", "oligodendroglioma_annotations_downsampled.txt", package = "infercnv"))

test_gencode <- read.delim(system.file("extdata", "gencode_downsampled.EXAMPLE_ONLY_DONT_REUSE.txt", package = "infercnv"))

infercnv_obj = CreateInfercnvObject(raw_counts_matrix=system.file("extdata", "oligodendroglioma_expression_downsampled.counts.matrix.gz", package = "infercnv"),
                                    annotations_file=system.file("extdata", "oligodendroglioma_annotations_downsampled.txt", package = "infercnv"),
                                    delim="\t",
                                    gene_order_file=system.file("extdata", "gencode_downsampled.EXAMPLE_ONLY_DONT_REUSE.txt", package = "infercnv"),
                                    ref_group_names=c("Microglia/Macrophage","Oligodendrocytes (non-malignant)")) 

infercnv_obj = infercnv::run(infercnv_obj,
                             cutoff=1, # cutoff=1 works well for Smart-seq2, and cutoff=0.1 works well for 10x Genomics
                             out_dir=tempfile(), 
                             cluster_by_groups=TRUE, 
                             denoise=TRUE,
                             HMM=TRUE)
```

## • InferCNV

### •• Prepping documents

```{r}
# get necessary inputs
counts <- GetAssayData(object = seu, slot = "counts", assay = 'RNA')
annotation <- seu[[c('orig.ident', 'global.cluster', 'orig.ident')]]

library(dplyr)

# Replace 'annotation' with your dataframe name
annotation <- annotation %>%
  # Create a new column 'cluster' and set it equal to 'global.cluster'
  mutate(cluster = global.cluster) %>%
  # Use 'case_when' to conditionally modify 'cluster' based on a prefix
  mutate(cluster = case_when(
    startsWith(cluster, 'Epithelial') ~ paste(cluster, orig.ident, sep = '_'),
    TRUE ~ cluster # keep 'cluster' as is if no prefix matches
  )) %>%
  # Use 'coalesce' to set 'cluster' to 'global.cluster' if 'cluster' is NULL
  mutate(cluster = coalesce(cluster, global.cluster)) %>%
  # Create new columns 'barcode' and 'cellname' using 'rowname' and 'orig.ident'
  mutate(barcode = rownames(.),cellname = paste(barcode, orig.ident, sep = '_')) %>%
  # Select only the columns you want to keep in the final dataframe
  select(barcode, cluster)

write.csv(counts, file = './inputs/InferCNV/Vulvar_3_samples/raw_counts_matrix.csv')
write.table(annotation, file = './inputs/InferCNV/Vulvar_3_samples/annotations_file.txt', row.names = FALSE, col.names = FALSE, sep = '\t')

save(counts, annotation, ref_group_names, seu, file = './inputs/InferCNV/Vulvar_3_samples/inferCNV_image_05082023.rda')
```

### •• InferCNV mini run

```{r}
library(infercnv)
library(DescTools)
load('./inputs/InferCNV/Vulvar_3_samples/inferCNV_image_05082023.rda')

# counts mini
counts_small <- counts[,1:800]
cells <- colnames(counts_small)
ncol(counts_small) # 800 cells
nrow(counts_small) # 17716 genes

ann_small <- ann[ann$V1 %in% cells,]
rnames <- ann_small$V1
ann_small$V1 <- NULL
rownames(ann_small) <- rnames
ncol(ann_small) # 2 column
nrow(ann_small) # 800 cells

ref_groups <- unique(ann_small$V2)[which(!unique(ann_small$V2) %like% '%Tumor%')]

infercnv_obj = infercnv::CreateInfercnvObject(raw_counts_matrix=counts_small,annotations_file=ann_small, delim="\t", gene_order_file= './inputs/InferCNV/Vulvar_3_samples/hg38_gencode_v27_AEGedits.txt', ref_group_names=ref_groups)

infercnv_obj = infercnv::run(infercnv_obj,
              cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics
              out_dir="./outputs/InferCNV/MINI_Vulvar_3_samples_with_HMM/",  # dir is auto-created for storing outputs
              cluster_by_groups=T, HMM = T, HMM_type = c('i3'))

```

### •• InferCNV run

```{r}
library(infercnv)
library(DescTools)
load('./inputs/InferCNV/Vulvar_3_samples/inferCNV_image_05082023.rda')

ann <- read.delim('./inputs/InferCNV/Vulvar_3_samples/annotations_file.txt', header = F) #19627 cells
ncol(annotation) #2 columns
nrow(annotation) #19267 cells
counts_load <- read.csv(file = '/Volumes/hqdinh2/Projects/Anogenital_p30/inputs/InferCNV/Vulvar_3_samples/raw_counts_matrix.csv', header = T)
colnames(counts_load) <- gsub('\\.', '-', colnames(counts_load))

#ref_group_names <- unique(annotation$cluster[!annotation$cluster %like% 'Epithelial%'])
#gene_order_file = read.delim('./inputs/InferCNV/Vulvar_3_samples/hg38_gencode_v27.txt', header = FALSE)
#gene_order_file <- gene_order_file[gene_order_file$V1 %in% rownames(counts),] #17716 genes

#counts <- counts[rownames(counts) %in% gene_order_file$V1,]
nrow(counts_load) #17716 genes
ncol(counts) #19267 cells

#write.csv(counts, file = '/Volumes/hqdinh2/Projects/Anogenital_p30/inputs/InferCNV/Vulvar_3_samples/raw_counts_matrix.csv', row.names = F, col.names = T)

infercnv_obj = infercnv::CreateInfercnvObject(raw_counts_matrix=counts,annotations_file='./inputs/InferCNV/Vulvar_3_samples/annotations_file.txt', delim="\t", gene_order_file= './inputs/InferCNV/Vulvar_3_samples/hg38_gencode_v27_AEGedits.txt', ref_group_names=ref_group_names)

# perform infercnv operations to reveal cnv signal
infercnv_obj = infercnv::run(infercnv_obj,
                             cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics
                             out_dir="./outputs/InferCNV/Vulvar_3_samples_with_HMM_V3/",  # dir is auto-created for storing outputs
                             cluster_by_groups=T, 
                             HMM = T
                             )

saveRDS(infercnv_obj, file = './outputs/InferCNV/Vulvar_3_samples/initial_infercnv_obj_2023-05-09.RDS')
```

### •• InferCNV interface

```{r}
infercnv_obj <- './outputs/InferCNV/Vulvar_3_samples/final_infercnv_obj.RDS'

# where the abnormal cells are stored
infercnv_obj@observation_grouped_cell_indices

infernv_obj@reference_grouped_cell_indices
```

## • Testing SoupX

```{r}
library(SoupX)
seurat <- seu

# Extract raw counts
raw_counts <- GetAssayData(object = seu, assay = "RNA", slot = "counts")

# Extract filtered counts
filtered_counts <- GetAssayData(object = seu, assay = "RNA", slot = "data")

cm = SoupChannel(raw_counts, filtered_counts, calcSoupProfile = F)

length(cm$nDropUMIs[cm$nDropUMIs < 100])
# 0
length(cm$nDropUMIs[cm$nDropUMIs < 400])
# 2
length(cm$nDropUMIs[cm$nDropUMIs < 500])
# 122
length(cm$nDropUMIs[cm$nDropUMIs < 600])
# 707
length(cm$nDropUMIs[cm$nDropUMIs < 700])
# 1236
length(cm$nDropUMIs[cm$nDropUMIs < 800])
# 1717
length(cm$nDropUMIs[cm$nDropUMIs < 900])
# 2157
length(cm$nDropUMIs[cm$nDropUMIs < 1000])
# 2545

cm <- estimateSoup(cm, soupRange = c(0, 500)) #none of our data had <350 UMI, so I increased this range

clusters <- seu$global.cluster

cm = setClusters(cm, clusters)

cm = autoEstCont(cm)
#898 genes passed tf-idf cut-off and 230 soup quantile filter.  Taking the top 100.
#Using 1151 independent estimates of rho.
#Estimated global rho of 0.01

out = adjustCounts(cm)

#  We can get a sense for what has been the most strongly decreased by looking at the fraction of cells that were non-zero now set to zero after correction
cntSoggy = rowSums(cm$toc > 0)
cntStrained = rowSums(out > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 20)
mostZeroed
#SELE      LGR6     ZPBP2     WFDC9      NXF5    TUBA3C    ZNF233    BTBD16 
#0.6205788 0.6212121 0.6470588 0.6666667 0.6666667 0.7000000 0.7017544 0.7058824 
#C5orf67   C7orf57       FLG     SPTA1    KLHL10     DCDC1     TAAR5      CTSE 
#0.7142857 0.7183099 0.7380952 0.7500000 0.7741935 0.8461538 0.8571429 0.8800000 
#DKKL1     OR2H2        GC   LRRC74A 
#0.8947368 0.9583333 1.0000000 1.0000000 


#If on the other hand we focus on genes for which there is a quantitative difference, we see the markers that are over-represented in the original data and corrected as soup
tail(sort(rowSums(cm$toc > out)/rowSums(cm$toc > 0)), n = 20)
# GAB3    DKC1    MPP1    CMC4   BRCC3    VBP1   TMLHE   VAMP7    IL9R  MT-ND1  MT-ND2 
#1       1       1       1       1       1       1       1       1       1       1 
# MT-CO2 MT-ATP6  MT-CO3  MT-ND3 MT-ND4L  MT-ND4  MT-ND5  MT-ND6  MT-CYB 
# 1       1       1       1       1       1       1       1       1 

match(rownames(raw_counts), rownames(out))
#[1]    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16   17   18   19   20   21   22   23   24   25   26   27   28   29   30   31   32

match(colnames(raw_counts), colnames(out))
#[1]    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16   17   18   19   20   21   22   23   24   25   26   27   28   29   30   31   32

# Assuming you have a matrix called "soupX_data" with the SoupX expression data for each cell
new_assay_name <- "SoupX_counts"
seu <- AddAssay(object = seu, assay = new_assay_name)
seu[[new_assay_name]] <- CreateAssayObject(data = out, assay = new_assay_name)

```

![](images/paste-B1C3F620.png)

Looking at the low end counts for for lowest UMI count clusters. We shouldn't really make this any higher than 500UMIs.

![](images/Screenshot%202023-04-10%20at%204.19.49%20PM.png)

![](images/paste-FC097698.png)

### •• Pre-processing SoupX assay

```{r}
DefaultAssay(seu) <- 'SoupX_counts'

seu <- seu_pp1(seu, harmony = T, assay = 'SoupX_counts', normalize = F)
seu <- seu_pp2(seu, reduction = 'harmonySoupX_counts', assay = 'SoupX_counts', npcs = 75)

DimPlot(seu, group.by = 'global.cluster')
```

## • Deconvolution of Kolitz et al bulk RNA seq data

### •• Prepping the object

```{r}
seu <- readRDS('./seurat_objs/vulvar_combined.RDS')
#tnk <- readRDS('./seurat_objs/vulvar_TNK.RDS')
#my <- readRDS('./seurat_objs/vulvar_myeloid.RDS')

# moving T subsets to the global object
#ind <- match(rownames(tnk@meta.data), rownames(seu@meta.data))
#seu$decon.clusters[ind] <- tnk$global.cluster2

# moving myeloid subsets to the global object
#ind <- match(rownames(my@meta.data), rownames(seu@meta.data))
#seu$decon.clusters[ind] <- my$global.cluster2

DimPlot(seu, group.by = 'decon.clusters')

library(DescTools)

seu$decon.clusters[seu$decon.clusters %like any% c('Mac%', 'Mon%', 'Myeloid%')] <- 'Mon/Mac'
seu$decon.clusters[seu$decon.clusters %like any% c('%DC%', 'Dendritic%')] <- 'Dendritic cells'
seu$decon.clusters[seu$decon.clusters %like any% c('%CD8%')] <- 'CD8 T cells'
seu$decon.clusters[seu$decon.clusters %like any% c('%Treg%')] <- 'CD4 T cells'
#seu$decon.clusters[seu$decon.clusters == 'Myeloid cells'] 

#removing the tiny weird fibroblast/myeloid cluster for these purposes
seu_mini <- seu[,!seu$decon.clusters %in% c('Myeloid cells', 'T_cell_contam_from_myeloid')]
DimPlot(seu, group.by = 'decon.clusters')
```

### •• [DON'T USE] Running the deconvolution

```{r}
seu <- readRDS('./seurat_objs/vulvar_combined_2023-03-27.RDS')

## Code adapted from Anqi melanoma project
library("writexl")

immune.combined2 = seu 
DimPlot(immune.combined2, group.by = 'decon.clusters')
Idents(immune.combined2) <- "decon.clusters"
immune.combined2@active.assay = "RNA"
exp <- as.matrix(GetAssayData(object = immune.combined2, slot = "counts"))
samples <-unique(immune.combined2$orig.ident)
sampleID <- rep(1, ncol(exp))
#sampleID <- match(immune.combined2$orig.ident, samples)

cellType = as.character(Idents(immune.combined2))
unique(cellType)
cellTypeID <- match(cellType, unique(cellType))

pd <- data.frame(sampleID, cellTypeID, cellType)
rownames(pd) <- colnames(exp)

metaData <- data.frame(labelDescription=c("Numbers", "Numbers", "Factor levels"))

pheno = new("AnnotatedDataFrame", data=pd, varMetadata = metaData)

Markers <- vector('list', length(unique(pd$cellTypeID)))
names(Markers) <- unique(pd$cellType)

#save(exp, immune.combined2, Markers, metaData, pd, pheno, seu, file = './inputs/bulk_RNAseq_deconvolution_cpm.rda')

gene_name_all <- FindAllMarkers(immune.combined2)

#save(exp, immune.combined2, Markers, metaData, pd, pheno, seu, gene_name_all, file = './inputs/bulk_RNAseq_deconvolution_cpm.rda')

gene_name <- gene_name_all
gene_name <- gene_name[which(gene_name$avg_log2FC>1),]
gene_name <- gene_name[which(gene_name$cluster %in% names(Markers)),]
for (i in 1:length(Markers)) {
  Markers[[i]] <- gene_name$gene[which(gene_name$cluster == names(Markers)[i])]
}
count <- unlist(lapply(1:length(Markers), function(x) length(Markers[[x]])))
Markers <- Markers[which(count>0)]
for (i in 1:length(Markers)) Markers[[i]] <- Markers[[i]][1:100]

exp2 <- exp[which(rownames(exp) %in% unlist(Markers)),]
eislet = new("ExpressionSet", exprs=exp2, phenoData=pheno)

#save(exp, immune.combined2, Markers, metaData, pd, pheno, seu, gene_name_all, gene_name, count, exp2, eislet, file = './inputs/bulk_RNAseq_deconvolution_cpm.rda')

####read in bulk RNASeq data as raw counts (for MuSiC at least)
e2 <- data.frame(read.delim("/Volumes/hqdinh2/Projects/Public_Data/Kolitz_Vulvar_cancer/GSE183454_gene_count_aeg_rename.txt"))

# converting th# 1. Convert from ensembl.gene to gene.symbol
ensembl.genes <- e2$gene
ensembl.genes2 <- sapply(strsplit(ensembl.genes,"\\."), `[`, 1)
e2$gene <- ensembl.genes2

library(EnsDb.Hsapiens.v79)
library(biomaRt)
geneIDs1 <- ensembldb::select(EnsDb.Hsapiens.v79, keys= ensembl.genes2, keytype = "GENEID", columns = c("SYMBOL","GENEID"))

e2 <- merge(geneIDs1, e2, by.x = 'GENEID', by.y = 'gene')

save(exp, immune.combined2, Markers, metaData, pd, pheno, seu, tnk, gene_name, ensembl.genes, ensembl.genes2, e2, eislet, count, file = './inputs/bulk_RNAseq_deconvolution.rda')

load('./inputs/bulk_RNAseq_deconvolution.rda')

gene_list <- unique(as.character(unlist(Markers)))
e2 <- e2[,2:ncol(e2)]
e2 <- e2[which(e2[,1] %in% gene_list),]
e2 <- aggregate(x = e2[,2:ncol(e2)], by = list(e2$SYMBOL), FUN = sum)
rownames(e2) = e2[,1]
e2 <- e2[,-1]

#Add meta data classifying samples as "A" or "B" treatment categories 
rownames(pd) <- colnames(e2)
metaData <- data.frame(labelDescription=c("Character"))
pheno = new("AnnotatedDataFrame", data=pd, varMetadata = metaData)

e2 <- edgeR::cpm(e2)

eset = new("ExpressionSet", exprs=as.matrix(e2), phenoData=pheno)

require(bseqsc)
plotCellTotals(eislet, 'cellType', 'sampleID')
B <- bseqsc_basis(eislet, Markers, clusters = 'cellType', samples = 'sampleID', ct.scale = TRUE)
plotBasis(B, Markers, Colv = NA, Rowv = NA, col = 'Blues')

bseqsc_config('~/CIBERSORT files/CIBERSORT.R')
fit <- bseqsc_proportions(eset, B, verbose = TRUE)

pData(eset) <- cbind(pData(eset), t(coef(fit)))
pData(eset) <- cbind(is_T = rep('Y', nrow(pData(eset))), pData(eset))

results <- as.matrix(t(coef(fit)))
results2 = as.data.frame(results)
#results3 = add_metadata(results2, metadata2022, xid = NULL, mid = NULL, drop_mid = T)


ITICproportionsOG <- as.data.frame(results)
write_xlsx(ITICproportionsOG,"~/CIBERSORT files/Melanoma_ITICproportionsOG.xlsx")

#results <- results[which(fit$stats[,1]<0.05),]
fit$stats

png('./figures/results_Kolitz_deconvolution_boxplot_cpmV2.png', width = 8, height = 8, units = 'in', res = 300)
par(mar=c(15.1,4.1,4.1,2.1))
boxplot(results, las = 2, ylab = "Proportion", main = "ITIC Deconvolution")
dev.off()

#boxplot with individual dots
results2 %>% 
  rownames_to_column(var="sample") %>%
  pivot_longer(cols = !contains("sample")) %>%
  ggplot(aes(x = name, y = value)) + 
  geom_boxplot(aes(fill = name), outlier.size = -1, alpha = 0.6) +
  geom_point(position = position_jitter()) + 
  theme_classic() + 
  theme(axis.text.x = element_text(color = "grey20", size = 14, angle = 30, hjust = .5, vjust = .5, face = "plain"),
                          axis.text.y = element_text(color = "grey20", size = 14, angle = 0, hjust = 1, vjust = 0, face = "plain"),  
                          axis.title.x = element_text(color = "grey20", size = 14, angle = 0, hjust = .5, vjust = 0, face = "plain"),
                          axis.title.y = element_text(color = "grey20", size = 14, angle = 90, hjust = .5, vjust = .5, face = "plain")) +
  theme(legend.text=element_text(size=16))

ggsave("ggsave/results2.pdf", results2, width = 10, height = 5)

#install.packages("ggpubr"); library(ggpubr)
#meta_boxplot = cbind(sample = colnames(results), treatment = eset$subtype) %>% as.data.frame()
results
long_df = results %>% as.data.frame() %>% rownames_to_column("Name") %>%
  mutate(Name = stringr::str_remove(Name, "_T_H7VJ5BGXL")) %>%
  gather("sample","proportion_of_sample", 
         `B cells`:`CD4 and CD8 Memory T cells`)
#long_df_meta = long_df %>% left_join(meta_boxplot)

final_df = long_df %>% left_join(metadata2022)
ggpaired(final_df, x="sample", y = "proportion_of_sample" )

#boxplots separated by "A" and "B" arms
meta_boxplot = cbind(sample = colnames(coef(fit)), treatment = eset$subtype) %>% as.data.frame()
long_df = coef(fit) %>% as.data.frame() %>% rownames_to_column("cell_type") %>% 
  gather("sample","proportion_of_sample", 
         Sondel_RNA_A_T_HJTKFBGX3:Sondel_RNA_S_T_HJ2H5BGX3)
long_df_meta = long_df %>% left_join(meta_boxplot)

ggplot7clusters <- ggplot(long_df_meta, aes(x = cell_type, y = proportion_of_sample, color = treatment)) + 
  geom_boxplot() + theme(axis.text.x = element_text(angle = 0, vjust = 0.5, hjust=0.5, size = 12)) +
  theme(axis.text.y = element_text(angle = 0, vjust = 0.5, hjust=1, size = 12)) +
  ggtitle("Deconvolution based on 3 clusters") + xlab("Cell type") + ylab("Proportion")

ggplot7clusters + theme(
  plot.title = element_text(color="black", size=20, face="bold"),
  axis.title.x = element_text(color="black", size=14, face="bold"),
  axis.title.y = element_text(color="black", size=14, face="bold")
) + theme(plot.title = element_text(hjust = 0.5))

colnames(pd) <- 'HPV status'

#require(pheatmap)
png('./figures/Kolitz_deconvolution_cpmV2.png', width = 8, height = 6, units = 'in', res = 300)
pheatmap::pheatmap(results[order(results$`Epithelial cells`),], show_rownames = F, annotation_row = pd, cluster_rows = F)
dev.off()

write.csv(results, file = './outputs/results_Kolitz_deconvolution-cpmV2.csv')

save(eset, eislet, file = "for_deconvolution_3celltypes.rds")
```

### •• Huy's code for deconvolution

```{r}
load('/Volumes/hqdinh2/Projects/Anogenital_p30//inputs/bulk_RNAseq_deconvolution_cpmV2.rda')
#require(bseqsc)
#plotCellTotals(eislet, 'cellType', 'sampleID')
#B <- bseqsc_basis(eislet, Markers, clusters = 'cellType', samples = #'sampleID', ct.scale = TRUE)
#plotBasis(B, Markers, Colv = NA, Rowv = NA, col = 'Blues')
#library(edgeR)
#f <- plyr::count(e2$SYMBOL)
#e2 <- e2[-which(e2$SYMBOL %in% f$x[which(f$freq>1)]),]
#e2_count <- e2[,3:ncol(e2)]
#rownames(e2_count) <- e2$SYMBOL
#e2_cpm <- cpm(as.matrix(e2_count))
#Add meta data classifying samples as “A” or “B” treatment categories
#rownames(pd) <- colnames(e2_cpm)
#metaData <- data.frame(labelDescription=c('Character'))
#pheno = new('AnnotatedDataFrame', data=pd, varMetadata = metaData)
#eset = new('ExpressionSet', exprs=as.matrix(e2_cpm), phenoData=pheno)
#bseqsc_config('./inputs/CIBERSORT files/CIBERSORT.R')
#fit <- bseqsc_proportions(eset, B, verbose = TRUE)
#pData(eset) <- cbind(pData(eset), t(coef(fit)))
#pData(eset) <- cbind(is_T = rep('Y', nrow(pData(eset))), pData(eset))
#results <- as.matrix(t(coef(fit)))
#results2 = as.data.frame(results)
#par(mar=c(15.1,4.1,4.1,2.1))
#boxplot(results, las = 2, ylab = 'Proportion', main = 'ITIC Deconvolution')

results$sample <- rownames(results)
results2 <- pivot_longer(results, cols = colnames(results[,1:ncol(results)-1]), names_to= 'cluster')
results2[c('hpv_status', 'orig.ident')] <- str_split_fixed(results2$sample, '_', 2)

results2$hpv_status <- factor(results2$hpv_status,
                    levels = c('HPVpos', 'HPVneg'))

ggplot(results2, aes(x=hpv_status, y=value, fill = hpv_status, colour = hpv_status)) + 
    geom_boxplot(position = position_dodge(), alpha = 0.5, outlier.color = NA) + facet_wrap(~cluster, scale="free", nrow = 2) + geom_point(aes(x = hpv_status, y = value, color = hpv_status), alpha = 0.8, position = position_jitterdodge())  + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:5)

ggsave('./figures/Kolitz_deconvolution_freq_boxplot.png', width = 15, height = 10, units = 'in', dpi = 300)
write.csv(results2, './outputs/Kolitz_deconvolution_freq_boxplot.csv')
```

#### ••• Wilcox test

```{r}
results2 <- read.csv('./outputs/Kolitz_deconvolution_freq_boxplot.csv')

results2_wc <- results2[results2$cluster == 'CD8 T cells',]
wc_hpvcode <- pairwise.wilcox.test(results2_wc$value, results2_wc$hpv_status, p.adjust.method = 'BH')

wc_hpvcode$p.value
#          HPVneg
# HPVpos 0.3041831

results2_wc <- results2[results2$cluster == 'Neutrophils',]
wc_hpvcode <- pairwise.wilcox.test(results2_wc$value, results2_wc$hpv_status, p.adjust.method = 'BH')

wc_hpvcode$p.value
#           HPVneg
# HPVpos 0.04174578
```

## • [ACTIVE] Plotting Megan's nCounter deconvolution

### •• Necessary files to load in

```{r}
# metadata file
ann_df <- read.csv('/Volumes/hqdinh2/Projects/Collaborations/MeganFitzpatrick/Annotation_files/Vulva_annotation_05.30.23.csv')


# deconvolution results
nc <- read.csv('/Volumes/hqdinh2/Projects/Collaborations/MeganFitzpatrick/nCounter_vulvars_deconvolution_Huy.csv')

nc$sample <- nc$X
nc$X <- NULL
nc2 <- pivot_longer(nc, cols = colnames(nc[,1:ncol(nc)-1]), names_to= 'cluster')
n3 <- merge(nc2, ann_df, by.x = 'sample', by.y = 0)

n3$p16[n3$p16 == 'N'] <- 'p16neg'
n3$p16[n3$p16 == 'P'] <- 'p16pos'

n3$p16 <- factor(n3$p16, levels = c('p16pos', 'p16neg'))
```

### •• Boxplot split by p16 status

```{r}
# split by p16 status
ggplot(n3, aes(x=p16, y=value, fill = p16, colour = p16)) + 
    geom_boxplot(position = position_dodge(), alpha = 0.5, outlier.color = NA) + facet_wrap(~cluster, scale="free", nrow = 2) + geom_point(aes(x = p16, y = value, color = p16), alpha = 0.8, position = position_jitterdodge())  + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:5)

ggsave('./figures/MBFncounter_deconvolution_freq_boxplot.png', width = 15, height = 10, units = 'in', dpi = 300)
write.csv(n3, './outputs/p16_MBFncounter_deconvolution_freq_boxplot.png')
```

### •• Split by PRECURSOR.INVASION

#### ••• Boxplot

```{r}
# split by cluster from heatmap
test <- table(ann_df$Case.ID)
matched <- names(test[test >= 2])
n3 <- n3[n3$Case.ID %in% matched,]
# removing one case where there are two precursor lesions
n3 <- n3[!n3$Case.ID %in% c(3,5),]

n3$Case.ID <- as.factor(n3$Case.ID)

ggplot(n3, aes(x=cluster.y, y=value, fill = cluster.y)) + 
    geom_boxplot(position = position_dodge(), alpha = 0.5, outlier.color = NA) + facet_wrap(~cluster.x, scale="free", nrow = 2) + geom_point(aes(x = cluster.y, y = value, color = cluster.y, shape = Case.ID), alpha = 0.8, size = 3)  + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:5)  + scale_shape_manual(values=1:nlevels(n3$Case.ID))



# split by PRECURSOR.INVASION
n3 <- n3[!n3$PRECURSOR.INVASION %in% c('LS', 'POST SURGERY NEGATIVE'),]
test <- table(ann_df$Case.ID)
matched <- names(test[test >= 2])
n3 <- n3[n3$Case.ID %in% matched,]
# removing one case where there are two precursor lesions
n3 <- n3[!n3$Case.ID %in% c(3,5),]

n3$PRECURSOR.INVASION[n3$PRECURSOR.INVASION == 'FOCAL INVASIVE'] <- 'INVASIVE'

n3$PRECURSOR.INVASION <- factor(n3$PRECURSOR.INVASION, levels = c('PRECURSOR', 'FOCAL INVASIVE', 'INVASIVE'))

table(n3$PRECURSOR.INVASION, n3$Case.ID)

n3$Case.ID <- as.factor(n3$Case.ID)

ggplot(n3, aes(x=PRECURSOR.INVASION, y=value, fill = PRECURSOR.INVASION)) + 
    geom_boxplot(position = position_dodge(), alpha = 0.5, outlier.color = NA) + facet_wrap(~cluster, scale="free", nrow = 2) + geom_point(aes(x = PRECURSOR.INVASION, y = value, color = PRECURSOR.INVASION, shape = Case.ID), alpha = 0.8, size = 3)  + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:5) + geom_line(aes(group=Case.ID), alpha = 0.2) + scale_shape_manual(values=1:nlevels(n3$Case.ID))

ggsave('./figures/PRECURSOR.INVASION_MBFncounter_deconvolution_freq_boxplot.png', width = 17, height = 10, units = 'in', dpi = 300)
write.csv(n3, './outputs/PRECURSOR.INVASION,_MBFncounter_deconvolution_freq_boxplot.csv')
```

#### ••• Heatmap

```{r}
pd <- ann_df[c('PRECURSOR.INVASION')]
pd$ID <- rownames(pd)
n4 <- n3[c('sample', 'cluster', 'value')] %>% pivot_wider(names_from = 'cluster')
rnames <- n4$sample
n4$sample <- NULL
rownames(n4) <- rnames

pd <- pd[rownames(pd) %in% rnames,]
pd$PRECURSOR.INVASION[pd$PRECURSOR.INVASION == 'FOCAL INVASIVE'] <- 'INVASIVE'
colmax<- data.frame(apply(n4, 2, function(x) x / max(x)))

pheatmap::pheatmap(colmax, show_rownames = F, annotation_row = pd, cluster_rows = T)
dev.off()
```

#### ••• Scatterplots

```{r}
n4 <- n3[c('Case.ID', 'PRECURSOR.INVASION', 'cluster', 'value')]

# calculating dendritic cell delta
n4_dc_pre <- n4[n4$cluster == c('Dendritic.cells') & n4$PRECURSOR.INVASION == c('PRECURSOR') & !n4$Case.ID == 3,c('Case.ID', 'value')] 
colnames(n4_dc_pre)[2] <- 'DC_precursor'

n4_dc_inv <- n4[n4$cluster == c('Dendritic.cells') & n4$PRECURSOR.INVASION %in% c('FOCAL INVASIVE', 'INVASIVE') & !n4$Case.ID == 3,c('Case.ID', 'value')] 
colnames(n4_dc_inv)[2] <- 'DC_invasive'

dc_delta <- merge(n4_dc_pre, n4_dc_inv, by = 'Case.ID')
dc_delta$DC_delta <- dc_delta$DC_precursor - dc_delta$DC_invasive
dc_delta <- dc_delta[c('Case.ID', 'DC_delta')]



# calculating cd4 T cell delta
n4_cd4_pre <- n4[n4$cluster == c('CD4.T.cells') & n4$PRECURSOR.INVASION == c('PRECURSOR') & !n4$Case.ID == 3,c('Case.ID', 'value')] 
colnames(n4_cd4_pre)[2] <- 'cd4_precursor'

n4_cd4_inv <- n4[n4$cluster == c('CD4.T.cells') & n4$PRECURSOR.INVASION %in% c('FOCAL INVASIVE', 'INVASIVE') & !n4$Case.ID == 3,c('Case.ID', 'value')] 
colnames(n4_cd4_inv)[2] <- 'cd4_invasive'

cd4_delta <- merge(n4_cd4_pre, n4_cd4_inv, by = 'Case.ID')
cd4_delta$cd4_delta <- cd4_delta$cd4_precursor - cd4_delta$cd4_invasive
cd4_delta <- cd4_delta[c('Case.ID', 'cd4_delta')]



# calculating cd8 T cell delta
n4_cd8_pre <- n4[n4$cluster == c('CD8.T.cells') & n4$PRECURSOR.INVASION == c('PRECURSOR') & !n4$Case.ID == 3,c('Case.ID', 'value')] 
colnames(n4_cd8_pre)[2] <- 'cd8_precursor'

n4_cd8_inv <- n4[n4$cluster == c('CD8.T.cells') & n4$PRECURSOR.INVASION %in% c('FOCAL INVASIVE', 'INVASIVE') & !n4$Case.ID == 3,c('Case.ID', 'value')] 
colnames(n4_cd8_inv)[2] <- 'cd8_invasive'

cd8_delta <- merge(n4_cd8_pre, n4_cd8_inv, by = 'Case.ID')
cd8_delta$cd8_delta <- cd8_delta$cd8_precursor - cd8_delta$cd8_invasive
cd8_delta <- cd8_delta[c('Case.ID', 'cd8_delta')]



# calculating mon.mac T cell delta
n4_mon.mac_pre <- n4[n4$cluster == c('Mon.Mac') & n4$PRECURSOR.INVASION == c('PRECURSOR') & !n4$Case.ID == 3,c('Case.ID', 'value')] 
colnames(n4_mon.mac_pre)[2] <- 'mon.mac_precursor'

n4_mon.mac_inv <- n4[n4$cluster == c('Mon.Mac') & n4$PRECURSOR.INVASION %in% c('FOCAL INVASIVE', 'INVASIVE') & !n4$Case.ID == 3,c('Case.ID', 'value')] 
colnames(n4_mon.mac_inv)[2] <- 'mon.mac_invasive'

mon.mac_delta <- merge(n4_mon.mac_pre, n4_mon.mac_inv, by = 'Case.ID')
mon.mac_delta$mon.mac_delta <- mon.mac_delta$mon.mac_precursor - mon.mac_delta$mon.mac_invasive
mon.mac_delta <- mon.mac_delta[c('Case.ID', 'mon.mac_delta')]


# plotting each of the combinations
cd4_mon.mac <- merge(cd4_delta, mon.mac_delta, by = 'Case.ID')
p1 <- ggplot(cd4_mon.mac, aes(x=cd4_delta, y=mon.mac_delta)) + geom_point() + theme_classic() + geom_smooth(method="lm", se=FALSE) 

cd8_mon.mac <- merge(cd8_delta, mon.mac_delta, by = 'Case.ID')
p2 <- ggplot(cd8_mon.mac, aes(x=cd8_delta, y=mon.mac_delta)) + geom_point() + theme_classic() + geom_smooth(method="lm", se=FALSE) 

cd4_dc <- merge(cd4_delta, dc_delta, by = 'Case.ID')
p3 <- ggplot(cd4_dc, aes(x=cd4_delta, y=DC_delta)) + geom_point() + theme_classic() + geom_smooth(method="lm", se=FALSE) 

cd8_dc <- merge(cd8_delta, dc_delta, by = 'Case.ID')
p4 <- ggplot(cd8_dc, aes(x=cd8_delta, y=DC_delta)) + geom_point() + theme_classic() + geom_smooth(method="lm", se=FALSE)

png('./figures/T_cell_to_myeloid_preVSinvasive_delta_scatterplots.png', width = 6, height = 5, units = 'in', res = 300)
cowplot::plot_grid(p1, p2, p3, p4)
dev.off()

```

### •• Boxplot split by HPV code

```{r}
# split by HPV_code
n3$HPV_code <- as.character(n3$HPV_code)
ggplot(n3[!n3$HPV_code == '4',], aes(x=HPV_code, y=value, fill =HPV_code, colour = HPV_code,6)) + geom_boxplot(position = position_dodge(), alpha = 0.5, outlier.color = NA) + facet_wrap(~cluster.x, scale="free", nrow = 2) + geom_point(aes(x = HPV_code, y = value, color = HPV_code), alpha = 0.8, position = position_jitterdodge())  + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:5) + theme(legend.key.size = unit(1, 'cm'), legend.title = element_text(size=30), legend.text = element_text(size=20), axis.text=element_text(size=16),axis.title=element_text(size=30)) + labs(y= "Cell Type Frequency")

ggsave('./figures/HPV_code_MBFncounter_deconvolution_freq_boxplot_allsamples.png', width = 19, height = 10, units = 'in', dpi = 300)
write.csv(n3, './outputs/HPV_code_MBFncounter_deconvolution_freq_boxplot_allsamples.csv')
```

#### ••• Wilcox test

```{r}
n3_wc <- n3[n3$cluster.x == 'CD8.T.cells',]
wc_hpvcode <- pairwise.wilcox.test(n3_wc$value, n3_wc$HPV_code, p.adjust.method = 'BH')

wc_hpvcode$p.value
#         0         1        2         3
#1 0.095644        NA       NA        NA
#2 0.095644 1.0000000       NA        NA
#3 0.095644 1.0000000 1.000000        NA
#4 1.000000 0.6835377 0.407561 0.6835377

wc_hpvcode_noadj <- pairwise.wilcox.test(n3_wc$value, n3_wc$HPV_code, p.adjust.method = 'none')

wc_hpvcode_noadj$p.value # pvals for CD8 T cell cluster
#           0         1         2         3
#1 0.02463230        NA        NA        NA
#2 0.02405158 0.8354845        NA        NA
#3 0.02869320 1.0000000 0.8403422        NA
#4 1.00000000 0.4101226 0.1630244 0.3931534

n3_wc <- n3[n3$cluster.x == 'Dendritic.cells',]
wc_hpvcode <- pairwise.wilcox.test(n3_wc$value, n3_wc$HPV_code, p.adjust.method = 'BH')

wc_hpvcode$p.value # pvals for dendritic cells cluster
#           0          1          2         3
#1 0.56729114         NA         NA        NA
#2 0.56729114 0.84484031         NA        NA
#3 0.08852346 0.04608605 0.04608605        NA
#4 0.56729114 0.56729114 0.56729114 0.8230296

n3_wc <- n3[n3$cluster.x == 'Epithelial.cells',]
wc_hpvcode <- pairwise.wilcox.test(n3_wc$value, n3_wc$HPV_code, p.adjust.method = 'BH')

wc_hpvcode$p.value # pvals for Epithelial cells cluster
#           0          1          2         3
#1 0.56729114         NA         NA        NA
#2 0.56729114 0.84484031         NA        NA
#3 0.08852346 0.04608605 0.04608605        NA
#4 0.56729114 0.56729114 0.56729114 0.8230296
```

## • Myeloid only

### •• Load in/save objects

```{r}
saveRDS(my, './seurat_objs/vulvar_myeloid.RDS')
my <- readRDS('./seurat_objs/vulvar_myeloid.RDS')
```

### •• global myeloid clustering

```{r}
#my <- seu[, seu$global.cluster %in% c('Myeloid cells', 'Dendritic cells', 'Neutrophils')]
#DimPlot(my, group.by = 'orig.ident')

#my <- seu_pp1(my, harmony = T, npcs = 100) # 30 PCS
#my <- seu_pp2(my, npcs = 30, reduction = 'harmony')
table(my$RNA_snn_res.0.5)
DimPlot(my, group.by = 'RNA_snn_res.0.5', label = T, label.box = T, repel = T)
DimPlot(my, group.by = 'global.cluster')

my_mini <- my[grep('KRT|COL', rownames(my), invert = T)]
marks <- de_dotplot(my_mini, metadata = 'RNA_snn_res.0.5', n_marks = 10, min_pct = 0.2)

Idents(my) <- my$RNA_snn_res.0.5
#cillo markers (my clustering)
VlnPlot(my, group.by = 'RNA_snn_res.0.5', features = c('SPP1', 'C1QA', 'CD14', 'FCGR3A', 'LAMP3', 'CCR7', 'CLEC9A', 'CLEC10A', 'CD1C', 'LYZ', 'IL3RA', 'CD123', 'CXCL9', 'PLTP', 'CD68', 'CD86', 'IL18R', 'CD123', 'CCR3', 'TLR2', 'TLR4', 'FCER1A', 'CLEC10A', 'CD1C', 'CCL20', 'CXCL8', 'CXCL3', 'CCL3L3', 'CXCL2', 'HSPB1', 'C15orf48', 'HSPA1A', 'MRC1', 'FN1', 'FBP1', 'TPSAB1', 'TPSB2', 'CPA3', 'HPGDS', 'CLU', 'S100A12', 'S100A8', 'S100A9', 'VCAN', 'MNDA', 'ID1', 'HES1', 'MAFB', 'DUSP6', 'IRF2BP2', 'CH17-373J23.1', 'NFKB1A', 'LINC00936', 'NFKBIZ', 'IL1R2', 'THBS1', 'FKBP5', 'LYZ'), pt.size = 0, ncol = 8)

#luoma markers
VlnPlot(my, group.by = 'RNA_snn_res.0.5', features = c('S100A8', 'FCN1', 'C1QB', 'C1QC', 'CXCL8', 'CCL3L1', 'IL32', 'FXYD5', 'FCER1A', 'CD1C', 'GZMB', 'PTGDS', 'CHI3L1', 'SPP1', 'FSCN1', 'CCL22', 'CPVL', 'S100B', 'APOC1', ''), pt.size = 0)
```

![](images/paste-6ED95CDA.png)

### •• Transfer annotations back to global cluster to check for contamination

```{r}
temp_meta <- ifelse(colnames(seu) %in% colnames(my),
                     as.character(my$RNA_snn_res.0.5),
                     as.character(seu$global.cluster))

seu$global.cluster2 <- temp_meta
Idents(seu) <- temp_meta

VlnPlot(seu, features = c('CD3E', 'CD4', 'CD8A'))

#checking myeloid cluster 2 markers in global merged object
VlnPlot(seu, group.by = 'global.cluster2', features = c('SFN', 'S100A2', 'SERPINE1', 'MMP1', 'DSP', 'PERP', 'NDRG1', 'COL17A1'))

#checking myeloid cluster 10 markers in the global merged object
VlnPlot(seu, group.by = 'global.cluster2', features = c('COL1A1', 'COL1A2', 'SPARC', 'COL3A1', 'COL6A2', 'COL6A1', 'COL6A3', 'COL4A1', 'BGN'))
```

### •• Destiny

```{r}
Idents(my) <- my$RNA_snn_res.0.5

ind <- 1:ncol(my@assays$RNA) #(ind variable here is for all cells, but if you have too many you can downsample using the sample function)
gene_ind <- which(rownames(my@assays$RNA) %in% VariableFeatures(my))
test <- as.matrix(my@assays$RNA[gene_ind,ind])
rownames(test) = rownames(my@assays$RNA)[gene_ind]
colnames(test) = colnames(my@assays$RNA)[ind]
library(destiny)
dif <- DiffusionMap(t(test))
color_clusters <- c('#DC050C', '#FB8072', '#1965B0', '#7BAFDE', '#882E72',
                             '#B17BA6', '#FF7F00', '#FDB462', '#E7298A', '#E78AC3',
                             '#33A02C', '#B2DF8A', '#55A1B1', '#8DD3C7', '#A6761D',
                             '#E6AB02', '#7570B3', '#BEAED4', '#666666', '#999999',
                             '#AA8282', '#D4B7B7', '#8600BF', '#BA5CE3', '#808000',
                             '#AEAE5C', '#1E90FF', '#00BFFF', '#56FF0D', '#FFFF00', 'red', 'black', 'gray')
plot(dif@eigenvectors[,1], dif@eigenvectors[,2], col = color_clusters[Idents(my)[ind]], cex = 0.7, pch = 16, xlab = 'Diffusion-1', ylab = 'Diffusion-2')
legend('bottomright', legend = unique(Idents(my)[ind]), pch = 15, col = color_clusters[unique(Idents(my)[ind])], ncol = 2)
plot(dif@eigenvectors[,2], dif@eigenvectors[,3], col = color_clusters[Idents(my)[ind]], cex = 0.7, pch = 16, xlab = 'Diffusion-2', ylab = 'Diffusion-3')
legend('bottomright', legend = unique(Idents(my)[ind]), pch = 15, col = color_clusters[unique(Idents(my)[ind])], ncol = 2)
```

### •• Monocle3

```{r}
library(monocle3)
library(Seurat)
library(SeuratData)
library(SeuratWrappers)
library(ggplot2)
library(patchwork)
library(magrittr)
cds <- as.cell_data_set(my)
cds <- cluster_cells(cds)
p1 <- plot_cells(cds, show_trajectory_graph = FALSE)
p2 <- plot_cells(cds, color_cells_by = "partition", show_trajectory_graph = FALSE)
wrap_plots(p1, p2)

integrated.sub <- subset(as.Seurat(cds), monocle3_partitions == 1)
cds <- as.cell_data_set(integrated.sub)
cds <- learn_graph(cds)
plot_cells(cds, label_groups_by_cluster = FALSE, label_leaves = FALSE, label_branch_points = FALSE)
```

### •• Transferring annotations to global [DONE]

```{r}
my$global.cluster2 <- as.character(my$RNA_snn_res.0.5)
my$global.cluster2[my$global.cluster2 == '0'] <- 'Mac_PLTP'
my$global.cluster2[my$global.cluster2 == '1'] <- 'Mac_CCL4'
my$global.cluster2[my$global.cluster2 == '2'] <- 'Mac_FABP5'
my$global.cluster2[my$global.cluster2 == '3'] <- 'Mac_MKI67'
my$global.cluster2[my$global.cluster2 == '4'] <- 'Monocytes'
my$global.cluster2[my$global.cluster2 == '5'] <- 'Neutrophils'
my$global.cluster2[my$global.cluster2 == '6'] <- 'Mac_IL1B'
my$global.cluster2[my$global.cluster2 == '7'] <- 'mregDC_LAMP3'
my$global.cluster2[my$global.cluster2 == '8'] <- 'T_cell_contam'
my$global.cluster2[my$global.cluster2 == '9'] <- 'cDC2_CD1C'
my$global.cluster2[my$global.cluster2 == '10'] <- 'Mac_FABP5'
my$global.cluster2[my$global.cluster2 == '11'] <- 'cDC1_CLEC9A'
my$global.cluster2[my$global.cluster2 == '12'] <- 'Mast cells'

# moving myeloid subsets to the global object
ind <- match(rownames(my@meta.data), rownames(seu@meta.data))
seu$global.cluster2[ind] <- my$global.cluster2
```

### •• Figure Plotting

```{r}
DimPlot(my[,!my$global.cluster2 == 'T_cell_contam'], group.by = 'global.cluster2', cols = color_clusters)
ggsave(file = './figures/myeloid_global.cluster2_UMAP.png', units = 'in', dpi = 300)

freq_boxplot(my[,!my$global.cluster2 == 'T_cell_contam'], metadata = 'global.cluster2', split.meta = 'hpv_status', out = './figures/myeloid_global.cluster2_')

my_mini <- my[,!my$global.cluster2 == 'T_cell_contam']

freq_plot(my_mini[,!my_mini$orig.ident == 'Vulva_#4_AdjNorm'], metadata = 'global.cluster2', split.meta = 'orig.ident', graph_type = 'stacked_barplot', out = './figures/my_global.cluster2_', out_width = 6, out_height = 8)

# M1/M2 feature set in myeloid object
my[['M1_signature_pct']] <- PercentageFeatureSet(my, features = c( 'CD80',   'CXCL9', 'CXCL10', 'CXCL11', 'CD86', 'IL1A', 'IL1B', 'IL6', 'TNF', 'CCL5', 'IRF5', 'IRF1', 'CD40', 'IDO1', 'KYNU', 'CCR7'))

my[['M2_signature_pct']] <- PercentageFeatureSet(my, features = c('ARG2', 'IL10',  'CD163', 'FCER2', 'CD200R1', 'PDCD1LG2', 'CD274', 'MARCO', 'CSF1R', 'MRC1', 'IL1RN', 'IL1R2', 'IL4R', 'CCL4', 'CCL13', 'CCL20', 'CCL17', 'CCL18', 'CCL22', 'CCL24', 'LYVE1', 'VEGFA', 'VEGFB', 'VEGFC',  'EGF', 'CTSA', 'CTSB', 'CTSC', 'CTSD', 'TGFB1', 'TGFB2', 'TGFB3', 'MMP14', 'MMP19', 'MMP9', 'CLEC7A', 'WNT7B',  'TNFSF12', 'TNFSF8', 'CD276',  'MSR1', 'FN1', 'IRF4'))

FeaturePlot(my[,!my$global.cluster2 == 'T_cell_contam'], features = c('M1_signature_pct', 'M2_signature_pct'), ncol = 1, cols = c('light grey', 'red'), max.cutoff = 6)
```

\

## • Fibroblasts only

### •• Load in/save objects

```{r}
saveRDS(fbs, './seurat_objs/vulvar_fibroblasts.RDS')
fbs <- readRDS('./seurat_objs/vulvar_fibroblasts.RDS')
```

```{r}
fbs <- seu[,grep('Fibroblasts', seu$global.cluster, value = F)]
DimPlot(fbs, group.by = 'global.cluster')
DimPlot(fbs, group.by = 'orig.ident')

fbs <- seu_pp1(fbs, harmony = T, npcs = 100) # 20 PCS
fbs <- seu_pp2(fbs, npcs = 20, reduction = 'harmony')
DimPlot(fbs, group.by = 'RNA_snn_res.0.6')

fbs_mini <- fbs[grep('KRT', rownames(fbs), invert = T)]
fbs_marks <- de_dotplot(fbs_mini, metadata = 'RNA_snn_res.0.5', n_marks = 10)

fbs_micro <- fbs_mini[,fbs_mini$RNA_snn_res.0.5 %in% c(0, 1, 2)]
fbs_marks_012 <- de_dotplot(fbs_micro, metadata = 'RNA_snn_res.0.5', n_marks = 10)

Idents(my) <- my$RNA_snn_res.0.1
```

![](images/paste-CF3B5764.png)

```{r}

# Buechler et al fibroblast markers (use this one)
VlnPlot(fbs, group.by = 'RNA_snn_res.0.5', features = c('PI16', 'CD34', 'HAS1', 'PLIN2', 'NPNT', 'CES1', 'ADH1B', 'FGFR4', 'LRRC15', 'COL11A1', 'CTHRC1', 'ADAMDEC1', 'CXCL14', 'CCL19', 'GREM1', 'TNFSF13B', 'COL3A1', 'POSTN', 'CTHRC1'), pt.size = 0)

# kurten fibroblast markers for CAF and elastic fibroblasts
VlnPlot(fbs, group.by = 'RNA_snn_res.0.5', features = c('SLC16A3', 'WNT5A', 'INHBA', 'COL5A2', 'LUM', 'FTH1', 'COL6A1', 'PHLDA1', 'HLA-C', 'TPM2', 'LGALS1', 'SERPINH1', 'MMP1', 'LOXL2', 'GPM6B', 'POSTN', 'CTSK', 'HLA-A', 'GREM1', 'TMEM45A', 'HLA-B', 'CEBPB', 'EVA1A', 'CLEC11A', 'IL24', 'PLA2G2A', 'PLAC9', 'GSN', 'CFD', 'MFAP5', 'TNXB', 'CCDC80', 'CST3', 'FGF7', 'IGFBP6', 'EFEMP1', 'SOD3', 'PI16', 'SLPI', 'MGP', 'SELENOP', 'C3', 'CXCL14', 'RARRES1', 'PTGIS', 'FBLN1', 'MFAP4', 'PRELP', 'LEPR', 'LTBP4'), ncol = 10, pt.size = 0)

# Deng et al fibroblast markers (mcaf and icaf)
VlnPlot(fbs, group.by = 'RNA_snn_res.0.5', features = c('PTGDS', 'LUM', 'CFD', 'FBLN1', 'APOD', 'DCN', 'CXCL14', 'SFRP2', 'MMP2', 'PLA2G2A', 'RGS5', 'NDUFA4L2', 'ADIRF', 'CRIP1', 'PPP1R14A', 'MYL9', 'MYH11'), ncol = 6, pt.size = 0)

# Galbo et al fibroblast markers (myCAF, dCAF, iCAF, nCAF, pCAF)
VlnPlot(fbs, group.by = 'RNA_snn_res.0.5', features = c('PARM1', 'CSPG4', 'CD36', 'SUSD2', 'SDC1', 'TREM1', 'THY1', 'BAMBI', 'P2RY62', 'CD34', 'PI16', 'CGC3', 'RAMP2', 'LRRN3', 'CLDN1', 'ICAM1', 'MUSK', 'CXCL4', 'NCAM1', 'EFNB1', 'EBP'), ncol = 6, pt.size = 0)
```

### •• Transfer annotations back to global clusters

```{r}
fbs$global.cluster2 <- as.character(fbs$RNA_snn_res.0.5)
fbs$global.cluster2[fbs$global.cluster2 == 'Fib_1'] <- 'myCAF_1'
fbs$global.cluster2[fbs$global.cluster2 == 'myCAF'] <- 'myCAF_2'
fbs$global.cluster2[fbs$global.cluster2 == 'Fib_2'] <- 'myCAF_3'
fbs$global.cluster2[fbs$global.cluster2 == '3'] <- 'iCAF'
fbs$global.cluster2[fbs$global.cluster2 == '4'] <- 'pericytes/vCAF'
fbs$global.cluster2[fbs$global.cluster2 == '5'] <- 'apCAF'

ind <- match(rownames(fbs@meta.data), rownames(seu@meta.data))
seu$global.cluster2[ind] <- fbs$global.cluster2
```

### •• Figure Plotting

```{r}
DimPlot(fbs, group.by = 'global.cluster2', cols = color_clusters)
ggsave(file = './figures/fibroblast_global.cluster2_UMAP.png', units = 'in', dpi = 300, width = 5, height = 4)

#freq_boxplot(fbs, metadata = 'global.cluster2', split.meta = 'hpv_status', out = './figures/fibroblast_global.cluster2_', out_width = 6, out_height = 8)

freq_plot(fbs[,!fbs$orig.ident == 'Vulva_#4_AdjNorm'], metadata = 'global.cluster2', split.meta = 'orig.ident', graph_type = 'stacked_barplot', out = './figures/fbs_global.cluster2_', out_width = 4, out_height = 6)

FeaturePlot(fbs, features = c('ACTA2', 'IL6', 'CD74', 'TGFB1', 'RGS5', 'COL1A1'), ncol = 2, cols = c("grey", "red"))
```

## • T cells only

```{r}
saveRDS(t, file = './seurat_objs/vulvar_TNK.RDS')
tnk <- readRDS(file = './seurat_objs/vulvar_TNK.RDS')
```

### •• Identifying T/NK clusters

```{r}
t <- seu[,grep('T/NK', seu$global.cluster, value = F)]
DimPlot(t, group.by = 'global.cluster')
DimPlot(t, group.by = 'orig.ident')

t <- seu_pp1(t, harmony = T, npcs = 100) # 25 PCS
t <- seu_pp2(t, npcs = 25, reduction = 'harmony')
DimPlot(t, group.by = 'RNA_snn_res.0.6')
table(t$RNA_snn_res.0.6)

de_dotplot(t, metadata = 'RNA_snn_res.0.6', n_marks = 10)

VlnPlot(t, features = c('NCR1', 'CD4', 'CD8A', 'CD8B1', 'CD3E', 'FOXP3', 'TNFSF8', 'IFITM1', 'IFIT1', 'IFIT2', 'IFIT3', 'CTLA4', 'HVARC2', 'PDCD1', 'TIGIT', 'IL2RA', 'TOX', 'TCF7', 'S1PR1', 'CCL5', 'IFNG', 'CCR7', 'IL7R', 'GZMK', 'GZMB', 'GZMH', 'GZMA', 'IL1R1', 'FN1'), group.by = 'RNA_snn_res.0.6')

t$global.cluster2 <- as.character(t$RNA_snn_res.0.6)
t$global.cluster2[t$global.cluster2 == '0'] <- 'Treg_IL1R1+FN1lo'
t$global.cluster2[t$global.cluster2 == '1'] <- 'Naive_CD8'
t$global.cluster2[t$global.cluster2 == '2'] <- 'Transitional_CD8'
t$global.cluster2[t$global.cluster2 == '3'] <- 'Treg_IL1R1+FN1hi'
t$global.cluster2[t$global.cluster2 == '4'] <- 'Exhausted_CD8'
t$global.cluster2[t$global.cluster2 == '5'] <- 'NK cells'
t$global.cluster2[t$global.cluster2 == '6'] <- 'Treg_IL1R1-'
t$global.cluster2[t$global.cluster2 == '7'] <- 'Dysfunctional_CD8'
Idents(t) <- t$global.cluster2
DimPlot(t)
```

### •• Transferring annotations to global [DONE]

```{r}
ind <- match(rownames(tnk@meta.data), rownames(seu@meta.data))
seu$global.cluster2[ind] <- tnk$global.cluster2
```

### •• Figure Plotting

```{r}
DimPlot(tnk, group.by = 'global.cluster2', cols = color_clusters)
ggsave(file = './figures/tnk_global.cluster2_UMAP.png', units = 'in', dpi = 300, width = 5, height = 4)

freq_plot(tnk[,!tnk$orig.ident == 'Vulva_#4_AdjNorm'], metadata = 'global.cluster2', split.meta = 'orig.ident', graph_type = 'stacked_barplot', out = './figures/tnk_global.cluster2_', out_width = 6, out_height = 8)

FeaturePlot(tnk, features = c('FOXP3', 'CTLA4', 'LAG3', 'GZMK'), ncol = 1)
```

# BLAST

## Extract 1,000,000 lines as a test

```{r}
# Set the path to the FASTQ file
fastq_file <- '/Volumes/hqdinh2/Projects/Anogenital_p30/Vulva_fastqs_from_BAM/V4T/bamtofastq_S1_L001_R1_001.fastq.gz'

# Open the file connection
con <- file(fastq_file, "r")

# Read the first 1000000 lines of the file
first_1000000_lines <- readLines(con, n = 1000000)

# Close the file connection
close(con)

writeLines(first_1000000_lines, "/Volumes/hqdinh2/Projects/Anogenital_p30/Vulva_fastqs_from_BAM/MINI_bamtofastq_S1_L001_R1_001.fastq.gz")
```

# [HUMAN + VIRUS GENOME] Vulvar scRNA-seq

## • Remove FASTA portion from gff3

```{r}
# set the working directory
setwd("/Users/agolfinos/AGAT/gffs/")

# get a list of all files in the directory
file_list <- list.files()

# loop through each file in the list
for (file_name in file_list) {
  # read in the file
  gff_file <- readLines(file_name)
  
  # find the index of the line starting with ">"; this marks the start of the FASTA portion
  fasta_start <- grep("^>", gff_file)
  
  # remove the FASTA portion from the end of the file
  gff_file <- gff_file[1:(fasta_start-2)]
  
  # create the output file name
  output_file <- paste0("new_", file_name)
  
  # write out the modified file
  writeLines(gff_file, output_file)
}

```

## • Create modified HPV (according to Fan et al)

```{r}
hpv.gtf <- read.table('./CellRanger_references/HPV/HPV.gtf', sep = "\t", header = FALSE, stringsAsFactors = FALSE)

hpv.gtf$V1 <- paste(hpv.gtf$V1, 'REF|lcl|Human', sep= '')

write.table(hpv.gtf, './CellRanger_references/HPV/HPV.gtf', sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
```

## • Split FASTQ into R1 and R2

```{r}
# Load the necessary libraries
library(stringr)

# Set the file path and file name
filepath <- '/Users/agolfinos/txg-macos-v1.3.1/Vulvar_scRNA-seq/per_sample_outs/V4T/count/'
filename <- "V4T.fastq"

# Read in the fastq file
fastq <- readLines(paste0(filepath, filename))

# Initialize the R1 and R2 fastq files
R1_fastq <- list()
R2_fastq <- list()

# Loop through the fastq file
for (i in seq(1, length(fastq), by=4)) {
  header <- fastq[i]
  seq <- fastq[i+1]
  qual_header <- fastq[i+2]
  qual <- fastq[i+3]
  
  # Get the read identifier
  read_id <- str_extract(header, "^[^ ]+")
  
  # Check if the read identifier ends in "/1" or "/2"
  if (str_detect(read_id, "/1$")) {
    R1_fastq <- append(R1_fastq, list(header, seq, qual_header, qual))
  } else if (str_detect(read_id, "/2$")) {
    R2_fastq <- append(R2_fastq, list(header, seq, qual_header, qual))
  }
}

# Write the R1 and R2 fastq files
writeLines(R1_fastq, paste0(filepath, "R1_", filename))
writeLines(R2_fastq, paste0(filepath, "R2_", filename))
```

## • Make EBV files

### •• GTF file

```{r}
ebv1 <- read.table('./CellRanger_references/EBV/DQ279927.1.gff3', header = FALSE, stringsAsFactors = FALSE, sep = "\t", comment.char = "#", quote = "",)[1,]

ebv2 <- read.table('./CellRanger_references/EBV/NC_007605.1.gff3',header = FALSE, stringsAsFactors = FALSE, sep = "\t", comment.char = "#", quote = "")[1,]

template <- read.table('./CellRanger_references/Fanetal_HPV_example/HPV.gtf', header = FALSE, stringsAsFactors = FALSE, sep = "\t", comment.char = "#", quote = "")[1:3,]

t1 <-  data.frame(lapply(template, function(x) {gsub("HPV16", ebv1$V1, x)}))
t1$V2 <- ebv1$V2
t1$V5 <- ebv1$V5

t2 <- data.frame(lapply(template, function(x) {gsub("HPV16", ebv2$V1, x)}))
t2$V2 <- ebv2$V2
t2$V5 <- ebv2$V5

ebv_all <- rbind(t1, t2)

write.table(ebv_all, file = "./CellRanger_references/EBV/EBV.gtf", quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)
```

### •• Concatenated FASTA file

```{r}
library(seqinr)
library(Biostrings)

# Define a list of FASTA files to concatenate
fasta_files <- c("./CellRanger_references/EBV/DQ279927.1.fasta", "./CellRanger_references/EBV/NC_007605.1.fasta")

# Create an empty list to hold the sequences and names
seqs <- list()
names <- character()

# Loop over the input files and read in the sequences
for (i in seq_along(fasta_files)) {
  file <- fasta_files[i]
  seqs_i <- readDNAStringSet(file = file)
  names_i <- paste(basename(file), names(seqs_i), sep = "_")
  seqs[[i]] <- seqs_i
  names <- c(names, names_i)
}

# Combine the sequences and names into a single DNAStringSet object
seqs <- do.call(c, seqs)
names(seqs) <- names

# Write the concatenated sequences to a new FASTA file
writeXStringSet(seqs, file = "./CellRanger_references/EBV/EBV.fa", format = "fasta")
```

# nCounter (48 vulvar patients)

## • Wilcox test for demographic information

```{r}
ncounter <- read.csv('/Volumes/hqdinh2/Projects/Collaborations/MeganFitzpatrick/Annotation_files/Vulva_annotation_04.12.22.csv')

table(ncounter$p16)
# N  P 
#29 19 

# average overall survival for p16- patients
mean(ncounter$Overall.survival..dg.death...months[ncounter$p16 == 'N'], na.rm = T)
#[1] 78.77423

# average overall survival for p16+ patients
mean(ncounter$Overall.survival..dg.death...months[ncounter$p16 == 'P'], na.rm = T)
#[1] 73.48615

wc0 <- wilcox.test(ncounter$Overall.survival..dg.death...months[ncounter$p16 == 'N'], ncounter$Overall.survival..dg.death...months[ncounter$p16 == 'P'], alternative = 'two.sided', paired = F) # average overall survival is not significant between HPV+ and HPV-



# average survival after surgery for p16- patients
mean(ncounter$Survival.after.surgery[ncounter$p16 == 'N'], na.rm = T)
#[1] 54

# average survival after surgery for p16+ patients
mean(ncounter$Survival.after.surgery[ncounter$p16 == 'P'], na.rm = T)
#[1] 73.01158

wc1 <- wilcox.test(ncounter$Survival.after.surgery[ncounter$p16 == 'N'], ncounter$Survival.after.surgery[ncounter$p16 == 'P'], alternative = 'two.sided', paired = F) # average survival after surgery is not significant between HPV+ and HPV-



# average time to recurrence for p16- patients
mean(ncounter$Time.to.recurrence[ncounter$p16 == 'N'], na.rm = T)
# [1] 63.42438

# average time to recurrence for p16+ patients
mean(ncounter$Time.to.recurrence[ncounter$p16 == 'P'], na.rm = T)
# [1] 44.79556

wc2 <- wilcox.test(ncounter$Time.to.recurrence[ncounter$p16 == 'N'], ncounter$Time.to.recurrence[ncounter$p16 == 'P'], alternative = 'two.sided', paired = F) # average time to recurrence is not significant between HPV+ and HPV-
```

## • Wilcox test for demographic information between heatmap clusters

```{r}
n3_wc <- n3[n3$cluster.x == 'CD8.T.cells',]

wc_hmcluster <- pairwise.wilcox.test(n3_wc$value, n3_wc$cluster.y, p.adjust.method = 'BH')

wc_hmcluster$p.value
#          cluster 1 cluster 2 cluster 3
#cluster 2 0.7382931        NA        NA
#cluster 3 0.7382931 0.7382931        NA
#cluster 4 0.7382931 0.7382931 0.7382931

wc_hmcluster_noadj <- pairwise.wilcox.test(n3_wc$value, n3_wc$cluster.y, p.adjust.method = 'none')

wc_hmcluster_noadj$p.value
#          cluster 1 cluster 2 cluster 3
#cluster 2 0.7294843        NA        NA
#cluster 3 0.5277227 0.3102504        NA
#cluster 4 0.3923715 0.7382931 0.1991287
```
