------------------------------------------------------------------------

------------------------------------------------------------------------

title: "HNC_Myeloid_figs" author: "Athena Golfinos" date: "2/21/2022" output: html_document ---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(tidyverse)
library(gridExtra)
library(Seurat)
library(scales)
library(ggdendro)
library(dendextend)
library(DescTools)
library(splitstackshape)
library(nichenetr)
library(tidyr)
library(stringr)
library(circlize)
library(magrittr)
library(data.table)
library(dplyr)

color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3",
"#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", "#AA8282", "#D4B7B7", "#8600BF", "#BA5CE3", "#808000","#AEAE5C", "#1E90FF", "#00BFFF", "#56FF0D", "#FFFF00")

#seurat objects
#hnc <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_all_annot_2023-02-03.RDS')

#load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/Seurat_MOC2_K17KO_2021-11-01**.rda') #object called "seurat"
```

# [FIGURE 1] Basic UMAPs and dotplots

```{r figure 1}
color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3","#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", "#AA8282", "#D4B7B7", "#8600BF", "#BA5CE3", "#808000","#AEAE5C", "#1E90FF", "#00BFFF", "#56FF0D", "#FFFF00")


# start mouse figs here

load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/Seurat_MOC2_K17KO_MonMacDC_2021-09-02.rda')

hnc2$tissue <- as.character(hnc2$tissue)
hnc2$tissue[hnc2$tissue == 'EX'] <- 'K17KOMOC2'
hnc2$tissue[hnc2$tissue == 'MO'] <- 'MOC2'
hnc2$tissue <- as.factor(hnc2$tissue)

DimPlot(hnc2, group.by = 'AEG_res2_annotations', split.by = 'tissue', cols = color_clusters)
ggsave('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/UMAP_MOC2_split.png', height = 5, width = 8, units = 'in', dpi = 300)

# reorder clusters
hnc2$AEG_res2_annotations <- factor(hnc2$AEG_res2_annotations, 
                            levels=c("DC_Ccr7", "DC_cDC1",
                                     "pDC", "DC_moDC_cDC2", 
                                     "Mon_Ccr5", "Mon_Intermediate",
                                     "Mon_Classical", "Mon_NonClassical",
                                     "Mac_Trem2", "Mac_Fn1",
                                     "Mac_Cxcl9", "Mac_Lyve1"))

de_dotplot(hnc2, metadata = 'AEG_res2_annotations', n_marks = 5)
ggsave('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Dotplot_MOC2.png', height = 10, width = 7, units = 'in', dpi = 300)

freq_boxplot(SEU = hnc2, metadata = 'AEG_res2_annotations', split.meta = 'tissue', out = '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/moc2_freq_boxplot.png', celltype = unique(hnc2$AEG_res2_annotations), out_width = 13)


# start human figs here

load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_myeloid_2021.rda')

hnc2$global.cluster[hnc2$global.cluster == 'pDC'] <- 'DCs'
hnc2$global.cluster[hnc2$global.cluster == 'Immature_Myeloids'] <- 'CD14+ Monocytes'
hnc2$global.cluster[hnc2$global.cluster == 'CD16_Monocytes'] <- 'CD16+ Monocytes'
hnc2$global.cluster4[hnc2$global.cluster4 == 'DC3_LAMP3'] <- 'mregDC_LAMP3'

hnc2$global.cluster4 <- factor(hnc2$global.cluster4, 
                               levels = c('cDC1_CLEC9A', "cDC2_CD1C", "cDC2_CD33", "mregDC_LAMP3", "DC_pDC", 
                                          "Mono_CD14", "Mono_TIL", "Mono_Int", "Mono_CD16", 
                                          "Mono_CD14_IL1B", "Mono_CD14_THBS1", "Mono_CD14_ID1",
                                          "Mac_SPP1", "Mac_C1QA", "Mac_IL1B", "Mac_SPP1_PLTP", 'Mast'))

DimPlot(hnc2, group.by = 'global.cluster4', cols = color_clusters, split.by = 'tissue') + ggtitle('')
ggsave('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/UMAP_MyeloidDimPlot.png', width = 8, height = 4)

de_dotplot(hnc2, metadata = 'global.cluster4')
ggsave('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/dotplot_cillo.png', width = 10, height = 14)

freq_boxplot(SEU = hnc2, metadata = 'global.cluster4', split.meta = 'tissue', out = '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/cillo_freq_boxplot.png', celltype = unique(hnc2$global.cluster4), out_width = 19, out_height = 10)

luoma <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/luoma_tumor_myeloid_plusmetadata.RDS')

DimPlot(luoma, group.by = 'predicted.id')
ggsave('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/UMAP_Luoma_tumor.png', height = 5, width = 8, units = 'in', dpi = 300)

```

# [FIGURE 2] Mouse to human comparison

## • a) UMAP of transferred subsets

```{r}
color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3","#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", "#AA8282", "#D4B7B7", "#8600BF", "#BA5CE3", "#808000","#AEAE5C", "#1E90FF", "#00BFFF", "#56FF0D", "#FFFF00")

query <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/Seurat/Without_cDC2_CD33/With_SPP1_PLTP_Macrophages/01182023_seurat_label_transfer_human_to_mouse_merged_subsets.RDS')

query$tissue[query$tissue == 'MO'] <- 'MOC2'
query$tissue[query$tissue == 'EX'] <- 'K17KOMOC2'

query <- query[,!query$predicted.id == 'Mast']

table(query$predicted.id)
#  cDC1_CLEC9A     cDC2_CD1C        DC_pDC      Mac_C1QA      Mac_IL1B      Mac_SPP1 
#          110          1114            32          1908           586          1237 
# Mac_SPP1_PLTP     Mono_CD14      Mono_TIL  mregDC_LAMP3 
#          718           398           385           248 

DimPlot(query, group.by = 'predicted.id', cols = color_clusters) + ggtitle('')
ggsave(paste('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/Seurat/Without_cDC2_CD33/With_SPP1_PLTP_Macrophages/', Sys.Date(), '_seurat_label_transfer_human_to_mouse_merged_subsets.png', sep = ''), width = 6, height =5, units = 'in', dpi = 300)

DimPlot(query, group.by = 'predicted.id', cols = color_clusters, split.by = 'tissue') + ggtitle('')
ggsave(paste('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/Seurat/Without_cDC2_CD33/With_SPP1_PLTP_Macrophages/', Sys.Date(), '_seurat_label_transfer_human_to_mouse_merged_subsets_split.png', sep = ''), width = 9, height =5, units = 'in', dpi = 300)
```

## • b) Heatmap of cluster distribution human \> mouse

```{r}
transferred <- readRDS(file = '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/Seurat/Without_cDC2_CD33/With_SPP1_PLTP_Macrophages/01182023_seurat_label_transfer_human_to_mouse_merged_subsets.RDS')

out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/Seurat/Without_cDC2_CD33/With_SPP1_PLTP_Macrophages/'

# __________________________________________________________

subsets <- unique(transferred[['predicted.id']][,1])

master <- data.frame(AEG_res2_annotations = character(), 
                     Freq = numeric(), 
                     Pct = numeric(), 
                     predicted_id = character())
for (x in subsets){
  # make plots of the distribution of each subset
  meta <- transferred[['predicted.id']]
  m2 <- setNames(as.character(meta[,1]), rownames(meta))
  trans_mini <- transferred[,m2 == x]
  #p2 <- DimPlot(trans_mini, group.by = 'predicted.id')
  #p1 <- DimPlot(trans_mini, group.by = 'AEG_res2_annotations')
  #cowplot::plot_grid(p1, p2)
  #ggsave(paste(out, x, 'cluster_comparison.png'), width = 12, height = 5)
  
  # calculating the percentage distribution
  tbl <- table(trans_mini[['AEG_res2_annotations']])
  tbl <- as.data.frame(tbl)
  sm <- as.numeric(sum(tbl$Freq))
  tbl$Pct <- tbl$Freq/sm
  tbl$predicted_id <- x
  master <- rbind(master, tbl)
}
master_mini <- master[,c('AEG_res2_annotations', 'Pct', 'predicted_id')]
master_wide <- pivot_wider(master_mini, names_from = predicted_id, values_from = 'Pct')
rnms <- master_wide[['AEG_res2_annotations']]
master_wide[['AEG_res2_annotations']] <- NULL
master_wide[is.na(master_wide)] <- 0
rownames(master_wide) <- rnms

r_order <- match(c("DC_cDC1", "DC_moDC_cDC2", "DC_Ccr7", "pDC", "Mon_Classical", "Mon_Intermediate", "Mon_NonClassical", "Mon_Ccr5", "Mac_Cxcl9", "Mac_Fn1", "Mac_Lyve1", "Mac_Trem2"), rownames(master_wide))
c_order <- match(c("cDC1_CLEC9A", "cDC2_CD1C", "mregDC_LAMP3", "DC_pDC", "Mono_CD14", "Mono_TIL", "Mac_C1QA", "Mac_SPP1", "Mac_SPP1_PLTP", "Mac_IL1B"), colnames(master_wide))

master_wide <- master_wide[r_order, c_order]
rownames(master_wide) <- c("DC_cDC1", "DC_moDC_cDC2", "DC_Ccr7", "pDC", "Mon_Classical", "Mon_Intermediate", "Mon_NonClassical", "Mon_Ccr5", "Mac_Cxcl9", "Mac_Fn1", "Mac_Lyve1", "Mac_Trem2")

png(paste(out, 'Heatmap_Seurat_transfer_learning_mouse_to_human_comparison.png'), width = 7, height = 5, units = 'in', res = 300)

ComplexHeatmap::Heatmap(master_wide, column_title = 'Predicted Human Cluster ID', column_title_side = c('bottom'), row_title = 'Original Mouse Cluster ID', row_title_side = 'left', heatmap_legend_param = list(title = "Percent"), cluster_rows = F, cluster_columns = F, row_split = c('DC', 'DC', 'DC', 'DC', 'Mon', 'Mon', 'Mon', 'Mon', 'Mac', 'Mac', 'Mac', 'Mac'), column_split = c('DC', 'DC', 'DC', 'DC', 'Mon', 'Mon', 'Mac', 'Mac', 'Mac', 'Mac'))

dev.off()
```

### •• [ORIGINAL] Heatmap of log2 fc of each subset from human to mouse (ORIGINAL)

Using the proportions from the Cillo object based on the ORIGINAL version. Meaning, the percentages are based on including all subsets, even those that WERE NOT inferred in the mouse data

```{r}
transferred <- readRDS(file = '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/Seurat/Without_cDC2_CD33/With_SPP1_PLTP_Macrophages/01182023_seurat_label_transfer_human_to_mouse_merged_subsets.RDS')

out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/Seurat/Without_cDC2_CD33/With_SPP1_PLTP_Macrophages/'

load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_myeloid_2021.rda')

# __________________________________________________________

q_distribution <- as.data.frame(table(transferred$predicted.id))
q_distribution$pct <- q_distribution$Freq/sum(q_distribution$Freq)
r_distribution <- as.data.frame(table(hnc2$global.cluster4))
r_distribution$pct <- r_distribution$Freq/sum(r_distribution$Freq)

all <- merge(r_distribution, q_distribution, by = 'Var1')
all$log2fc <- log2(all$pct.x/all$pct.y)
all <- all[!all$Var1 == 'Mast',]

png(paste(out, 'log2fc_barchart_Seurat_transfer_learning_mouse_to_human_comparison.png'), width = 6, height = 4, units = 'in', res = 300)

ggplot(all, aes(x=Var1, y=log2fc)) + geom_bar(stat = "identity", aes(fill = log2fc)) + coord_flip() + theme_classic() + scale_fill_gradient2(low='blue', mid='snow', high='red') + xlab('Myeloid subset') + ylab('<---- Higher in mouse data ----                 ---- Higher in human data ---->')

dev.off()
```

### •• [MINI] Heatmap of log2fc of each subset from mouse to human

Using proportions of Cillo object as if it only had the subsets we inferred in the mouse data (aka proportions from subsetting Cillo to only include the subsets we inferred in mouse data)

```{r}
transferred <- readRDS(file = '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/Seurat/Without_cDC2_CD33/With_SPP1_PLTP_Macrophages/01182023_seurat_label_transfer_human_to_mouse_merged_subsets.RDS')

out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/Seurat/Without_cDC2_CD33/With_SPP1_PLTP_Macrophages/'

load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_myeloid_2021.rda')

# __________________________________________________________

q_distribution <- as.data.frame(table(transferred$predicted.id))
q_distribution$pct <- q_distribution$Freq/sum(q_distribution$Freq)
r_distribution <- as.data.frame(table(hnc2$global.cluster4))
r_distribution <- r_distribution[r_distribution$Var1 %in% q_distribution$Var1,]
r_distribution$pct <- r_distribution$Freq/sum(r_distribution$Freq)

all <- merge(r_distribution, q_distribution, by = 'Var1')
all$log2fc <- log2(all$pct.x/all$pct.y)
all <- all[!all$Var1 == 'Mast',]

png(paste(out, 'mini_log2fc_barchart_Seurat_transfer_learning_mouse_to_human_comparison.png'), width = 6, height = 4, units = 'in', res = 300)

ggplot(all, aes(x=Var1, y=log2fc)) + geom_bar(stat = "identity", aes(fill = log2fc)) + coord_flip() + theme_classic() + scale_fill_gradient2(low='blue', mid='snow', high='red') + xlab('Myeloid subset') + ylab('<---- Higher in mouse data ----                 ---- Higher in human data ---->')

dev.off()
```

## • c) UMAP of label transferred Luoma

```{r}
query <- readRDS(file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/UMAP_Luoma_labeltransfer_01182023.RDS')

color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3","#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", "#AA8282", "#D4B7B7", "#8600BF", "#BA5CE3", "#808000","#AEAE5C", "#1E90FF", "#00BFFF", "#56FF0D", "#FFFF00")

query <- RunUMAP(query, dims = 1:30)

query$Stage <- factor(query$Stage, levels=c('Pre-Tx', 'Post-Tx'))

query$predicted.id[query$predicted.id == 'Mono_CD14'] <- 'Blood_CD14_Mono'
DimPlot(query, group.by = 'predicted.id', cols = color_clusters, label = T, label.box = T, repel = T) + ggtitle('')
ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/merged_UMAP_Luoma_labeltransfer_01182023.png', width = 7, height = 5, units = 'in', dpi = 300)

DimPlot(query, group.by = 'predicted.id', cols = color_clusters, split.by = 'Stage')
ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/UMAP_Luoma_labeltransfer_01182023.png', width = 11, height =5, units = 'in', dpi = 300)
```

## • d) Luoma frequency boxplots

```{r}
luoma <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/UMAP_Luoma_labeltransfer_01172023.RDS')

SEU <- luoma[,luoma$predicted.id %like any% c('%Mac%', '%Mon%', '%DC%')]
SEU$orig.ident <- paste(SEU$Patient_ID, SEU$Stage)
SEU$meta <- paste(SEU$Patient_ID, SEU$Stage)
SEU$Path_response[SEU$Path_response %in% c('Medium', 'High')] <- 'Medium/High'
SEU$treatment_path_response <- paste(SEU$Stage, SEU$Path_response)
metadata <- 'predicted.id'

get_freqs <- function(immune.combined, cluster_res) {
  Idents(immune.combined) <- cluster_res
  new.ident <- sort(unique(Idents(immune.combined)))
  samples <- unique(immune.combined@meta.data$orig.ident)
  tmp <- match(immune.combined$orig.ident, samples)
  sample_ind <- unique(tmp)
  ids <- Idents(immune.combined)
  tmp_v <- matrix(0, nrow = length(samples), ncol = length(new.ident))
  rownames(tmp_v) <- samples
  total_in_sample <- rep(0, length(samples))
  tmp <- immune.combined$orig.ident
  tmp <- plyr::count(tmp)
  total_in_sample = tmp$freq
  names(total_in_sample) <- tmp$x
  total_in_sample <- total_in_sample[match(samples, names(total_in_sample))]
  for (i in 1:length(new.ident)) {
    tmp <- ids[which(ids == new.ident[i])]
    tmp <- immune.combined$orig.ident[match(names(tmp), rownames(immune.combined@meta.data))]
    tmp <- plyr::count(tmp)
    for (j in 1:nrow(tmp)) {
      ind <- which(rownames(tmp_v) == tmp$x[j])
      tmp_v[ind,i] <- tmp$freq[j]
    }
  }
  colnames(tmp_v) <- new.ident
  ind <- order(colnames(tmp_v))
  tmp_v
}

t <- get_freqs(SEU, metadata)
t2 <- t
for (i in 1:nrow(t2)) {
  t2[i,] = t2[i,]/sum(t2[i,])
}
t2 <- as.matrix(t2)

meta <- SEU@meta.data[,c('meta', 'treatment_path_response')]
rownames(meta) <- NULL
meta <- unique(meta)
df <- NULL
for (i in 1:nrow(t2)) {
  for (j in 1:ncol(t2)) {
    df <- rbind(df, c(rownames(t2)[i], meta$treatment_path_response[which(meta$meta == rownames(t2)[i])], colnames(t2)[j], as.numeric(t2[i,j])))
  }
}
colnames(df) <- c('Sample', 'Group', 'Cluster', 'Freqs')
df <- as.data.frame(df)
df$paired <- sapply(strsplit(df$Sample," "), `[`, 1)

df$Group <- as.character(df$Group)
df$Freqs <- as.numeric(as.character(df$Freqs))
df <- df[df$Group %like any% c('%Low%', '%Medium%', '%High%'),]
df$Group <- factor(df$Group, levels = c('Pre-Tx Low', 'Post-Tx Low', 'Pre-Tx Medium/High', 'Post-Tx Medium/High'))

# with the paired lines
ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + geom_line(aes(x = Group, y = Freqs, group = paired), alpha = 0.3) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
#ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/01172023Luoma_PBMC_2responses_boxplot_myeloid_stage_pathresponse.png', width = 14, height = 8)

# without the paired lines
ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
#ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/01172023Luoma_PBMC_2responses_boxplot_myeloid_stage_pathresponse_NOpairlines.png', width = 14, height = 8)
```

### •• d) Wilcox test for frequency boxplot

```{r}
wilcox_by_condition(df, out = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/Boxplots/')
```

## • e) Scatterplots of freq comparison Luoma \> tranferred mouse

```{r}
transferred <- readRDS(file = '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/Seurat/Without_cDC2_CD33/With_SPP1_PLTP_Macrophages/01182023_seurat_label_transfer_human_to_mouse_merged_subsets.RDS')
out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/Seurat/Without_cDC2_CD33/With_SPP1_PLTP_Macrophages/'
query_meta <- 'predicted.id'
original_meta <- 'AEG_res2_annotations'

luoma <- readRDS( '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/UMAP_Luoma_labeltransfer_01172023.RDS')
luoma_meta <- 'predicted.id'

#transferred <- transferred[,transferred$tissue == 'MOC2']
table(transferred$tissue)

#luoma <- luoma[,luoma$Path_response %in% c('Low')]
table(luoma$Path_response)

#status <- 'Non-responders'
# __________________________________________________________

subsets <- unique(transferred[[query_meta]][,1])

# calculating the percentage distribution
tbl <- table(transferred$predicted.id, transferred$tissue)
tbl <- as.data.frame(tbl)
tbl <- tbl %>% mutate(Pct = case_when(
  Var2 == 'K17KOMOC2' ~ tbl$Freq/sum(tbl$Freq[tbl$Var2 == 'K17KOMOC2']), 
  Var2 == 'MOC2' ~ tbl$Freq/sum(tbl$Freq[tbl$Var2 == 'MOC2'])))

tbl_moc2 <- tbl[tbl$Var2 == 'MOC2',]
tbl_k17ko <- tbl[tbl$Var2 == 'K17KOMOC2',]
tbl <- merge(tbl_moc2, tbl_k17ko, by = 'Var1')

#getting percentage diff in R vs. NR in mice
tbl$delta <- tbl$Pct.y - tbl$Pct.x # more positive -> more enriched in K17KO

luoma$Path_response[luoma$Path_response %in% c('Medium', 'High')] <- 'Medium/High'

tbl1 <- table(luoma$predicted.id, luoma$Path_response)
tbl1 <- as.data.frame(tbl1)
tbl1 <- tbl1 %>% mutate(Pct = case_when(
  Var2 == c('Medium/High') ~ tbl1$Freq/sum(tbl1$Freq[tbl1$Var2 == c('Medium/High')]),   Var2 == 'Low' ~ tbl1$Freq/sum(tbl1$Freq[tbl1$Var2 == 'Low'])))

tbl1_r <- tbl1[tbl1$Var2 == 'Medium/High',]
tbl1_nr <- tbl1[tbl1$Var2 == 'Low',]
tbl1 <- merge(tbl1_r, tbl1_nr, by = 'Var1')

tbl1$delta <- tbl1$Pct.x - tbl1$Pct.y # more positive -> more enriched in responders

setdiff(tbl$Var1, tbl1$Var1) # in mouse and not in Luoma
# [1] "Mac_SPP1_PLTP" "Mast"
setdiff(tbl1$Var1, tbl$Var1) # in Luoma and not in mouse
# [1] "Mono_CD16" "Mono_Int"

tbl <- tbl[,c('Var1', 'delta')]
tbl1 <- tbl1[,c('Var1', 'delta')]

all <- merge(tbl, tbl1, by = 'Var1', all = TRUE)

all[is.na(all)] <- 0
colnames(all) <- c('cluster', 'k17ko_delta', 'R_delta')


ggplot(all, aes(x=k17ko_delta, y=R_delta)) + geom_point() + theme_classic() +
  xlab("Mouse Subset Proportion Delta [% in K17KOMOC2 - % in MOC2]") + ylab("Human Subset Proportion Delta [% in R - % in NR]") + 
  ggrepel::geom_text_repel(aes(x=k17ko_delta, y=R_delta), label = all$cluster) +
  geom_line(stat = 'smooth', method = 'lm', alpha = 0.3) + 
  ggtitle(paste('Subset Proportion Comparison', sep = ' '))

ggsave(file = paste(out, Sys.Date(), '_Human_mouse_subset_comparison.png', sep = ''), width = 5, height = 5, units = 'in', dpi = 300)
```

# [FIGURE 3] CXCL9 as a marker for ICI

## • a) CXCL9 and CD274 Cristescu comparison by response status

```{r}
litch <- readr::read_delim('/Volumes/hqdinh2/Projects/HNC_SPORE/AACR2022/Using_Litchfield_Data/meta_analysis_input_data.txt')
litch <- litch[litch$study == 'CRISTESCU_SCIENCE_2018',]
litch <- litch[litch$histology == 'HEAD AND NECK',]

litch_mini <- litch[,c('CXCL9', 'CD274', 'response')]

r <- litch_mini[litch_mini$response == 'response',]
nr <- litch_mini[litch_mini$response == 'no_response',]
  
wilcox <- wilcox.test(r$CXCL9, nr$CXCL9, alternative = c('two.sided'))

subtitle1 <- paste('p=', round(wilcox$p.value, digits = 5), sep = '')
outname <- paste('/Volumes/hqdinh2/Projects/HNC_SPORE/AACR2022/Using_Litchfield_Data/', as.character(Sys.Date()), 'CXCL9_CD274_RVsNRBoxplot.png', sep = '')

litch_mini$response <- factor(litch_mini$response, levels=c('response', 'no_response'))

p1 <- ggplot(litch_mini) +geom_boxplot(aes(x = response, y = CXCL9, color = response, fill = response), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = response, y = CXCL9, color = response), alpha = 0.8, position = position_jitterdodge()) + theme_bw() + theme(legend.title=element_text(size=14), legend.text=element_text(size=14), axis.text=element_text(size=14), axis.title=element_text(size=16), axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 18), legend.position = "none", plot.title = element_text(hjust = 0.5, size = 18), plot.subtitle = element_text(hjust = 0.5, size = 16)) + ggtitle('CXCL9 expression', subtitle = subtitle1)

wilcox <- wilcox.test(r$CD274, nr$CD274, alternative = c('two.sided'))
subtitle2 <- paste('p=', round(wilcox$p.value, digits = 3), sep = '')

p2 <- ggplot(litch_mini) +geom_boxplot(aes(x = response, y = CD274, color = response, fill = response), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = response, y = CD274, color = response), alpha = 0.8, position = position_jitterdodge()) + theme_bw() + theme(legend.title=element_text(size=14), legend.text=element_text(size=14), axis.text=element_text(size=14), axis.title=element_text(size=16), axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 18), plot.title = element_text(hjust = 0.5, size = 18), plot.subtitle = element_text(hjust = 0.5, size = 16)) + ggtitle('CD274 expression', subtitle = subtitle2)

png(filename = outname, units = 'in', res = 300, width = 8, height = 7) 
cowplot::plot_grid(p1, p2, rel_widths = c(1,1.6))
dev.off()
 
#ggsave(file = outname, width = 10, height = 15) 
```

## • b) Mouse CXCL9 VlnPlot

```{r}
load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/Seurat_MOC2_K17KO_2021-11-01**.rda')

VlnPlot(seurat, group.by = 'global.cluster', features = 'Cxcl9', split.by = 'tissue')
ggsave('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Violin_plots/VlnPlot_mouse_CXCL9.png', units = 'in', width = 7, height = 5, dpi = 300)
```

# Label transfer comparison

## • Prep seurat object (mouse with human genes)

```{r}
load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/Seurat_MOC2_K17KO_MonMacDC_2021-09-02.rda')
seurat <- hnc2
human <- read_csv('/Volumes/hqdinh2/Projects/HNC_SPORE/Resources/mart_export.csv')[,c(1,2)]

human <- unique(human)

mouse <- as.character(human$`Gene name`) 
hu <- as.character(human$`Human gene name`)

length(intersect(mouse, rownames(seurat))) #12284
length(setdiff(mouse, rownames(seurat))) #1858

# only keeping genes that were found in Biomart for both mouse and human
seurat_mini <- seurat[rownames(seurat) %in% intersect(mouse, rownames(seurat))]

# reorder the biomart genes to match the genes in the original dataset
order <- match(rownames(seurat_mini), mouse)
hu <- hu[order]

# now creating a new assay with the human names

# RenameGenesSeurat  -----------------------------------------------------------
RenameGenesSeurat <- function(obj = ls.Seurat[[i]], newnames = HGNC.updated[[i]]$Suggested.Symbol) { # Replace gene names in slots of a Seurat object. Run this before integration. Only changes obj@assays$RNA@counts, @data and @scale.data.
  print("Run this before integration. It only changes obj@assays$RNA@counts, @data and @scale.data.")
  RNA <- obj@assays$RNA

  if (nrow(RNA) == length(newnames)) {
    if (length(RNA@counts)) RNA@counts@Dimnames[[1]]            <- newnames
    if (length(RNA@data)) RNA@data@Dimnames[[1]]                <- newnames
    #if (length(RNA@scale.data)) RNA@scale.data@Dimnames[[1]]    <- newnames
  } else {"Unequal gene sets: nrow(RNA) != nrow(newnames)"}
  obj@assays$RNA <- RNA
  return(obj)
}
# RenameGenesSeurat(obj = SeuratObj, newnames = HGNC.updated.genes)

seurat_mini_test <- RenameGenesSeurat(obj = seurat_mini, newnames = hu)

# test whether it was successful

p1 <- VlnPlot(seurat_mini_test, features = 'CXCL9')
p2 <- VlnPlot(seurat, features = 'Cxcl9')
cowplot::plot_grid(p1, p2)

# prep and save this object
seu_mini <- seurat_mini_test
saveRDS(seu_mini, '/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/moc2_human_genes_01112022.RDS')
```

## • Label transfer

### •• [seurat] All subsets

```{r}
# load in target datasets
query <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/moc2_human_genes_01112022.RDS')

hnc <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_all_annot_2022-08-16.RDS')
ref <- hnc[,hnc$mreg_cxcl9_globalcluster4 %like any% c('%DC%', 'Mac%', 'Mon%')]
ref <- ref[,ref$tissue == c('TIL')]
ref$mreg_cxcl9_globalcluster4[ref$mreg_cxcl9_globalcluster4 == 'mregDC_CXCL9lo'] <- 'mregDC_LAMP3'
ref$mreg_cxcl9_globalcluster4[ref$mreg_cxcl9_globalcluster4 == 'mregDC_CXCL9hi'] <- 'mregDC_LAMP3'

table(ref$tissue)
table(ref$mreg_cxcl9_globalcluster4)

#query <- DietSeurat(query, counts = T, data = T, scale.data = T)
#VariableFeatures(query) <- NULL

# myeloid

query <- NormalizeData(query)
#query <- FindVariableFeatures(query, selection.method = "vst", nfeatures = 2000)
#query <- ScaleData(query, features = rownames(query))
query <- RunPCA(query, features = VariableFeatures(object = query), npcs = 100)
ElbowPlot(query, ndims = 100)

my.anchors <- FindTransferAnchors(reference = ref, query = query, dims = 1:30)
my.predictions <- TransferData(anchorset = my.anchors, refdata = ref$mreg_cxcl9_globalcluster4, dims = 1:30)
query <- AddMetaData(query, metadata = my.predictions)

color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3","#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", "#AA8282", "#D4B7B7", "#8600BF", "#BA5CE3", "#808000","#AEAE5C", "#1E90FF", "#00BFFF", "#56FF0D", "#FFFF00")

DimPlot(query, group.by = 'predicted.id', cols = color_clusters)
ggsave('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/seurat_label_transfer_human_to_mouse.png', width = 7, height =5, units = 'in', dpi = 300)

saveRDS(query, file = '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/seurat_label_transfer_human_to_mouse_all_subsets.RDS')

save(hnc, my.anchors, my.predictions, query, ref, color_clusters, file="/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/ENVIRONMENT_seurat_label_transfer_human_to_mouse_all_subsets.RData")
```

### •• [seurat] Partially merged subsets

```{r}
# load in target datasets
query <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/moc2_human_genes_01112022.RDS')

load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_myeloid_2021.rda')
ref <- hnc2[,hnc2$tissue == c('TIL')]
ref <- ref[,!ref$global.cluster4 == 'cDC2_CD33']
ref$global.cluster4[ref$global.cluster4 == 'DC3_LAMP3'] <- 'mregDC_LAMP3'
#ref$global.cluster4[ref$global.cluster4 == c('Mac_SPP1_PLTP')] <- 'Mac_SPP1'
ref$global.cluster4[ref$global.cluster4 == c('Mono_CD14_IL1B')] <- 'Mono_CD14'
ref$global.cluster4[ref$global.cluster4 == c('Mono_CD14_THBS1')] <- 'Mono_CD14'
ref$global.cluster4[ref$global.cluster4 == c('Mono_CD14_ID1')] <- 'Mono_CD14'

table(ref$tissue)
# TIL 
# 9526

table(ref$global.cluster4)
# cDC1_CLEC9A     cDC2_CD1C        DC_pDC      Mac_C1QA      Mac_IL1B 
#         251           939           847          1216           774 
#     Mac_SPP1 Mac_SPP1_PLTP          Mast     Mono_CD14     Mono_CD16 
#         1837           573           304          1115           160 
#     Mono_Int      Mono_TIL  mregDC_LAMP3 
#           87           961           462 

hnc <- NULL
seurat <- NULL
hnc2 <- NULL
m <- NULL

ref_mini <- ref[, sample(colnames(ref), size = 5000, replace=F)]
ref <- NULL

# myeloid

query <- NormalizeData(query)
#query <- FindVariableFeatures(query, selection.method = "vst", nfeatures = 2000)
#query <- ScaleData(query, features = rownames(query))
query <- RunPCA(query, features = VariableFeatures(object = query), npcs = 100)
ElbowPlot(query, ndims = 100)

my.anchors <- FindTransferAnchors(reference = ref_mini, query = query, dims = 1:30)
my.predictions <- TransferData(anchorset = my.anchors, refdata = ref_mini$global.cluster4, dims = 1:30)
query <- AddMetaData(query, metadata = my.predictions)

color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3","#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", "#AA8282", "#D4B7B7", "#8600BF", "#BA5CE3", "#808000","#AEAE5C", "#1E90FF", "#00BFFF", "#56FF0D", "#FFFF00")

#query <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/Seurat/Without_cDC2_CD33/seurat_label_transfer_human_to_mouse_merged_subsets.RDS')

query$tissue[query$tissue == 'EX'] <- 'K17KOMOC2'
query$tissue[query$tissue == 'MO'] <- 'MOC2'

DimPlot(query, group.by = 'predicted.id', cols = color_clusters, split.by = 'tissue')
ggsave('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/01182023_seurat_label_transfer_human_to_mouse_merged_subsets_split.png', width = 12, height =5, units = 'in', dpi = 300)

saveRDS(query, file = '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/01182023_seurat_label_transfer_human_to_mouse_merged_subsets.RDS')

save(ref_mini, my.anchors, my.predictions, query, ref, color_clusters, file="/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/01182023_ENVIRONMENT_seurat_label_transfer_human_to_mouse_merged_subsets.RData")
```

### •• [seurat] Assess cluster distribution

```{r}
transferred <- readRDS(file = '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/seurat_label_transfer_human_to_mouse_merged_subsets.RDS')

out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/'

# __________________________________________________________

subsets <- unique(transferred[['predicted.id']][,1])

master <- data.frame(AEG_res2_annotations = character(), 
                     Freq = numeric(), 
                     Pct = numeric(), 
                     predicted_id = character())
for (x in subsets){
  # make plots of the distribution of each subset
  meta <- transferred[['predicted.id']]
  m2 <- setNames(as.character(meta[,1]), rownames(meta))
  trans_mini <- transferred[,m2 == x]
  p2 <- DimPlot(trans_mini, group.by = 'predicted.id')
  p1 <- DimPlot(trans_mini, group.by = 'AEG_res2_annotations')
  cowplot::plot_grid(p1, p2)
  ggsave(paste(out, x, 'cluster_comparison.png'), width = 12, height = 5)
  
  # calculating the percentage distribution
  tbl <- table(trans_mini[['AEG_res2_annotations']])
  tbl <- as.data.frame(tbl)
  sm <- as.numeric(sum(tbl$Freq))
  tbl$Pct <- tbl$Freq/sm
  tbl$predicted_id <- x
  master <- rbind(master, tbl)
}
master_mini <- master[,c('AEG_res2_annotations', 'Pct', 'predicted_id')]
master_wide <- pivot_wider(master_mini, names_from = predicted_id, values_from = 'Pct')
rnms <- master_wide[['AEG_res2_annotations']]
master_wide[['AEG_res2_annotations']] <- NULL
master_wide[is.na(master_wide)] <- 0
rownames(master_wide) <- rnms

r_order <- match(c("DC_cDC1", "DC_moDC_cDC2", "DC_Ccr7", "pDC", "Mon_Classical", "Mon_Intermediate", "Mon_NonClassical", "Mon_Ccr5", "Mac_Cxcl9", "Mac_Fn1", "Mac_Lyve1", "Mac_Trem2"), rownames(master_wide))
c_order <- match(c("cDC1_CLEC9A", "cDC2_CD1C", "mregDC_LAMP3", "DC_pDC", "Mono_CD14", "Mono_TIL", "Mac_C1QA", "Mac_SPP1", "Mac_IL1B"), colnames(master_wide))

master_wide <- master_wide[r_order, c_order]
rownames(master_wide) <- c("DC_cDC1", "DC_moDC_cDC2", "DC_Ccr7", "pDC", "Mon_Classical", "Mon_Intermediate", "Mon_NonClassical", "Mon_Ccr5", "Mac_Cxcl9", "Mac_Fn1", "Mac_Lyve1", "Mac_Trem2")

png(paste(out, 'Heatmap_Seurat_transfer_learning_mouse_to_human_comparison.png'), width = 7, height = 5, units = 'in', res = 300)

ComplexHeatmap::Heatmap(master_wide, column_title = 'Predicted Human Cluster ID', column_title_side = c('bottom'), row_title = 'Original Mouse Cluster ID', row_title_side = 'left', heatmap_legend_param = list(title = "Percent"), cluster_rows = F, cluster_columns = F, row_split = c('DC', 'DC', 'DC', 'DC', 'Mon', 'Mon', 'Mon', 'Mon', 'Mac', 'Mac', 'Mac', 'Mac'), column_split = c('DC', 'DC', 'DC', 'DC', 'Mon', 'Mon', 'Mac', 'Mac', 'Mac'))

dev.off()
```

### •• [pre-tl.ingest] Convert ref and query to anndata

```{r}
query <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/moc2_human_genes_01112022.RDS')

load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/Seurat_MOC2_K17KO_MonMacDC_2021-08-30.rda')
DimPlot(hnc2, group.by = 'AEG_res2_annotations')

load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_myeloid_2021.rda')
ref <- hnc2[,hnc2$tissue == c('TIL')]
ref <- ref[,!ref$global.cluster4 == 'cDC2_CD33']
ref$global.cluster4[ref$global.cluster4 == 'DC3_LAMP3'] <- 'mregDC_LAMP3'
ref$global.cluster4[ref$global.cluster4 == c('Mac_SPP1_PLTP')] <- 'Mac_SPP1'
ref$global.cluster4[ref$global.cluster4 == c('Mono_CD14_IL1B')] <- 'Mono_CD14'
ref$global.cluster4[ref$global.cluster4 == c('Mono_CD14_THBS1')] <- 'Mono_CD14'
ref$global.cluster4[ref$global.cluster4 == c('Mono_CD14_ID1')] <- 'Mono_CD14'

ref_mini <- ref[, sample(colnames(ref), size = 5000, replace=F)]

#saveRDS(ref_mini, '/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/human_transfer_reference_01112022.RDS')

seurat_to_adata(out = '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/human_reference_downsampled.h5seurat', seu = ref_mini)


seurat_to_adata(out = '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/mouse_query.h5seurat', seu = hnc2)
```

### •• [post-tl.ingest] Add annotations to the mouse object

```{r}
#out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/tl.Ingest/humanized_mouse_ingested.h5ad'
#SeuratDisk::Convert(out, dest = 'h5seurat')

#mouse_ingest <- SeuratDisk::LoadH5Seurat("/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/tl.Ingest/humanized_mouse_ingested.h5seurat")

ingest <- read.csv('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/tl.Ingest/mouse_ingest_results.csv')

load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/Seurat_MOC2_K17KO_MonMacDC_2021-09-02.rda')

seu_mini <- hnc2[,colnames(hnc2) %in% rownames(ingest)]
rmns <- ingest$X
ingest$X <- NULL
rownames(ingest) <- rmns

seu_mini <- AddMetaData(seu_mini, col.name = 'tl.ingest', metadata = ingest)
DimPlot(seu_mini, group.by = 'tl.ingest', cols = color_clusters)

seu_mini <- seu_pp1(seu_mini)
ElbowPlot(seu_mini, ndims = 50) # choosing 30
seu_mini <- seu_pp2(seu_mini, npcs = 30)

seu_mini$tissue[seu_mini$tissue == 'EX'] <- 'K17KOMOC2'
seu_mini$tissue[seu_mini$tissue == 'MO'] <- 'MOC2'

DimPlot(seu_mini, group.by = 'tl.ingest', cols = color_clusters, split.by = 'tissue') + ggtitle('')
ggsave('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/tl.Ingest/UMAP_mouse_splitbytissue.png', width = 9, height = 5, dpi = 300) 

saveRDS(seu_mini, '/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/moc2_myeloid_tlingest.RDS')
```

#### ••• [post-tl.ingest] Cluster distribution heatmap

```{r}
transferred <- readRDS(file = '/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/moc2_myeloid_tlingest.RDS')

out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/tl.Ingest/'

transfer_meta <- 'tl.ingest'

# __________________________________________________________

subsets <- unique(transferred[[transfer_meta]][,1])

master <- data.frame(AEG_res2_annotations = character(), 
                     Freq = numeric(), 
                     Pct = numeric(), 
                     predicted_id = character())
for (x in subsets){
  # make plots of the distribution of each subset
  meta <- transferred[[transfer_meta]]
  m2 <- setNames(as.character(meta[,1]), rownames(meta))
  trans_mini <- transferred[,m2 == x]
  
  # calculating the percentage distribution
  tbl <- table(trans_mini[['AEG_res2_annotations']])
  tbl <- as.data.frame(tbl)
  sm <- as.numeric(sum(tbl$Freq))
  tbl$Pct <- tbl$Freq/sm
  tbl$predicted_id <- x
  master <- rbind(master, tbl)
}
master_mini <- master[,c('AEG_res2_annotations', 'Pct', 'predicted_id')]
master_wide <- pivot_wider(master_mini, names_from = predicted_id, values_from = 'Pct')
rnms <- master_wide[['AEG_res2_annotations']]
master_wide[['AEG_res2_annotations']] <- NULL
master_wide[is.na(master_wide)] <- 0
rownames(master_wide) <- rnms

r_order <- match(c("DC_cDC1", "DC_moDC_cDC2", "DC_Ccr7", "pDC", "Mon_Classical", "Mon_Intermediate", "Mon_NonClassical", "Mon_Ccr5", "Mac_Cxcl9", "Mac_Fn1", "Mac_Lyve1", "Mac_Trem2"), rownames(master_wide))
c_order <- match(c("cDC1_CLEC9A", "cDC2_CD1C", "mregDC_LAMP3", "DC_pDC", "Mono_CD14", "Mono_TIL",  "Mac_SPP1"), colnames(master_wide))

master_wide <- master_wide[r_order, c_order]
rownames(master_wide) <- c("DC_cDC1", "DC_moDC_cDC2", "DC_Ccr7", "pDC", "Mon_Classical", "Mon_Intermediate", "Mon_NonClassical", "Mon_Ccr5", "Mac_Cxcl9", "Mac_Fn1", "Mac_Lyve1", "Mac_Trem2")

png(paste(out, 'Heatmap_ingest_transfer_learning_mouse_to_human_comparison.png'), width = 7, height = 5, units = 'in', res = 300)

ComplexHeatmap::Heatmap(master_wide, column_title = 'Predicted Human Cluster ID', column_title_side = c('bottom'), row_title = 'Original Mouse Cluster ID', row_title_side = 'left', heatmap_legend_param = list(title = "Percent"), cluster_rows = F, cluster_columns = F, row_split = c('DC', 'DC', 'DC', 'DC', 'Mon', 'Mon', 'Mon', 'Mon', 'Mac', 'Mac', 'Mac', 'Mac'), column_split = c('DC', 'DC', 'DC', 'DC', 'Mon', 'Mon', 'Mac'))

dev.off()
```

#### ••• [post-tl.ingest] Scatterplot of freq difference

```{r}
transferred <- readRDS(file = '/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/moc2_myeloid_tlingest.RDS')

out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/Seurat_label_transfer_mouse_to_human/tl.Ingest/'
query_meta <- 'tl.ingest'
original_meta <- 'AEG_res2_annotations'

luoma <- readRDS( '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/UMAP_Luoma_labeltransfer_01172023.RDS')
luoma_meta <- 'predicted.id'

# __________________________________________________________

# calculating the percentage distribution
tbl <- table(transferred$tl.ingest, transferred$tissue)
tbl <- as.data.frame(tbl)
tbl <- tbl %>% mutate(Pct = case_when(
  Var2 == 'K17KOMOC2' ~ tbl$Freq/sum(tbl$Freq[tbl$Var2 == 'K17KOMOC2']), 
  Var2 == 'MOC2' ~ tbl$Freq/sum(tbl$Freq[tbl$Var2 == 'MOC2'])))

tbl_moc2 <- tbl[tbl$Var2 == 'MOC2',]
tbl_k17ko <- tbl[tbl$Var2 == 'K17KOMOC2',]
tbl <- merge(tbl_moc2, tbl_k17ko, by = 'Var1')

#getting percentage diff in R vs. NR in mice
tbl$delta <- tbl$Pct.y - tbl$Pct.x # more positive -> more enriched in K17KO

luoma$Path_response[luoma$Path_response %in% c('Medium', 'High')] <- 'Medium/High'

tbl1 <- table(luoma$predicted.id, luoma$Path_response)
tbl1 <- as.data.frame(tbl1)
tbl1 <- tbl1 %>% mutate(Pct = case_when(
  Var2 == c('Medium/High') ~ tbl1$Freq/sum(tbl1$Freq[tbl1$Var2 == c('Medium/High')]),   Var2 == 'Low' ~ tbl1$Freq/sum(tbl1$Freq[tbl1$Var2 == 'Low'])))

tbl1_r <- tbl1[tbl1$Var2 == 'Medium/High',]
tbl1_nr <- tbl1[tbl1$Var2 == 'Low',]
tbl1 <- merge(tbl1_r, tbl1_nr, by = 'Var1')

tbl1$delta <- tbl1$Pct.x - tbl1$Pct.y # more positive -> more enriched in responders

setdiff(tbl$Var1, tbl1$Var1) # in mouse and not in Luoma
# character(0)
setdiff(tbl1$Var1, tbl$Var1) # in Luoma and not in mouse
# [1] "Mac_C1QA" "Mac_IL1B"

tbl <- tbl[,c('Var1', 'delta')]
tbl1 <- tbl1[,c('Var1', 'delta')]

all <- merge(tbl, tbl1, by = 'Var1', all = TRUE)

all[is.na(all)] <- 0
colnames(all) <- c('cluster', 'k17ko_delta', 'R_delta')


ggplot(all, aes(x=k17ko_delta, y=R_delta)) + geom_point() + theme_classic() +
  xlab("Mouse Subset Proportion Delta [% in K17KOMOC2 - % in MOC2]") + ylab("Human Subset Proportion Delta [% in R - % in NR]") + 
  ggrepel::geom_text_repel(aes(x=k17ko_delta, y=R_delta), label = all$cluster) +
  geom_line(stat = 'smooth', method = 'lm', alpha = 0.3) + 
  ggtitle(paste('Subset Proportion Comparison', sep = ' '))

ggsave(file = paste(out, Sys.Date(), '_Human_mouse_subset_comparison.png', sep = ''), width = 5, height = 5, units = 'in', dpi = 300)
```

# VlnPlot/DotPlot

```{r}
id <- unique(seurat$global.cluster2[seurat$global.cluster2 %like any% c('Mac%', 'Mon%', '%DC%')])
#id <- unique(seurat$global.cluster2[seurat$global.cluster2 %like any% c('CD8%', '%CD4%', 'Treg%')])

feats <- stringr::str_to_title(c('PDCD1LG2', 'LGALS9', 'FN1', 'CXCL9', 'CXCL10', 'CCL7', 'CCL6', 'SPP1'))

VlnPlot(seurat, idents = id, features = c(feats), split.by = 'tissue') + theme(axis.text.x = element_text(angle = 90))
DotPlot(seurat, idents = id, features = c(feats), split.by = 'tissue', col.min = 0) + theme(axis.text.x = element_text(angle = 90)) + coord_flip()
```

# 

# Stacked barplot

```{r stacked bar plot figure 1e}
library(dplyr)

load('/Volumes/hdlab/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_myeloid_2021.rda') #object is named 'hnc2'

color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3","#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", "#AA8282", "#D4B7B7", "#8600BF", "#BA5CE3", "#808000","#AEAE5C", "#1E90FF", "#00BFFF", "#56FF0D", "#FFFF00")

hnc2$tissue.cluster <- paste(hnc2$tissue_hpv, hnc2$global.cluster4, sep = '|')
split <- as.data.frame(table(hnc2$tissue.cluster))
split <- split %>% tidyr::separate(Var1, c("dem", "cluster"), sep = '\\|')
split2 <- split %>% tidyr::separate(dem, c('donor_type', 'tissue', 'hpv_status'), sep = ':')

#PBMC
#making pbmc data sets
pbmc <- split2[stringr::str_detect(split2$tissue, "PBMC"), ] 

#HPV pos data sets
pbmc_hpvpos <- pbmc[pbmc$hpv_status == 'HPV+',]
pbmc_hpvpos$freqs <- pbmc_hpvpos$Freq/sum(pbmc_hpvpos$Freq)

#HPV neg HNSCC data set
pbmc_hnc_hpvneg <- pbmc[pbmc$donor_type == 'HNSCC',]
pbmc_hnc_hpvneg <- pbmc_hnc_hpvneg[pbmc_hnc_hpvneg$donor_type == 'HNSCC',]
pbmc_hnc_hpvneg <- pbmc_hnc_hpvneg[pbmc_hnc_hpvneg$hpv_status == 'HPV-',]
pbmc_hnc_hpvneg$freqs <- pbmc_hnc_hpvneg$Freq/sum(pbmc_hnc_hpvneg$Freq)

#HPV neg HD data set
pbmc_hd <- pbmc[pbmc$donor_type == 'HD',]
pbmc_hd$freqs <- pbmc_hd$Freq/sum(pbmc_hd$Freq)

#TIL
#making TIL data sets
til <- split2[stringr::str_detect(split2$tissue, 'TIL'),]
til_hpvpos <- til[til$hpv_status == 'HPV+',]
til_hpvpos$freqs <- til_hpvpos$Freq/sum(til_hpvpos$Freq)
til_hpvneg <- til[til$hpv_status == 'HPV-',]
til_hpvneg$freqs <- til_hpvneg$Freq/sum(til_hpvneg$Freq)

#HD tonsil
tonsil <- split2[stringr::str_detect(split2$tissue, 'Tonsil'),]
tonsil$freqs <- tonsil$Freq/sum(tonsil$Freq)

all1 <- rbind(pbmc_hpvpos, pbmc_hnc_hpvneg, pbmc_hd, til_hpvpos, til_hpvneg, tonsil)
all1$tissue_hpv <- paste(all1$tissue, all1$donor_type, all1$hpv_status)

ggplot(all1, aes(fill=cluster, y=freqs, x=tissue_hpv, cex.axis = 1.5, cex.lab = 4)) + geom_bar(position="stack", stat="identity") + coord_flip() + theme_bw(base_size = 32) + scale_fill_manual(values = color_clusters) + theme(axis.text = element_text(size = 20))
ggsave('/Volumes/hdlab/Projects/HNC_SPORE/Golfinosetal2022/tissue_HPV_stackedbarplot.png', width = 24, height = 12)
```

![](images/paste-0DD27DD5.png)

# Frequency boxplots

```{r}
f <- freq_boxplot(hnc, celltype = 'CD4_1', metadata = 'mreg_cxcl9_globalcluster4')

ggplot(f) +
    geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:5) #needs to be equal to the number of samples that you have
```

# Dotplot of LAMP3 DC markers

```{r dotplot of LAMP3 markers}
DotPlot(hnc, features = c('CD40', 'CD80', 'CD86', 'Relb', 'CD83'), idents = c('DC3_LAMP3', 'cDC2_CD33', 'cDC2_CD1C', 'cDC1_CLEC9A'), col.min = 0) + ggtitle('Maturation')

DotPlot(hnc, features = c('CD274', 'PDCD1LG2', 'CD200', 'FAS', 'ALDH1A2', 'SOCS1', 'SOCS2'), idents = c('DC3_LAMP3', 'cDC2_CD33', 'cDC2_CD1C', 'cDC1_CLEC9A'), col.min = 0) + ggtitle('Regulatory') + theme(axis.text.x = element_text(angle = 90))

DotPlot(hnc, features = c('CCR7', 'MYO1G', 'CXCL16', 'ADAM8', 'ICAM1', 'MARCKS', 'MARCKSL1'), idents = c('DC3_LAMP3', 'cDC2_CD33', 'cDC2_CD1C', 'cDC1_CLEC9A'), col.min = 0) + ggtitle('Migration') + theme(axis.text.x = element_text(angle = 90))
```

# GSEA

```{r}
library(cerebroApp)
gmt <- qusage::read.gmt('/Volumes/hqdinh2/Projects/HNC_SPORE/Resources/mSigDB_hallmarkSigs/c5.go.bp.v7.5.1.symbols.gmt')

#Basic function to convert mouse to human gene names
require("biomaRt")
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")

human = useEnsembl('ensembl', dataset = 'hsapiens_gene_ensembl', mirror = 'uswest')
mouse = useEnsembl('ensembl', dataset = 'mmusculus_gene_ensembl', mirror = 'uswest')

new_gmt <- list()
for (x in 1:length(gmt)){
  genesV2 = biomaRt::getLDS(attributes = c("mgi_symbol"), filters = "mgi_symbol", values = genes, mart = mouse, attributesL = c("hgnc_symbol"), martL = human, uniqueRows=T)
  #test <- str_to_title(unlist(gmt[[x]]))
  #new_gmt[[names(gmt[x])]] <- str_to_title(as.character(unlist(gmt[x])))
}

library(LeviRmisc)
LeviRmisc::writeGMT(new_gmt, fname = paste('/Volumes/hqdinh2/Projects/HNC_SPORE/Resources/mouse_', 'c5.go.bp.v7.5.1.symbols.gmt', sep = ''))


## if you do not need to convert anything start here####

gmt <- qusage::read.gmt('/Volumes/hqdinh2/Projects/HNC_SPORE/Resources/m5.go.bp.v2022.1.Mm.symbols.gmt')

test <- Seurat::AddModuleScore(seurat, features = gmt, name = 'GSEA')

test <- cerebroApp::performGeneSetEnrichmentAnalysis(seurat, assay = 'RNA', GMT_file = paste('/Volumes/hqdinh2/Projects/HNC_SPORE/Resources/', 'm5.go.bp.v2022.1.Mm.symbols.gmt', sep = ''), groups = 'tissue', name = 'GSVA')

test1 <- test[[c("global.cluster2", "tissue", 'GSEA1')]]
#bxp <- ggplot(test1, aes(x=global.cluster2 , y=GSEA1, fill = tissue)) + geom_boxplot()
bxp <- ggboxplot(test1, x = 'global.cluster2', y='GSEA1', color = 'tissue') + coord_flip()

stat.test <- test1 %>% group_by(global.cluster2) %>% t_test(GSEA1 ~ tissue) %>%
  adjust_pvalue(method = "bonferroni") %>% add_significance("p.adj")
stat.test <- stat.test %>% add_xy_position(x = "global.cluster2", dodge = 0.8)
bxp + stat_pvalue_manual(stat.test,  label = "p", tip.length = 0) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))


DotPlot(test, features = 'GSEA1', split.by = 'tissue', col.min = 0, cols = c('blue', 'red'))
```

# _____ Zhou et al MOC HNC models _________

## • Generate seurat object

```{r}
#moc1e <- read.delim('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/GSE153383_MOC1E_scRNA_counts.txt', header = TRUE)
#moc22 <- read.delim('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/GSE153383_MOC22_scRNA_counts.txt', header = TRUE)

#moc1e <- CreateSeuratObject(moc1e)
#moc1e$orig.ident <- 'MOC1E'
#moc22 <- CreateSeuratObject(moc22)
#moc22$orig.ident <- 'MOC22'

#moc <- merge(moc1e, moc22)

#moc1e <- seu_pp1(moc1e)
#moc1e <- seu_pp2(moc1e)
#DimPlot(moc1e, group.by = 'orig.ident')

#moc22 <- seu_pp1(moc22)
#moc22 <- seu_pp2(moc22)
#DimPlot(moc22, group.by = 'orig.ident')

#saveRDS(moc1e, '/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/zhou_etal_moc1e.RDS')
#saveRDS(moc22, '/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/zhou_etal_moc22.RDS')

#DimPlot(moc, group.by = 'orig.ident')

#saveRDS(moc, '/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/zhou_etal_seurat.RDS')
```

## • Integrate samples
```{r}
moc1e <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/zhou_etal_moc1e.RDS')
moc22 <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/zhou_etal_moc22.RDS')

ifnb.list <- list(moc1e, moc22)

# normalize and identify variable features for each dataset independently
ifnb.list <- lapply(X = ifnb.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = ifnb.list)

# select features that are repeatedly variable across datasets for integration
immune.anchors <- FindIntegrationAnchors(object.list = ifnb.list, anchor.features = features)
immune.combined <- IntegrateData(anchorset = immune.anchors)

DimPlot(immune.combined, group.by = 'orig.ident')

immune.combined <- seu_pp1(immune.combined, normalize = F)
immune.combined <- seu_pp2(immune.combined)
DimPlot(immune.combined, group.by = 'orig.ident')

#saveRDS(immune.combined, '/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/zhou_etal_combined.RDS')
```

## • Further QC and doublet removal
```{r}
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

immune.combined <- seu_pp0(immune.combined)

DimPlot(immune.combined, group.by = 'orig.ident')

#saveRDS(immune.combined, '/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/zhou_etal_combined.RDS')
```

## • Label transfer [Reference: Wang et al 2022]
```{r}
load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/Seurat_MOC2_K17KO_2021-11-01**.rda')

query <- zhou
anchors <- FindTransferAnchors(reference = seurat, query = query, dims = 1:20)
predictions <- TransferData(anchorset = anchors, refdata = seurat$global.cluster2, dims = 1:20)
immune.combined <- AddMetaData(query, metadata = predictions)

DimPlot(immune.combined, group.by = 'predicted.id')

#saveRDS(immune.combined, '/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/zhou_etal_combined.RDS')
```

## • Myeloid only

```{r}
myeloid <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Objects/zhou_etal_combined_myeloid.RDS')

# Ccl7 DCs/mregDC markers
VlnPlot(myeloid, group.by = 'predicted.id', features = c('Ccr7', 'Fscn1', 'Ccl22'))

# cDC1 markers
VlnPlot(myeloid, group.by = 'predicted.id', features = c('Plet1', 'H2-DMb2'))

# cDC2 markers
VlnPlot(myeloid, group.by = 'predicted.id', features = c('Cd209a', 'Cd7'))

# Mac CXCL9 markers
VlnPlot(myeloid, group.by = 'predicted.id', features = c('Cxcl9', 'Cxcl10'))

# Mac Fn1 markers
VlnPlot(myeloid, group.by = 'predicted.id', features = c('Mmp12', 'Arg1', 'Bsg'))

# Mac Lyve1 markers
VlnPlot(myeloid, group.by = 'predicted.id', features = c('Cbr2', 'Gas6', 'Ccl8'))

# Mac Trem2 markers
VlnPlot(myeloid, group.by = 'predicted.id', features = c('Bsg', 'Arg1', 'Ms4a7'))

# Mon Ccr5 markers
VlnPlot(myeloid, group.by = 'predicted.id', features = c('Vcan', 'Dmkn', 'Itga1'))

# Mon Classical markers
VlnPlot(myeloid, group.by = 'predicted.id', features = c('Ifitm6', 'Mgst1', 'Hp'))

# Mon Intermediate markers
VlnPlot(myeloid, group.by = 'predicted.id', features = c('Cxcr2', 'Hdc', 'Pou2f2'))

# Mon Non Classical markers
VlnPlot(myeloid, group.by = 'predicted.id', features = c('Ace', 'Adgre4', 'Acod1'))
```


### •• Label transfer Wang myeloid --> Zhou myeloid
```{r}
zhou <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Objects/zhou_etal_combined.RDS')

DimPlot(zhou, group.by = 'orig.ident')

zhou_myeloid <- zhou[,zhou$predicted.id %like any% c('%Mac%', '%Mon%', '%DC%')]
DimPlot(zhou_myeloid, group.by = 'predicted.id')
VlnPlot(zhou_myeloid, features = c("nCount_RNA", "nFeature_RNA"), group.by = 'orig.ident')

load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/Seurat_MOC2_K17KO_MonMacDC_2021-09-02.rda')

zhou_myeloid <- seu_pp1(zhou_myeloid, normalize = FALSE)
zhou_myeloid <- seu_pp2(zhou_myeloid)
DimPlot(zhou_myeloid)
zhou <- NULL

anchors <- FindTransferAnchors(reference = hnc2, query = zhou_myeloid, dims = 1:20)
predictions <- TransferData(anchorset = anchors, refdata = hnc2$AEG_res2_annotations, dims = 1:20)
zhou_myeloid <- AddMetaData(zhou_myeloid, metadata = predictions)

DimPlot(zhou_myeloid, group.by = 'predicted.id')
#ggsave('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Figures/myeloid_predicted.id.png', width = 7, height = 5, units = 'in', dpi = 300)
#saveRDS(zhou_myeloid, file = '/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/zhou_etal_combined_myeloid.RDS')

zhou_myeloid$tissue <- zhou_myeloid$orig.ident
zhou_myeloid$AEG_res2_annotations <- zhou_myeloid$predicted.id

all <- merge(hnc2, zhou_myeloid)

all <- seu_pp1(all)
all <- seu_pp2(all)
DimPlot(all, group.by = 'AEG_res2_annotations')

all <- all[,! all$tissue == c('MOC22')]
all$tissue[all$tissue == 'EX'] <- 'K17KOMOC2'
all$tissue[all$tissue == 'MO'] <- 'MOC2'
 
#df <- freq_boxplot(all,  metadata = 'AEG_res2_annotations', out = '/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Figures/zhou_myeloid_freq_boxplots.png', split.meta = 'tissue')

#saveRDS(zhou_myeloid, file = paste('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Objects/', Sys.Date(), 'zhou_etal_combined_myeloid.RDS', sep = ''))
```

## • Neutrophil only

```{r}
neu <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Objects/2023-02-09_zhou_etal_combined_neutrophils.RDS')
VlnPlot(neu, group.by = 'orig.ident', features = c('nCount_RNA', 'nFeature_RNA'))
VlnPlot(neu, group.by = 'predicted.id', features = 'Cxcl9')
neu <- subset(neu, subset = nFeature_RNA > 200 & nFeature_RNA < 2000)
neu <- subset(neu, subset = nCount_RNA < 4000)

FeaturePlot(neu, features = c('nCount_RNA', 'nFeature_RNA'))

saveRDS(neu, paste('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Objects/', Sys.Date(), 'zhou_etal_combined_neutrophils.RDS', sep = ''))

DimPlot(neu)
```

### •• Label transfer Wang neutrophils --> Zhou neutrophils
```{r}
zhou_myeloid <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Objects/2023-02-09_zhou_etal_combined_neutrophils.RDS')

DimPlot(zhou, group.by = 'orig.ident')

zhou_myeloid <- zhou[,zhou$predicted.id %like any% c('%N0%', '%N1%', '%N2%', '%N3%')]
DimPlot(zhou_myeloid, group.by = 'predicted.id')
VlnPlot(zhou_myeloid, features = c("nCount_RNA", "nFeature_RNA"), group.by = 'orig.ident')

load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/Seurat_MOC2_K17KO_2021-11-01**.rda')

hnc2 <- seurat[,seurat$global.cluster2 %like any% c('%N0%', '%N1%', '%N2%', '%N3%')]
DimPlot(hnc2, group.by = 'global.cluster2')

zhou_myeloid <- seu_pp1(zhou_myeloid, normalize = FALSE)
zhou_myeloid <- seu_pp2(zhou_myeloid)
DimPlot(zhou_myeloid)
zhou <- NULL

anchors <- FindTransferAnchors(reference = hnc2, query = zhou_myeloid, dims = 1:20)
predictions <- TransferData(anchorset = anchors, refdata = hnc2$global.cluster2, dims = 1:20)
zhou_myeloid <- AddMetaData(zhou_myeloid, metadata = predictions)

DimPlot(zhou_myeloid, group.by = 'predicted.id')
ggsave('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Figures/neutrophil_predicted.id.png', width = 7, height = 5, units = 'in', dpi = 300)

#saveRDS(zhou_myeloid, file = '/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/zhou_etal_combined_neutrophils.RDS')

zhou_myeloid$tissue <- zhou_myeloid$orig.ident
zhou_myeloid$global.cluster2 <- zhou_myeloid$predicted.id

all <- merge(hnc2, zhou_myeloid)

all <- seu_pp1(all)
all <- seu_pp2(all)
DimPlot(all, group.by = 'global.cluster2')

all <- all[,! all$tissue == c('MOC22')]
all$tissue[all$tissue == 'EX'] <- 'K17KOMOC2'
all$tissue[all$tissue == 'MO'] <- 'MOC2'
 
df <- freq_boxplot(all,  metadata = 'global.cluster2', out = '/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Figures/zhou_neutrophil_freq_boxplots.png', split.meta = 'tissue', out_width = 6)

saveRDS(zhou_myeloid, file = paste('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Objects/', Sys.Date(), 'zhou_etal_combined_myeloid.RDS', sep = ''))
```

## • Transfer neutrophil annotations back to main object
```{r}

zhou_neu <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Objects/2023-02-09_zhou_etal_combined_neutrophils.RDS')

zhou_all <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Objects/zhou_etal_combined.RDS')

# basically this loops through and checks if the cell belongs to both seurat objects, then it takes the metadata from the smaller if they intersect

temp_meta <- ifelse(colnames(zhou_all) %in% colnames(zhou_neu),
                     as.character(zhou_neu@meta.data$predicted.id),
                     as.character(zhou_all@meta.data$predicted.id))

# assigning new metadata to your larger seurat object and setting as active identity
zhou_all@meta.data$global.cluster2 <- temp_meta
Idents(zhou_all) <- temp_meta
```

## • Transfer myeloid predictions back to the main object
```{r}
temp_meta <- ifelse(colnames(zhou_all) %in% colnames(zhou_myeloid),
                     as.character(zhou_myeloid@meta.data$predicted.id),
                     as.character(zhou_all@meta.data$predicted.id))

# assigning new metadata to your larger seurat object and setting as active identity
zhou_all@meta.data$global.cluster2 <- temp_meta
Idents(zhou_all) <- temp_meta
```


## • VlnPlot
```{r}
zhou_neu <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Objects/2023-02-09_zhou_etal_combined_neutrophils.RDS')

zhou_all <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Objects/zhou_etal_combined.RDS')
saveRDS(zhou_all, file = '/Volumes/hqdinh2/Projects/HNC_SPORE/Zhou_HNC_mice/Objects/zhou_etal_combined.RDS')


VlnPlot(zhou_all, group.by = 'predicted.id', features = c('Tnfsf14', 'Ltbr'))
```


# ________Choi et al HNC progression________
## • Load and save object
```{r}
choi <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Choi_HNC_progression/2023-03-02_Choi_combined.RDS')

choi_my <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Choi_HNC_progression/2023-03-02_Choi_myeloid.RDS')


#saveRDS(choi, paste('/Volumes/hqdinh2/Projects/Public_Data/Choi_HNC_progression/', Sys.Date(), '_Choi_combined.RDS', sep = ''))

#saveRDS(choi_my, paste('/Volumes/hqdinh2/Projects/Public_Data/Choi_HNC_progression/', Sys.Date(), '_Choi_myeloid.RDS', sep = ''))
```


## • Myeloid only
```{r}
choi_my <- choi[,choi$cell.type %in% c('Dendritic.cells', 'Macrophages')]

choi_my <- seu_pp1(choi_my)
choi_my <- seu_pp2(choi_my, npcs = 40)

DimPlot(choi_my)
```

## • Label transfer (Cillo > Choi)
```{r}
query <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Choi_HNC_progression/2023-03-02_Choi_myeloid.RDS')

load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_myeloid_2021.rda')
ref <- hnc2[,hnc2$tissue == c('TIL')]
ref <- ref[,!ref$global.cluster4 == 'cDC2_CD33']
ref <- ref[,!ref$global.cluster4 == 'Mono_Int']
ref$global.cluster4[ref$global.cluster4 == 'DC3_LAMP3'] <- 'mregDC_LAMP3'
#ref$global.cluster4[ref$global.cluster4 == c('Mac_SPP1_PLTP')] <- 'Mac_SPP1'
ref$global.cluster4[ref$global.cluster4 == c('Mono_CD14_IL1B')] <- 'Mono_CD14'
ref$global.cluster4[ref$global.cluster4 == c('Mono_CD14_THBS1')] <- 'Mono_CD14'
ref$global.cluster4[ref$global.cluster4 == c('Mono_CD14_ID1')] <- 'Mono_CD14'

table(ref$tissue)
# TIL 
#9439 

table(ref$global.cluster4)
#  cDC1_CLEC9A     cDC2_CD1C        DC_pDC      Mac_C1QA      Mac_IL1B 
#          251           939           847          1216           774 
#     Mac_SPP1 Mac_SPP1_PLTP          Mast     Mono_CD14     Mono_CD16 
#         1837           573           304          1115           160 
#     Mono_TIL  mregDC_LAMP3 
#         961           462 

hnc2 <- NULL
m <- NULL

ref_mini <- ref[, sample(colnames(ref), size = 5000, replace=F)]

table(ref_mini$global.cluster4)

ref <- NULL

my.anchors <- FindTransferAnchors(reference = ref_mini, query = query, dims = 1:30)
my.predictions <- TransferData(anchorset = my.anchors, refdata = ref_mini$global.cluster4, dims = 1:30)
query <- AddMetaData(query, metadata = my.predictions)

#saveRDS(query, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/UMAP_Luoma_labeltransfer_01182023.RDS')

color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3","#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", "#AA8282", "#D4B7B7", "#8600BF", "#BA5CE3", "#808000","#AEAE5C", "#1E90FF", "#00BFFF", "#56FF0D", "#FFFF00")


DimPlot(query, group.by = 'predicted.id', cols = color_clusters) + ggtitle('')
ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/Stage_UMAP_Luoma_labeltransfer_01182023.png', width = 11, height =5, units = 'in', dpi = 300)

query$Path_response[query$Path_response %in% c('Medium', 'High')] <- 'Medium/High'

DimPlot(query, group.by = 'predicted.id', cols = color_clusters, split.by = 'Path_response') + ggtitle('')
ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/PathResponse_UMAP_Luoma_labeltransfer_01182023.png', width = 11, height =5, units = 'in', dpi = 300)
```


## • Global: Pre-processing (done)
```{r}
counts <- read.delim('/Volumes/hqdinh2/Projects/Public_Data/Choi_HNC_progression/GSE181919_UMI_counts.txt')

meta <- read.delim('/Volumes/hqdinh2/Projects/PUblic_Data/Choi_HNC_progression/GSE181919_Barcode_metadata.txt')

# making sure both objects have the same cellIDs 
rownames(meta) <- gsub('-', '\\.', rownames(meta))

choi <- CreateSeuratObject(counts = counts, project = "Choi_HNC", min.cells = 3, min.features = 200)

choi <- AddMetaData(choi, meta)

choi <- seu_pp1(choi)
choi <- seu_pp2(choi, npcs = 75)

saveRDS(choi, paste('/Volumes/hqdinh2/Projects/Public_Data/Choi_HNC_progression/', Sys.Date(), '_Choi_combined.RDS'))
```

# _____Wigerblad healthy neutrophils_____
```{r}
wbd <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Wigerblad_healthy_neutrophils/neutrophil_merge_processed.rds')

DimPlot(wbd, label = T)

wbd$seurat_clusters <- as.character(wbd$seurat_clusters)
wbd.neu <- wbd[,wbd$seurat_clusters %in% c('1', '2', '3', '4', '5', '6')]


```

# _______________Kurten_______________

# Merge CD45- cells from Kurten into Cillo data

```{r}
load('/Volumes/hqdinh2-1/Projects/HNC_SPORE/Golfinosetal2022/Kurten_reanalysis/R_objects/kurten_seuratobject.RData')

kurten_cd45neg <- kurten[,kurten$global.cluster %in% c('Endothelial cells', 'Epithelial cells', 'Fibroblasts', 'Pericytes')]

kurten_cd45neg$mreg_cxcl9_globalcluster4 <- kurten_cd45neg$global.cluster

DimPlot(kurten_cd45neg, group.by = 'global.cluster')

#now merging the CD45- cells in with the Cillo data
hnc2 <- merge(hnc, kurten_cd45neg, add.cell.ids = c('cillo', 'kurten'), project = 'CD45pos_neg')

hnc2 <- NormalizeData(hnc2)

DimPlot(hnc2, group.by = 'mregDC_CXCL9_globalcluster4')

hnc2 <- FindVariableFeatures(hnc2, selection.method = "vst", nfeatures = 2000)

hnc2 <- ScaleData(hnc2)

hnc2 <- RunPCA(hnc2, features = VariableFeatures(object = hnc2), npcs = 100)

ElbowPlot(hnc2, ndims = 100)

hnc2 <- FindNeighbors(hnc2, dims = 1:50)
hnc2 <- FindClusters(hnc2, resolution = 0.5)

hnc2 <- RunUMAP(hnc2, dims = 1:10)

DimPlot(hnc2, reduction = "umap", group.by = 'mreg_cxcl9_globalcluster4')

saveRDS(hnc2, '/Volumes/hqdinh2-1/Projects/HNC_SPORE/Seurat_Objs/HNC_human/cillo_kurten.RDS')

hnc2$mreg_cxcl9_globalcluster4[hnc2$mreg_cxcl9_globalcluster4 == 'cDC2_CD1C'] <- 'DC3_CD1C' 

hnc2 <- hnc2[,!hnc2$mreg_cxcl9_globalcluster4 %in% c('NKs', 'pDC', 'Macrophages', 'KRT17+', 'DCs', 'CD4', 'CD8', 'CD16_Monocytes', 'Immature_Myeloids')]

saveRDS(hnc2, '/Volumes/hqdinh2-1/Projects/HNC_SPORE/Seurat_Objs/HNC_human/cillo_kurten.RDS')
```

# Making deconvolution input files (Cytospace)
```{r}
ref <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/HNC_human/cillo_kurten.RDS')

#ref_mini <- subset(ref, downsample = 5000)
ref_mini <- ref[, sample(colnames(ref), size = 10000, replace=F)]
DimPlot(ref_mini)

ref_counts <- GetAssayData(object = ref_mini, slot = "counts")

ref_meta <- ref_mini[[c('decon.clusters')]]

write.csv(ref_meta, '/Volumes/hqdinh2/Projects/HNC_SPORE/Cytospace/kurten_mini_meta.csv')

write.csv(ref_counts, '/Volumes/hqdinh2/Projects/HNC_SPORE/Cytospace/kurten_mini_counts.csv')
```

# _________HNC CITE-SEQ Blood___________


# [DE genes between subsets]

```{r}
DefaultAssay(neu) <- 'FB'

de_dotplot(neu, 'wsnn_res.0.4')

plot_tib<-data_frame(RNA_Count=colSums(seu@assays$RNA),
                     FB_Count=colSums(seu@assays$FB),
                     Cell_Type=seu$global.cluster)

ggplot(data=plot_tib,aes(x=RNA_Count,y=FB_Count,colour=Cell_Type),cols=c('red','blue','green','yellow','violet'))+geom_point()
```

# [Plotting protein markers]

```{r}
#seu <- readRDS('/Volumes/hqdinh2/Projects/RawData_FromNovoGene/HNCBld_CITESeq_Nov2022/DeepSequencing_2022-12-16/04.SeuratObjects/2023-02-09_neutrophils.RDS')
DefaultAssay(seu) <- 'FB'
rownames(seu)

FeaturePlot(seu, features = "Hu.CD66b")

FeaturePlot(seu, features = "Hu.CD8")
FeaturePlot(seu, features = "Hu.CD4-RPA.T4")


FeaturePlot(seu, features = "Hu.CD11b")
FeaturePlot(seu, features = "Hu.CD11c")
FeaturePlot(seu, features = "Hu.CD14-M5E2")
FeaturePlot(seu, features = "Hu.CX3CR1")

FeaturePlot(seu, features = "Hu.IgM")
FeaturePlot(seu, features = "Hu.KLRG1")
FeaturePlot(seu, features = "Hu.TCR.AB")

DimPlot(seu, group.by = 'wsnn_res.0.1', label = T)

DefaultAssay(seu) <- 'integratedRNA'

de_dotplot(seu, metadata = 'wsnn_res.0.1')
VlnPlot(seu, features = 'nCount_RNA', group.by = 'global.cluster')
```

# Neutrophils only

## • Load (and save) neutrophil object
```{r}
neu <- readRDS('/Volumes/hqdinh2/Projects/RawData_FromNovoGene/HNCBld_CITESeq_Nov2022/DeepSequencing_2022-12-16/04.SeuratObjects/2023-03-02_neutrophils.RDS')

seu

#saveRDS(neu, paste('/Volumes/hqdinh2/Projects/RawData_FromNovoGene/HNCBld_CITESeq_Nov2022/DeepSequencing_2022-12-16/04.SeuratObjects/', Sys.Date(), '_neutrophils.RDS', sep = ''))
```

## • Remove doublets
```{r}
d70774 <- neu[,neu$orig.ident == 'D70774']

require(DoubletFinder)
sweep.res <- paramSweep_v3(d70774) 
sweep.stats <-summarizeSweep(sweep.res, GT = FALSE) 
bcmvn <- find.pK(sweep.stats)
barplot(bcmvn$BCmetric, names.arg = bcmvn$pK, las=2)

# define the expected number of doublet cellscells.
nExp <- round(ncol(d70774) * 0.015)  # expect 1.5% doublets
d70774 <- doubletFinder_v3(d70774, pN = 0.25, pK = 0.005, nExp = nExp, PCs = 1:10)


d70774 <- d70774[,d70774$DF.classifications_0.25_0.005_28 == 'Singlet']

d70792 <- neu[,neu$orig.ident == 'D70792']
require(DoubletFinder)
sweep.res <- paramSweep_v3(d70792) 
sweep.stats <-summarizeSweep(sweep.res, GT = FALSE) 
bcmvn <- find.pK(sweep.stats)
barplot(bcmvn$BCmetric, names.arg = bcmvn$pK, las=2)

# define the expected number of doublet cellscells.
nExp <- round(ncol(d70792) * 0.02)  # expect 2% doublets
d70792 <- doubletFinder_v3(d70792, pN = 0.25, pK = 0.01, nExp = nExp, PCs = 1:10)

d70792 <- d70792[,d70792$DF.classifications_0.25_0.01_52 == 'Singlet']

neu <- merge(d70792,d70774)
```


## • Characterizing neutrophil subsets

```{r}
#seu$global.cluster <- seu$wsnn_res.0.1
#seu$global.cluster <- as.character(seu$global.cluster)
#seu$global.cluster[seu$global.cluster %in% c('1', '0', '5')] <- 'Neutrophils'
#seu$global.cluster[seu$global.cluster %in% c('3', '6')] <- 'T/NK cells'
#seu$global.cluster[seu$global.cluster == '10'] <- 'Platelets'
#seu$global.cluster[seu$global.cluster %in% c('2', '4')] <- 'Myeloid cells'

#DimPlot(seu, group.by = 'global.cluster', label = T)

#neu <- seu[,seu$global.cluster == 'Neutrophils']

# RNA
neu@active.assay <- "RNA"
neu <- NormalizeData(neu)
neu <- FindVariableFeatures(neu)
neu <- ScaleData(neu)
neu <- RunPCA(neu, assay.use = "RNA", reduction.name = "NEUrnapca", features = rownames(neu@assays$RNA@scale.data))
neu <- harmony::RunHarmony(neu, assay.use = "RNA", group.by.vars = "orig.ident", reduction = "NEUrnapca", reduction.save = "harmonyrna")

# FB
neu@active.assay <- "FB"
neu <- NormalizeData(neu, normalization.method = "CLR", margin = 2)
VariableFeatures(neu, assay = "FB") <- grep("Iso", rownames(neu), value = TRUE, invert = TRUE)
neu <- ScaleData(neu)
neu <- RunPCA(neu, assay = "FB", reduction.name = "NEUfbpca")
neu <- harmony::RunHarmony(neu, assay.use = "FB", group.by.vars = "orig.ident", reduction = "NEUfbpca",reduction.save = "harmonyfb")

#DimPlot(neu, reduction = "NEUfbpca", group.by = "orig.ident")
ElbowPlot(neu, reduction = "NEUfbpca", ndims = 50) # 20
ElbowPlot(neu, reduction = "NEUrnapca", ndims = 50) # 20
neu <- FindMultiModalNeighbors(neu, reduction.list = list("harmonyrna", "harmonyfb"), dims.list = list(1:20, 1:20), modality.weight.name = "RNA.weight")

neu <- RunUMAP(neu, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
neu <- FindClusters(neu, graph.name = "wsnn", algorithm = 3, resolution = seq(0.1, 1, 0.1), verbose = TRUE)
DimPlot(neu, group.by = c("orig.ident"), reduction = "wnn.umap", ncol = 2)


DimPlot(neu, reduction = 'wnn.umap', label = TRUE, repel = TRUE, label.size = 2.5, group.by = 'orig.ident') + NoLegend()

saveRDS(neu, paste('/Volumes/hqdinh2/Projects/RawData_FromNovoGene/HNCBld_CITESeq_Nov2022/DeepSequencing_2022-12-16/04.SeuratObjects/', Sys.Date(), '_neutrophils.RDS', sep = ''))
```

## • Plotting all protein markers
```{r}
metadata <- 'integrated_snn_res.0.2'
neu <- seu

# prepping the gene by cluster dataframe
#s <- as.data.frame(t(GetAssayData(neu, assay = 'FB', slot = 'data'))) # get gene expression
#s <- s[,!colnames(s) %like any% c('%Isotype%')] # remove isotype genes
#meta <- neu[[metadata]] # get cluster metadata
#all <- merge(s, meta, by = 0) # merge objects
#all1 <- aggregate(all, by = list(all$integrated_snn_res.0.2), FUN=mean) # aggregate by cluster
#all1 <- all1[,colSums(is.na(all1))<nrow(all1)] # remove columns with N/A
#row.names(all1) <- all1$Group.1  # put gene in `row`
#all1 <- all1[,-1] #drop gene column as now in rows
#DefaultAssay(neu) <- 'FB'
#Idents(neu) <- metadata
#all_markers <- FindAllMarkers(object = neu, group.by = metadata)

png('/Volumes/hqdinh2/Projects/RawData_FromNovoGene/HNCBld_CITESeq_Nov2022/DeepSequencing_2022-12-16/clustered_dot_plot_integrated_wsnn_res.0.2.png', width = 7, height = 17, units = 'in', res = 300)
scCustomize::Clustered_DotPlot(seurat_object = neu, features = rownames(neu)[!rownames(neu) %like any% '%Isotype%'], group.by = 'wsnn_res.0.2', assay = 'FB', exp_color_min = 0)
dev.off()

# order of clusters from Clustered_DotPlot is 3, 5, 1, 4, 6, 2, 0, 7
#neu$integrated_snn_res.0.2 <- factor(x = neu$integrated_snn_res.0.2, levels = c(3, 5, 1, 4, 6, 2, 0, 7))
```

```{r test function}
seurat_object = seu
features = rownames(seu)[!rownames(seu) %like any% "%Isotype%"]
colors_use_exp = c('grey', 'light blue', 'blue')
exp_color_min = 0
exp_color_middle = NULL
exp_color_max = 2 
print_exp_quantiles = FALSE
colors_use_idents = NULL
x_lab_rotate = TRUE
flip = FALSE
k = 1
feature_km_repeats = 1000
ident_km_repeats = 1000
row_label_size = 8
raster = FALSE
plot_km_elbow = TRUE 
elbow_kmax = NULL
assay = NULL
group.by = 'integrated_snn_res.0.9'
idents = NULL
show_parent_dend_line = TRUE
ggplot_default_colors = FALSE 
color_seed = 123
seed = 123 
  
features_unique <- unique(x = features)
  
seurat_plot <- DotPlot(object = seurat_object, features = features, 
    assay = assay, group.by = group.by, scale = TRUE, idents = idents, col.min = NULL, col.max = NULL)
data <- seurat_plot$data
exp_mat <- data %>% select(-any_of(c("pct.exp", "avg.exp"))) %>% 
  pivot_wider(names_from = .data[["id"]], values_from = .data[["avg.exp.scaled"]]) %>% 
  as.data.frame()

exp_mat <- as.data.frame(exp_mat)
exp_mat <- na.omit(exp_mat)

row.names(x = exp_mat) <- exp_mat$features.plot
exp_mat <- exp_mat[, -1] %>% as.matrix()
percent_mat <- data %>% select(-any_of(c("avg.exp", "avg.exp.scaled"))) %>% 
  pivot_wider(names_from = .data[["id"]], values_from = .data[["pct.exp"]]) %>% 
  as.data.frame()
row.names(x = percent_mat) <- percent_mat$features.plot
percent_mat <- percent_mat[, -1] %>% as.matrix()
assay <- assay %||% DefaultAssay(object = seurat_object)
group_by_length <- length(x = unique(x = seurat_object@meta.data[[group.by]]))
if (is.null(x = colors_use_idents)) {
  colors_use_idents <- scCustomize_Palette(num_groups = group_by_length, 
    ggplot_default_colors = ggplot_default_colors, color_seed = color_seed)
}
colors_use_idents <- colors_use_idents[1:group_by_length]
if (inherits(x = colors_use_idents, what = "colors")) {
  colors_use_idents <- as.vector(colors_use_idents)
}
Identity <- colnames(exp_mat)
identity_colors <- colors_use_idents
names(identity_colors) <- Identity
identity_colors_list <- list(Identity = identity_colors)
if (flip) {
  column_ha <- ComplexHeatmap::rowAnnotation(Identity = Identity, 
    col = identity_colors_list, na_col = "grey", name = "Identity", 
    show_legend = FALSE)
}
if (!flip) {
  column_ha <- ComplexHeatmap::HeatmapAnnotation(Identity = Identity, 
    col = identity_colors_list, na_col = "grey", name = "Identity", 
    show_legend = FALSE)
}
if (is.null(x = exp_color_middle)) {
  exp_color_middle <- (exp_color_min + exp_color_max) / 2 
}
palette_length <- length(colors_use_exp)
palette_middle <- (0 + palette_length)/2 
col_fun = colorRamp2(c(exp_color_min, exp_color_middle, 
  exp_color_max), colors_use_exp[c(1, palette_middle, 
  palette_length)])

if (flip) {
  cell_fun_flip = function(i, j, x, y, w, h, fill) {
    grid.rect(x = x, y = y, width = w, height = h, 
      gp = gpar(col = NA, fill = NA))
    grid.circle(x = x, y = y, r = sqrt(percent_mat[i, 
      j]/100) * unit(2, "mm"), gp = gpar(fill = col_fun(exp_mat[i, 
      j]), col = NA))
  }
}
if (!flip) {
  cell_fun = function(j, i, x, y, w, h, fill) {
    grid.rect(x = x, y = y, width = w, height = h, 
      gp = gpar(col = NA, fill = NA))
    grid.circle(x = x, y = y, r = sqrt(percent_mat[i, 
      j]/100) * unit(2, "mm"), gp = gpar(fill = col_fun(exp_mat[i, 
      j]), col = NA))
  }
}
lgd_list = list(ComplexHeatmap::Legend(at = Identity, title = "Identity", 
  legend_gp = gpar(fill = identity_colors_list[[1]])), 
  ComplexHeatmap::Legend(labels = c(0.25, 0.5, 0.75, 1), 
    title = "Percent Expressing", graphics = list(function(x, 
      y, w, h) grid.circle(x = x, y = y, r = sqrt(0.25) * 
      unit(2, "mm"), gp = gpar(fill = "black")), function(x, 
      y, w, h) grid.circle(x = x, y = y, r = sqrt(0.5) * 
      unit(2, "mm"), gp = gpar(fill = "black")), function(x, 
      y, w, h) grid.circle(x = x, y = y, r = sqrt(0.75) * 
      unit(2, "mm"), gp = gpar(fill = "black")), function(x, 
      y, w, h) grid.circle(x = x, y = y, r = 1 * unit(2, 
      "mm"), gp = grid::gpar(fill = "black")))))
if (is.numeric(x = x_lab_rotate)) {
  x_lab_rotate <- x_lab_rotate
}
if (isTRUE(x = x_lab_rotate)) {
  x_lab_rotate <- 45
}
if (isFALSE(x = x_lab_rotate)) {
  x_lab_rotate <- 0
}
  set.seed(seed = seed)

if (flip) {
  cluster_dot_plot <- ComplexHeatmap::Heatmap(t(exp_mat), 
    heatmap_legend_param = list(title = "Expression"), 
    col = col_fun, rect_gp = gpar(type = "none"), 
    cell_fun = cell_fun_flip, row_names_gp = gpar(fontsize = row_label_size), 
    column_km = k, row_km_repeats = ident_km_repeats, 
    border = "black", left_annotation = column_ha, 
    column_km_repeats = feature_km_repeats, show_parent_dend_line = show_parent_dend_line, 
    column_names_rot = x_lab_rotate)
}
if (!flip) {
  cluster_dot_plot <- ComplexHeatmap::Heatmap(exp_mat, 
    heatmap_legend_param = list(title = "Expression"), 
    col = col_fun, rect_gp = gpar(type = "none"), 
    cell_fun = cell_fun, row_names_gp = gpar(fontsize = row_label_size), 
    row_km = k, row_km_repeats = feature_km_repeats, 
    border = "black", top_annotation = column_ha, 
    column_km_repeats = ident_km_repeats, show_parent_dend_line = show_parent_dend_line, 
    column_names_rot = x_lab_rotate)
}
  
ComplexHeatmap::draw(cluster_dot_plot, annotation_legend_list = lgd_list)
```


## • GSEA
### •• Preparing expression dataset files
```{r}
seu <- readRDS('/Volumes/hqdinh2/Projects/RawData_FromNovoGene/HNCBld_CITESeq_Nov2022/DeepSequencing_2022-12-16/04.SeuratObjects/neutrophils_01252022.RDS')

seu <- seu_pp1(seu, batch_id = 'orig.ident')

DimPlot(seu, group.by = 'RNA_snn_res.0.7', split.by = 'orig.ident', cols = color_clusters)

seu_mini <- seu[,seu$RNA_snn_res.0.7 == 0]
seu_big <- seu[,!seu$RNA_snn_res.0.7 == 0]

big <- as.data.frame(GetAssayData(object = seu_big, slot = "counts"))
big1 <- rowMeans(big)
mini <- as.data.frame(GetAssayData(object = seu_mini, slot = 'counts'))

saveRDS(seu, '/Volumes/hqdinh2/Projects/RawData_FromNovoGene/HNCBld_CITESeq_Nov2022/DeepSequencing_2022-12-16/04.SeuratObjects/neutrophils_01252023.RDS')
```


## • Wigerblad et al healthy neutrophil comparison

```{r}
DefaultAssay(neu) <- 'RNA'

VlnPlot(neu, features = c('S100A12', 'S100A6', 'VIM', 'MME', 'CD10', 'GADPH', 'HMGB1', 'JAK1', 'S100P', 'AIF1', 'TXNIP', 'CXCR2', 'MALAT1', 'NEAT1', 'CSF3R', 'IFIT1', 'HERC5', 'ISG15', 'IFI16', 'MX1', 'MX2', 'IFIT2', 'IFIT3'), group.by = 'wsnn_res.0.2', pt.size = 0)
```

## • Xue et al liver cancer (Zemin Zhang) comparison
```{r}
VlnPlot(neu, features = c('LTF', 'LCN2', 'RTN', 'MMP8', 'CAMP', 'MMP9', 'HMGB2', 'CST7', 'IL17RA', 'ENTPD1', 'IFITM1', 'IFIT1', 'IFIT2', 'IFIT3', 'HERC5', 'TXNIP', 'RIPOR2', 'FKBP5', 'RGS18', 'KCNJ2', 'SAMSN1', 'FRMD4B', 'PRKDC', 'ELL2', 'SYAP1'), group.by = 'wsnn_res.0.2', pt.size = 0)

VlnPlot(neu, features = c('PTGS2', 'PROK2', 'IL1RN', 'NR4A3', 'APOA2', 'ALB', 'RBP4', 'APOA1', 'IGKC', 'CD74', 'RSAD2', 'MX1', 'IFI6', 'HES4', 'SPP1', 'S100A10', 'DYNLL1', 'LGALS3', 'PTMA', 'CCL3', 'CCL4', 'CCL4L4', 'CCL3L1'), group.by = 'wsnn_res.0.2', pt.size = 0)

seu$Gender <- 'Male'
seu$hpv_status <- 'HPV-'

saveRDS(seu, '/Volumes/hqdinh2/Projects/RawData_FromNovoGene/HNCBld_CITESeq_Nov2022/DeepSequencing_2022-12-16/04.SeuratObjects/neutrophils_01252023.RDS')
```

## • Salcher et al (NSCLC) neutrophil comparison
```{r}
# normal associated neutrophils (NANs)
VlnPlot(neu, features = c('PADI4', 'MMP9', 'S100A12', 'CDA', 'RBP7', 'ATG16L2', 'ARHGEF40', 'SLC8A1', 'ALPL', 'HRH2', 'SECTM1', 'GBP1', 'IFIT2', 'GBP5', 'APOL2'), group.by = 'wsnn_res.0.2', pt.size = 0)

# tumor associated neutrophils (TANs)
VlnPlot(neu, features = c('IL1RN', 'CD44', 'RHOH', 'ELOC', 'RIPK2', 'HLA-DRA', 'HLA-DRB1', 'CD74', 'SNRPG', 'HLA-DMB', 'ASAH1', 'CSTB', 'LGALS3', 'GNPDA1', 'PLEKHB2', 'ATP5MC2', 'RPL23', 'RPN2', 'RPL3', 'RPS12'), group.by = 'wsnn_res.0.2', pt.size = 0)
```

# Global clusters/all data

##  • Testing
```{r}
seu.data <- Read10X(data.dir = "/Volumes/hqdinh2/Projects/RawData_FromNovoGene/HNCBld_CITESeq_Nov2022/DeepSequencing_2022-12-16/03.CellRanger_output/HPC_cluster/D70774/outs/filtered_feature_bc_matrix/")
rownames(x = seu.data[["Antibody Capture"]]) <- gsub(pattern = "_[control_]*TotalSeqB", replacement = "",
    x = rownames(x = seu.data[["Antibody Capture"]]))

seu <- CreateSeuratObject(counts = seu.data[["Gene Expression"]], min.cells = 3, min.features = 200)
seu <- NormalizeData(seu)
seu[["FB"]] <- CreateAssayObject(seu.data[["Antibody Capture"]][, colnames(x = seu)])
seu <- NormalizeData(seu, assay = "FB", normalization.method = "CLR")
seu$orig.ident <- 'D70774'

seu1.data <- Read10X(data.dir = "/Volumes/hqdinh2/Projects/RawData_FromNovoGene/HNCBld_CITESeq_Nov2022/DeepSequencing_2022-12-16/03.CellRanger_output/HPC_cluster/D70792/outs/filtered_feature_bc_matrix/")
rownames(x = seu1.data[["Antibody Capture"]]) <- gsub(pattern = "_[control_]*TotalSeqB", replacement = "",
    x = rownames(x = seu1.data[["Antibody Capture"]]))

seu1 <- CreateSeuratObject(counts = seu1.data[["Gene Expression"]], min.cells = 3, min.features = 200)
seu1 <- NormalizeData(seu1)
seu1[["FB"]] <- CreateAssayObject(seu1.data[["Antibody Capture"]][, colnames(x = seu1)])
seu1 <- NormalizeData(seu1, assay = "FB", normalization.method = "CLR")
seu1$orig.ident <- 'D70792'

seu <- merge(seu, seu1)

# RNA
seu@active.assay <- "RNA"
seu <- NormalizeData(seu)
seu <- FindVariableFeatures(seu)
seu <- ScaleData(seu)
seu <- RunPCA(seu, assay.use = "RNA", reduction.name = "rnapca")
seu <- harmony::RunHarmony(seu, assay.use = "RNA", group.by.vars = "orig.ident", reduction = "rnapca", reduction.save = "harmonyrna")
# FB
seu@active.assay <- "FB"
seu <- NormalizeData(seu, normalization.method = "CLR", margin = 2)
VariableFeatures(seu, assay = "FB") <- grep("Iso", rownames(seu), value = TRUE, invert = TRUE)
seu <- ScaleData(seu)
seu <- RunPCA(seu, assay = "FB", reduction.name = "fbpca")
seu <- harmony::RunHarmony(seu, assay.use = "FB", group.by.vars = "orig.ident", reduction = "fbpca",reduction.save = "harmonyfb")

DimPlot(seu, reduction = "fbpca", group.by = "orig.ident")
ElbowPlot(seu, reduction = "fbpca", ndims = 50) # 20
ElbowPlot(seu, reduction = "rnapca", ndims = 50) # 20
seu <- FindMultiModalNeighbors(seu, reduction.list = list("harmonyrna", "harmonyfb"), dims.list = list(1:20, 1:20),
                                 modality.weight.name = "RNA.weight")
seu <- RunUMAP(seu, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
seu <- FindClusters(seu, graph.name = "wsnn", algorithm = 3, resolution = seq(0.1, 1, 0.1), verbose = TRUE)
DimPlot(seu, group.by = c("orig.ident"), reduction = "wnn.umap", ncol = 2)

plot_tib<-data_frame(RNA_Count=colSums(seu@assays$RNA),
                     FB_Count=colSums(seu@assays$FB))

ggplot(data=plot_tib,aes(x=RNA_Count,y=FB_Count),cols=c('red','blue','green','yellow','violet'))+geom_point()

saveRDS(seu, file = paste('/Volumes/hqdinh2/Projects/RawData_FromNovoGene/HNCBld_CITESeq_Nov2022/DeepSequencing_2022-12-16/04.SeuratObjects/', Sys.Date(), '_combined.RDS', sep = ''))
```


## • Load object

```{r}
seu <- readRDS('/Volumes/hqdinh2/Projects/RawData_FromNovoGene/HNCBld_CITESeq_Nov2022/DeepSequencing_2022-12-16/04.SeuratObjects/D70774_D70792_02232023.RDS')
```






## • Pre-processing merged object (Josh method)
```{r}
seu <- readRDS('/Volumes/hqdinh2/Projects/RawData_FromNovoGene/HNCBld_CITESeq_Nov2022/DeepSequencing_2022-12-16/04.SeuratObjects/D70774_D70792_02232023.RDS')

# RNA
seu@active.assay <- "RNA"
seu <- NormalizeData(seu)
seu <- FindVariableFeatures(seu)
seu <- ScaleData(seu)
seu <- RunPCA(seu, assay.use = "RNA", reduction.name = "rnapca")
seu <- harmony::RunHarmony(seu, assay.use = "RNA", group.by.vars = "orig.ident", reduction = "rnapca", reduction.save = "harmonyrna")
# FB
seu@active.assay <- "FB"
seu <- NormalizeData(seu, normalization.method = "CLR", margin = 2)
VariableFeatures(seu, assay = "FB") <- grep("Iso", rownames(seu), value = TRUE, invert = TRUE)
seu <- ScaleData(seu)
seu <- RunPCA(seu, assay = "FB", reduction.name = "fbpca")
seu <- harmony::RunHarmony(seu, assay.use = "FB", group.by.vars = "orig.ident", reduction = "fbpca",reduction.save = "harmonyfb")

DimPlot(seu, reduction = "fbpca", group.by = "orig.ident")
ElbowPlot(seu, reduction = "fbpca", ndims = 50) # 20
ElbowPlot(seu, reduction = "rnapca", ndims = 50) # 20
seu <- FindMultiModalNeighbors(seu, reduction.list = list("harmonyrna", "harmonyfb"), dims.list = list(1:20, 1:20),
                                 modality.weight.name = "RNA.weight")
seu <- RunUMAP(seu, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
seu <- FindClusters(seu, graph.name = "wsnn", algorithm = 3, resolution = seq(0.1, 1, 0.1), verbose = TRUE)
DimPlot(seu, group.by = c("orig.ident"), reduction = "wnn.umap", ncol = 2)

#saveRDS(seu, file = '/Volumes/hqdinh2/Projects/RawData_FromNovoGene/HNCBld_CITESeq_Nov2022/DeepSequencing_2022-12-16/04.SeuratObjects/D70774_D70792_02272023.RDS')
```

# __________MOC2 MOUSE__________

# Plotting VENN DIAGRAM markers in Luoma

## • Plotting k17ko specific markers in RESPONDERS

```{r}
load('/Volumes/hqdinh2/Lab Manuscripts/HNC_Wangetal2021/CellphoneDB_heatmap/10282021_1neutrophil/Final/CellphoneDB_heatmap.RData')
#load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/Seurat_MOC2_K17KO_2021-11-01**.rda')

k17lr <- unique(k17ko.all$lr)
moc2lr <- unique(moc2.all$lr)

intersect(k17lr, moc2lr)
setdiff(k17lr, moc2lr)
setdiff(moc2lr, k17lr)

luoma <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/luoma_tumor_myeloid_T_plusmetadata.RDS')

luoma_mini <- luoma[,luoma$Path_response %in% c('Medium', 'High')]

obj <- luoma_mini
name <- 'luoma'
ligands <- sapply(strsplit(k17lr," "), `[`, 1)
receptors <- sapply(strsplit(k17lr, ' '), `[`, 2)

df <- data.frame()

assay <- as.data.frame(obj@assays$RNA@data) 

lr <- data.frame(ligands, receptors)

#making an object that will hold the running total of the sum of all spots
total_score <- 0
for (x in 1:nrow(lr)){
  total_score <- 0
  # for one spot
  for (y in 1:ncol(assay)){
    # get ligand/receptor names
    lig <- lr[x,1]
    rec <- lr[x,2]
    #get a dataframe of just that single spot with gene names
    assay_mini <- data.frame(assay[,y])
    rownames(assay_mini) <- rownames(assay)
    
    #getting the ligand receptor score
    lrexp <- assay_mini[lr[x,1],] * assay_mini[lr[x,2],]
    
    # scaling the ligand receptor score based on the percentage of immune cells in the spot
    lr_scaled <- lrexp
    
    #preventing the value from becoming NA in the case that it's zero
    if(is.na(lr_scaled)){
      lr_scaled <- 0
    }
    
    #adding the score to the total score
    total_score <- total_score + lr_scaled
    
    #jump back to the top to do the rest of the spots
  }
    #now dividing the total score by the number of spots in the downsized object
    total_score_norm <- total_score / as.numeric(ncol(obj))
    
    # adding the row to the dataframe
    rname <- paste(lig, rec, sep = '_')
    row <- c(rname, total_score_norm)
    df <- rbind(df, row)

}

colnames(df) <- c('lr_pair', 'responder_score')
df[is.na(df)] <- 0
df$ck17_5_score <- as.numeric(df$responder_score)

r1 <- df

```

```{r}

```

# Plotting markers in mouse subsets

```{r}
seu <- seurat
metadata <- 'global.cluster2'
feature <- 'Sdc1'
split_by <- 'tissue'
out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/zzz_Miscellaneous/'

marker_plotting <- function(seu, metadata, feature, split_by, out){
  a <- DimPlot(seu, group.by = metadata)
  b <- FeaturePlot(seu, features = feature, split.by = split_by)
  c <- VlnPlot(seu, group.by = metadata, features = feature, split.by = split_by)
  d <- DotPlot(seu, group.by = metadata, features = feature, split.by = split_by, col.min = 0)

  png(paste(out, feature, '_marker_plotting.png', sep = ''), width = 20, height = 12, units = 'in', res = 300)
  cowplot::plot_grid(a, b, c, d)
  dev.off()
}


```

# Plotting SDC1 in mouse subsets

```{r}
load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/Seurat_MOC2_K17KO_2021-11-01**.rda')

DimPlot(seurat, group.by = 'global.cluster')
VlnPlot(seurat, group.by = 'global.cluster', features = 'Sdc1', split.by = 'tissue')
DotPlot(seurat, group.by = 'global.cluster', features = 'Sdc1', split.by = 'tissue')
FeaturePlot(seurat, features = 'Sdc1', split.by = 'tissue')
```

## Manually evaluating SDC1 expression

```{r}
x <- GetAssayData(seurat)
x <- x[which(rownames(x) == 'Sdc1'),]
x <- as.data.frame(x)

# getting metadata from the object for all cells
y <- seurat[[c("global.cluster", 'tissue')]]
y$cluster_tissue <- paste(y$global.cluster, y$tissue)
y$global.cluster2 <- NULL
y$tissue <- NULL

# merging both the expression and metadata objects
z <- merge(x, y, by = 0)
z$x <- as.numeric(z$x)
z$cluster_tissue <- as.factor(z$cluster_tissue)

z1 <- aggregate(z$x, list(z$cluster_tissue), FUN=mean) 
z1$x <- as.factor(z1$x)

z$x <- as.numeric(z$x)
z$x <- scale(z$x)

p<-ggplot(z, aes(x=cluster_tissue, y=x)) + geom_dotplot(binaxis='y', stackdir='center', stackratio=1.5, dotsize=0.3, binwidth = 0.1) + stat_summary(fun.y=mean, geom="point", shape=18, size=3, color="red") + geom_boxplot(notch = TRUE) + stat_summary(fun.data=mean_sdl, fun.args = list(mult=1),geom="pointrange", color="red") + theme_classic() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
p
```

# Plotting Xue et al markers (2022) in mouse neutrophils

```{r}
load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/Seurat_MOC2_K17KO_2021-11-01**.rda')

seu <- seurat[,seurat$global.cluster2 %like% c('^N\\d+')]
DimPlot(seu, group.by = 'global.cluster2')

# Ltf Neutrophils
fs <- stringr::str_to_title(c('MMP8', 'BMX', 'TSPO', 'S100A6', 'PICALM', 'CXCR2', 'IRF2', 'SH3GLB1', 'CCNDBP1', 'FBXL5', 'IL17RA'))
DotPlot(seu, group.by = 'global.cluster2', features = fs, col.min = 0) + coord_flip()

# Ngp Neutrophils
fs <- stringr::str_to_title(c('MMP8', 'MMP9', 'LCN2', 'BMX', 'TSPO', 'S1PR4', 'S100A6'))
DotPlot(seu, group.by = 'global.cluster2', features = fs, col.min = 0) + coord_flip()

# Gm2a/Pabpc1 Neutrophils
fs <- stringr::str_to_title(c('PICALM', 'CXCR2', 'IRF2', 'TSC22D3', 'SH3GLB1', 'CCNDBP1', 'FBXL5', 'IL17RA'))
DotPlot(seu, group.by = 'global.cluster2', features = fs, col.min = 0) + coord_flip()

# Ifit3 Neutrophils
fs <- stringr::str_to_title(c('PICALM', 'CXCR2', 'IRF2', 'TSC22D3', 'SH3GLB1', 'CCNDBP1', 'FBXL5', 'IL17RA', 'PSMB9', 'STAT1', 'STAT2', 'IFIT2'))
DotPlot(seu, group.by = 'global.cluster2', features = fs, col.min = 0) + coord_flip()

# Mmp8 Neutrophils
fs <- stringr::str_to_title(c('MMP8', 'MMP9', 'LCN2', 'S1PR4', 'CYBB', 'CAMP', 'ELL2'))
DotPlot(seu, group.by = 'global.cluster2', features = fs, col.min = 0) + coord_flip()

# Actg1 Neutrophils
fs <- stringr::str_to_title(c('C5AR1', 'ICAM1', 'SLC3A2'))
DotPlot(seu, group.by = 'global.cluster2', features = fs, col.min = 0) + coord_flip()

# Marco Neutrophils
fs <- stringr::str_to_title(c('ICAM1', 'G0S2'))
DotPlot(seu, group.by = 'global.cluster2', features = fs, col.min = 0) + coord_flip()

# Ifit1 Neutrophils
fs <- stringr::str_to_title(c('PSMB9', 'STAT1', 'STAT2', 'IFIT2', 'C5AR1', 'ICAM1', 'G0S2', 'IL1A', 'IL1RN', 'CCL3', 'CCL4', 'CD74', 'IRAK2', 'CD69', 'CKS2', 'APOA1'))
DotPlot(seu, group.by = 'global.cluster2', features = fs, col.min = 0) + coord_flip()

# Apoa2 Neutrophils
fs <- stringr::str_to_title(c('APOA2', 'FABP1', 'TTR', 'LGALS3'))
DotPlot(seu, group.by = 'global.cluster2', features = fs, col.min = 0) + coord_flip()

# Spp1 Neutrophils
fs <- stringr::str_to_title(c('CCL4', 'CD74', 'IRAK2', 'CD69', 'LGALS3', 'LDHA', 'DYNLL1', 'SLC3A2', 'LTF', 'SPP1', 'OST4'))
DotPlot(seu, group.by = 'global.cluster2', features = fs, col.min = 0) + coord_flip()

# Ccl4 Neutrophils
fs <- stringr::str_to_title(c('C5AR1', 'ICAM1', 'IL1A', 'IL1RN', 'CCL3', 'CCL4', 'CD74', 'IRAK2', 'CD69', 'CKS2', 'APOA1', 'LDHA', 'DYNLL1', 'SLC3A2', 'LTF', 'SPP1', 'OST4'))
DotPlot(seu, group.by = 'global.cluster2', features = fs, col.min = 0) + coord_flip()

```

# PseudotimeDE myeloid trajectory

```{r}
suppressPackageStartupMessages(library(PseudotimeDE))
suppressPackageStartupMessages(library(SingleCellExperiment))
suppressPackageStartupMessages(library(slingshot))
suppressPackageStartupMessages(library(tibble))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(scales))
suppressPackageStartupMessages(library(irlba))

#myeloid <- seurat[,seurat$global.cluster2 %like any% c('%Mon%', '%Mac%', '%DC%')]
#moc2 <- myeloid[,myeloid$tissue == 'MOC2']
#test <- monocle3::learn_graph(moc2, use_partition = TRUE, verbose = FALSE)

data(LPS_sce, package = "PseudotimeDE")

# this is the first two principal components
rd <- irlba::prcomp_irlba(t(logcounts(LPS_sce)), scale. = FALSE)$x[, 1:2]

#putting new PC information in the single cell object
reducedDims(LPS_sce) <- SimpleList(PCA = rd)

colData(LPS_sce)$cl <- 1


fit_ori <- slingshot(LPS_sce, reducedDim = 'PCA', clusterLabels = "cl")

LPS_ori_tbl <- tibble(cell = colnames(LPS_sce), pseudotime = rescale(colData(fit_ori)$slingPseudotime_1))

set.seed(123)
## Set the cores for parallelization. Note that mclapply doesnot work on Windows.
options(mc.cores = 2)

## Number of subsmaples
n = 2

## Ganerate random subsamples
LPS_index <- parallel::mclapply(seq_len(n), function(x) {
  sample(x = c(1:dim(LPS_sce)[2]), size = 0.8*dim(LPS_sce)[2], replace = FALSE)
})

LPS_sub_tbl <- parallel::mclapply(LPS_index, function(x, sce) {
  sce <- sce[, x]
  rd <- irlba::prcomp_irlba(t(logcounts(sce)), scale. = FALSE)$x[, 1:2]
  reducedDims(sce) <- SimpleList(PCA = rd)

  fit <- slingshot(sce, reducedDim = 'PCA', clusterLabels = "cl")
  tbl <- tibble(cell = colnames(sce), pseudotime = rescale(colData(fit)$slingPseudotime_1))
  
  ## Make sure the direction of pseudotime is the same as the original pseudotime
  merge.tbl <- left_join(tbl, LPS_ori_tbl, by = "cell")
  
  if(cor(merge.tbl$pseudotime.x, merge.tbl$pseudotime.y) < 0) {
    tbl <- dplyr::mutate(tbl, pseudotime = 1-pseudotime)
  }
  tbl
}, sce = LPS_sce)
```

# LR-Heatmap

## Mouse interactions in mice

```{r}
# ligands in senders--from mouse in mouse

seurat$meta <- paste(seurat$global.cluster2, seurat$tissue)
l_avg <- AverageExpression(seurat, features = str_to_title(c('PDCD1LG2', 'LGALS9', 'FN1', 'CXCL9', 'CXCL10', 'CCL7', 'CCL6', 'SPP1')), group.by = 'meta')
l_avg <- as.data.frame(l_avg$RNA)
l_avg <- l_avg[,colnames(l_avg) %like any% c('Mac%', 'Mon%', '%DC%')]
l_avg <- l_avg/do.call(pmax, l_avg)

#reordering the column names
col_order <- c(3,4,5,6,9,10,11,12,13,14,15,16,17,18,21,22,23,24, 1,2,7,8,19,20)
l_avg <- l_avg[, col_order]

png('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/cellchat_l_expression_frommouse_inmouse.png', units = 'in', width = 10, height = 6, res = 300)
ComplexHeatmap::Heatmap(l_avg, cluster_rows = F, cluster_columns = F)
dev.off()


# receptors in receivers--from mouse in mouse

seurat$meta <- paste(seurat$global.cluster2, seurat$tissue)
r_avg <- AverageExpression(seurat, features = str_to_title(c('PDCD1', 'PTPRC', 'CD44', 'CXCR3', 'CXCR3', 'CCR2', 'CCR2', 'CD44')), group.by = 'meta')
r_avg <- as.data.frame(r_avg$RNA)
r_avg <- r_avg[,colnames(r_avg) %like any% c('CD8%', 'CD4%', 'Treg%', 'NK%', '^B%')]

r_avg <- rbind(r_avg[1,], r_avg[2,], r_avg[3,], r_avg[4,], r_avg[4,], r_avg[5,], r_avg[5,], r_avg[3,])
r_avg <- r_avg/do.call(pmax, r_avg)
col_order <- c(3,4,5,6,9,10,1,2,7,8)
r_avg <- r_avg[,col_order]

png('/Volumes/hqdinh2/Projects/HNC_SPORE/Golfinosetal2022/cellchat_r_expression_frommouse_inmouse.png', units = 'in', width = 4, height = 4, res = 300)
ComplexHeatmap::Heatmap(r_avg, cluster_rows = F, cluster_columns = F)
dev.off()
```

# CellChat

## MERAD CELL-CELL COMMUNICATION HEATMAP

```{r}
setwd("/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/MOC2vsK17KO/Myeloid_to_T/")
list.files()

# cellchat object
cc <- readRDS("MOC2_cellchat.Rds")

# your interactions scores are held here: x axis are senders, y axis are receivers
cc@net$prob

# the names of the interactions:
ints <- cc@LR$LRsig[,1]

# extracting one interaction:
cc@net$prob[,,ints[1]] # array slices are [sender, receiver, interaction]

all_interactions <- list()
for (int in ints){
  message(int)
  this_int <- cc@net$prob[,,int] %>% as.data.frame()
  
  this_df <- this_int %>% 
                    rownames_to_column(var = "sender") %>%
                    reshape2::melt() %>%
                    mutate(int)
 
  colnames(this_df) <- c("sender", "receiver", "score", "interaction")
   
  all_interactions[[int]] <- this_df
}

# collapsing to single dataframe:
data1 <- do.call(rbind, all_interactions)

r <- data1
nr <- data1

# here I would merge the two object (one form HPV+, HPV-)
# merge(data1, data2, by = c("sender", "receiver", "interaction")
# you'll end up with two score columns and then can plot them with ggplot

#to_plot <- c('PDCD1LG2_PDCD1%', 'LGALS9_CD45%', 'FN1_ITGAV_ITGB8%', 'FN1_ITGAV_ITGB3%', 'CXCL9_CXCR3%', 'CD274_PDCD1%', 'CCL7_CCR2%')
to_plot <- c('CD274_PDCD1%', 'LGALS9_CD45%', 'CXCL9_CXCR3%', 'CCL7_CCR2%')
n1 <- nr[(nr$interaction %like any% to_plot & nr$sender %in% c('DC_Ccr7', 'Mac_Cxcl9', 'Mon_Intermediate') & nr$receiver %like any% c('CD8%', 'CD4%', 'Treg%')),]
n1$group <- 'MOC2'

r1 <- r[(r$interaction %like any% to_plot & r$sender %in% c('DC_Ccr7', 'Mac_Cxcl9', 'Mon_Intermediate') & r$receiver %like any% c('CD8%', 'CD4%', 'Treg%')),]
r1$group <- 'K17KOMOC2'

all <- rbind(n1, r1)
all$srlr <- paste(all$sender, all$receiver, all$interaction)
all1 <- all %>% separate(interaction, c('ligand', 'receptor'), sep = '_')
all1 <- all1 %>% arrange(srlr)
all1$lg <- paste(all1$ligand, all1$group, sep = '_')
all1$rg <- paste(all1$receptor, all1$group, sep = '_')


# making ligand moc2 heatmap

all_l1 <- reshape(all1, idvar = "sender", timevar = "lg", direction = "wide", drop = c('receiver', 'ligand', 'receptor', 'group', 'srlr', 'rg'))
rownames(all_l1) <- all_l1$sender
all_l1$sender <- NULL
all_l1 <- t(all_l1)
png("/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/MOC2vsK17KO/Myeloid_to_T/srlr_score_heatmap.png", units = 'in', height  = 6, width = 7, res = 300)
ComplexHeatmap::Heatmap(all_l1, cluster_rows = F, cluster_columns = F, heatmap_legend_param = list(title = 'SRLR score'))
dev.off()


all_l2 <- reshape(all1, idvar = "receiver", timevar = "rg", direction = "wide", drop = c('sender', 'ligand', 'receptor', 'group', 'srlr', 'lg'))
rownames(all_l2) <- all_l2$receiver
all_l2$receiver <- NULL
all_l2 <- t(all_l2)
png("/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/MOC2vsK17KO/Myeloid_to_T/srlr_score_heatmap_receptors.png", units = 'in', height  = 6, width = 7, res = 300)
ComplexHeatmap::Heatmap(all_l2, cluster_rows = F, cluster_columns = F, heatmap_legend_param = list(title = 'SRLR score'))
dev.off()
```

## FUNCTION: Running CellChat (split by tissue)

```{r}
sender <- c('%DC%', 'Mon%', 'Mac%')
#sender <- c('^N\\d+%')
#receiver <- unique(seurat$global.cluster2)[!unique(seurat$global.cluster2) %like% sender]
receiver <- c('^N\\d+%')
metadata = 'global.cluster2'
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/MOC2vsK17KO/Myeloid_to_neutrophil/'


library(Seurat)
library(CellChat)
hnc <- load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/Seurat_MOC2_K17KO_2021-11-01**.rda')
hnc <- seurat

hnc_til_mono <- hnc[,hnc$global.cluster2 %like% c(receiver, sender)] 
hnc <- NULL

ccs <- list()
for (x in unique(hnc_til_mono$tissue)){
  input_seurat <- hnc_til_mono[,hnc_til_mono$tissue == x]
  data.input<-GetAssayData(input_seurat,assay = "RNA", slot = "data")
  labels<-input_seurat$global.cluster2   #extract cell type labels
  meta <- data.frame(group = labels, row.names = names(labels))
  cellchat <- createCellChat(object = data.input, meta = meta, group.by = "group")
  cellchat@DB <- CellChatDB.mouse
  cellchat <- subsetData(cellchat)
  cellchat <- identifyOverExpressedGenes(cellchat,thresh.p = 0.1)
  cellchat <- identifyOverExpressedInteractions(cellchat)
  cellchat <- computeCommunProb(cellchat)
  cellchat <- computeCommunProbPathway(cellchat)
  cellchat <- filterCommunication(cellchat, min.cells = 5)
  cellchat <- aggregateNet(cellchat)
  groupSize <- as.numeric(table(cellchat@idents))
  par(mfrow = c(1,2), xpd=TRUE)
  mat <- cellchat@net$weight
  ccs[x] <- cellchat
}

hpvp <- ccs[['MOC2']]
hpvn <- ccs[['K17KOMOC2']]
saveRDS(hpvp, paste(out, 'MOC2_cellchat.RDS', sep = ''))
saveRDS(hpvn, paste(out, 'K17KOMOC2_cellchat.RDS', sep = ''))

senders <- unique(input_seurat[[metadata]][[1]])[unique(input_seurat[[metadata]][[1]]) %like% sender]
receivers <- unique(input_seurat[[metadata]][[1]])[unique(input_seurat[[metadata]][[1]]) %like% receiver]

#pathways.show <- c("CXCL") 
pathways.show <- c('CXCL', 'CD80', 'CCL', "CD45", "FN1", "THBS", "ICAM")
pdf(paste(out, 'signaling.pdf', sep = ''), width = 12, height = 6)

for (y in pathways.show){
  p1 <- netVisual_aggregate(hpvp, signaling = y, layout = 'circle', sources.use = senders, targets.use = receivers, signaling.name = paste('MOC2', y, sep = ' '))
  netAnalysis_contribution(hpvp, signaling = y, signaling.name = paste('MOC2', y, sep = ' '))
  p3 <- netVisual_aggregate(hpvn, signaling = y, layout = 'circle', sources.use = senders, targets.use = receivers, signaling.name = paste('K17KOMOC2', y, sep = ' '))
  netAnalysis_contribution(hpvn, signaling = y, signaling.name = paste('MOC2', y, sep = ' '))
}  
dev.off()
```

## Heatmap of differential \# of interactions

```{r}
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/MOC2vsK17KO/Myeloid_to_neutrophil/'
moc2 <- readRDS(file = paste(out, 'MOC2_cellchat.RDS', sep = ''))
k17ko <- readRDS(file = paste(out, 'K17KOMOC2_cellchat.RDS', sep = ''))
moc2 <- netAnalysis_computeCentrality(moc2)
k17ko <- netAnalysis_computeCentrality(k17ko)
object.list <- list(MOC2 =  moc2, K17KOMOC2 = k17ko)
cellchat <- mergeCellChat(object.list, add.names = names(object.list))

s <- as.character(unique(cellchat@idents[[1]]))
s1 <- s[s %like any% c('%DC%', 'Mon%', 'Mac%')]
r <- s[s %like any% c('^N\\d+%')]
gg2 <- netVisual_heatmap(cellchat, measure = "count", sources.use = s1, targets.use = r, title.name = paste('Differential number of interactions (', names(cellchat@idents[2]), ' (red)', ' vs ', names(cellchat@idents[1]), ' (blue)', ')', sep = ''))

png(paste(out, 'heatmap.png', sep = ''), width = 5.5, height = 4.5, units = 'in', res = 300)
gg2
```

## Plotting markers (DC_Ccr7 & Mac_Cxcl9)

```{r}
p1 <- VlnPlot(seurat, group.by = 'global.cluster2', split.by = 'tissue', idents = c('Mac_Cxcl9', 'DC_Ccr7', 'Mac_Fn1', 'Mac_Trem2', 'Mon_Ccr5', 'Mon_Classical', 'Mon_Intermediate', "Mon_NonClassical"), features = c('Ccl6'))

p2 <- VlnPlot(seurat, group.by = 'global.cluster2', split.by = 'tissue', idents = c('Mac_Cxcl9', 'DC_Ccr7', 'Mac_Fn1', 'Mac_Trem2', 'Mon_Ccr5', 'Mon_Classical', 'Mon_Intermediate', "Mon_NonClassical"), features = c('Ccl7'))

p1 + p2

VlnPlot(seurat, group.by = 'global.cluster2', split.by = 'tissue', idents = c('CD4', 'CD8', 'Treg'), features = c('Ccr2'))

p1 <- VlnPlot(seurat, group.by = 'global.cluster2', split.by = 'tissue', idents = c('Mon_Intermediate', 'Mon_Classical', 'Mon_NonClassical'), features = c('Ccl7'))

p2 <- VlnPlot(seurat, group.by = 'global.cluster2', split.by = 'tissue', idents = c('Mon_Intermediate', 'Mon_Classical', 'Mon_NonClassical'), features = c('Ccl7'))

p1 + p2
```

```{r}
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/MOC2vsK17KO/Neut_to_all/'
moc2 <- readRDS(file = paste(out, 'MOC2_cellchat.RDS', sep = ''))
k17ko <- readRDS(file = paste(out, 'K17KOMOC2_cellchat.RDS', sep = ''))
moc2 <- netAnalysis_computeCentrality(moc2)
k17ko <- netAnalysis_computeCentrality(k17ko)
object.list <- list(MOC2 =  moc2, K17KOMOC2 = k17ko)
cellchat <- mergeCellChat(object.list, add.names = names(object.list))
cellchat
```

## Plotting single signaling pathways (Circos)

```{r}
path <- '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/MOC2vsK17KO/Myeloid_to_neutrophil/'
s <- as.character(unique(k17ko@idents))
s1 <- s[s %like any% c('%DC%', 'Mon%', 'Mac%')]
r <- s[s %like any% c('^N\\d+%')]
pathways.show <- c("CXCL", 'CCL', 'BST2', 'ANNEXIN', 'TGFb', 'ITGAL-ITGB2', 'THBS', 'MIF', 'SELPLG', 'CD45', 'LIGHT', 'SPP1')
objname <- 'moc2'
obj <- moc2

for (pathway in pathways.show){
  png(paste(path, objname, pathway, 'signaling.png', sep = ''), width = 8.5, height = 8.5, units = 'in', res = 300)
  netVisual_aggregate(obj, signaling = pathway, layout = "circle", sources.use = s1, targets.use = r)
  dev.off()
  
  netAnalysis_contribution(obj, signaling = pathway)
}

```

## Compute different \# of interactions

```{r}
gg1 <- compareInteractions(cellchat, show.legend = F, group = c(1,2))
gg2 <- compareInteractions(cellchat, show.legend = F, group = c(1,2), measure = "weight")

png(paste(out, 'Overall_interaction_num_strength.png', sep = ''), width = 8, height = 5, units = 'in', res = 300)
gg1 + gg2
dev.off()
```

## Differential \# of interactions/interaction strength in HPV+ vs. HPV-

```{r}
group.cellType <- c(rep("CD4", 1), rep("CD8", 1), rep("DC", 3), rep("Mac", 4), rep("Mono", 4), rep("DC", 1), rep("Treg", 1))
group.cellType <- factor(group.cellType, levels = c('CD8', 'CD4', 'Treg', 'DC', 'Mono', 'Mac'))
object.list <- lapply(object.list, function(x) {mergeInteractions(x, group.cellType)})
cellchat_test <- mergeCellChat(object.list, add.names = names(object.list))

weight.max <- getMaxWeight(object.list, slot.name = c("idents", "net", "net"), attribute = c("idents","count", "count.merged"))

png(paste(out, 'Global_#interaction.png', sep = ''), width = 8, height = 5, units = 'in', res = 300)
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_circle(object.list[[i]]@net$count.merged, weight.scale = T, label.edge= T, edge.weight.max = weight.max[3], edge.width.max = 12, title.name = paste0("Number of interactions - ", names(object.list)[i]), sources.use = c('Mac', 'Mono', 'DC'), targets.use = c('CD8', 'CD4', 'Treg'))
}
dev.off()

png(paste(out, '#signaling_interactions.png', sep = ''), width = 8, height = 5, units = 'in', res = 300)
netVisual_heatmap(cellchat, measure = 'count', cluster.rows = F, cluster.cols = F, sources.use = s, targets.use = c('CD8', 'CD4', 'Treg'))
dev.off()
```

## Signaling changes of particular subsets

```{r}
pdf(paste(out, 'Signaling_changes_per_subset.pdf'))
for (x in unique(cellchat@idents$K17KOMOC2)){
  gg <- netAnalysis_signalingChanges_scatter(cellchat, idents.use = x)
  plot(gg)
}
dev.off()
```

## Information flow + bubble plot

```{r}
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/MOC2vsK17KO/Myeloid_to_neutrophil/'
s <- as.character(unique(k17ko@idents))
s1 <- s[s %like any% c('%DC%', 'Mon%', 'Mac%')]
r <- s[s %like any% c('^N\\d+%')]

gg1 <- rankNet(cellchat, mode = "comparison", stacked = T, do.stat = TRUE, sources.use = s1, targets.use = r)
gg2 <- rankNet(cellchat, mode = "comparison", stacked = F, do.stat = TRUE, sources.use = s1, targets.use = r)

png(paste(out, 'Relative_information_flow.png', sep = ''), width = 5, height = 8, units = 'in', res = 300)
gg1
dev.off()

pdf(paste(out, 'Bubble_plots_signaling_diffs.pdf', sep = ''), height = 5, width = 5)
for (x in s1){
  gg <- netVisual_bubble(cellchat, sources.use = x, targets.use = r,  comparison = c(1, 2), angle.x = 45)
  plot(gg)
}
dev.off()
```

## Identify dysfunctional signaling by using differential expression analysis

```{r}
# define a positive dataset, (dataset w/ positive FC against the other)
pos.dataset = "K17KOMOC2"

# define a name used for storing the results of DE analysis
features.name = pos.dataset

# perform differential expression analysis
cellchat <- identifyOverExpressedGenes(cellchat, group.dataset = "datasets", pos.dataset = pos.dataset, features.name = features.name, only.pos = FALSE, thresh.pc = 0.1, thresh.fc = 0.1, thresh.p = 1)

#> Use the joint cell labels from the merged CellChat object
# map the results of DE analysis onto the inferred cell-cell communications to manage/subset the LR pairs of interest
net <- netMappingDEG(cellchat, features.name = features.name)

# extract the ligand-receptor pairs with upregulated ligands in HPVpos
net.up <- subsetCommunication(cellchat, net = net, datasets = "K17KOMOC2",ligand.logFC = 0.2, receptor.logFC = NULL)

# extract the ligand-receptor pairs with upregulated ligands and upregulated recetptors in NL, i.e.,downregulated in LS
net.down <- subsetCommunication(cellchat, net = net, datasets = "MOC2",ligand.logFC = -0.1, receptor.logFC = -0.1)

gene.up <- extractGeneSubsetFromPair(net.up, cellchat)
gene.down <- extractGeneSubsetFromPair(net.down, cellchat)

# Chord diagram
pdf(paste(out, 'Chord_diagram_netup_netdown.pdf', sep = ''), width = 12)
for (x in s){
par(mfrow = c(1,2), xpd=TRUE)
netVisual_chord_gene(object.list[[2]], sources.use = x, targets.use = c('CD8', 'CD4', 'Treg'), slot.name = 'net', net = net.up, lab.cex = 0.8, small.gap = 3.5, title.name = paste0("Up-regulated signaling in ", names(object.list)[2]))
netVisual_chord_gene(object.list[[1]], sources.use = x, targets.use = c('CD8', 'CD4', 'Treg'), slot.name = 'net', net = net.down, lab.cex = 0.8, small.gap = 3.5, title.name = paste0("Down-regulated signaling in ", names(object.list)[2]))
}
dev.off()
```

# _______________LUOMA______________

# LUOMA figure prep --BOTH

# • DimPlot

```{r}
pbmc <- readRDS(file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/luoma_PBMC_myeloid_plusmetadata_fivesubsets.RDS')

color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3", "#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", "#AA8282", "#D4B7B7", "#8600BF", "#BA5CE3", "#808000","#AEAE5C", "#1E90FF", "#00BFFF", "#56FF0D", "#FFFF00")

DimPlot(pbmc, group.by = 'predicted.id', cols = color_clusters, label = T, label.box = T)
ggsave(file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/luoma_PBMC_myeloid_plusmetadata_fivesubsets.png', width = 7, height = 5, units = 'in', dpi = 300)
```

## • Freq comparison: only patients w/ pre-tx biopsies

### •• PBMC

#### ••• Out of all myeloid

```{r}
pbmc <- readRDS(file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/luoma_PBMC_myeloid_plusmetadata_fivesubsets.RDS')

pbmc_mini <- pbmc[,pbmc$Patient_ID %in% c('P18', 'P23', 'P24', 'P27', 'P29', 'P32')]

table(pbmc_mini$Patient_ID)
# P18  P23  P24  P32 
# 551 2105 2184 4017 

# input
SEU <- pbmc_mini[,pbmc_mini$predicted.id %like any% c('%Mac%', '%Mon%', '%DC%')]
metadata <- 'predicted.id'

SEU$orig.ident <- paste(SEU$Patient_ID, SEU$Stage)
SEU$meta <- paste(SEU$Patient_ID, SEU$Stage)
SEU$Path_response[SEU$Path_response %in% c('Medium', 'High')] <- 'Medium/High'
SEU$treatment_path_response <- paste(SEU$Stage, SEU$Path_response)

t <- get_freqs(SEU, metadata)
t2 <- t
for (i in 1:nrow(t2)) {
  t2[i,] = t2[i,]/sum(t2[i,])
}
t2 <- as.matrix(t2)

meta <- SEU@meta.data[,c('meta', 'treatment_path_response')]
rownames(meta) <- NULL
meta <- unique(meta)
df <- NULL
for (i in 1:nrow(t2)) {
  for (j in 1:ncol(t2)) {
    df <- rbind(df, c(rownames(t2)[i], meta$treatment_path_response[which(meta$meta == rownames(t2)[i])], colnames(t2)[j], as.numeric(t2[i,j])))
  }
}
colnames(df) <- c('Sample', 'Group', 'Cluster', 'Freqs')
df <- as.data.frame(df)
df$paired <- sapply(strsplit(df$Sample," "), `[`, 1)

df$Group <- as.character(df$Group)
df$Freqs <- as.numeric(as.character(df$Freqs))
df <- df[df$Group %like any% c('%Low%', '%Medium%', '%High%'),]
df$Group <- factor(df$Group, levels = c('b1 Low', 'b2 Low', 'b3 Low', 'b1 Medium/High', 'b2 Medium/High', 'b3 Medium/High'))

# with the paired lines
ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + geom_line(aes(x = Group, y = Freqs, group = paired), alpha = 0.3) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_2responses_boxplot_myeloid_stage_pathresponse_PAIREDONLY.png', width = 8, height = 8)

# without the paired lines
ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
#ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_2responses_boxplot_myeloid_stage_pathresponse_NOpairlines.png', width = 14, height = 8)
```

#### ••• Out of each separate myeloid populations

```{r}
pbmc <- readRDS(file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/luoma_PBMC_myeloid_plusmetadata.RDS')

pbmc_mini <- pbmc[,pbmc$Patient_ID %in% c('P18', 'P23', 'P24', 'P27', 'P29', 'P32')]

table(pbmc_mini$Patient_ID)
# P18  P23  P24  P32 
# 551 2105 2184 4017 

# input

seu <- pbmc_mini[,pbmc_mini$predicted.id %like any% c('%Mac%', '%Mon%', '%DC%')]
metadata <- 'predicted.id'

for (x in c('Mac', 'Mon', 'DC')){
  
  Idents(seu) <- metadata
  
  id <- paste('%', x, '%', sep = '')
  
  subsets <- grep(id, unique(pbmc_mini[[metadata]][,1]), value = T)
  
  SEU <- seu[,seu@active.ident %like any% id]

  SEU$orig.ident <- paste(SEU$Patient_ID, SEU$Stage)
  SEU$meta <- paste(SEU$Patient_ID, SEU$Stage)
  SEU$Path_response[SEU$Path_response %in% c('Medium', 'High')] <- 'Medium/High'
  SEU$treatment_path_response <- paste(SEU$Stage, SEU$Path_response)
  
  t <- get_freqs(SEU, metadata)
  t2 <- t
  for (i in 1:nrow(t2)) {
    t2[i,] = t2[i,]/sum(t2[i,])
  }
  t2 <- as.matrix(t2)
  
  meta <- SEU@meta.data[,c('meta', 'treatment_path_response')]
  rownames(meta) <- NULL
  meta <- unique(meta)
  df <- NULL
  for (i in 1:nrow(t2)) {
    for (j in 1:ncol(t2)) {
      df <- rbind(df, c(rownames(t2)[i], meta$treatment_path_response[which(meta$meta == rownames(t2)[i])], colnames(t2)[j], as.numeric(t2[i,j])))
    }
  }
  colnames(df) <- c('Sample', 'Group', 'Cluster', 'Freqs')
  df <- as.data.frame(df)
  df$paired <- sapply(strsplit(df$Sample," "), `[`, 1)
  
  df$Group <- as.character(df$Group)
  df$Freqs <- as.numeric(as.character(df$Freqs))
  df <- df[df$Group %like any% c('%Low%', '%Medium%', '%High%'),]
  df$Group <- factor(df$Group, levels = c('Pre-Tx Low', 'Post-Tx Low', 'Pre-Tx Medium/High', 'Post-Tx Medium/High'))
  
  # with the paired lines
  ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + geom_line(aes(x = Group, y = Freqs, group = paired), alpha = 0.3) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
  ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_tumor_2responses_boxplot_myeloid_stage_pathresponse_PAIREDONLY.png', width = 14, height = 8)
}
```

### •• Tumor

#### ••• Out of all myeloid

```{r}
luoma <- readRDS(file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/luoma_tumor_myeloid_plusmetadata.RDS')

# identifying which patients have pre-treatment biopsies
table(luoma$Stage, luoma$Patient_ID)
    #P13 P14 P15 P16 P18 P19 P20 P21 P22 P23 P24 P25 P26 P27 P28 P29 P30 P31 P32
#Pre  0   0   0   0  322  0   0   0   0  518 1569 0   0  198  0   76  0   0 1163

# making mini objects with just the patients who have pre-treatment biopsies
luoma_mini <- luoma[,luoma$Patient_ID %in% c('P18', 'P23', 'P24', 'P27', 'P29', 'P32')]

table(luoma_mini$Patient_ID)
# P18  P23  P24  P27  P32 
# 942 1209 2233  548 1580 

# input
SEU <- luoma_mini[,luoma_mini$predicted.id %like any% c('%Mac%', '%Mon%', '%DC%')]
metadata <- 'predicted.id'

SEU$orig.ident <- paste(SEU$Patient_ID, SEU$Stage)
SEU$meta <- paste(SEU$Patient_ID, SEU$Stage)
SEU$Path_response[SEU$Path_response %in% c('Medium', 'High')] <- 'Medium/High'
SEU$treatment_path_response <- paste(SEU$Stage, SEU$Path_response)

t <- get_freqs(SEU, metadata)
t2 <- t
for (i in 1:nrow(t2)) {
  t2[i,] = t2[i,]/sum(t2[i,])
}
t2 <- as.matrix(t2)

meta <- SEU@meta.data[,c('meta', 'treatment_path_response')]
rownames(meta) <- NULL
meta <- unique(meta)
df <- NULL
for (i in 1:nrow(t2)) {
  for (j in 1:ncol(t2)) {
    df <- rbind(df, c(rownames(t2)[i], meta$treatment_path_response[which(meta$meta == rownames(t2)[i])], colnames(t2)[j], as.numeric(t2[i,j])))
  }
}
colnames(df) <- c('Sample', 'Group', 'Cluster', 'Freqs')
df <- as.data.frame(df)
df$paired <- sapply(strsplit(df$Sample," "), `[`, 1)

df$Group <- as.character(df$Group)
df$Freqs <- as.numeric(as.character(df$Freqs))
df <- df[df$Group %like any% c('%Low%', '%Medium%', '%High%'),]
df$Group <- factor(df$Group, levels = c('Pre-Tx Low', 'Post-Tx Low', 'Pre-Tx Medium/High', 'Post-Tx Medium/High'))

# with the paired lines
ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + geom_line(aes(x = Group, y = Freqs, group = paired), alpha = 0.3) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_tumor_2responses_boxplot_myeloid_stage_pathresponse_PAIREDONLY.png', width = 14, height = 8)

# without the paired lines
#ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
#ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_2responses_boxplot_myeloid_stage_pathresponse_NOpairlines.png', width = 14, height = 8)
```

#### ••• Out of each separate myeloid population

```{r}
luoma <- readRDS(file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/luoma_tumor_myeloid_plusmetadata.RDS')

# identifying which patients have pre-treatment biopsies
table(luoma$Stage, luoma$Patient_ID)
    #P13 P14 P15 P16 P18 P19 P20 P21 P22 P23 P24 P25 P26 P27 P28 P29 P30 P31 P32
#Pre  0   0   0   0  322  0   0   0   0  518 1569 0   0  198  0   76  0   0 1163

# making mini objects with just the patients who have pre-treatment biopsies
luoma_mini <- luoma[,luoma$Patient_ID %in% c('P18', 'P23', 'P24', 'P27', 'P29', 'P32')]

table(luoma_mini$Patient_ID)
# P18  P23  P24  P27  P32 
# 942 1209 2233  548 1580 

# input

SEU <- luoma_mini[,luoma_mini$predicted.id %like any% c('%Mac%', '%Mon%', '%DC%')]
#SEU <- luoma_mini[,luoma_mini$predicted.id %like any% c('%Mac%')]
metadata <- 'predicted.id'

for (x in c('Mac', 'Mon', 'DC')){
  
  Idents(SEU) <- metadata
  
  id <- paste('%', x, '%', sep = '')
  
  SEU <- SEU[,SEU@active.ident %like any% id]

  SEU$orig.ident <- paste(SEU$Patient_ID, SEU$Stage)
  SEU$meta <- paste(SEU$Patient_ID, SEU$Stage)
  SEU$Path_response[SEU$Path_response %in% c('Medium', 'High')] <- 'Medium/High'
  SEU$treatment_path_response <- paste(SEU$Stage, SEU$Path_response)
  
  t <- get_freqs(SEU, metadata)
  t2 <- t
  for (i in 1:nrow(t2)) {
    t2[i,] = t2[i,]/sum(t2[i,])
  }
  t2 <- as.matrix(t2)
  
  meta <- SEU@meta.data[,c('meta', 'treatment_path_response')]
  rownames(meta) <- NULL
  meta <- unique(meta)
  df <- NULL
  for (i in 1:nrow(t2)) {
    for (j in 1:ncol(t2)) {
      df <- rbind(df, c(rownames(t2)[i], meta$treatment_path_response[which(meta$meta == rownames(t2)[i])], colnames(t2)[j], as.numeric(t2[i,j])))
    }
  }
  colnames(df) <- c('Sample', 'Group', 'Cluster', 'Freqs')
  df <- as.data.frame(df)
  df$paired <- sapply(strsplit(df$Sample," "), `[`, 1)
  
  df$Group <- as.character(df$Group)
  df$Freqs <- as.numeric(as.character(df$Freqs))
  df <- df[df$Group %like any% c('%Low%', '%Medium%', '%High%'),]
  df$Group <- factor(df$Group, levels = c('Pre-Tx Low', 'Post-Tx Low', 'Pre-Tx Medium/High', 'Post-Tx Medium/High'))
  
  # with the paired lines
  ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + geom_line(aes(x = Group, y = Freqs, group = paired), alpha = 0.3) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
  ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_tumor_2responses_boxplot_myeloid_stage_pathresponse_PAIREDONLY.png', width = 14, height = 8)
}
```

## • Freq comparison: all patients

### •• PBMC

#### ••• Out of all myeloid

```{r}
pbmc <- readRDS(file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/03.WorkingSeurat/luoma_PBMC_myeloid_plusmetadata_fivesubsets.RDS')

# input
SEU <- pbmc[,pbmc$predicted.id %like any% c('%Mac%', '%Mon%', '%DC%')]
metadata <- 'predicted.id'

SEU$orig.ident <- paste(SEU$Patient_ID, SEU$Stage)
SEU$meta <- paste(SEU$Patient_ID, SEU$Stage)
SEU$Path_response[SEU$Path_response %in% c('Medium', 'High')] <- 'Medium/High'
SEU$treatment_path_response <- paste(SEU$Stage, SEU$Path_response)

t <- get_freqs(SEU, metadata)
t2 <- t
for (i in 1:nrow(t2)) {
  t2[i,] = t2[i,]/sum(t2[i,])
}
t2 <- as.matrix(t2)

meta <- SEU@meta.data[,c('meta', 'treatment_path_response')]
rownames(meta) <- NULL
meta <- unique(meta)
df <- NULL
for (i in 1:nrow(t2)) {
  for (j in 1:ncol(t2)) {
    df <- rbind(df, c(rownames(t2)[i], meta$treatment_path_response[which(meta$meta == rownames(t2)[i])], colnames(t2)[j], as.numeric(t2[i,j])))
  }
}
colnames(df) <- c('Sample', 'Group', 'Cluster', 'Freqs')
df <- as.data.frame(df)

df$Group <- as.character(df$Group)
df$Freqs <- as.numeric(as.character(df$Freqs))
df <- df[df$Group %like any% c('%Low%', '%Medium%', '%High%'),]
df$Group <- factor(df$Group, levels = c('b1 Low', 'b2 Low', 'b3 Low', 'b1 Medium/High', 'b2 Medium/High', 'b3 Medium/High'))

df$Group <- str_replace_all(df$Group, '/', '_')

# without the paired lines
ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_2responses_boxplot_myeloid_stage_pathresponse_NOpairlines_5myeloid.png', width = 8, height = 8)
```

#### •••• Wilcoxan test

```{r}
# comparing low and med/high from each timepoint
wc <- freq_wilcox(df = df, cond1 = 'b1 Low', cond2 = 'b1 Medium_High', out = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/')

wc <- freq_wilcox(df = df, cond1 = 'b2 Low', cond2 = 'b2 Medium_High', out = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/')

wc <- freq_wilcox(df = df, cond1 = 'b3 Low', cond2 = 'b3 Medium_High', out = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/')


# comparing different timepoints in low responders
wc <- freq_wilcox(df = df, cond1 = 'b1 Low', cond2 = 'b2 Low', out = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/') # cd16 monocytes are significantly different

wc <- freq_wilcox(df = df, cond1 = 'b2 Low', cond2 = 'b3 Low', out = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/')

wc <- freq_wilcox(df = df, cond1 = 'b1 Low', cond2 = 'b3 Low', out = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/')


# comparing different timepoints in medium/high responders
wc <- freq_wilcox(df = df, cond1 = 'b1 Medium_High', cond2 = 'b2 Medium_High', out = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/') # pDC is significantly different 

wc <- freq_wilcox(df = df, cond1 = 'b2 Medium_High', cond2 = 'b3 Medium_High', out = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/')

wc <- freq_wilcox(df = df, cond1 = 'b1 Medium_High', cond2 = 'b3 Medium_High', out = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/') # cdc2_cd33 is significantly different
```

# LUOMA figure prep -- BLOOD

## • DE genes b/t CD14 monocytes in low and high responders

```{r}
pbmc <- readRDS(file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/03.WorkingSeurat/luoma_PBMC_myeloid_plusmetadata_fivesubsets.RDS')

pbmc$var <- paste(pbmc$predicted.id, pbmc$Path_response, sep = '_')
pbmc$var[pbmc$var %like any% c('Mono_CD14_Medium', 'Mono_CD14_High')] <- 'Mono_CD14_MedHigh'
table(pbmc$var)

volcPlot(pbmc, seu_name = 'luoma_CD14_mon_hivsloresponse', meta = 'var', cell1 = 'Mono_CD14_Low', cell2 = 'Mono_CD14_MedHigh', out = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/04.Figures/')
```

## • Heatmap of subsets (y-axis) and patients (x-axis)

```{r}

SEU <- luoma[,luoma$predicted.id %like any% c('%Mac%', '%Mon%', '%DC%')]
SEU$orig.ident <- paste(SEU$Patient_ID, SEU$Stage)
SEU$meta <- paste(SEU$Patient_ID, SEU$Stage)
SEU$treatment_path_response <- paste(SEU$Stage, SEU$Path_response)
metadata <- 'predicted.id'

t <- get_freqs(SEU, metadata)
t2 <- t
for (i in 1:nrow(t2)) {
  t2[i,] = t2[i,]/sum(t2[i,])
}
t2 <- as.matrix(t2)

meta <- SEU@meta.data[,c('meta', 'treatment_path_response')]
rownames(meta) <- NULL
meta <- unique(meta)
df <- NULL
for (i in 1:nrow(t2)) {
  for (j in 1:ncol(t2)) {
    df <- rbind(df, c(rownames(t2)[i], meta$treatment_path_response[which(meta$meta == rownames(t2)[i])], colnames(t2)[j], as.numeric(t2[i,j])))
  }
}
colnames(df) <- c('Sample', 'Group', 'Cluster', 'Freqs')
df <- as.data.frame(df)
df$paired <- sapply(strsplit(df$Sample," "), `[`, 1)
df$Group <- as.character(df$Group)
df$Freqs <- as.numeric(as.character(df$Freqs))
df <- df[df$Group %like any% c('%Low%', '%Medium%', '%High%'),]

df1 <- pivot_wider(df, names_from = 'Sample', values_from = 'Freqs', values_fn=mean)
df1 <- df1[,!colnames(df1) %in% c('Group', 'paired')]
df1[is.na(df1)] = 0
df2 <- aggregate(df1[,c(2:ncol(df1))], by = list(df1$Cluster), FUN = sum)

rnames <- df2$Group.1
df2$Group.1 <- NULL
rownames(df2) <- rnames
meta2 <- luoma[[c('Patient_ID', 'Stage', 'Path_response')]]
meta2$pt_stage <- paste(meta2$Patient_ID, meta2$Stage)
meta2$Patient_ID <- NULL
rownames(meta2) <- NULL
meta2 <- meta2[!is.na(meta2$Path_response),]
meta2 <- meta2[!duplicated(meta2),]
rnames <- meta2$pt_stage
meta2$pt_stage <- NULL


column_ha <- ComplexHeatmap::HeatmapAnnotation(df = meta2, col = list(Stage = c('B1' = 'light grey', 'B2' = 'dark grey', 'B3' = 'black'), Path_response = c('High' = 'indian red', 'Medium' = 'light yellow 2', 'Low' = 'light blue')))

png('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/luoma_heatmap_subset_by_patient.png', width = 9, height = 5, units = 'in', res = 300)
ComplexHeatmap::Heatmap(df2, top_annotation = column_ha)
dev.off()
```

## • Density plot: Plotting monocyte markers in Luoma CD14 monocytes

```{r}
#VlnPlot(luoma, features = c('THBS1', 'IL1B', 'ID1'), group.by = 'predicted.id')

min <- -0.05
max <- 1

# density plot of the scores for the different monocyte subsets
plot(density(luoma$prediction.score.Mono_CD14_THBS1), col = 'blue', xlim = c(min, max), main = 'Density of monocyte subset prediction score')

# additional lines beyond the first one
lines(density(luoma$prediction.score.Mono_CD14), col = 'red', xlim = c(min, max))
lines(density(luoma$prediction.score.Mono_CD14_ID1), col = 'green', xlim = c(min, max))
lines(density(luoma$prediction.score.Mono_TIL), col = 'orange', xlim = c(min, max))
lines(density(luoma$prediction.score.Mono_CD14_IL1B), col = 'purple', xlim = c(min, max))

# legend
legend("topright", c('Mono_CD14_THBS1', 'Mono_CD14', 'Mono_CD14_ID1', 'Mono_TIL', 'Mono_CD14_IL1B'), fill=c("blue","red", 'green', 'orange', 'purple'))
```

## • Boxplots (Stage + path response) (3 path response groups)

```{r}
luoma <- readRDS(file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/luoma_PBMC_myeloid_plusmetadata.RDS')

SEU <- luoma[,luoma$predicted.id %like any% c('%Mac%', '%Mon%', '%DC%')]
SEU$orig.ident <- paste(SEU$Patient_ID, SEU$Stage)
SEU$meta <- paste(SEU$Patient_ID, SEU$Stage)
SEU$treatment_path_response <- paste(SEU$Stage, SEU$Path_response)
metadata <- 'predicted.id'

get_freqs <- function(immune.combined, cluster_res) {
  Idents(immune.combined) <- cluster_res
  new.ident <- sort(unique(Idents(immune.combined)))
  samples <- unique(immune.combined@meta.data$orig.ident)
  tmp <- match(immune.combined$orig.ident, samples)
  sample_ind <- unique(tmp)
  ids <- Idents(immune.combined)
  tmp_v <- matrix(0, nrow = length(samples), ncol = length(new.ident))
  rownames(tmp_v) <- samples
  total_in_sample <- rep(0, length(samples))
  tmp <- immune.combined$orig.ident
  tmp <- plyr::count(tmp)
  total_in_sample = tmp$freq
  names(total_in_sample) <- tmp$x
  total_in_sample <- total_in_sample[match(samples, names(total_in_sample))]
  for (i in 1:length(new.ident)) {
    tmp <- ids[which(ids == new.ident[i])]
    tmp <- immune.combined$orig.ident[match(names(tmp), rownames(immune.combined@meta.data))]
    tmp <- plyr::count(tmp)
    for (j in 1:nrow(tmp)) {
      ind <- which(rownames(tmp_v) == tmp$x[j])
      tmp_v[ind,i] <- tmp$freq[j]
    }
  }
  colnames(tmp_v) <- new.ident
  ind <- order(colnames(tmp_v))
  tmp_v
}

t <- get_freqs(SEU, metadata)
t2 <- t
for (i in 1:nrow(t2)) {
  t2[i,] = t2[i,]/sum(t2[i,])
}
t2 <- as.matrix(t2)

meta <- SEU@meta.data[,c('meta', 'treatment_path_response')]
rownames(meta) <- NULL
meta <- unique(meta)
df <- NULL
for (i in 1:nrow(t2)) {
  for (j in 1:ncol(t2)) {
    df <- rbind(df, c(rownames(t2)[i], meta$treatment_path_response[which(meta$meta == rownames(t2)[i])], colnames(t2)[j], as.numeric(t2[i,j])))
  }
}
colnames(df) <- c('Sample', 'Group', 'Cluster', 'Freqs')
df <- as.data.frame(df)
df$paired <- sapply(strsplit(df$Sample," "), `[`, 1)

df$Group <- as.character(df$Group)
df$Freqs <- as.numeric(as.character(df$Freqs))
df <- df[df$Group %like any% c('%Low%', '%Medium%', '%High%'),]
df$Group <- factor(df$Group, levels = c('B1 Low', 'B2 Low', 'B3 Low', 'B1 Medium', 'B2 Medium',  'B3 Medium', 'B1 High',  'B2 High', 'B3 High'))

ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + geom_line(aes(x = Group, y = Freqs, group = paired), alpha = 0.3) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
#ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_boxplot_myeloid_stage_pathresponse.png', width = 14, height = 8)
```

## • [1/17/2023] Boxplots (Stage + path response) (2 path responses)

```{r}
luoma <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/UMAP_Luoma_labeltransfer_01172023.RDS')

SEU <- luoma[,luoma$predicted.id %like any% c('%Mac%', '%Mon%', '%DC%')]
SEU$orig.ident <- paste(SEU$Patient_ID, SEU$Stage)
SEU$meta <- paste(SEU$Patient_ID, SEU$Stage)
SEU$Path_response[SEU$Path_response %in% c('Medium', 'High')] <- 'Medium/High'
SEU$treatment_path_response <- paste(SEU$Stage, SEU$Path_response)
metadata <- 'predicted.id'

get_freqs <- function(immune.combined, cluster_res) {
  Idents(immune.combined) <- cluster_res
  new.ident <- sort(unique(Idents(immune.combined)))
  samples <- unique(immune.combined@meta.data$orig.ident)
  tmp <- match(immune.combined$orig.ident, samples)
  sample_ind <- unique(tmp)
  ids <- Idents(immune.combined)
  tmp_v <- matrix(0, nrow = length(samples), ncol = length(new.ident))
  rownames(tmp_v) <- samples
  total_in_sample <- rep(0, length(samples))
  tmp <- immune.combined$orig.ident
  tmp <- plyr::count(tmp)
  total_in_sample = tmp$freq
  names(total_in_sample) <- tmp$x
  total_in_sample <- total_in_sample[match(samples, names(total_in_sample))]
  for (i in 1:length(new.ident)) {
    tmp <- ids[which(ids == new.ident[i])]
    tmp <- immune.combined$orig.ident[match(names(tmp), rownames(immune.combined@meta.data))]
    tmp <- plyr::count(tmp)
    for (j in 1:nrow(tmp)) {
      ind <- which(rownames(tmp_v) == tmp$x[j])
      tmp_v[ind,i] <- tmp$freq[j]
    }
  }
  colnames(tmp_v) <- new.ident
  ind <- order(colnames(tmp_v))
  tmp_v
}

t <- get_freqs(SEU, metadata)
t2 <- t
for (i in 1:nrow(t2)) {
  t2[i,] = t2[i,]/sum(t2[i,])
}
t2 <- as.matrix(t2)

meta <- SEU@meta.data[,c('meta', 'treatment_path_response')]
rownames(meta) <- NULL
meta <- unique(meta)
df <- NULL
for (i in 1:nrow(t2)) {
  for (j in 1:ncol(t2)) {
    df <- rbind(df, c(rownames(t2)[i], meta$treatment_path_response[which(meta$meta == rownames(t2)[i])], colnames(t2)[j], as.numeric(t2[i,j])))
  }
}
colnames(df) <- c('Sample', 'Group', 'Cluster', 'Freqs')
df <- as.data.frame(df)
df$paired <- sapply(strsplit(df$Sample," "), `[`, 1)

df$Group <- as.character(df$Group)
df$Freqs <- as.numeric(as.character(df$Freqs))
df <- df[df$Group %like any% c('%Low%', '%Medium%', '%High%'),]
df$Group <- factor(df$Group, levels = c('Pre-Tx Low', 'Post-Tx Low', 'Pre-Tx Medium/High', 'Post-Tx Medium/High'))

# with the paired lines
ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + geom_line(aes(x = Group, y = Freqs, group = paired), alpha = 0.3) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
#ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/01172023Luoma_PBMC_2responses_boxplot_myeloid_stage_pathresponse.png', width = 14, height = 8)

# without the paired lines
ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
#ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/01172023Luoma_PBMC_2responses_boxplot_myeloid_stage_pathresponse_NOpairlines.png', width = 14, height = 8)
```

### •• (NS) Wilcox test (B1 high vs low)

```{r, warning=FALSE}
wilcox <- data.frame(Cluster=character(), pval = numeric())
for (x in unique(df$Cluster)){
  df2 <- df[df$Cluster == x,]
  l <- df2[df2$Group == 'B1 Low', 'Freqs']
  h <- df2[df2$Group == c('B1 Medium/High'), 'Freqs']
  wc <- wilcox.test(l, h)
  p <- wc$p.value
  row <- c(x, p)
  wilcox <- rbind(wilcox, row)
}
#write.csv(wilcox, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_boxplot_myeloid_2responses_B1.csv')
```

### •• (NS) Wilcox test (B2 high vs low)

```{r, warning=FALSE}
wilcox <- data.frame(Cluster=character(), pval = numeric())
for (x in unique(df$Cluster)){
  df2 <- df[df$Cluster == x,]
  l <- df2[df2$Group == 'B2 Low', 'Freqs']
  h <- df2[df2$Group == c('B2 Medium/High'), 'Freqs']
  wc <- wilcox.test(l, h)
  p <- wc$p.value
  row <- c(x, p)
  wilcox <- rbind(wilcox, row)
}
write.csv(wilcox, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_boxplot_myeloid_2responses_B2.csv')
```

### •• (NS) Wilcox test (B3 high vs low)

```{r, warning=FALSE}
wilcox <- data.frame(Cluster=character(), pval = numeric())
for (x in unique(df$Cluster)){
  df2 <- df[df$Cluster == x,]
  l <- df2[df2$Group == 'B3 Low', 'Freqs']
  h <- df2[df2$Group == c('B3 Medium/High'), 'Freqs']
  wc <- wilcox.test(l, h)
  p <- wc$p.value
  row <- c(x, p)
  wilcox <- rbind(wilcox, row)
}
Ewrite.csv(wilcox, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_boxplot_myeloid_2responses_B3.csv')
```

### •• Wilcox test (B1 high vs. B2 high) (pDC lower \@ B2)

```{r, warning = FALSE}
wilcox <- data.frame(Cluster=character(), pval = numeric())
for (x in unique(df$Cluster)){
  df2 <- df[df$Cluster == x,]
  l <- df2[df2$Group == 'B1 Medium/High', 'Freqs']
  h <- df2[df2$Group == c('B2 Medium/High'), 'Freqs']
  wc <- wilcox.test(l, h)
  p <- wc$p.value
  row <- c(x, p)
  wilcox <- rbind(wilcox, row)
}
#write.csv(wilcox, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_boxplot_myeloid_2responses_B1highvsB2high.csv')
```

### •• Wilcox test (B2 high vs. B3 high) (CD14 mon \^ in B3)

```{r, warning = FALSE}
wilcox <- data.frame(Cluster=character(), pval = numeric())
for (x in unique(df$Cluster)){
  df2 <- df[df$Cluster == x,]
  l <- df2[df2$Group == 'B2 Medium/High', 'Freqs']
  h <- df2[df2$Group == c('B3 Medium/High'), 'Freqs']
  wc <- wilcox.test(l, h)
  p <- wc$p.value
  row <- c(x, p)
  wilcox <- rbind(wilcox, row)
}
#write.csv(wilcox, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_boxplot_myeloid_2responses_B2highvsB3high.csv')
```

### •• Wilcox test (B1 high vs. B3 high) (CD14 mon \^ in B3)

```{r}
wilcox <- data.frame(Cluster=character(), pval = numeric())
for (x in unique(df$Cluster)){
  df2 <- df[df$Cluster == x,]
  l <- df2[df2$Group == 'B1 Medium/High', 'Freqs']
  h <- df2[df2$Group == c('B3 Medium/High'), 'Freqs']
  wc <- wilcox.test(l, h)
  p <- wc$p.value
  row <- c(x, p)
  wilcox <- rbind(wilcox, row)
}
#write.csv(wilcox, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_boxplot_myeloid_2responses_B1highvsB3high.csv')
```

### •• Wilcox test (B1 low vs. B2 low)

```{r}
wilcox <- data.frame(Cluster=character(), pval = numeric())
for (x in unique(df$Cluster)){
  df2 <- df[df$Cluster == x,]
  l <- df2[df2$Group == 'B1 Low', 'Freqs']
  h <- df2[df2$Group == c('B2 Low'), 'Freqs']
  wc <- wilcox.test(l, h)
  p <- wc$p.value
  row <- c(x, p)
  wilcox <- rbind(wilcox, row)
}
#write.csv(wilcox, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_boxplot_myeloid_2responses_B1lowvsB2low.csv')
```

### •• (NS) Wilcox test (B2 low vs. B3 low)

```{r}
wilcox <- data.frame(Cluster=character(), pval = numeric())
for (x in unique(df$Cluster)){
  df2 <- df[df$Cluster == x,]
  l <- df2[df2$Group == 'B2 Low', 'Freqs']
  h <- df2[df2$Group == c('B3 Low'), 'Freqs']
  wc <- wilcox.test(l, h)
  p <- wc$p.value
  row <- c(x, p)
  wilcox <- rbind(wilcox, row)
}
write.csv(wilcox, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_boxplot_myeloid_2responses_B2lowvsB3low.csv')
```

### •• Wilcox test (B1 low vs. B3 low)

```{r}
wilcox <- data.frame(Cluster=character(), pval = numeric())
for (x in unique(df$Cluster)){
  df2 <- df[df$Cluster == x,]
  l <- df2[df2$Group == 'B1 Low', 'Freqs']
  h <- df2[df2$Group == c('B3 Low'), 'Freqs']
  wc <- wilcox.test(l, h)
  p <- wc$p.value
  row <- c(x, p)
  wilcox <- rbind(wilcox, row)
}
write.csv(wilcox, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_boxplot_myeloid_2responses_B1lowvsB3low.csv')
```

## • Boxplots (Path response)

```{r}
SEU <- luoma[,luoma$predicted.id %like any% c('%Mac%', '%Mon%', '%DC%')]
SEU$orig.ident <- paste(SEU$Patient_ID, SEU$Stage)
SEU$meta <- paste(SEU$Patient_ID, SEU$Stage)
metadata <- 'predicted.id'

get_freqs <- function(immune.combined, cluster_res) {
  Idents(immune.combined) <- cluster_res
  new.ident <- sort(unique(Idents(immune.combined)))
  samples <- unique(immune.combined@meta.data$orig.ident)
  tmp <- match(immune.combined$orig.ident, samples)
  sample_ind <- unique(tmp)
  ids <- Idents(immune.combined)
  tmp_v <- matrix(0, nrow = length(samples), ncol = length(new.ident))
  rownames(tmp_v) <- samples
  total_in_sample <- rep(0, length(samples))
  tmp <- immune.combined$orig.ident
  tmp <- plyr::count(tmp)
  total_in_sample = tmp$freq
  names(total_in_sample) <- tmp$x
  total_in_sample <- total_in_sample[match(samples, names(total_in_sample))]
  for (i in 1:length(new.ident)) {
    tmp <- ids[which(ids == new.ident[i])]
    tmp <- immune.combined$orig.ident[match(names(tmp), rownames(immune.combined@meta.data))]
    tmp <- plyr::count(tmp)
    for (j in 1:nrow(tmp)) {
      ind <- which(rownames(tmp_v) == tmp$x[j])
      tmp_v[ind,i] <- tmp$freq[j]
    }
  }
  colnames(tmp_v) <- new.ident
  ind <- order(colnames(tmp_v))
  tmp_v
}

t <- get_freqs(SEU, metadata)
t2 <- t
for (i in 1:nrow(t2)) {
  t2[i,] = t2[i,]/sum(t2[i,])
}
t2 <- as.matrix(t2)

meta <- SEU@meta.data[,c('meta', 'Path_response')]
rownames(meta) <- NULL
meta <- unique(meta)
df <- NULL
for (i in 1:nrow(t2)) {
  for (j in 1:ncol(t2)) {
    df <- rbind(df, c(rownames(t2)[i], meta$Path_response[which(meta$meta == rownames(t2)[i])], colnames(t2)[j], as.numeric(t2[i,j])))
  }
}
colnames(df) <- c('Sample', 'Group', 'Cluster', 'Freqs')
df <- as.data.frame(df)
df$paired <- sapply(strsplit(df$Sample," "), `[`, 1)

df$Group <- as.character(df$Group)
df$Group[df$Group %like% 'Medium'] <- 'Medium/High'
df$Group[df$Group %like% 'High'] <- 'Medium/High'
df$Freqs <- as.numeric(as.character(df$Freqs))
df <- df[df$Group %like any% c('%Low%', '%Medium%', '%High%'),]
df$Group <- factor(df$Group, levels = c('Low', 'Medium/High'))

ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
#ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_boxplot_myeloid_pathresponse.png', width = 14, height = 8)
```

### •• Wilcox test

```{r}
wilcox <- data.frame(Cluster=character(), pval = numeric())
for (x in unique(df$Cluster)){
  df2 <- df[df$Cluster == x,]
  l <- df2[df2$Group == 'Low', 'Freqs']
  h <- df2[df2$Group == c('Medium/High'), 'Freqs']
  wc <- wilcox.test(l, h)
  p <- wc$p.value
  row <- c(x, p)
  wilcox <- rbind(wilcox, row)
}
#write.csv(wilcox, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Luoma_PBMC_boxplot_myeloid_pathresponse.csv')
```

# LUOMA figure prep -- TUMOR

```{r}
luoma <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/luoma_tumor_myeloid_T_plusmetadata.RDS')

DimPlot(luoma, group.by = 'predicted.id', cols = color_clusters)

luoma$global_cluster2 <- luoma$predicted.id
luoma$global_cluster2[luoma$global_cluster2 %like% c('CD4%')] <- 'CD4'
luoma$global_cluster2[luoma$global_cluster2 %like% c('CD8%')] <- 'CD8'
luoma$global_cluster2[luoma$global_cluster2 %like% c('Treg%')] <- 'Treg'

#png('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/luoma_tumor_myeloid_T_plusmetadata_UMAP.png', units = 'in', width = 10, height = 7, res = 300)
DimPlot(luoma, group.by = 'global_cluster2', cols = color_clusters, label = TRUE, label.box = TRUE, repel = TRUE)
#dev.off()

luoma$path_response2 <- luoma$Path_response
luoma$path_response2[luoma$path_response2 %like% 'Medium%'] <- 'Medium/High'
luoma$path_response2[luoma$path_response2 %like% 'High%'] <- 'Medium/High'

png('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/luoma_tumor_myeloid_T_plusmetadata_vlnplot.png', units = 'in', width = 7, height = 5, res = 300)
VlnPlot(luoma[,luoma$predicted.id %like any% c('Mon%', 'Mac%', '%DC%')], features = c('CXCL9'), group.by = 'global_cluster2', split.by = 'path_response2')
dev.off()
```

## • Volcano plot of one group vs. the other group (path response)

```{r}
luoma$path_response2 <- luoma$Path_response
luoma$path_response2[luoma$path_response2 %like% 'Medium%'] <- 'Medium/High'
luoma$path_response2[luoma$path_response2 %like% 'High%'] <- 'Medium/High'

seu = luoma[,luoma$predicted.id == 'TIL_Mono']
seu_name = 'luoma'
meta = 'path_response2'
cell1 = 'Low'
cell2 = 'Medium/High'
out = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/'
test.use = 'wilcox'
logfc = 0.1
pval_adj = 0.05
fccutoff = 0.1
pcutoff = 0.05

Seurat::Idents(object = seu) <- seu[[meta]]

ids <- unique(seu@active.ident)

marks_all <- Seurat::FindMarkers(seu, ident.1 = cell1, ident.2 = cell2, group.by = meta, test.use = test.use, logfc.threshold = logfc, only.pos = FALSE)

marks_all$p_val_adj <- as.numeric(marks_all$p_val_adj)
marks_all <- marks_all[!rownames(marks_all) %like any% c('%^RP%', '%^RPS%', '%^RPL%'),]

xlab <- paste('<-----', cell2, 'genes', '     log2 fold change     ', cell1, 'genes', '-----> ', sep = ' ')

nums <- as.data.frame(table(seu@active.ident))
nums$fin <- paste(nums$Var1, nums$Freq, sep = '=')
nums2 <- as.character(nums$fin)

EnhancedVolcano(marks_all, lab = rownames(marks_all), x = 'avg_log2FC',y = 'p_val_adj', title = paste(as.character(cell2), 'vs.', as.character(cell1), sep = ' '), selectLab = intersect(rownames(marks_all), lr$ligands), subtitle = paste('DE logfc =', logfc, 'Vln fc-cutoff =', fccutoff, 'Vln p-cutoff =', pcutoff), caption = paste(nums2[1], nums2[2], sep = "   "), FCcutoff = fccutoff, pCutoff = pcutoff, legendLabels = NULL, legendIconSize = 0, legendDropLevels = TRUE, legendPosition = 'right', xlab = xlab)
ggsave(paste(out, seu_name, paste(as.character(ids[1]), 'vs.', as.character(ids[2]), 'logfc', logfc, 'pvaladj', pval_adj, 'volcano_plot.png', sep = '_')), width = 12, height = 12)
write.csv(marks_all, file = paste(out, seu_name, paste(as.character(ids[1]), 'vs.', as.character(ids[2]), 'logfc', logfc, 'pvaladj', pval_adj, 'volcano_plot.csv', sep = '_')))

```

------------------------------------------------------------------------

## • Volcano plot of one group vs. the other group (pre- and post- treatment)

```{r}
seu = luoma[,luoma$predicted.id == 'Mac_C1QA']
seu_name = 'luoma'
meta = 'Stage'
cell1 = 'Post-Tx'
cell2 = 'Pre-Tx'
out = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/'
test.use = 'wilcox'
logfc = 0.1
pval_adj = 0.05
fccutoff = 0.1
pcutoff = 0.05

Seurat::Idents(object = seu) <- seu[[meta]]

ids <- unique(seu@active.ident)

marks_all <- Seurat::FindMarkers(seu, ident.1 = cell1, ident.2 = cell2, group.by = meta, test.use = test.use, logfc.threshold = logfc, only.pos = FALSE)

marks_all$p_val_adj <- as.numeric(marks_all$p_val_adj)
marks_all <- marks_all[!rownames(marks_all) %like any% c('%^RP%', '%^RPS%', '%^RPL%'),]

xlab <- paste('<-----', cell2, 'genes', '     log2 fold change     ', cell1, 'genes', '-----> ', sep = ' ')

nums <- as.data.frame(table(seu@active.ident))
nums$fin <- paste(nums$Var1, nums$Freq, sep = '=')
nums2 <- as.character(nums$fin)

EnhancedVolcano(marks_all, lab = rownames(marks_all), x = 'avg_log2FC',y = 'p_val_adj', title = paste(as.character(cell2), 'vs.', as.character(cell1), sep = ' '), selectLab = intersect(rownames(marks_all), lr$ligands), subtitle = paste('DE logfc =', logfc, 'Vln fc-cutoff =', fccutoff, 'Vln p-cutoff =', pcutoff), caption = paste(nums2[1], nums2[2], sep = "   "), FCcutoff = fccutoff, pCutoff = pcutoff, legendLabels = NULL, legendIconSize = 0, legendDropLevels = TRUE, legendPosition = 'right', xlab = xlab)
ggsave(paste(out, seu_name, paste(as.character(ids[1]), 'vs.', as.character(ids[2]), 'logfc', logfc, 'pvaladj', pval_adj, 'volcano_plot.png', sep = '_')), width = 12, height = 12)
write.csv(marks_all, file = paste(out, seu_name, paste(as.character(ids[1]), 'vs.', as.character(ids[2]), 'logfc', logfc, 'pvaladj', pval_adj, 'volcano_plot.csv', sep = '_')))
```

## • DE genes

```{r}
luoma_mini <- luoma[,luoma$global_cluster2 %like% c('Mac_SPP1%')]

all_g <- c()
for (y in sort(unique(luoma_mini$global_cluster2))){
  fm <- FindMarkers(luoma_mini, ident.1 = y, group.by = 'global_cluster2', logfc.threshold = 0.1, test.use = 'wilcox', min.pct = 0.1, min.diff.pct = 0.05, only.pos = TRUE)
  fm <- fm[order(-fm$avg_log2FC, fm$p_val_adj),]
  g <- rownames(fm[1:3,])
  all_g <- c(all_g, g)
}
all_g <- unique(all_g)
DotPlot(luoma_mini, features = rev(all_g), col.min = 0, group.by = 'global_cluster2') + coord_flip() +  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
ggsave('/Volumes/hdlab/Projects/HNC_SPORE/Golfinosetal2022/mouse_myeloid_dotplot.png', width = 7, height = 9)
```

## • Make myeloid cell only object + DimPlot

```{r}
# make myeloid cell only
luoma <- luoma[,! luoma$global_cluster2 %like% c('%CD8%', '%CD4%', '%Treg%')]
luoma <- ScaleData(luoma, features = rownames(luoma))
luoma <- FindVariableFeatures(luoma, selection.method = "vst", nfeatures = 2000)
luoma <- RunPCA(luoma, features = VariableFeatures(object = luoma), npcs = 100)
ElbowPlot(luoma, ndims = 50)

luoma <- FindNeighbors(luoma, dims = 1:30)
luoma <- RunUMAP(luoma, dims = 1:30)

png('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/luoma_dimplot_myeloid_only.png', units = 'in', width = 8, height = 5, res = 300)
DimPlot(luoma, reduction = "umap", group.by = 'predicted.id', cols = color_clusters, label = T, label.box = T, repel = T)
dev.off()
```

![](images/paste-E52F8630.png)

![](images/paste-667E8D1D.png)

![](images/paste-065738E5.png) \

## • [1/17/2023] Boxplots (Stage + path response) (2 path responses)

```{r}
luoma <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/UMAP_Luoma_labeltransfer_01182023.RDS')
luoma$predicted.id[luoma$predicted.id %like any% '%CD14%'] == 'Blood_CD14_Mono' 

SEU <- luoma[,luoma$predicted.id %like any% c('%Mac%', '%Mon%', '%DC%')]
SEU$orig.ident <- paste(SEU$Patient_ID, SEU$Stage)
SEU$meta <- paste(SEU$Patient_ID, SEU$Stage)
SEU$Path_response[SEU$Path_response %in% c('Medium', 'High')] <- 'Medium/High'
SEU$treatment_path_response <- paste(SEU$Stage, SEU$Path_response)
metadata <- 'predicted.id'

get_freqs <- function(immune.combined, cluster_res) {
  Idents(immune.combined) <- cluster_res
  new.ident <- sort(unique(Idents(immune.combined)))
  samples <- unique(immune.combined@meta.data$orig.ident)
  tmp <- match(immune.combined$orig.ident, samples)
  sample_ind <- unique(tmp)
  ids <- Idents(immune.combined)
  tmp_v <- matrix(0, nrow = length(samples), ncol = length(new.ident))
  rownames(tmp_v) <- samples
  total_in_sample <- rep(0, length(samples))
  tmp <- immune.combined$orig.ident
  tmp <- plyr::count(tmp)
  total_in_sample = tmp$freq
  names(total_in_sample) <- tmp$x
  total_in_sample <- total_in_sample[match(samples, names(total_in_sample))]
  for (i in 1:length(new.ident)) {
    tmp <- ids[which(ids == new.ident[i])]
    tmp <- immune.combined$orig.ident[match(names(tmp), rownames(immune.combined@meta.data))]
    tmp <- plyr::count(tmp)
    for (j in 1:nrow(tmp)) {
      ind <- which(rownames(tmp_v) == tmp$x[j])
      tmp_v[ind,i] <- tmp$freq[j]
    }
  }
  colnames(tmp_v) <- new.ident
  ind <- order(colnames(tmp_v))
  tmp_v
}

t <- get_freqs(SEU, metadata)
t2 <- t
for (i in 1:nrow(t2)) {
  t2[i,] = t2[i,]/sum(t2[i,])
}
t2 <- as.matrix(t2)

meta <- SEU@meta.data[,c('meta', 'treatment_path_response')]
rownames(meta) <- NULL
meta <- unique(meta)
df <- NULL
for (i in 1:nrow(t2)) {
  for (j in 1:ncol(t2)) {
    df <- rbind(df, c(rownames(t2)[i], meta$treatment_path_response[which(meta$meta == rownames(t2)[i])], colnames(t2)[j], as.numeric(t2[i,j])))
  }
}
colnames(df) <- c('Sample', 'Group', 'Cluster', 'Freqs')
df <- as.data.frame(df)
df$paired <- sapply(strsplit(df$Sample," "), `[`, 1)

df$Group <- as.character(df$Group)
df$Freqs <- as.numeric(as.character(df$Freqs))
df <- df[df$Group %like any% c('%Low%', '%Medium%', '%High%'),]
df$Group <- factor(df$Group, levels = c('Pre-Tx Low', 'Post-Tx Low', 'Pre-Tx Medium/High', 'Post-Tx Medium/High'))

# with the paired lines
ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + geom_line(aes(x = Group, y = Freqs, group = paired), alpha = 0.3) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
#ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/01172023Luoma_PBMC_2responses_boxplot_myeloid_stage_pathresponse.png', width = 14, height = 8)

# without the paired lines
ggplot(df) + geom_boxplot(aes(x = Group, y = Freqs, color = Group, fill = Group), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Group, y = Freqs, color = Group), alpha = 0.8, position = position_jitterdodge()) + facet_wrap(~ Cluster, scales = 'free', nrow = 2) + theme_bw() + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), strip.text = element_text(size = 12)) + scale_shape_manual(values = 1:63) 
#ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/01172023Luoma_PBMC_2responses_boxplot_myeloid_stage_pathresponse_NOpairlines.png', width = 14, height = 8)
```

### •• [1/17/2023] Frequency boxplot wilcox test

```{r, warning = FALSE}
wilcox_by_condition(df, out = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/Boxplots/')
```



## • [1/17/2023] Label transfer

```{r}
# load in target datasets
query <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/luoma_tumor_myeloid_plusmetadata.RDS')
query <- query[,query$global_cluster == 'Myeloid']

load('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_myeloid_2021.rda')
ref <- hnc2[,hnc2$tissue == c('TIL')]
ref <- ref[,!ref$global.cluster4 == 'cDC2_CD33']
ref$global.cluster4[ref$global.cluster4 == 'DC3_LAMP3'] <- 'mregDC_LAMP3'
#ref$global.cluster4[ref$global.cluster4 == c('Mac_SPP1_PLTP')] <- 'Mac_SPP1'
ref$global.cluster4[ref$global.cluster4 == c('Mono_CD14_IL1B')] <- 'Mono_CD14'
ref$global.cluster4[ref$global.cluster4 == c('Mono_CD14_THBS1')] <- 'Mono_CD14'
ref$global.cluster4[ref$global.cluster4 == c('Mono_CD14_ID1')] <- 'Mono_CD14'

table(ref$tissue)
# TIL 
#9526 

table(ref$global.cluster4)
#  cDC1_CLEC9A     cDC2_CD1C        DC_pDC      Mac_C1QA      Mac_IL1B 
#          251           939           847          1216           774 
#     Mac_SPP1 Mac_SPP1_PLTP          Mast     Mono_CD14     Mono_CD16 
#         1837           573           304          1115           160 
#     Mono_Int      Mono_TIL  mregDC_LAMP3 
#           87           961           462 

hnc <- NULL
seurat <- NULL
hnc2 <- NULL
m <- NULL


ref_mini <- ref[, sample(colnames(ref), size = 5000, replace=F)]

ref <- NULL

# myeloid

query <- NormalizeData(query)
query <- FindVariableFeatures(query, selection.method = "vst", nfeatures = 2000)
query <- ScaleData(query, features = rownames(query))
query <- RunPCA(query, features = VariableFeatures(object = query), npcs = 100)
ElbowPlot(query, ndims = 100)

my.anchors <- FindTransferAnchors(reference = ref_mini, query = query, dims = 1:30)
my.predictions <- TransferData(anchorset = my.anchors, refdata = ref_mini$global.cluster4, dims = 1:30)
query <- AddMetaData(query, metadata = my.predictions)

saveRDS(query, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/UMAP_Luoma_labeltransfer_01182023.RDS')

color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3","#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", "#AA8282", "#D4B7B7", "#8600BF", "#BA5CE3", "#808000","#AEAE5C", "#1E90FF", "#00BFFF", "#56FF0D", "#FFFF00")

query <- RunUMAP(query, dims = 1:30)

query$Stage <- factor(query$Stage, levels=c('Pre-Tx', 'Post-Tx'))

DimPlot(query, group.by = 'predicted.id', cols = color_clusters, split.by = 'Stage') + ggtitle('')
ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/Stage_UMAP_Luoma_labeltransfer_01182023.png', width = 11, height =5, units = 'in', dpi = 300)

query$Path_response[query$Path_response %in% c('Medium', 'High')] <- 'Medium/High'

DimPlot(query, group.by = 'predicted.id', cols = color_clusters, split.by = 'Path_response') + ggtitle('')
ggsave('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/Tumor_figs/PathResponse_UMAP_Luoma_labeltransfer_01182023.png', width = 11, height =5, units = 'in', dpi = 300)
```

## • Label transfer (original)

```{r}
# load in target datasets
lmy <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/luoma_pbmc_globalmetadata.RDS')
lmy <- lmy[,lmy$CellType_ID == 'Myeloid']

hnc <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_all_annot_2022-06-30.RDS')

# load in reference datasets
cmy <- hnc[,hnc$mreg_cxcl9_globalcluster4 %like% c('%DC%', 'Mac%', 'Mon%')]


# check to be sure it is tumor only
#table(ccd8$tissue)
#table(ccd4$tissue)
table(cmy$tissue)

# group monocytes together as 3 subsets


hnc <- NULL
seurat <- NULL

# cd8

lcd8 <- NormalizeData(lcd8)
lcd8 <- FindVariableFeatures(lcd8, selection.method = "vst", nfeatures = 2000)
lcd8 <- ScaleData(lcd8, features = rownames(lcd8))
lcd8 <- RunPCA(lcd8, features = VariableFeatures(object = lcd8), npcs = 100)
ElbowPlot(lcd8, ndims = 100)

cd8.anchors <- FindTransferAnchors(reference = ccd8, query = lcd8, dims = 1:30)
cd8.predictions <- TransferData(anchorset = cd8.anchors, refdata = ccd8$mreg_cxcl9_globalcluster4, dims = 1:30)
lcd8 <- AddMetaData(lcd8, metadata = cd8.predictions)

DimPlot(lcd8, reduction = 'umap_CD8', group.by = 'predicted.id', cols = color_clusters)

saveRDS(lcd8, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/luoma_tumor_CD8_plusmetadata.RDS')

# cd4

lcd4 <- NormalizeData(lcd4)
lcd4 <- FindVariableFeatures(lcd4, selection.method = "vst", nfeatures = 2000)
lcd4 <- ScaleData(lcd4, features = rownames(lcd4))
lcd4 <- RunPCA(lcd4, features = VariableFeatures(object = lcd4), npcs = 100)
ElbowPlot(lcd4, ndims = 100)

cd4.anchors <- FindTransferAnchors(reference = ccd4, query = lcd4, dims = 1:30)
cd4.predictions <- TransferData(anchorset = cd4.anchors, refdata = ccd4$mreg_cxcl9_globalcluster4, dims = 1:30)
lcd4 <- AddMetaData(lcd4, metadata = cd4.predictions)

DimPlot(lcd4, reduction = 'umap_CD4', group.by = 'predicted.id', cols = color_clusters)

saveRDS(lcd4, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/luoma_tumor_CD4_plusmetadata.RDS')

# myeloid

lmy <- NormalizeData(lmy)
lmy <- FindVariableFeatures(lmy, selection.method = "vst", nfeatures = 2000)
lmy <- ScaleData(lmy, features = rownames(lmy))
lmy <- RunPCA(lmy, features = VariableFeatures(object = lmy), npcs = 100)
ElbowPlot(lmy, ndims = 100)

my.anchors <- FindTransferAnchors(reference = cmy, query = lmy, dims = 1:30)
my.predictions <- TransferData(anchorset = my.anchors, refdata = cmy$mreg_cxcl9_globalcluster4, dims = 1:30)
lmy <- AddMetaData(lmy, metadata = my.predictions)

DimPlot(lmy, reduction = 'umap_myeloid', group.by = 'predicted.id', cols = color_clusters)

saveRDS(lmy, file = '/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/luoma_tumor_myeloid_plusmetadata.RDS')
```



# ____________Luoma Cellchat______________

# RUNNING CELLCHAT

### •• [Low] Split by path response
```{r}
library(Seurat)
library(CellChat)
seu1 <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/luoma_tumor_myeloid_T_plusmetadata.RDS')
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/'
#path_response = 'Low'

seu <- seu1[,seu1$Path_response %in% path_response]

seu$cluster <- seu$predicted.id
seu$cluster[seu$cluster %like any% 'CD4%'] <- 'CD4'
seu$cluster[seu$cluster %like any% 'CD8%'] <- 'CD8'
seu$cluster[seu$cluster %like any% 'Treg%'] <- 'Treg'

metadata = 'cluster'

patients <- unique(seu$Patient_ID)
ccc_count <- vector('list', length(patients))
for (i in 1:length(patients)) {
  tmp <- subset(seu, Patient_ID == patients[i])
  tmp_cellchat <- creat_cellchat(tmp)
  ccc_count[[i]] <- tmp_cellchat@net$count
}

tab <- NULL
for (i in 1:nrow(ccc_count[[1]])) {
  for (j in 1:ncol(ccc_count[[1]])) {
    v <- c()
    for (k in 1:length(ccc_count)) {
      if (i<=nrow(ccc_count[[k]]) && j<=ncol(ccc_count[[k]])) {v <- c(v, ccc_count[[k]][i,j])} 
      else {v <- c(v, NA)}
    }
    tab <- rbind(tab, v)
  }
}
colnames(tab) <- patients
row_name = NULL
for (i in 1:nrow(ccc_count[[1]])) {
  for (j in 1:ncol(ccc_count[[1]])) {
    row_name <- c(row_name, paste(rownames(ccc_count[[1]])[i], colnames(ccc_count[[1]])[j],sep =  ":"))
  }
}

rownames(tab) <- row_name
write.csv(tab, file = paste(out, path_response[1], '_patient_by_interaction_table.csv', sep = ''))

# making a normalized version
#tab2 <- tab/apply(tab,1,max)
#meta <- seu[[c('Stage', 'Path_response', 'Patient_ID')]]
```

### •• [Medium/High] Split by path response
```{r}
library(Seurat)
library(CellChat)
seu1 <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/luoma_tumor_myeloid_T_plusmetadata.RDS')
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/'
path_response = c('Medium', 'High')

seu <- seu1[,seu1$Path_response %in% path_response]

seu$cluster <- seu$predicted.id
seu$cluster[seu$cluster %like any% 'CD4%'] <- 'CD4'
seu$cluster[seu$cluster %like any% 'CD8%'] <- 'CD8'
seu$cluster[seu$cluster %like any% 'Treg%'] <- 'Treg'

metadata = 'cluster'

patients <- unique(seu$Patient_ID)
ccc_count <- vector('list', length(patients))
for (i in 1:length(patients)) {
  tmp <- subset(seu, Patient_ID == patients[i])
  tmp_cellchat <- creat_cellchat(tmp)
  ccc_count[[i]] <- tmp_cellchat@net$count
}

tab <- NULL
for (i in 1:nrow(ccc_count[[1]])) {
  for (j in 1:ncol(ccc_count[[1]])) {
    v <- c()
    for (k in 1:length(ccc_count)) {
      if (i<=nrow(ccc_count[[k]]) && j<=ncol(ccc_count[[k]])) {v <- c(v, ccc_count[[k]][i,j])} 
      else {v <- c(v, NA)}
    }
    tab <- rbind(tab, v)
  }
}
colnames(tab) <- patients
row_name = NULL
for (i in 1:nrow(ccc_count[[1]])) {
  for (j in 1:ncol(ccc_count[[1]])) {
    row_name <- c(row_name, paste(rownames(ccc_count[[1]])[i], colnames(ccc_count[[1]])[j],sep =  ":"))
  }
}

rownames(tab) <- row_name
write.csv(tab, file = paste(out, path_response, '_patient_by_interaction_table.csv', sep = ''))

# making a normalized version
#tab2 <- tab/apply(tab,1,max)
#meta <- seu[[c('Stage', 'Path_response', 'Patient_ID')]]
```

### •• [Pre-Tx] Split by treatment stage
```{r}
library(Seurat)
library(CellChat)
seu1 <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/luoma_tumor_myeloid_T_plusmetadata.RDS')
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/'
stage <- 'Pre-Tx'

seu <- seu1[,seu1$Stage %in% stage]

seu$cluster <- seu$predicted.id
seu$cluster[seu$cluster %like any% 'CD4%'] <- 'CD4'
seu$cluster[seu$cluster %like any% 'CD8%'] <- 'CD8'
seu$cluster[seu$cluster %like any% 'Treg%'] <- 'Treg'

metadata = 'cluster'

patients <- unique(seu$Patient_ID)
ccc_count <- vector('list', length(patients))
for (i in 1:length(patients)) {
  tmp <- subset(seu, Patient_ID == patients[i])
  tmp_cellchat <- creat_cellchat(tmp)
  ccc_count[[i]] <- tmp_cellchat@net$count
}

tab <- NULL
for (i in 1:nrow(ccc_count[[1]])) {
  for (j in 1:ncol(ccc_count[[1]])) {
    v <- c()
    for (k in 1:length(ccc_count)) {
      if (i<=nrow(ccc_count[[k]]) && j<=ncol(ccc_count[[k]])) {v <- c(v, ccc_count[[k]][i,j])} 
      else {v <- c(v, NA)}
    }
    tab <- rbind(tab, v)
  }
}
colnames(tab) <- patients
row_name = NULL
for (i in 1:nrow(ccc_count[[1]])) {
  for (j in 1:ncol(ccc_count[[1]])) {
    row_name <- c(row_name, paste(rownames(ccc_count[[1]])[i], colnames(ccc_count[[1]])[j],sep =  ":"))
  }
}

rownames(tab) <- row_name
write.csv(tab, file = paste(out, stage, '_patient_by_interaction_table.csv', sep = ''))
```

### •• [Post-Tx] Split by treatment stage
```{r}
library(Seurat)
library(CellChat)
seu1 <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/luoma_tumor_myeloid_T_plusmetadata.RDS')
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/'
stage <- 'Post-Tx'

seu <- seu1[,seu1$Stage %in% stage]

seu$cluster <- seu$predicted.id
seu$cluster[seu$cluster %like any% 'CD4%'] <- 'CD4'
seu$cluster[seu$cluster %like any% 'CD8%'] <- 'CD8'
seu$cluster[seu$cluster %like any% 'Treg%'] <- 'Treg'

metadata = 'cluster'

patients <- unique(seu$Patient_ID)
ccc_count <- vector('list', length(patients))
for (i in 1:length(patients)) {
  tmp <- subset(seu, Patient_ID == patients[i])
  tmp_cellchat <- creat_cellchat(tmp)
  ccc_count[[i]] <- tmp_cellchat@net$count
}

tab <- NULL
for (i in 1:nrow(ccc_count[[1]])) {
  for (j in 1:ncol(ccc_count[[1]])) {
    v <- c()
    for (k in 1:length(ccc_count)) {
      if (i<=nrow(ccc_count[[k]]) && j<=ncol(ccc_count[[k]])) {v <- c(v, ccc_count[[k]][i,j])} 
      else {v <- c(v, NA)}
    }
    tab <- rbind(tab, v)
  }
}
colnames(tab) <- patients
row_name = NULL
for (i in 1:nrow(ccc_count[[1]])) {
  for (j in 1:ncol(ccc_count[[1]])) {
    row_name <- c(row_name, paste(rownames(ccc_count[[1]])[i], colnames(ccc_count[[1]])[j],sep =  ":"))
  }
}

rownames(tab) <- row_name
write.csv(tab, file = paste(out, stage, '_patient_by_interaction_table.csv', sep = ''))
```


### •• Testing Patient 18
```{r}
seu <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/luoma_tumor_myeloid_T_plusmetadata.RDS')

seu_pretx <- seu[,seu$Patient_ID == 'P23' & seu$Stage == 'Pre-Tx']
seu_posttx <- seu[,seu$Patient_ID == 'P23' & seu$Stage == 'Post-Tx']

seu_pretx$cluster <- seu_pretx$predicted.id
seu_pretx$cluster[seu_pretx$cluster %like any% 'CD4%'] <- 'CD4'
seu_pretx$cluster[seu_pretx$cluster %like any% 'CD8%'] <- 'CD8'
seu_pretx$cluster[seu_pretx$cluster %like any% 'Treg%'] <- 'Treg'

pretx <- creat_cellchat(seu_pretx)

seu_posttx$cluster <- seu_posttx$predicted.id
seu_posttx$cluster[seu_posttx$cluster %like any% 'CD4%'] <- 'CD4'
seu_posttx$cluster[seu_posttx$cluster %like any% 'CD8%'] <- 'CD8'
seu_posttx$cluster[seu_posttx$cluster %like any% 'Treg%'] <- 'Treg'

posttx <- creat_cellchat(seu_posttx)

pretx <- netAnalysis_computeCentrality(pretx)
posttx <- netAnalysis_computeCentrality(posttx)
object.list <- list(pretx =  pretx, posttx = posttx)
cellchat <- mergeCellChat(object.list, add.names = names(object.list))

s <- as.character(unique(cellchat@idents[[1]]))
s1 <- s[s %like any% c('%DC%', '%Mon%', 'Mac%')]
r <- s[s %like any% c('CD8%', 'CD4%', 'Treg%')]
netVisual_heatmap(cellchat, measure = "count", sources.use = s1, targets.use = r, title.name = paste('Differential number of interactions (', names(cellchat@idents[2]), ' (red)', ' vs ', names(cellchat@idents[1]), ' (blue)', ')', sep = ''))

png(paste(out, 'pre_tx_low_vs_high_response_differential_number_interactions_heatmap.png', sep = ''), width = 6, height = 4.5, units = 'in', res = 300)
gg2
dev.off()
```

# MAKING VOLCANO PLOTS
### •• Huy volcano plot (R vs. NR)
#### ••• Low response: Pre vs. Post-Tx (nothing significant)
```{r}
t <- read.csv('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Low_patient_by_interaction_table.csv')
path <- '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/'
rownames(t) <- t$X
t <- t[,-1]
meta <- read.csv('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/patient_metadata.csv')
meta <- unique(meta[,2:4])

p.val = fc = c()
ind1 <- which(colnames(t) %in% meta$Patient_ID[which(meta$Stage == 'Post-Tx')])
ind2 <- which(colnames(t) %in% meta$Patient_ID[which(meta$Stage == 'Pre-Tx')])
for (i in 1:nrow(t)) {
  tmp <- wilcox.test(as.numeric(t[i, ind1]), as.numeric(t[i, ind2]), na.rm=T)
  p.val <- c(p.val, tmp$p.value)
  fc <- c(fc, log2(mean(as.numeric(t[i, ind2]))/mean(as.numeric(t[i, ind1]))))
}

ind <- intersect(which(p.val < 0.05), which(fc > 0.5))
print(paste(length(ind), 'significant interactions found positively enriched'))
ind2 <- intersect(which(p.val < 0.05), which(fc < -0.5))
print(paste(length(ind), 'significant interactions found negatively enriched'))
tab <- data.frame(fc = fc, pval = -log10(p.val))
rownames(tab) <- rownames(t)

tab[is.na(tab)] <- 0

plot(tab$fc, tab$pval, xlab = 'log2FC(#interactions)', ylab = '-log10(p)')
points(tab$fc[ind], tab$pval[ind], col = 'red', pch = 16)
if (length(ind) > 0){
  text(tab$fc[ind], tab$pval[ind], rownames(tab)[ind], cex = 0.5)
}
points(tab$fc[ind2], tab$pval[ind2], col = 'blue', pch = 16)
if (length(ind2) > 0){
  text(tab$fc[ind2], tab$pval[ind2], rownames(tab)[ind2], cex = 0.5)
}

saveRDS(tab, file = paste(path, 'Low_prevspost_tx_volcanoplot_input.csv'))
#test <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/luoma_tumor_myeloid_T_plusmetadata.RDS') 
```

#### ••• Medium/High: Pre vs. Post-Tx

```{r}
t <- read.csv('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/patient_by_interaction_table_v2.csv')
path <- '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/'
#t <- tab
stage <- 'Pre-Tx'
rownames(t) <- t$X
t <- t[,-1]
meta <- read.csv('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/patient_metadata.csv')
meta <- unique(meta[,2:4])

p.val = fc = c()
#ind1 <- which(colnames(t) %in% meta$Patient_ID[which(meta$Path_response == 'Low')])
ind1 <- which(colnames(t) %in% meta$Patient_ID[which(meta$Stage == 'Post-Tx')])
#ind1 <- intersect(ind1_1, ind1_2)
#ind2_1 <- which(colnames(t) %in% meta$Patient_ID[which(meta$Path_response %in% c('High', 'Medium'))])
ind2 <- which(colnames(t) %in% meta$Patient_ID[which(meta$Stage == 'Pre-Tx')])
#ind2 <- intersect(ind2_1, ind2_2)
for (i in 1:nrow(t)) {
  tmp <- wilcox.test(as.numeric(t[i, ind1]), as.numeric(t[i, ind2]), na.rm=T)
  p.val <- c(p.val, tmp$p.value)
  fc <- c(fc, log2(mean(as.numeric(t[i, ind2]))/mean(as.numeric(t[i, ind1]))))
}

ind <- intersect(which(p.val < 0.05), which(fc > 0.5))
ind2 <- intersect(which(p.val < 0.05), which(fc < -0.5))
tab <- data.frame(fc = fc, pval = -log10(p.val))
rownames(tab) <- rownames(t)

tab[is.na(tab)] <- 0

png(paste(path,'MedHigh_prevsposttx_volcano_plot.png', sep = ''), width = 7, height = 7, res = 300, units = 'in')
EnhancedVolcano::EnhancedVolcano(tab, lab = rownames(tab), x = 'fc', y = 'pval', legendLabels = NULL, legendIconSize = 0,legendDropLevels = TRUE, legendPosition = 'right')
dev.off()

plot(tab$fc, tab$pval, xlab = 'log2FC(#interactions)', ylab = '-log10(p)')
points(tab$fc[ind], tab$pval[ind], col = 'red', pch = 16)
if (length(ind) > 0){
  text(tab$fc[ind], tab$pval[ind], rownames(tab)[ind], cex = 0.5)
}
points(tab$fc[ind2], tab$pval[ind2], col = 'blue', pch = 16)
if (length(ind2) > 0){
  text(tab$fc[ind2], tab$pval[ind2], rownames(tab)[ind2], cex = 0.5)
}
saveRDS(tab, file = paste(path, 'MedHigh_prevspost_tx_volcanoplot_input.csv'))
#test <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/luoma_tumor_myeloid_T_plusmetadata.RDS') 
```



### •• Huy volcano plot (pre vs. post treatment)
#### ••• Post-Tx: R vs. NR
```{r}
t <- read.csv('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Post-Tx_patient_by_interaction_table.csv')
path <- '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/'
#t <- tab
rownames(t) <- t$X
condition_level2 <- 'Low'
condition_level1 <- c('Medium', 'High')
t <- t[,-1]
meta <- read.csv('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/patient_metadata.csv')
meta <- unique(meta[,2:4])

p.val = fc = c()
ind1 <- which(colnames(t) %in% meta$Patient_ID[which(meta$Path_response == condition_level2)])
#ind1 <- which(colnames(t) %in% meta$Patient_ID[which(meta$Stage == 'Post-Tx')])
#ind1 <- intersect(ind1_1, ind1_2)
ind2 <- which(colnames(t) %in% meta$Patient_ID[which(meta$Path_response %in% condition_level1)])
#ind2 <- which(colnames(t) %in% meta$Patient_ID[which(meta$Stage == 'Pre-Tx')])
#ind2 <- intersect(ind2_1, ind2_2)
for (i in 1:nrow(t)) {
  tmp <- wilcox.test(as.numeric(t[i, ind1]), as.numeric(t[i, ind2]), na.rm=T)
  p.val <- c(p.val, tmp$p.value)
  fc <- c(fc, log2(mean(as.numeric(t[i, ind2]))/mean(as.numeric(t[i, ind1]))))
}

ind <- intersect(which(p.val < 0.05), which(fc > 0.5))
ind2 <- intersect(which(p.val < 0.05), which(fc < -0.5))
tab <- data.frame(fc = fc, pval = -log10(p.val))
rownames(tab) <- rownames(t)

tab[is.na(tab)] <- 0

tab <- tab %>%
  mutate(Color = case_when(
    pval < 0.05 ~ "p < 0.05",
    pval < 0.01 ~ "p < 0.01", 
    pval > 0.05 ~ 'NS'))

sub <- tab[!tab$Color == 'NS',]

png(paste(path,'Post-Tx_RvsNR_volcano_plot.png', sep = ''), width = 7, height = 7, res = 300, units = 'in')
ggplot(tab, aes(x = fc, y = pval, color = Color, label = rownames(tab)) + geom_vline(xintercept = c(0.5, -0.5), lty = "dashed") + geom_hline(yintercept = -log10(0.05), lty = "dashed") + geom_point() + labs(x = paste("Enriched in ", condition_level2, " <-----  log2(FC)  ------> Enriched in ",   condition_level1, sep = ''), y = "Significance, -log10(pval)", color = "Color") + scale_color_manual(values = c(`p < 0.01` = "dodgerblue", `p < 0.05` = "lightblue", `NS` = "gray"), guide = guide_legend(override.aes = list(size = 4))) + scale_y_continuous(expand = expansion(mult = c(0,0.05))) + ggrepel::geom_text_repel(data = sub, size = 4, point.padding = 0.15, color = "black", min.segment.length = .1, box.padding = .2, lwd = 2,max.overlaps = 50) + theme_classic(base_size = 16) #+ theme(legend.position = "bottom") #+ ggtitle(paste('Significantly upregulated genes in', topmeta_oi, sep = ' '), subtitle = paste(count1, '|', count2, '|', 'paired =', paired))
#EnhancedVolcano::EnhancedVolcano(tab, lab = rownames(tab), x = 'fc', y = 'pval', legendLabels = NULL, legendIconSize = 0,legendDropLevels = TRUE, legendPosition = 'right')
dev.off()

plot(tab$fc, tab$pval, xlab = 'log2FC(#interactions)', ylab = '-log10(p)')
points(tab$fc[ind], tab$pval[ind], col = 'red', pch = 16)
if (length(ind) > 0){
  text(tab$fc[ind], tab$pval[ind], rownames(tab)[ind], cex = 0.5)
}
points(tab$fc[ind2], tab$pval[ind2], col = 'blue', pch = 16)
if (length(ind2) > 0){
  text(tab$fc[ind2], tab$pval[ind2], rownames(tab)[ind2], cex = 0.5)
}
saveRDS(tab, file = paste(path, 'MedHigh_prevspost_tx_volcanoplot_input.csv'))
#test <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/luoma_tumor_myeloid_T_plusmetadata.RDS') 
```

#### ••• Non-responders
```{r}
t <- read.csv('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/patient_by_interaction_table_v2.csv')
path <- '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/'
#t <- tab
response <- c('Low')
rownames(t) <- t$X
t <- t[,-1]
meta <- read.csv('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/patient_metadata.csv')
meta <- unique(meta[,2:4])

p.val = fc = c()
ind1_1 <- which(colnames(t) %in% meta$Patient_ID[which(meta$Path_response %in% response)])
ind1_2 <- which(colnames(t) %in% meta$Patient_ID[which(meta$Stage == 'Pre-Tx')])
ind1 <- intersect(ind1_1, ind1_2)
ind2_1 <- which(colnames(t) %in% meta$Patient_ID[which(meta$Path_response %in% response)])
ind2_2 <- which(colnames(t) %in% meta$Patient_ID[which(meta$Stage == 'Post-Tx')])
ind2 <- intersect(ind2_1, ind2_2)
for (i in 1:nrow(t)) {
  tmp <- wilcox.test(as.numeric(t[i, ind1]), as.numeric(t[i, ind2]), na.rm=T)
  p.val <- c(p.val, tmp$p.value)
  fc <- c(fc, log2(mean(as.numeric(t[i, ind2]))/mean(as.numeric(t[i, ind1]))))
}

ind <- intersect(which(p.val < 0.05), which(fc > 0.5))
ind2 <- intersect(which(p.val < 0.05), which(fc < -0.5))
tab <- data.frame(fc = fc, pval = -log10(p.val))
rownames(tab) <- rownames(t)

tab[is.na(tab)] <- 0

#png(paste(path, response, '_volcano_plot.png', sep = ''), width = 7, height = 7, res = 300, units = 'in')
#EnhancedVolcano::EnhancedVolcano(tab, lab = rownames(tab), x = 'fc', y = 'pval', legendLabels = NULL, legendIconSize = 0,legendDropLevels = TRUE, legendPosition = 'right')
#dev.off()

plot(tab$fc, tab$pval, xlab = 'log2FC(#interactions)', ylab = '-log10(p)')
points(tab$fc[ind], tab$pval[ind], col = 'red', pch = 16)
if (length(ind) > 0){
  text(tab$fc[ind], tab$pval[ind], rownames(tab)[ind], cex = 0.5)}
points(tab$fc[ind2], tab$pval[ind2], col = 'blue', pch = 16)
if (length(ind2) > 0){
  text(tab$fc[ind2], tab$pval[ind2], rownames(tab)[ind2], cex = 0.5)}
```


### •• Add metadata
```{r}
seu <- readRDS('/Volumes/hqdinh2/Projects/Public_Data/Luoma_HNC/Tumor/luoma_tumor_myeloid_T_plusmetadata.RDS')

pre_nr <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Pre%_Low_cellchat.RDS')
pre_r <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/pre_MedHigh_cellchat.RDS')
post_nr <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Post%_Low_cellchat.RDS')
post_r <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/post_MedHigh_cellchat.RDS')

seu1 <- seu[,colnames(seu) %in% intersect(colnames(pre_nr@data.signaling), colnames(seu))]
seu1 <- seu1@meta.data

test <- CellChat::addMeta(pre_nr, meta = seu1)
```


## • Comparing # of interactions

### •• Pre-Tx: Low vs. med/high response

```{r}
out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/'
nr <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Pre%_Low_cellchat.RDS')
r <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/pre_MedHigh_cellchat.RDS')

nr <- updateCellChat(nr)
r <- updateCellChat(r)

nr <- netAnalysis_computeCentrality(nr)
r <- netAnalysis_computeCentrality(r)
object.list <- list(NR =  nr, R = r)
cellchat <- mergeCellChat(object.list, add.names = names(object.list))

s <- as.character(unique(cellchat@idents[[1]]))
s1 <- s[s %like any% c('%DC%', '%Mon%', 'Mac%')]
r <- s[s %like any% c('CD8%', 'CD4%', 'Treg%')]
gg2 <- netVisual_heatmap(cellchat, measure = "count", sources.use = s1, targets.use = r, title.name = paste('Differential number of interactions (', names(cellchat@idents[2]), ' (red)', ' vs ', names(cellchat@idents[1]), ' (blue)', ')', sep = ''))

png(paste(out, 'pre_tx_low_vs_high_response_differential_number_interactions_heatmap.png', sep = ''), width = 6, height = 4.5, units = 'in', res = 300)
gg2
dev.off()
```

### •• Post-Tx: Low vs. med/high response
```{r}
out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/'
nr <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Post%_Low_cellchat.RDS')
r <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/post_MedHigh_cellchat.RDS')

nr <- updateCellChat(nr)
r <- updateCellChat(r)

nr <- netAnalysis_computeCentrality(nr)
r <- netAnalysis_computeCentrality(r)
object.list <- list(NR =  nr, R = r)
cellchat <- mergeCellChat(object.list, add.names = names(object.list))

s <- as.character(unique(cellchat@idents[[1]]))
s1 <- s[s %like any% c('%DC%', '%Mon%', 'Mac%')]
r <- s[s %like any% c('CD8%', 'CD4%', 'Treg%')]
gg2 <- netVisual_heatmap(cellchat, measure = "count", sources.use = s1, targets.use = r, title.name = paste('Differential number of interactions (', names(cellchat@idents[2]), ' (red)', ' vs ', names(cellchat@idents[1]), ' (blue)', ')', sep = ''))

png(paste(out, 'post_tx_low_vs_high_response_differential_number_interactions_heatmap.png', sep = ''), width = 6, height = 4.5, units = 'in', res = 300)
gg2
dev.off()
```

## ••• Responders: Pre- vs. Post-Tx
```{r}
out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/'
pre <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/pre_MedHigh_cellchat.RDS')
post <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/post_MedHigh_cellchat.RDS')

pre <- updateCellChat(pre)
post <- updateCellChat(post)

pre <- netAnalysis_computeCentrality(pre)
post <- netAnalysis_computeCentrality(post)
object.list <- list(pre =  pre, post = post)
cellchat <- mergeCellChat(object.list, add.names = names(object.list))

s <- as.character(unique(cellchat@idents[[1]]))
s1 <- s[s %like any% c('%DC%', '%Mon%', 'Mac%')]
r <- s[s %like any% c('CD8%', 'CD4%', 'Treg%')]
gg2 <- netVisual_heatmap(cellchat, measure = "count", sources.use = s1, targets.use = r, title.name = paste('Differential number of interactions (', names(cellchat@idents[2]), ' (red)', ' vs ', names(cellchat@idents[1]), ' (blue)', ')', sep = ''), cluster.cols = TRUE)

#, row.show = c('Blood_CD14_Mono', 'CD16_Mono', 'cDC1_CLEC9A', 'cDC2_CD1C', 'cDC2_CD33', 'DC_pDC', 'Mac_C1QA', 'Mac_IL1B', 'Mac_SPP1', 'Mac_SPP1_PLTP', 'mregDC_LAMP3', 'TIL_Mono'), col.show = c('CD4', 'CD8', 'Treg')

png(paste(out, 'Responders_pre_vs_post_tx_differential_number_interactions_heatmap.png', sep = ''), width = 6, height = 4.5, units = 'in', res = 300)
gg2
dev.off()
```


### •• Non-responders: Pre- vs. Post-Tx
```{r}
out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/'
pre <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Pre%_Low_cellchat.RDS')
post <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Post%_Low_cellchat.RDS')

pre <- updateCellChat(pre)
post <- updateCellChat(post)

pre <- netAnalysis_computeCentrality(pre)
post <- netAnalysis_computeCentrality(post)
object.list <- list(pre =  pre, post = post)
cellchat <- mergeCellChat(object.list, add.names = names(object.list))

s <- as.character(unique(cellchat@idents[[1]]))
s1 <- s[s %like any% c('%DC%', '%Mon%', 'Mac%')]
r <- s[s %like any% c('CD8%', 'CD4%', 'Treg%')]
gg2 <- netVisual_heatmap(cellchat, measure = "count", sources.use = s1, targets.use = r, title.name = paste('Differential number of interactions (', names(cellchat@idents[2]), ' (red)', ' vs ', names(cellchat@idents[1]), ' (blue)', ')', sep = ''), cluster.cols = TRUE)

#, row.show = c('Blood_CD14_Mono', 'CD16_Mono', 'cDC1_CLEC9A', 'cDC2_CD1C', 'cDC2_CD33', 'DC_pDC', 'Mac_C1QA', 'Mac_IL1B', 'Mac_SPP1', 'Mac_SPP1_PLTP', 'mregDC_LAMP3', 'TIL_Mono'), col.show = c('CD4', 'CD8', 'Treg')

png(paste(out, 'Non-responders_pre_vs_post_tx_differential_number_interactions_heatmap.png', sep = ''), width = 6, height = 4.5, units = 'in', res = 300)
gg2
dev.off()
```


### •• Log2FC of pre- and post-Tx sample counts
```{r}
pre_tx <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/luoma_PreTx_combined.RDS')
post_tx <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/luoma_PostTx_combined.RDS')


s <- as.character(unique(pre_tx@idents[[1]]))
s1 <- s[s %like any% c('%DC%', '%Mon%', 'Mac%')]
r <- s[s %like any% c('CD8%', 'CD4%', 'Treg%')]


# generate matrices with pre and post treatment data

pre1 <- netVisual_heatmap(pre_tx, measure = "count", sources.use = s1, targets.use = r, title.name = paste('Differential number of interactions (', names(pre_tx@idents[2]), ' (red)', ' vs ', names(pre_tx@idents[1]), ' (blue)', ')', sep = ''))
pre2 <- pre1@matrix 
pre3 <- pre2[rowSums(is.na(pre2))<ncol(pre2),colSums(is.na(pre2))<nrow(pre2)]
pre3[is.na(pre3)] = 0

post1 <- netVisual_heatmap(post_tx, measure = "count", sources.use = s1, targets.use = r, title.name = paste('Differential number of interactions (', names(post_tx@idents[2]), ' (red)', ' vs ', names(post_tx@idents[1]), ' (blue)', ')', sep = ''))
post2 <- post1@matrix
post3 <- post2[rowSums(is.na(post2))<ncol(post2),colSums(is.na(post2))<nrow(post2)]
post3[is.na(post3)] = 0


# manipulating data to deal with zeroes in log2fc matrix

fc <- (post3-pre3)/ abs(pre3)
max <- max(abs(fc)[abs(fc) != max(abs(fc))])  
fc[fc == Inf] <- max + 1
fc[fc == -Inf] <- max - 1
test <- log2(fc)


png('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Figures/pre_vs_post_foldchange.png', width = 4.5, height = 5, units = 'in', res = 300)
ComplexHeatmap::Heatmap(fc, row_names_side = 'left', row_dend_side = 'right', rect_gp = grid::gpar(col = "white", lwd = 2), row_title = 'Sending cell subset', column_title = 'Receiving cell subset', column_title_side = 'bottom', column_dend_height = unit(0.5, 'cm'), row_dend_width = unit(0.5, 'cm'), cluster_columns = FALSE, name = 'Fold change', col = c('dodgerblue3', 'white', 'firebrick3'))
dev.off()
```


## • Get unique interactions from the heatmap

```{r}
object <- r
sources.use = NULL
targets.use = NULL
signaling = NULL
pairLR.use = NULL
remove.isolate = TRUE
thresh = 0.05
return.object = TRUE

net <- object@net
if (is.null(sources.use) & is.null(targets.use) & is.null(signaling) & is.null(pairLR.use)) {
  prob <- net$prob
  pval <- net$pval
  pval[prob == 0] <- 1
  prob[pval >= thresh] <- 0
  net$count <- apply(prob > 0, c(1,2), sum)
  net$weight <- apply(prob, c(1,2), sum)
  net$weight[is.na(net$weight)] <- 0
  net$count[is.na(net$count)] <- 0
} else {
  df.net <- subsetCommunication(object, slot.name = "net", sources.use = sources.use, 
                                targets.use = targets.use, signaling = signaling,
                                pairLR.use = pairLR.use, thresh = thresh)
  df.net$source_target <- paste(df.net$source, df.net$target, sep = "_")
  df.net2 <- df.net %>% group_by(source_target) %>% summarize(count = n(), .groups = 'drop')
  df.net3 <- df.net %>% group_by(source_target) %>% summarize(prob = sum(prob), .groups = 'drop')
  df.net2$prob <- df.net3$prob
  a <- stringr::str_split(df.net2$source_target, "_", simplify = T)
  df.net2$source <- as.character(a[, 1])
  df.net2$target <- as.character(a[, 2])
  cells.level <- levels(object@idents)
  df.net2$source <- factor(df.net2$source, levels = cells.level)
  df.net2$target <- factor(df.net2$target, levels = cells.level)

  count <- tapply(df.net2[["count"]], list(df.net2[["source"]], df.net2[["target"]]), sum)
  prob <- tapply(df.net2[["prob"]], list(df.net2[["source"]], df.net2[["target"]]), sum)
  net$count <- count
  net$weight <- prob
  net$weight[is.na(net$weight)] <- 0
  net$count[is.na(net$count)] <- 0
}
```

## • Plotting individual interactions (CXCL9-CXCR3)

```{r}
#nr <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Post%_Low_cellchat.RDS')
#r <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Post%_MediumHigh_cellchat.RDS')

nr <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Pre%_Low_cellchat.RDS')
r <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Pre%_MediumHigh_cellchat.RDS')

seu <- nr

senders <- rownames(seu@netP$prob)[rownames(seu@netP$prob) %like any% c('%Mac%', '%Mon%', '%DC%')]

receivers <- rownames(seu@netP$prob)[rownames(seu@netP$prob) %like any% c('%CD4%', '%CD8%', '%Treg%')]


png('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Pre_Low_cellchat_CXCL9_CXCR3_circos.png', width = 7, height = 7, units = 'in', res = 500)
netVisual_individual(seu, signaling = 'CXCL', pairLR.use = 'CXCL9_CXCR3', sources.use = senders, targets.use = receivers)
dev.off()
```

## • Comparison bubble plot

```{r}
out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/'
nr <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Pre%_Low_cellchat.RDS')
r <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/pre_MedHigh_cellchat.RDS')

nr <- netAnalysis_computeCentrality(nr)
r <- netAnalysis_computeCentrality(r)
object.list <- list(NR =  nr, R = r)
cellchat <- mergeCellChat(object.list, add.names = names(object.list))

s <- as.character(unique(luoma$global_cluster2))
s1 <- s[s %like any% c('%DC%', '%Mon%', 'Mac%')]
r <- s[s %like any% c('CD8%', 'CD4%', 'Treg%')]

gg1 <- rankNet(cellchat, mode = "comparison", stacked = T, do.stat = TRUE, sources.use = s1, targets.use = r)
gg2 <- rankNet(cellchat, mode = "comparison", stacked = F, do.stat = TRUE, sources.use = s1, targets.use = r)

png(paste(out, 'Pre_treatment_Relative_information_flow.png', sep = ''), width = 5, height = 8, units = 'in', res = 300)
gg1
dev.off()

pdf(paste(out, 'Pre_treatment_Bubble_plots_signaling_diffs.pdf', sep = ''), height = 6, width = 5)
for (x in s1){
  gg <- netVisual_bubble(cellchat, sources.use = x, targets.use = r,  comparison = c(1, 2), angle.x = 45)
  plot(gg)
}
dev.off()
```

# _____CILLO ALL CELLCHAT_____

```{r}
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/'
hpvpos <- readRDS(file = paste(out, 'HPV+_cellchat.RDS', sep = ''))
hpvneg <- readRDS(file = paste(out, 'HPV-_cellchat.RDS', sep = ''))
hpvpos <- netAnalysis_computeCentrality(hpvpos)
hpvneg <- netAnalysis_computeCentrality(hpvneg)
object.list <- list(HPVpos =  hpvpos, HPVneg = hpvneg)
cellchat <- mergeCellChat(object.list, add.names = names(object.list))
cellchat
```

## • Heatmap of differential \# of interactions

```{r}
s <- as.character(unique(cellchat@idents$HPVpos))
s1 <- s[s %like any% c('%DC%', 'Mac%', 'Mon%')]
r <- s[s %like any% c('CD8%', 'CD4%', 'Treg%')]
gg2 <- netVisual_heatmap(cellchat, measure = "count", sources.use = s1, targets.use = r, title.name = paste('Differential number of interactions (', names(cellchat@idents[2]), ' (red)', ' vs ', names(cellchat@idents[1]), ' (blue)', ')', sep = ''))

out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/'
png(paste(out, 'myeloid_to_T_heatmap.png', sep = ''), width = 7, height = 6, units = 'in', res = 300)
gg2
```

## • Plotting single datasets

```{r}
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/'
hpvpos <- readRDS(file = paste(out, 'HPV+_cellchat.RDS', sep = ''))
obj <- hpvpos
pathways.show <- c("CXCL") 

s <- as.character(unique(obj@idents))
s1 <- s[s %like any% c('%DC%', 'Mac%', 'Mon%')]
r <- s[s %like any% c('CD8%', 'CD4%')]

png(paste(out, pathways.show, 'signaling.png', sep = ''), width = 9, height = 9, units = 'in', res = 300)
netVisual_aggregate(obj, signaling = pathways.show, layout = "circle", sources.use = s1, targets.use = r)
dev.off()

netAnalysis_contribution(obj, signaling = pathways.show)
```

## • Compute different \# of interactions

```{r}
gg1 <- compareInteractions(cellchat, show.legend = F, group = c(1,2))
gg2 <- compareInteractions(cellchat, show.legend = F, group = c(1,2), measure = "weight")

png(paste(out, 'Overall_interaction_num_strength.png', sep = ''), width = 8, height = 5, units = 'in', res = 300)
gg1 + gg2
dev.off()
```

## • Differential \# of interactions/interaction strength in HPV+ vs. HPV-

```{r}
group.cellType <- c(rep("CD8", 5), rep("CD4", 4), rep("Treg", 2), rep("DC", 6), rep("Mono", 7), rep("Mac", 4))
group.cellType <- factor(group.cellType, levels = c('CD8', 'CD4', 'Treg', 'DC', 'Mono', 'Mac'))
object.list <- lapply(object.list, function(x) {mergeInteractions(x, group.cellType)})
cellchat_test <- mergeCellChat(object.list, add.names = names(object.list))

weight.max <- getMaxWeight(object.list, slot.name = c("idents", "net", "net"), attribute = c("idents","count", "count.merged"))

png(paste(out, 'Global_#interaction.png', sep = ''), width = 7, height = 5, units = 'in', res = 300)
par(mfrow = c(1,2), xpd=TRUE)
for (i in 1:length(object.list)) {
  netVisual_circle(object.list[[i]]@net$count.merged, weight.scale = T, label.edge= T, edge.weight.max = weight.max[3], edge.width.max = 12, title.name = paste0("Number of interactions - ", names(object.list)[i]), sources.use = c('Mac', 'Mono', 'DC'), targets.use = c('CD8', 'CD4', 'Treg'))
}
dev.off()
```

## • Signaling changes of particular subsets

```{r}
pdf(paste(out, 'Signaling_changes_per_subset.pdf'))
for (x in unique(cellchat@idents$HPVneg)){
  gg <- netAnalysis_signalingChanges_scatter(cellchat, idents.use = x, signaling.exclude = "MIF")
  plot(gg)
}
dev.off()
```

## • Bubble plot of signaling differences

```{r}
s <- as.character(unique(cellchat@idents$HPVneg))
s <- s[s %like any% c('%DC%', 'Mac%', 'Mon%')]

pdf(paste(out, 'Bubble_plots_signaling_diffs.pdf', sep = ''))
for (x in s){
  gg <- netVisual_bubble(cellchat, sources.use = x, targets.use = c('CD8_1', 'CD8_2', 'CD8_3', 'CD8_4', 'CD8_5', 'CD4_1', 'CD4_2', 'CD4_3', 'CD4_4', 'Treg_1', 'Treg_2'),  comparison = c(1, 2), angle.x = 45)
  plot(gg)
}
dev.off()
```

## • Bubble plot of up and downregulated signaling per sender

```{r}
pdf(paste(out, 'Bubble_plots_upregulated_downregulated.pdf', sep = ''), width = 12)
for (x in s){
  gg1 <- netVisual_bubble(cellchat, sources.use = x, targets.use = c('CD8_1', 'CD8_2', 'CD8_3', 'CD8_4', 'CD8_5', 'CD4_1', 'CD4_2', 'CD4_3', 'CD4_4', 'Treg_1', 'Treg_2'),  comparison = c(1, 2), max.dataset = 2, title.name = "Increased signaling in HPV-", angle.x = 45, remove.isolate = T)
  gg2 <- netVisual_bubble(cellchat, sources.use = x, targets.use = c('CD8_1', 'CD8_2', 'CD8_3', 'CD8_4', 'CD8_5', 'CD4_1', 'CD4_2', 'CD4_3', 'CD4_4', 'Treg_1', 'Treg_2'),  comparison = c(1, 2), max.dataset = 1, title.name = "Decreased signaling in HPV-", angle.x = 45, remove.isolate = T)
  plot(gg1 + gg2)
}
dev.off()
```


# CellChat

## • JOSH COMPARISON SCATTERPLOT

```{r}
library(dplyr)
library(tidyverse)
library(CellChat)

setwd("/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/Low_pathological_response/")
list.files()

# cellchat object
cc <- readRDS("cellchat.Rds")

# your interactions scores are held here:
# x axis are senders, y axis are receivers
cc@net$prob

# the names of the interzctions:
ints <- cc@LR$LRsig[,1]

# extracting one interaction:
cc@net$prob[,,ints[1]] # array slices are [sender, receiver, interaction]

all_interactions <- list()
for (int in ints){
  message(int)
  this_int <- cc@net$prob[,,int] %>% as.data.frame()
  
  this_df <- this_int %>% 
                    rownames_to_column(var = "sender") %>%
                    reshape2::melt() %>%
                    mutate(int)
 
  colnames(this_df) <- c("sender", "receiver", "score", "interaction")
   
  all_interactions[[int]] <- this_df
}

# collapsing to single dataframe:
data1 <- do.call(rbind, all_interactions)

#r <- data1
#r$R_score <- r$score
#r$score <- NULL

#nr <- data1
#nr$NR_score <- nr$score
#nr$score <- NULL

# here I would merge the two object (one form HPV+, HPV-)
# merge(data1, data2, by = c("sender", "receiver", "interaction")
# you'll end up with two score columns and then can plot them with ggplot

to_plot <- c('PDCD1LG2_PDCD1%', 'LGALS9_CD45%', 'FN1_ITGAV_ITGB8%', 'FN1_ITGAV_ITGB3%', 'CXCL9_CXCR3%', 'CD274_PDCD1%', 'CCL7_CCR2%', 'CCL6_CCR2%')
n1 <- nr[(nr$interaction %like any% to_plot & nr$sender %in% c('mregDC_CXCL9hi', 'Mac_C1QA', 'Mono_Int') & nr$receiver %like any% c('CD8%', 'CD4%', 'Treg%')),]
r1 <- r[(r$interaction %like any% to_plot & r$sender %in% c('mregDC_CXCL9hi', 'Mac_C1QA', 'Mono_Int') & r$receiver %like any% c('CD8%', 'CD4%', 'Treg%')),]

all <- merge(r1, n1, by = 0, all.x = FALSE, all.y = FALSE, no.dups = TRUE)
all$srlr <- paste(all$sender.x, all$receiver.x, all$interaction.x, sep = '_')
all_f <- all[c('srlr', 'R_score', 'NR_score')]

all_f$color <- 'shared'
all_f$color[all_f$R_score > 0 & all_f$NR_score == 0] <- 'R-specific'
all_f$color[all_f$NR_score > 0 & all_f$R_score == 0] <- 'NR-specific'

all_f %>% ggplot(aes(R_score, NR_score, color = color)) + 
  geom_point(alpha = 0.25, size = 8) + 
  labs(y = 'Non-Responder SRLR score', x = 'Responder SRLR score') + 
  theme_classic() + 
  ggrepel::geom_text_repel(data=subset(all_f, color == 'NR-specific' | color == 'R-specific'), aes(R_score, NR_score, label = srlr)) + 
  scale_color_manual(values = c("dodgerblue", "indianred", "gray50"))
ggsave('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/NR_vs_R_SRLR_scatterplot_comparison.png', units = 'in', width = 6, height = 5, dpi = 300)

write.csv(all, file = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_Luoma_myeloidtoT/NR_vs_R_SRLR_scatterplot_comparison.csv')
```

## • Plotting weights for myeloid to T interactions

```{r}
mon <- readRDS('/Volumes/hqdinh2-1/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/Mon_to_T/cellchat.RDS')

mac <- readRDS('/Volumes/hqdinh2-1/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/Mac_to_T/cellchat.RDS')

dc <- readRDS('/Volumes/hqdinh2-1/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/DC_to_T/cellchat.RDS')

cellchat <- mon
out <- '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/Mon_to_T/'

cellchat_weights_counts <- function(cellchat, out){
  pdf(paste(out, 'Cellchat_interaction_weights_counts.pdf', sep = ''))
  pheatmap::pheatmap(cellchat@net$weight, main = 'Interaction weights, Raw')
  pheatmap::pheatmap(scale(cellchat@net$weight), 
                     main = 'Interaction weights, Z-score normalized')
  pheatmap::pheatmap(cellchat@net$count, main = 'Interaction counts, Raw')
  pheatmap::pheatmap(scale(cellchat@net$count), 
                     main = 'Interaction counts, Z-score normalized')
  test <- cellchat@net$count *cellchat@net$weight
  pheatmap::pheatmap(test, main = "Interaction counts * weight, Raw")
  pheatmap::pheatmap(scale(test), main = "Interaction counts * weight, Z-score normalized")
  dev.off()
}

cellchat_weights_counts(mon, out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/Mon_to_T/')

cellchat_weights_counts(dc, out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/DC_to_T/')

cellchat_weights_counts(mac, out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/Mac_to_T/')
```

## • Testing functions

```{r}
object <- mon
signaling = NULL
pattern = c("all")
slot.name = "netP"
color.use = NULL 
color.heatmap = "BuGn"
title = NULL
width = 10
height = 8 
font.size = 8
font.size.title = 10
cluster.rows = FALSE

if (length(slot(object, slot.name)$centr) == 0) {
  stop("Please run `netAnalysis_computeCentrality` to compute the network centrality scores! ")
}
centr <- slot(object, slot.name)$centr
outgoing <- matrix(0, nrow = nlevels(object@idents), ncol = length(centr))
incoming <- matrix(0, nrow = nlevels(object@idents), ncol = length(centr))
dimnames(outgoing) <- list(levels(object@idents), names(centr))
dimnames(incoming) <- dimnames(outgoing)
for (i in 1:length(centr)) {
  outgoing[, i] <- centr[[i]]$outdeg
  incoming[, i] <- centr[[i]]$indeg
}
if (pattern == "outgoing") {
  mat <- t(outgoing)
  legend.name <- "Outgoing"
}
else if (pattern == "incoming") {
  mat <- t(incoming)
  legend.name <- "Incoming"
}
if (pattern == "all") {
  mat <- t(outgoing + incoming)
  legend.name <- "Overall"
}
if (is.null(title)) {
  title <- paste0(legend.name, " signaling patterns")
}
else {
  title <- paste0(paste0(legend.name, " signaling patterns"), 
    " - ", title)
}
if (!is.null(signaling)) {
  mat1 <- mat[rownames(mat) %in% signaling, , drop = FALSE]
  mat <- matrix(0, nrow = length(signaling), ncol = ncol(mat))
  idx <- match(rownames(mat1), signaling)
  mat[idx[!is.na(idx)], ] <- mat1
  dimnames(mat) <- list(signaling, colnames(mat1))
}
mat.ori <- mat
mat <- sweep(mat, 1L, apply(mat, 1, max), "/", check.margin = FALSE)
mat[mat == 0] <- NA
if (is.null(color.use)) {
  color.use <- scPalette(length(colnames(mat)))
}
color.heatmap.use = (grDevices::colorRampPalette((RColorBrewer::brewer.pal(n = 9, 
  name = color.heatmap))))(100)
df <- data.frame(group = colnames(mat))
rownames(df) <- colnames(mat)
names(color.use) <- colnames(mat)
col_annotation <- HeatmapAnnotation(df = df, col = list(group = color.use), 
  which = "column", show_legend = FALSE, show_annotation_name = FALSE, 
  simple_anno_size = grid::unit(0.2, "cm"))
ha2 = HeatmapAnnotation(Strength = anno_barplot(colSums(mat.ori), 
  border = FALSE, gp = gpar(fill = color.use, col = color.use)), 
  show_annotation_name = FALSE)
pSum <- rowSums(mat.ori)
pSum.original <- pSum
pSum <- -1/log(pSum)
pSum[is.na(pSum)] <- 0
idx1 <- which(is.infinite(pSum) | pSum < 0)
if (length(idx1) > 0) {
  values.assign <- seq(max(pSum) * 1.1, max(pSum) * 1.5, 
    length.out = length(idx1))
  position <- sort(pSum.original[idx1], index.return = TRUE)$ix
  pSum[idx1] <- values.assign[match(1:length(idx1), position)]
}
ha1 = rowAnnotation(Strength = anno_barplot(pSum, border = FALSE), 
  show_annotation_name = FALSE)
if (min(mat, na.rm = T) == max(mat, na.rm = T)) {
  legend.break <- max(mat, na.rm = T)
}
else {
  legend.break <- c(round(min(mat, na.rm = T), digits = 1), 
    round(max(mat, na.rm = T), digits = 1))
}
ht1 = Heatmap(mat, col = color.heatmap.use, na_col = "white", 
  name = "Relative strength", bottom_annotation = col_annotation, 
  top_annotation = ha2, right_annotation = ha1, cluster_rows = cluster.rows, 
  cluster_columns = cluster.rows, row_names_side = "left", 
  row_names_rot = 0, row_names_gp = gpar(fontsize = font.size), 
  column_names_gp = gpar(fontsize = font.size), width = unit(width, 
    "cm"), height = unit(height, "cm"), column_title = title, 
  column_title_gp = gpar(fontsize = font.size.title), 
  column_names_rot = 90, heatmap_legend_param = list(title_gp = gpar(fontsize = 8, 
    fontface = "plain"), title_position = "leftcenter-rot", 
    border = NA, at = legend.break, legend_height = unit(20, 
      "mm"), labels_gp = gpar(fontsize = 8), grid_width = unit(2, 
      "mm")))
return(ht1)

```

## • FUNCTION: Running CellChat

```{r}
library(Seurat)
library(CellChat)

sender <- 'Mon%'
hpv_status <- 'HPV+'
metadata = 'mreg_cxcl9_globalcluster4'
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/MontoT/'



hnc <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_all_annot_2022-06-30.RDS')

hnc_til_mono <- hnc[,hnc$mreg_cxcl9_globalcluster4 %like% c('CD4_%', 'CD8_%', 'Treg_%', sender)] 

hnc_til_mono$mreg_cxcl9_globalcluster4[hnc_til_mono$mreg_cxcl9_globalcluster4 %like% 'mreg%'] <- 'mregDC'

hnc_til_mono <- hnc_til_mono[,hnc_til_mono$tissue == 'TIL']

hnc_til_mono <- hnc_til_mono[,hnc_til_mono$hpv_status == hpv_status]

DimPlot(hnc_til_mono, group.by = 'mreg_cxcl9_globalcluster4')

input_seurat = hnc_til_mono

#assuming your input is RNA in data slot, run other assay/slot if it is different
data.input<-GetAssayData(input_seurat,assay = "RNA", slot = "data")

#extract cell type labels
labels<-input_seurat$mreg_cxcl9_globalcluster4
meta <- data.frame(group = labels, row.names = names(labels))
cellchat <- createCellChat(object = data.input, meta = meta, group.by = "group")
cellchat@DB <- CellChatDB.human
cellchat <- subsetData(cellchat)
cellchat <- identifyOverExpressedGenes(cellchat,thresh.p = 0.1)
cellchat <- identifyOverExpressedInteractions(cellchat)
cellchat <- computeCommunProb(cellchat)
cellchat <- computeCommunProbPathway(cellchat)
cellchat <- filterCommunication(cellchat, min.cells = 5)

cellchat <- aggregateNet(cellchat)

groupSize <- as.numeric(table(cellchat@idents))
par(mfrow = c(1,2), xpd=TRUE)

mat <- cellchat@net$weight

saveRDS(cellchat, paste(out, 'cellchat.RDS', sep = ''))

pathways.show <- c("CXCL") 
vertex.receiver = seq(1,4) # a numeric vector. 
netVisual_aggregate(cellchat, signaling = pathways.show)
# Circle plot
par(mfrow=c(1,1))

png(paste(out, 'CXCL_signaling.png', sep = ''), width = 500, height = 500)
netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle")
dev.off()

#dct <- cellchathelper(input_seurat = hnc_til_dct, metadata = 'mreg_cxcl9_globalcluster4', out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/DC_to_T/')


```

## • Re-running for global clusters (remanipulating the metadata)

```{r}
hnc$global <- hnc$mreg_cxcl9_globalcluster4
hnc$global[hnc$global %like% 'CD4%'] <- 'CD4'
hnc$global[hnc$global %like% 'CD8%'] <- 'CD8'
hnc$global[hnc$global %like% 'Mon%'] <- 'Mon'
hnc$global[hnc$global %like% 'Mac%'] <- 'Mac'
hnc$global[hnc$global %like% 'Treg%'] <- 'Treg'
hnc$global[hnc$global %like% '%DC%'] <- 'DC'
hnc$global[hnc$global %like% 'NK%'] <- 'NK'
hnc <- hnc[,!hnc$global %like% "KRT%"]

table(hnc$global)

saveRDS(hnc, '/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_all_annot_2022-08-16.RDS')
```

## • FUNCTION: Running CellChat (split by HPV status)

```{r}
library(Seurat)
library(CellChat)
sender <- c('%DC%', 'Mon%', 'Mac%')
metadata = 'mreg_cxcl9_globalcluster4'
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/'
hnc <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_all_annot_2022-08-16.RDS')

hnc_til_mono <- hnc[,hnc$mreg_cxcl9_globalcluster4 %like% c('CD4_%', 'CD8_%', 'Treg%', sender)] 
#hnc_til_mono$mreg_cxcl9_globalcluster4[hnc_til_mono$mreg_cxcl9_globalcluster4 %like% 'mreg%'] <- 'mregDC'
hnc_til_mono <- hnc_til_mono[,hnc_til_mono$tissue == 'TIL']
hnc <- NULL

ccs <- list()
for (x in unique(hnc_til_mono$hpv_status)){
  input_seurat <- hnc_til_mono[,hnc_til_mono$hpv_status == x]
  data.input<-GetAssayData(input_seurat,assay = "RNA", slot = "data")
  labels<-input_seurat$mreg_cxcl9_globalcluster4   #extract cell type labels
  meta <- data.frame(group = labels, row.names = names(labels))
  cellchat <- createCellChat(object = data.input, meta = meta, group.by = "group")
  cellchat@DB <- CellChatDB.human
  cellchat <- subsetData(cellchat)
  cellchat <- identifyOverExpressedGenes(cellchat,thresh.p = 0.1)
  cellchat <- identifyOverExpressedInteractions(cellchat)
  cellchat <- computeCommunProb(cellchat)
  cellchat <- computeCommunProbPathway(cellchat)
  cellchat <- filterCommunication(cellchat, min.cells = 5)
  cellchat <- aggregateNet(cellchat)
  groupSize <- as.numeric(table(cellchat@idents))
  par(mfrow = c(1,2), xpd=TRUE)
  mat <- cellchat@net$weight
  ccs[x] <- cellchat
}

hpvp <- ccs[['HPV+']]
hpvn <- ccs[['HPV-']]
saveRDS(hpvp, paste(out, 'HPV+_cellchat.RDS', sep = ''))
saveRDS(hpvn, paste(out, 'HPV-_cellchat.RDS', sep = ''))

senders <- unique(input_seurat[[metadata]][[1]])[unique(input_seurat[[metadata]][[1]]) %like% sender]
receivers <- unique(input_seurat[[metadata]][[1]])[unique(input_seurat[[metadata]][[1]]) %like% c('CD4%', 'CD8%', 'Treg%')]

#pathways.show <- c("CXCL") 
pathways.show <- c('CXCL', 'CLEC', 'CD86', 'MHC-I', 'MHC-II')
pdf(paste(out, 'signaling.pdf', sep = ''), width = 12, height = 6)

for (y in pathways.show){
  p1 <- netVisual_aggregate(hpvp, signaling = y, layout = 'circle', sources.use = senders, targets.use = receivers, signaling.name = paste('HPV+', y, sep = ' '))
  netAnalysis_contribution(hpvp, signaling = y, signaling.name = paste('HPV+', y, sep = ' '))
  p3 <- netVisual_aggregate(hpvn, signaling = y, layout = 'circle', sources.use = senders, targets.use = receivers, signaling.name = paste('HPV-', y, sep = ' '))
  netAnalysis_contribution(hpvn, signaling = y, signaling.name = paste('HPV+', y, sep = ' '))
}  
dev.off()
```

## • REF FIGURE: Cellchat figures

![](images/paste-1A733A5E.png)

```{r}
library(tidyverse)

df <- data.frame(hpv_pos = numeric(), hpv_neg = numeric())
df1 <- NULL
matrices_hpvpos <- list()
matrices_hpvneg <- list()
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/'
hpvpos <- readRDS(file = paste(out, 'HPV+_cellchat.RDS', sep = ''))
hpvneg <- readRDS(file = paste(out, 'HPV-_cellchat.RDS', sep = ''))
hpos <- as.data.frame(hpvpos@net$count)
hpos$cell_type <- rownames(hpos)
hneg <- as.data.frame(hpvneg@net$count)
hneg$cell_type <- rownames(hneg)
df[1,1] <- sum(hpvpos@net$count)
df[1,2] <- sum(hpvneg@net$count)

df1 <- as.data.frame(colSums(df))
df1$hpv_status <- rownames(df1)
colnames(df1) <- c('num_interactions', 'hpv_status')

y <- hpos
colnames(y)[colnames(y) %like% 'CD4%'] <- 'CD4'
colnames(y)[colnames(y) %like% 'CD8%'] <- 'CD8'
colnames(y)[colnames(y) %like% 'Treg%'] <- 'Treg'
colnames(y)[colnames(y) %like% 'Mac%'] <- 'Mac'
colnames(y)[colnames(y) %like% 'Mon%'] <- 'Mon'
colnames(y)[colnames(y) %like% '%DC%'] <- 'DC'
y$cell_type[y$cell_type %like% 'CD4%'] <- 'CD4'
y$cell_type[y$cell_type %like% 'CD8%'] <- 'CD8'
y$cell_type[y$cell_type %like% 'Treg%'] <- 'Treg'
y$cell_type[y$cell_type %like% 'Mac%'] <- 'Mac'
y$cell_type[y$cell_type %like% 'Mon%'] <- 'Mon'
y$cell_type[y$cell_type %like% '%DC%'] <- 'DC'
df2 <- aggregate(.~cell_type, data = y, FUN = sum)
rownames(df2) <- df2$cell_type
df2$cell_type <- NULL
hpos2 <- df2

y <- hneg
colnames(y)[colnames(y) %like% 'CD4%'] <- 'CD4'
colnames(y)[colnames(y) %like% 'CD8%'] <- 'CD8'
colnames(y)[colnames(y) %like% 'Treg%'] <- 'Treg'
colnames(y)[colnames(y) %like% 'Mac%'] <- 'Mac'
colnames(y)[colnames(y) %like% 'Mon%'] <- 'Mon'
colnames(y)[colnames(y) %like% '%DC%'] <- 'DC'
y$cell_type[y$cell_type %like% 'CD4%'] <- 'CD4'
y$cell_type[y$cell_type %like% 'CD8%'] <- 'CD8'
y$cell_type[y$cell_type %like% 'Treg%'] <- 'Treg'
y$cell_type[y$cell_type %like% 'Mac%'] <- 'Mac'
y$cell_type[y$cell_type %like% 'Mon%'] <- 'Mon'
y$cell_type[y$cell_type %like% '%DC%'] <- 'DC'
df2 <- aggregate(.~cell_type, data = y, FUN = sum)
rownames(df2) <- df2$cell_type
df2$cell_type <- NULL
hneg2 <- df2

aeg_circlize(hneg2)

# making figure a

png('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/inferred_myeloidTcell_interactions.png', width = 3, height = 3, units = 'in', res = 300)
ggplot(data=df1, aes(x=hpv_status, y=num_interactions, fill=hpv_status)) + geom_bar(stat="identity") + theme_classic() +  labs(y = "Number of inferred interactions", x = "HPV status")
dev.off()

mt_hp <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus_global/HPV+_cellchat.RDS')
mt_hn <- readRDS('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus_global/HPV-_cellchat.RDS')


# making figure c

p1 <- hpvpos@net$count
n1 <- hpvneg@net$count
diff <- p1 - n1
col.order <- c("CD4_1", "CD4_2", "CD4_3", "CD4_4", "CD8_1", "CD8_2", "CD8_3", 'CD8_4', 'CD8_5', "Treg_1", "Treg_2", 'cDC1_CLEC9A', "cDC2_CD1C", "cDC2_CD33", "mregDC_CXCL9hi", "mregDC_CXCL9lo", "DC_pDC", 'Mac_C1QA', "Mac_IL1B", "Mac_SPP1", "Mac_SPP1_PLTP", "Mono_CD14", "Mono_CD14_ID1", "Mono_CD14_IL1B", "Mono_CD14_THBS1", "Mono_CD16", "Mono_Int", "Mono_TIL")
diff <- diff[ col.order, col.order]

png('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/differential_num_of_interactions.png', width = 7, height = 6, units = 'in', res = 300)
pheatmap::pheatmap(diff, cluster_rows = F, cluster_cols = F, main = "Interactions enriched in HPV+")
dev.off()


# making figure d (see Test netAnalysis_signalingRole_scatter for actual figure creation)

hpvpos <- netAnalysis_computeCentrality(hpvpos)
hpvneg <- netAnalysis_computeCentrality(hpvneg)
test <- data.frame(slot(hpvpos, 'netP')$centr)
test1 <- netAnalysis_signalingRole_scatter(hpvpos)
hpvpos

# making figure e (see Test netAnalysis_signalingRole_scatter below for figure creation)

# making figure f 

# signaling pathway specific

signaling <- c('TNF')

png(paste('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/hpvpos_', signaling, '_signaling_circosplot.png', sep = ''), width = 8, height = 6, units = 'in', res = 300)
netVisual_chord_gene(hpvpos, signaling = signaling, legend.pos.x = 8, legend.pos.y = 50, sources.use = c('cDC1_CLEC9A', 'cDC2_CD1C', 'mregDC_CXCL9hi', 'mregDC_CXCL9lo', 'Mac_C1QA', 'Mac_IL1B', 'Mac_SPP1', 'Mac_SPP1_PLTP', 'Mono_CD16', 'Mono_TIL', 'Mono_CD14', 'Mono_CD14_ID1', 'Mono_CD14_IL1B', 'Mono_CD14_THBS1', 'Mono_Int'), targets.use = c('CD8_1', 'CD8_2', 'CD8_3', 'CD8_4', 'CD8_5', 'CD4_1', 'CD4_2', 'CD4_3', 'CD4_4', 'Treg_1', 'Treg_2'))
dev.off()
png(paste('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/hpvneg_', signaling, '_signaling_circosplot.png', sep = ''), width = 8, height = 6, units = 'in', res = 300)
netVisual_chord_gene(hpvneg, signaling = signaling, legend.pos.x = 8, legend.pos.y = 50, sources.use = c('cDC1_CLEC9A', 'cDC2_CD1C', 'mregDC_CXCL9hi', 'mregDC_CXCL9lo', 'Mac_C1QA', 'Mac_IL1B', 'Mac_SPP1', 'Mac_SPP1_PLTP', 'Mono_CD16', 'Mono_TIL', 'Mono_CD14', 'Mono_CD14_ID1', 'Mono_CD14_IL1B', 'Mono_CD14_THBS1', 'Mono_Int'), targets.use = c('CD8_1', 'CD8_2', 'CD8_3', 'CD8_4', 'CD8_5', 'CD4_1', 'CD4_2', 'CD4_3', 'CD4_4', 'Treg_1', 'Treg_2'))
dev.off()

library(NMF)
library(ggalluvial)
selectK(hpvpos, pattern = 'incoming')
nPatterns = 3
hpvpos <- identifyCommunicationPatterns(hpvpos, pattern = 'incoming', k = nPatterns)
selectK(hpvpos, pattern = 'outgoing')
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/'
saveRDS(hpvpos, file = paste(out, 'HPV+_cellchat.RDS', sep = ''))

netAnalysis_dot(hpvpos, pattern = 'incoming')
```

## • Test netAnalysis_signalingRole_scatter

```{r}
#function (object, signaling = NULL, color.use = NULL, slot.name = "netP", group = NULL, weight.MinMax = NULL, dot.size = c(2, 6), point.shape = c(21, 22, 24, 23, 25, 8, 3), label.size = 3, dot.alpha = 0.6, x.measure = "outdeg", y.measure = "indeg", xlabel = "Outgoing interaction strength", ylabel = "Incoming interaction strength", title = NULL, font.size = 10, font.size.title = 10, do.label = T, show.legend = T, show.axes = T) 

hpvpos <- readRDS(file = paste(out, 'HPV+_cellchat.RDS', sep = ''))
hpvneg <- readRDS(file = paste(out, 'HPV-_cellchat.RDS', sep = ''))

hpvpos <- netAnalysis_computeCentrality(hpvpos)
hpvneg <- netAnalysis_computeCentrality(hpvneg)

object <- hpvpos
#signaling = NULL 
signaling = c('TNF')
color.use = NULL
slot.name = "netP"
group = NULL
weight.MinMax = NULL
dot.size = c(2, 6)
point.shape = c(21, 22, 24, 23, 25, 8, 3)
label.size = 3
dot.alpha = 0.6
x.measure = "outdeg"
y.measure = "indeg"
xlabel = "Differential outgoing interaction strength"
ylabel = "Differential incoming interaction strength"
title = NULL
font.size = 10
font.size.title = 10
do.label = T
show.legend = T
show.axes = T
#{
  if (length(slot(object, slot.name)$centr) == 0) {
    stop("Please run `netAnalysis_computeCentrality` to compute the network centrality scores! ")
  }
  if (sum(c(x.measure, y.measure) %in% names(slot(object, 
    slot.name)$centr[[1]])) != 2) {
    stop(paste0("`x.measure, y.measure` should be one of ", 
      paste(names(slot(object, slot.name)$centr[[1]]), 
        collapse = ", "), "\n", "`outdeg_unweighted` is only supported for version >= 1.1.2"))
  }
  centr <- slot(object, slot.name)$centr
  outgoing <- matrix(0, nrow = nlevels(object@idents), ncol = length(centr))
  incoming <- matrix(0, nrow = nlevels(object@idents), ncol = length(centr))
  dimnames(outgoing) <- list(levels(object@idents), names(centr))
  dimnames(incoming) <- dimnames(outgoing)
  for (i in 1:length(centr)) {
    outgoing[, i] <- centr[[i]][[x.measure]]
    incoming[, i] <- centr[[i]][[y.measure]]
  }
  if (is.null(signaling)) {
    message("Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways")
  }
  if (!is.null(signaling)) {
    message("Signaling role analysis on the cell-cell communication network from user's input")
    signaling <- signaling[signaling %in% object@netP$pathways]
    if (length(signaling) == 0) {
      stop("There is no significant communication for the input signaling. All the significant signaling are shown in `object@netP$pathways`")
    }
    outgoing <- outgoing[, signaling, drop = FALSE]
    incoming <- incoming[, signaling, drop = FALSE]
  }
  outgoing.cells <- rowSums(outgoing)
  incoming.cells <- rowSums(incoming)
  num.link <- aggregateNet(object, signaling = signaling, 
    return.object = FALSE, remove.isolate = FALSE)$count
  num.link <- rowSums(num.link) + colSums(num.link) - diag(num.link)
  df <- data.frame(x = outgoing.cells, y = incoming.cells, 
    labels = names(incoming.cells), Count = num.link)
  df$labels <- factor(df$labels, levels = names(incoming.cells))
  if (!is.null(group)) {
    df$Group <- group
  }
  
  write.csv(df, '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/hpvpos_outgoing_incoming_signaling_TNF.csv')
    write.csv(df, '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/hpvneg_outgoing_incoming_signaling_TNF.csv')
  
  hpvpos <- read.csv('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/hpvpos_outgoing_incoming_signaling_TNF.csv', row.names = 1)
  hpvneg <- read.csv('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/hpvneg_outgoing_incoming_signaling_TNF.csv', row.names = 1)
  
  df <- NULL
  df$x <- hpvpos$x - hpvneg$x
  df$y <- hpvpos$y - hpvneg$y
  df$labels <- hpvpos$labels
  df$Count <- hpvpos$Count - hpvneg$Count
  df <- as.data.frame(df)
  rownames(df) <- rownames(hpvpos)
  
  if (is.null(color.use)) {
    color.use <- scPalette(nlevels(object@idents))
  }
  if (!is.null(group)) {
    gg <- ggplot(data = df, aes(x, y)) + geom_point(aes(size = Count, 
      colour = labels, fill = labels, shape = Group))
  }
  if (is.null(group)) {
    gg <- ggplot(data = df, aes(x, y)) + geom_point(aes(size = Count, 
      colour = labels, fill = labels))
  }
  gg <- gg + CellChat_theme_opts() + theme(text = element_text(size = font.size), 
    legend.key.height = grid::unit(0.15, "in")) + labs(title = title, 
    x = xlabel, y = ylabel) + theme(plot.title = element_text(size = font.size.title, 
    face = "plain")) + theme(axis.line.x = element_line(size = 0.25), 
    axis.line.y = element_line(size = 0.25))
  gg <- gg + scale_fill_manual(values = ggplot2::alpha(color.use, 
    alpha = dot.alpha), drop = FALSE) + guides(fill = FALSE)
  gg <- gg + scale_colour_manual(values = color.use, drop = FALSE) + 
    guides(colour = FALSE)
  if (!is.null(group)) {
    gg <- gg + scale_shape_manual(values = point.shape[1:length(unique(df$Group))])
  }
  if (is.null(weight.MinMax)) {
    gg <- gg + scale_size_continuous(range = dot.size)
  }
  if (!is.null(weight.MinMax)){
    gg <- gg + scale_size_continuous(limits = weight.MinMax, 
      range = dot.size)
  }
  if (do.label) {
    gg <- gg + ggrepel::geom_text_repel(mapping = aes(label = labels, 
      colour = labels), size = label.size, show.legend = F, 
      segment.size = 0.2, segment.alpha = 0.5)
  }
  if (!show.legend) {
    gg <- gg + theme(legend.position = "none")
  }
  if (!show.axes) {
    gg <- gg + theme_void()
  }
  
  png('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT_splitbyHPVstatus/differential_signaling_changes_TNFpathway.png', width = 8, height = 7, units = 'in', res = 300)
  gg + ggtitle('TNF signaling changes (HPV+ compared to HPV-)') + geom_hline(yintercept=0, linetype="dashed", color = "grey")
  dev.off()
#}
```



## • Plotting all significant interactions

```{r}
out1 = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/DC_to_T/'
cellchat1 <- readRDS(paste(out1, 'cellchat.RDS', sep = ''))
groupSize1 <- as.numeric(table(cellchat1@idents))
#par(mfrow = c(1,2), xpd=TRUE)
#netVisual_circle(cellchat1@net$count, vertex.weight = groupSize1, weight.scale = T, label.edge= F, title.name = "Number of interactions")
#netVisual_circle(cellchat1@net$weight, vertex.weight = groupSize1, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")

pathways.show <- c("PD-L1") 

pdf(file = paste('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/DC_to_T/', pathways.show, '_signaling.pdf', sep = ''), width = 7, height = 7)
netVisual_aggregate(cellchat1, signaling = pathways.show)
netVisual_heatmap(cellchat1, signaling = pathways.show, color.heatmap = "Reds")
netAnalysis_contribution(cellchat1, signaling = pathways.show)
dev.off()
#weights1 <- cellchat1@net$count
#ph1 <- pheatmap::pheatmap(weights1)



out2 = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/Mon_to_T/'
cellchat2 <- readRDS(paste(out2, 'cellchat.RDS', sep = ''))
groupSize2 <- as.numeric(table(cellchat2@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat2@net$count, vertex.weight = groupSize2, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat2@net$weight, vertex.weight = groupSize2, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")

pdf(file = paste('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/Mon_to_T/', pathways.show, '_signaling.pdf', sep = ''), width = 7, height = 7)
netVisual_aggregate(cellchat2, signaling = pathways.show)
netVisual_heatmap(cellchat2, signaling = pathways.show, color.heatmap = "Reds")
netAnalysis_contribution(cellchat2, signaling = pathways.show)
dev.off()
#weights2 <- cellchat2@net$count
#ph2 <- pheatmap::pheatmap(weights2)


out3 = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/Mac_to_T/'
cellchat3 <- readRDS(paste(out3, 'cellchat.RDS', sep = ''))
groupSize3 <- as.numeric(table(cellchat3@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat3@net$count, vertex.weight = groupSize3, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat3@net$weight, vertex.weight = groupSize3, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")

pdf(file = paste('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/Mac_to_T/', pathways.show, '_signaling.pdf', sep = ''), width = 7, height = 7)
netVisual_aggregate(cellchat3, signaling = pathways.show)
netVisual_heatmap(cellchat3, signaling = pathways.show, color.heatmap = "Reds")
netAnalysis_contribution(cellchat3, signaling = pathways.show)
dev.off()
#weights3 <- cellchat3@net$count
#ph3 <- pheatmap::pheatmap(weights3)
#cp <- cowplot::plot_grid(ph1$gtable, ph2$gtable, ph3$gtable)

cowplot::save_plot(filename = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/myeloid_to_T_count.png', plot = cp, base_width = 15, base_height = 12)
```

## • Plotting

```{r}
out = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/DC_to_T/'

#all_int <- c('ICOS', 'IL15', 'IL12', 'PVR', 'CD86', 'CD80', 'PDCD1LG2', 'CD274', 'LTB', 'CD40', 'IL21', 'CH25H')

cellchat <- readRDS(paste(out, 'cellchat.RDS', sep = ''))
all_int <- c('CD40')
for (y in all_int){
  interactions <- extractEnrichedLR(cellchat, signaling = c(y), geneLR.return = FALSE)
  for (x in 1:nrow(interactions)){
    LR.show <- interactions[x,] # show one ligand-receptor pair
    png(paste(out, LR.show, '_circosplot.png'), width = 7, height = 7, units = 'in', res = 600)
    netVisual_individual(cellchat, signaling = c(y),  pairLR.use = LR.show)
    dev.off()
  }
}
```

## • Extracting weights and number of interactions

```{r}
out1 = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/DC_to_T/'
cellchat1 <- readRDS(paste(out1, 'cellchat.RDS', sep = ''))
groupSize1 <- as.numeric(table(cellchat1@idents))
#par(mfrow = c(1,2), xpd=TRUE)
#netVisual_circle(cellchat1@net$count, vertex.weight = groupSize1, weight.scale = T, label.edge= F, title.name = "Number of interactions")
#netVisual_circle(cellchat1@net$weight, vertex.weight = groupSize1, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")

pathways.show <- c("CD80") 

pdf(file = paste('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/DC_to_T/', pathways.show, '_signaling.pdf', sep = ''), width = 7, height = 7)
netVisual_aggregate(cellchat1, signaling = pathways.show)
netVisual_heatmap(cellchat1, signaling = pathways.show, color.heatmap = "Reds")
netAnalysis_contribution(cellchat1, signaling = pathways.show)
dev.off()
#weights1 <- cellchat1@net$count
#ph1 <- pheatmap::pheatmap(weights1)



out2 = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/Mon_to_T/'
cellchat2 <- readRDS(paste(out2, 'cellchat.RDS', sep = ''))
groupSize2 <- as.numeric(table(cellchat2@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat2@net$count, vertex.weight = groupSize2, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat2@net$weight, vertex.weight = groupSize2, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")

pdf(file = paste('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/Mon_to_T/', pathways.show, '_signaling.pdf', sep = ''), width = 7, height = 7)
netVisual_aggregate(cellchat2, signaling = pathways.show)
netVisual_heatmap(cellchat2, signaling = pathways.show, color.heatmap = "Reds")
netAnalysis_contribution(cellchat2, signaling = pathways.show)
dev.off()
#weights2 <- cellchat2@net$count
#ph2 <- pheatmap::pheatmap(weights2)


out3 = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/Mac_to_T/'
cellchat3 <- readRDS(paste(out3, 'cellchat.RDS', sep = ''))
groupSize3 <- as.numeric(table(cellchat3@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat3@net$count, vertex.weight = groupSize3, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat3@net$weight, vertex.weight = groupSize3, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")

pdf(file = paste('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/Mac_to_T/', pathways.show, '_signaling.pdf', sep = ''), width = 7, height = 7)
netVisual_aggregate(cellchat3, signaling = pathways.show)
netVisual_heatmap(cellchat3, signaling = pathways.show, color.heatmap = "Reds")
netAnalysis_contribution(cellchat3, signaling = pathways.show)
dev.off()
#weights3 <- cellchat3@net$count
#ph3 <- pheatmap::pheatmap(weights3)
#cp <- cowplot::plot_grid(ph1$gtable, ph2$gtable, ph3$gtable)

cowplot::save_plot(filename = '/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_myeloidtoT/myeloid_to_T_count.png', plot = cp, base_width = 15, base_height = 12)
```

## • CellChat visualization

### •• Circos plot, heatmap with signaling pathway networks

```{r}
groupSize <- as.numeric(table(cellchat@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")

mat <- cellchat@net$weight

pathways.show <- c("CXCL") 
# Hierarchy plot
# Here we define `vertex.receive` so that the left portion of the hierarchy plot shows signaling to fibroblast and the right portion shows signaling to immune cells . 
netVisual_aggregate(cellchat, signaling = pathways.show)
# Circle plot
par(mfrow=c(1,1))

#png('/Volumes/hqdinh2/Projects/HNC_SPORE/CellChat/HNC_DCtoT/CXCL_signaling.png', width = 500, height = 500)
netVisual_aggregate(cellchat, signaling = pathways.show, layout = "circle")
#dev.off()

par(mfrow=c(1,1))
netVisual_heatmap(cellchat, signaling = pathways.show, color.heatmap = "Reds")

netAnalysis_contribution(cellchat, signaling = pathways.show)
```

### Plotting individual interactions

```{r}
pairLR.CXCL <- extractEnrichedLR(cellchat, signaling = pathways.show, geneLR.return = FALSE)
LR.show <- pairLR.CXCL[1,] # show one ligand-receptor pair
# Hierarchy plot
vertex.receiver = seq(1,4) # a numeric vector
netVisual_individual(cellchat, signaling = pathways.show,  pairLR.use = LR.show)
#> [[1]]
# Circle plot
netVisual_individual(cellchat, signaling = pathways.show, pairLR.use = LR.show, layout = "chord")
```

### Visualize dominent senders and receivers

```{r}
# Compute the network centrality scores
cellchat <- netAnalysis_computeCentrality(cellchat, slot.name = "netP") # the slot 'netP' means the inferred intercellular communication network of signaling pathways
# Visualize the computed centrality scores using heatmap, allowing ready identification of major signaling roles of cell groups
netAnalysis_signalingRole_network(cellchat, signaling = pathways.show, width = 8, height = 2.5, font.size = 10)

# Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
gg1 <- netAnalysis_signalingRole_scatter(cellchat)
#> Signaling role analysis on the aggregated cell-cell communication network from all signaling pathways
# Signaling role analysis on the cell-cell communication networks of interest
gg2 <- netAnalysis_signalingRole_scatter(cellchat, signaling = c("CXCL", "CCL"))
#> Signaling role analysis on the cell-cell communication network from user's input
gg1 + gg2
```

# Figure 2

```{r VlnPlot of regulatory molecules in LAMP3 DCs figure 2a}
VlnPlot(hnc, features = 'CD274')

#the regulatory genes are from the Maier paper
DotPlot(hnc, group.by = 'global.cluster4', features = c('CD274', 'PDCD1LG2', 'CD200', 'FAS', 'ALDH1A2', 'SOCS1', 'SOCS2'), idents = c('cDC1_CLEC9A', 'cDC2_CD1C', 'cDC2_CD33', 'DC_pDC', 'DC3_LAMP3'), col.min = 0) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + coord_flip() + ggtitle('Regulatory genes') + theme(plot.title = element_text(hjust = 0.5))
out2a <- paste(out, '/Figure2/DotPlot_LAMP3DC3_regulatorymolecules.png', sep = '')
ggsave(out2a, width = 10)

#the maturation genes are from the Maier paper
DotPlot(hnc, group.by = 'global.cluster4', features = c('CD40', 'CD80', 'CD86', 'RELB', 'CD83'), idents = c('cDC1_CLEC9A', 'cDC2_CD1C', 'cDC2_CD33', 'DC_pDC', 'DC3_LAMP3'), col.min = 0) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + coord_flip() + ggtitle('Maturation genes') + theme(plot.title = element_text(hjust = 0.5))
out2a2 <- paste(out, '/Figure2/DotPlot_LAMP3DC3_maturationmolecules.png', sep = '')
ggsave(out2a2, width = 10)

#the migratory markers are from the Maier paper
DotPlot(hnc, group.by = 'global.cluster4', features = c('CCR7', 'MYO1G', 'CXCL16', 'ADAM8', 'ICAM1', 'FSCN1', 'MARCKS', 'MARCKSL1'), idents = c('cDC1_CLEC9A', 'cDC2_CD1C', 'cDC2_CD33', 'DC_pDC', 'DC3_LAMP3'), col.min = 0) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + coord_flip() + ggtitle('Migratory genes') + theme(plot.title = element_text(hjust = 0.5))
out2a3 <- paste(out, '/Figure2/DotPlot_LAMP3DC3_migratorymolecules.png', sep = '')
ggsave(out2a3, width = 10)
```

```{r monocle3 function}
#SEU <- hpvpos_dcs
#ROOT <- 'cDC1_CLEC9A'
#out <- '/Volumes/hdlab/Projects/HNC_SPORE/Monocle3/cillohnc_HPV+dcs_pseudotime.png'

monocle3_pseudotime <- function(SEU, ROOT, out) {
  cds <- SeuratWrappers::as.cell_data_set(SEU)
  cds <- monocle3::cluster_cells(cds = cds, reduction_method = "UMAP")
  cds <- monocle3::learn_graph(cds, use_partition = TRUE)
  
  #subsetting so to get our root
  root <- subset(SEU, subset = global.cluster4 == ROOT)
  root_cells <- colnames(root)
  
  #computing and plotting the pseudotime graph
  cds <- monocle3::order_cells(cds, reduction_method = "UMAP", root_cells = root_cells)
  monocle3::plot_cells(cds = cds, color_cells_by = 'pseudotime', show_trajectory_graph = TRUE)
  ggsave(out)
  return(cds)
}

```

```{r monocle3 Dc trajectory}
dcs <- subset(hnc, idents = c('cDC2_CD33', 'cDC1_CLEC9A', 'cDC2_CD1C', 'DC3_LAMP3'))
dcs <- RunPCA(dcs, features = VariableFeatures(object = dcs), pcs = 100)
ElbowPlot(dcs, ndims = 100)
dcs <- FindNeighbors(dcs, dims = 1:20)
dcs <- RunUMAP(dcs, dims = 1:20)
DimPlot(dcs, group.by = 'global.cluster4')
ggsave('/Volumes/hdlab/Projects/HNC_SPORE/Monocle3/cillohnc_dcs_dimplot.png')
DimPlot(dcs, group.by = 'global.cluster4', split.by = 'tissue')
ggsave('/Volumes/hdlab/Projects/HNC_SPORE/Monocle3/cillohnc_dcs_dimplot_splitbytissue.png')
DimPlot(dcs, group.by = 'global.cluster4', split.by = 'hpv_status')
ggsave('/Volumes/hdlab/Projects/HNC_SPORE/Monocle3/cillohnc_dcs_dimplot_splitbyhpvstatus.png')

monocle3_pseudotime(SEU = dcs, ROOT = 'cDC1_CLEC9A', out = '/Volumes/hdlab/Projects/HNC_SPORE/Monocle3/cillohnc_dcs_pseudotime.png')

save(dcs, file = '/Volumes/hdlab/Projects/HNC_SPORE/Seurat_Objs/HNC_human/hnc_dcs_03072022.rda')

hpvpos_dcs <- subset(dcs, subset = hpv_status == 'HPV+')
hpvpos_dcs <- RunPCA(hpvpos_dcs, features = VariableFeatures(object = hpvpos_dcs), pcs = 100)
ElbowPlot(hpvpos_dcs, ndims = 100)
hpvpos_dcs <- FindNeighbors(hpvpos_dcs, dims = 1:20)
hpvpos_dcs <- RunUMAP(hpvpos_dcs, dims = 1:20)
DimPlot(hpvpos_dcs, group.by = 'global.cluster4')
ggsave('/Volumes/hdlab/Projects/HNC_SPORE/Monocle3/03072022_onlyHPVposhncDCs/cillohnc_hpvposDCs_dimplot.png')

monocle3_pseudotime(SEU = hpvpos_dcs, ROOT = 'cDC1_CLEC9A', out = '/Volumes/hdlab/Projects/HNC_SPORE/Monocle3/cillohnc_HPV+dcs_pseudotime.png')

hpvneg_dcs <- subset(dcs, subset = hpv_status == 'HPV-')
hpvneg_dcs <- RunPCA(hpvneg_dcs, features = VariableFeatures(object = hpvneg_dcs), pcs = 100)
ElbowPlot(hpvneg_dcs, ndims = 100)
hpvneg_dcs <- FindNeighbors(hpvneg_dcs, dims = 1:20)
hpvneg_dcs <- RunUMAP(hpvneg_dcs, dims = 1:20)
DimPlot(hpvneg_dcs, group.by = 'global.cluster4')
ggsave('/Volumes/hdlab/Projects/HNC_SPORE/Monocle3/03072022_onlyHPVneghncDCs/cillohnc_hpvnegDCs_dimplot.png')

monocle3_pseudotime(SEU = hpvneg_dcs, ROOT = 'cDC1_CLEC9A', out = '/Volumes/hdlab/Projects/HNC_SPORE/Monocle3/03072022_onlyHPVneghncDCs/cillohnc_HPV-dcs_pseudotime.png')
```

```{r mouse UMAP}
DimPlot(hnc2, group.by = 'AEG_res2_annotations', split.by = 'tissue', cols = color_clusters) + ggtitle('MOC2 vs. MOC2K17KO myeloid cell subsets')
ggsave('/Volumes/hdlab/Projects/HNC_SPORE/Golfinosetal2022/moc2_myeloidUMAP_splitbytissue.png', width = 9, height = 4)
```

```{r mouse myeloid genes dotplot}
load('/Volumes/hdlab/Projects/HNC_SPORE/Seurat_Objs/MOC2_mouse/Seurat_MOC2_K17KO_MonMacDC_2021-09-02.rda')

all_g <- c()
for (y in sort(unique(hnc2$AEG_res2_annotations))){
  fm <- FindMarkers(hnc2, ident.1 = y, group.by = 'AEG_res2_annotations', logfc.threshold = 0.3, test.use = 'wilcox', min.pct = 0.4, min.diff.pct = 0.2, only.pos = TRUE)
  fm <- fm[order(-fm$avg_log2FC, fm$p_val_adj),]
  g <- rownames(fm[1:3,])
  all_g <- c(all_g, g)
}
all_g <- unique(all_g)
DotPlot(hnc2, features = rev(all_g), col.min = 0, group.by = 'AEG_res2_annotations') + coord_flip() +  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
ggsave('/Volumes/hdlab/Projects/HNC_SPORE/Golfinosetal2022/mouse_myeloid_dotplot.png', width = 7, height = 9)
```


# 04/xx/2022 #stuff for AACR 2022 (e-poster)

```{r objects are gns and cris, out = cris}
library(DescTools)
library(tidyverse)
library(data.table)
library(dplyr)
library(stringr)

myeloid_id <- as.character(unique(Idents(seurat))[unique(Idents(seurat)) %like any% c('%Mon%', '%Mac%', '%DC%')])
myeloid_id <- as.character(unique(Idents(hnc))[unique(Idents(hnc)) %like any% c('%Mon%', '%Mac_%', '%DC%')])
myeloid_id <- myeloid_id[-17]
myeloid_id <- myeloid_id[-18]
mac_id <- as.character(unique(Idents(hnc))[unique(Idents(hnc)) %like any% c('%Mac_%')])
dc_id <- as.character(unique(Idents(hnc))[unique(Idents(hnc)) %like any% c('%DC%')])
mon_id <- as.character(unique(Idents(hnc))[unique(Idents(hnc)) %like any% c('%Mon%')])

hnc$hpv_cluster <- paste(hnc$global.cluster4, hnc$hpv_status)
hnc_mac <- hnc[,hnc$global.cluster4 %in% mac_id]

cris <- read.csv('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/Cristescu_supplemental1.csv')
gns <- read.csv('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022//Cristescu_supplemental2.csv')
#removing white space from the patient IDs in the gns dataframe
gns <- gns %>% 
  mutate(across(where(is.character), str_trim))

cris <- cris[cris$Tumor.Type %like% '%HNSCC%',]

#getting top 20% of our GEP and TMB values
cris$response_var <- cris$GEP. * cris$TMB
cris <- cris[order(cris$response_var, decreasing = TRUE),]
cris <- cris %>% drop_na(response_var)
n20 <- round(0.20*nrow(cris))

responders <- cris[1:n20,]
responders <- responders$Patient.ID

#filtering the gene expression data to only keep those rows where the patients were HNSCC
gns <- gns[gns$Patient.ID %in% cris$Patient.ID,]
colnames(gns) <- colnames(gns) %>% str_replace_all('\\.$', '')
colnames(gns) <- str_replace_all(colnames(gns), '\\.', '-')

nonresponders <- tail(cris, n = n20)
nonresponders <- nonresponders$Patient.ID

cris$Patient.ID <- as.character(cris$Patient.ID)

# Adding responder status based on the responder status column:
cris <- cris %>%
  mutate(response = case_when(
    Patient.ID %in% responders ~ "Responder",
    Patient.ID %in% nonresponders ~ "Non-responder" 
    ))

cris <- cris %>% drop_na(response)

write.csv(cris, file = '/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/responder_nonresponder_anndata.csv')
```

```{r plotting genes}
#now plotting the genes to see how they look in our DCs, Macs, and Mons

#first getting a list of all the genes we are interested in
g <- colnames(gns[2:ncol(gns)]) %>% str_replace_all('\\.$', '')
g1 <- str_replace_all(g, '\\.', '-')

#now getting subsets of our data
dc_id <- dc_id[-6]
dc_seu <-  hnc[,hnc@meta.data$global.cluster4 %in% c(dc_id)]

mons <- hnc[,hnc@meta.data$global.cluster4 %in% c(mon_id)]
macs <- hnc[,hnc@meta.data$global.cluster4 %in% c(mac_id)]

#now plotting each of the genes in our different subsets
DotPlot(subset(dc_seu, subset = tissue == 'TIL'), features = g1, col.min = 0, cols = c('yellow', 'blue')) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
vln <- VlnPlot(dc_seu, features = g1, pt.size = 0)
ggsave(vln, file = '/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/DC_vlnplot.png', height = 14, width = 16)

dc_genes <- list()
dc_genes$DC3_LAMP3 <- c('CCL5', 'CD274', 'CXCL9', 'IDO1', 'PDCD1LG2')
dc_genes$cDC2_CD1C <- c('CCL5', 'HLA-DQA1', 'HLA-DRB1', 'STAT1')
dc_genes$cDC1_CLEC9A <- c('IDO1','HLA-DQA1', 'HLA-DRB1', 'LAG3')

#DotPlot(mons, features = g1, col.min = 0, cols = c('yellow', 'blue')) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
#vln <- VlnPlot(mons, features = g1, pt.size = 0)
#ggsave(vln, file = '/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/mon_vlnplot.png', height = 14, width = 16)

#mon_genes <- list()
#mon_genes$CD16_Monocytes <- c('CCL5', 'CD27', 'CD8A', 'NKG7', 'PCD1LG2')
#mon_genes$Mono_TIL <- c('CD274', 'CD276', 'CMKLR1', 'CXCL9', 'HLA-DQA1', 'HLA-DRB1', 'IDO1', 'LAG3', 'STAT1', 'TIGIT')
#mon_genes$Mono_CD14_IL1B <- c('CXCR6')
#mon_genes$Mono_CD16 <- c('HLA-E', 'PSMB10')

DotPlot(macs, features = g1, col.min = 0, cols = c('yellow', 'blue')) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
vln <- VlnPlot(macs, features = g1, pt.size = 0)
ggsave(vln, file = '/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/mac_vlnplot.png', height = 14, width = 16)

mac_genes <- list()
mac_genes$Mac_C1QA <- c('CCL5', 'CMKLR1', 'CXCL9', 'HLA-DQA1', 'LAG3', 'STAT1', 'PSMB10')
mac_genes$Mac_SPP1 <- c('STAT1', 'PSMB10')

saveRDS(dc_genes, mon_genes, mac_genes, file = '/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/Cristescu_gene_assignments_persubset.RDS')
```

```{r QC}
gns <- mutate_all(gns, function(x) as.numeric(as.character(x)))
boxplot(as.matrix(t(gns[,2:ncol(gns)])), use.cols = TRUE, outline = FALSE) + geom_boxplot() + theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) 

gns[,2:ncol(gns)] <- log2(gns[,2:ncol(gns)])
boxplot(as.matrix(gns[,2:ncol(gns)]), use.cols = TRUE, outline = FALSE) + geom_boxplot() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
```

```{r scatter plots}
cris <- read.csv('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/Cristescu_supplemental1.csv')
cris <- cris[cris$Tumor.Type %like% '%HNSCC%',]
litch <- read.csv('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/litchfield.csv')
###############################################################################################################
#now making the boxplots
SUBSET <- 'Test CXCL9'
############################
#getting the genes that were assigned to the 'SUBSET' cell type
x <- gns[,c(1, which(colnames(gns) %in% mac_genes[[SUBSET]]))]

#making all the rows into numeric because they are currently characters
x <- mutate_all(x, function(x) as.numeric(as.character(x)))
x[,2:ncol(x)] <- log2(x[,2:ncol(x)])


#getting the average expression of all the genes divided by the number of genes as a new column
x$avg_exp <- rowMeans(x)/nc

#making a dataframe with patient id and average expression of the genes 
test <- as.data.frame(x[,c('Patient-ID', 'avg_exp')])

#now making a mini annotation dataframe that we can merge with our test dataframe
mini_ann <- as.data.frame(cris[,c('Patient.ID', 'TMB', 'GEP.')])
a <- merge(test, mini_ann, by.x = 'Patient-ID', by.y = 'Patient.ID')

xax <- paste(SUBSET, 'signature score', sep = ' ')
outname <- paste('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/', as.character(Sys.Date()), SUBSET, 'LinearRegression.png', sep = '')

require(gridExtra)
t <- ggplot(a, aes(x=avg_exp, y=TMB)) + geom_point(size=2) + labs(y= "TMB", x = xax) + theme_bw()  + stat_summary(fun.data=mean_cl_normal) + geom_smooth(method='lm', formula= y~x) + geom_text(x = 10, y = 3500, label = eq(a$avg_exp, a$TMB), parse = TRUE) + ggtitle(paste(xax, 'vs TMB', sep = ' '))

g <- ggplot(a, aes(x=avg_exp, y=GEP.)) + geom_point(size=2) + labs(y='GEP', x = xax) + theme_bw() + stat_summary(fun.data=mean_cl_normal) + geom_smooth(method='lm', formula= y~x) + geom_text(x = 10, y = -0.75, label = eq(a$avg_exp, a$GEP.), parse = TRUE) + ggtitle(paste(xax, 'vs GEP', sep = ' '))

d <- grid.arrange(t, g, ncol=2)
ggsave(d, file = outname, width = 18, height = 10)
```

```{r functions}
substrRight <- function(x, n){
  substr(x, nchar(x)-n+1, nchar(x))
}

eq <- function(x,y) {
  m <- lm(y ~ x)
  as.character(
    as.expression(
      substitute(italic(y) == a + b %.% italic(x)*","~~italic(r)^2~"="~r2, list(a = format(coef(m)[1], digits = 4),
                b = format(coef(m)[2], digits = 4), r2 = format(summary(m)$r.squared, digits = 3)))
    )
  )
}
```

```{r boxplots of responders and non-responders for each subset}
cris <- read.csv('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/responder_nonresponder_anndata.csv')
litch <- read.csv('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/litchfield.csv')
#############################################################################################
#now making the boxplots
SUBSET <- 'Test CXCL9'
############################

#for (l in dc_genes) {
  #print(names(l))
  #print(l)
#}

#getting the genes that were assigned to the 'SUBSET' cell type
x <- gns[,which(colnames(gns) %in% mac_genes[[SUBSET]])]
x <- gns[,c(1, which(colnames(gns) == 'CXCL9'))]
nc <- as.numeric(ncol(x))

#making all the rows into numeric because they are currently characters and taking the LOG2 of it
x <- log2(mutate_all(x, function(x) as.numeric(as.character(x))))

#getting the average expression of all the genes divided by the number of genes as a new column
x$avg_exp <- rowMeans(x)/nc
x$pat_id <- rownames(x)
x <- merge(x, litch, by.x = 'pat_id', by.y = 'Patient')

#making a dataframe with patient id and average expression of the genes 
a <- as.data.frame(x[,c('pat_id', 'avg_exp', 'Radiological.response.data')])

#now making a mini annotation dataframe that we can merge with our test dataframe
#mini_ann <- as.data.frame(cris[,c('Patient.ID', 'response')])
#a <- merge(test, mini_ann, by.x = 'pat_id', by.y = 'Patient.ID')

r <- a[a$Radiological.response.data == 'response',]
nr <- a[a$Radiological.response.data == 'no_response',]
  
wilcox <- wilcox.test(r$avg_exp, nr$avg_exp, alternative = c('two.sided'))

title <- paste(SUBSET, 'genes in HNC', sep = ' ')
subtitle <- paste('From Cristescu data   p=', wilcox$p.value, sep = '')
outname <- paste('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/', as.character(Sys.Date()), SUBSET, 'ResponderVsNonresponderBoxplot.png', sep = '')

#makes a boxplot of responders and non-responders for each subset
ggplot(a) +
    geom_boxplot(aes(x = Radiological.response.data, y = avg_exp, color = Radiological.response.data, fill = Radiological.response.data), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = Radiological.response.data, y = avg_exp, color = Radiological.response.data), alpha = 0.8, position = position_jitterdodge()) + theme_bw() + theme(plot.title = element_text(size = 30), axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), axis.text.y = element_text(size = 30), strip.text = element_text(size = 16), legend.text=element_text(size=30), legend.title = element_blank()) + scale_shape_manual(values = 1:5) + ggtitle(title, subtitle = subtitle) + theme(text=element_text(size=21)) + labs(y = "Log2 Counts")
ggsave(file = outname, width = 10, height = 15)
```

```{r testing litchfield data}

litch <- readr::read_delim('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/Using_Litchfield_Data/meta_analysis_input_data.txt')
litch <- litch[litch$study == 'CRISTESCU_SCIENCE_2018',]
litch <- litch[litch$histology == 'HEAD AND NECK',]

# CXCL9

litch_mini <- litch[,c('CXCL9', 'response')]

r <- litch_mini[litch_mini$response == 'response',]
nr <- litch_mini[litch_mini$response == 'no_response',]
  
wilcox <- wilcox.test(r$CXCL9, nr$CXCL9, alternative = c('two.sided'))

subtitle <- paste('p=', wilcox$p.value, sep = '')
outname <- paste('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/Using_Litchfield_Data/', as.character(Sys.Date()), 'CXCL9ResponderVsNonresponderBoxplot.png', sep = '')

ggplot(litch_mini) +
    geom_boxplot(aes(x = response, y = CXCL9, color = response, fill = response, color = response), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = response, y = CXCL9, color = response), alpha = 0.8, position = position_jitterdodge()) + theme_bw() + theme(plot.title = element_text(size = 30), axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), axis.text.y = element_text(size = 30), strip.text = element_text(size = 16), legend.text=element_text(size=30), legend.title = element_blank()) + scale_shape_manual(values = 1:5) + ggtitle('CXCL9 expression in ICI response', subtitle = subtitle) + theme(text=element_text(size=21)) + scale_fill_manual(breaks = c('response', 'no_response'), values = c("indianred1", "cyan3")) + scale_color_manual(values = c('response' = 'indianred1', 'no_response' = 'cyan3')) 

ggsave(file = outname, width = 10, height = 15) 

# CD274

litch_mini <- litch[,c('CD274', 'response')]

r <- litch_mini[litch_mini$response == 'response',]
nr <- litch_mini[litch_mini$response == 'no_response',]
  
wilcox <- wilcox.test(r$CD274, nr$CD274, alternative = c('two.sided'))

subtitle <- paste('p=', wilcox$p.value, sep = '')
outname <- paste('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/Using_Litchfield_Data/', as.character(Sys.Date()), 'CD274ResponderVsNonresponderBoxplot.png', sep = '')

ggplot(litch_mini) +
    geom_boxplot(aes(x = response, y = CD274, color = response, fill = response, color = response), position = position_dodge(), alpha = 0.5, outlier.color = NA) + geom_point(aes(x = response, y = CD274, color = response), alpha = 0.8, position = position_jitterdodge()) + theme_bw() + theme(plot.title = element_text(size = 30), axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.title.x = element_blank(), axis.text.y = element_text(size = 30), strip.text = element_text(size = 16), legend.text=element_text(size=30), legend.title = element_blank()) + scale_shape_manual(values = 1:5) + ggtitle('CD274 expression in ICI response', subtitle = subtitle) + theme(text=element_text(size=21)) + scale_fill_manual(breaks = c('response', 'no_response'), values = c("indianred1", "cyan3")) + scale_color_manual(values = c('response' = 'indianred1', 'no_response' = 'cyan3'))
ggsave(file = outname, width = 10, height = 15)
```

#AACR poster modification (for actual poster)

```{r cpdb functions}
cpdb_dir <- "/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/01242022_HNC_HPV+vsHPV-_PBMCvsTIL/HPV+"
my_seurat <- hnc

#Function 1: Reformats the cpdb files and merges the pvalues with the interaction data. This drops complex interactions and those that have 2 senders, 2 receivers, or no sender, receiver assigned.
#It takes the path of your cellphonedb out directory. You can filter at this point by telling which senders, receivers to include. Here’s some ways to filter these interactions:

###############################################################################################
cpdb_summary <- function(path = "", senders = "all", receivers = "all", drop = "none"){

  outs <- list()
  expected_files <- c("significant_means.txt", "pvalues.txt")
  
  if(sum(expected_files %in% list.files(path)) !=2){
    message(paste("missing file(s):", expected_files[!expected_files %in% list.files(path)]))
    break
  }
  
  # file read in (sm for significant means, pv for pvalues)
  sm <- read.table(paste(path, "/", expected_files[1], sep = ""), check.names = FALSE, header = TRUE, sep = "\t")
  pv <- read.table(paste(path, "/", expected_files[2], sep = ""), check.names = FALSE, header = TRUE, sep = "\t")

  # reshaping wide to long and merging sig_means
  sm2 <- sm %>% 
    pivot_longer(cols = contains("|"), names_to = "pops") %>%
    separate(col = "pops", into = c("pop1", "pop2"), sep = "\\|") %>% 
    mutate(sender = "", receiver = "", ligand = "", receptor = "") %>% 
    dplyr::filter(!is.na(value))
  
  # reshaping wide to long and merging pvalues
  pv2 <- pv %>%
    pivot_longer(cols = contains("|"), names_to = "pops") %>%
    separate(col = "pops", into = c("pop1", "pop2"), sep = "\\|") %>%
    mutate(sender = "", receiver = "", ligand = "", receptor = "") %>% 
    dplyr::filter(value != 1)
  
  # this is hard coded because SELL and SELPLG are backwards in cpdb database
  sm2[sm2$interacting_pair == "SELL_SELPLG","receptor_a"] <- "True"
  pv2[pv2$interacting_pair == "SELL_SELPLG","receptor_a"] <- "True"
  
  sm2[sm2$interacting_pair == "SELL_SELPLG","receptor_b"] <- "False"
  pv2[pv2$interacting_pair == "SELL_SELPLG","receptor_b"] <- "False"
  
  # separating ligands and receptors, separating senders and receivers
  # for significant means
  for (i in 1:nrow(sm2)){
    this_row <- sm2[i,]
    #status <- sum(this_row$receptor_a == "True", this_row$receptor_b == "True")
    #if (status != 1){
      #next
    #} else if (this_row$receptor_a == "True"){
      #sm2[i,"sender"] <- this_row$pop2
      #sm2[i,"receiver"] <- this_row$pop1
      #sm2[i, "ligand"] <- str_split(this_row$interacting_pair, pattern = "_", simplify = TRUE)[2]
      #sm2[i, "receptor"] <- str_split(this_row$interacting_pair, pattern = "_", simplify = TRUE)[1]
    #} else {
      #sm2[i,"sender"] <- this_row$pop1
      #sm2[i,"receiver"] <- this_row$pop2
      #sm2[i, "ligand"] <- str_split(this_row$interacting_pair, pattern = "_", simplify = TRUE)[1]
      #sm2[i, "receptor"] <- str_split(this_row$interacting_pair, pattern = "_", simplify = TRUE)[2]
    #}
    sm2[i,"sender"] <- this_row$pop1
    sm2[i,"receiver"] <- this_row$pop2
    sm2[i, "ligand"] <- str_split(this_row$interacting_pair, pattern = "_", simplify = TRUE)[1]
    sm2[i, "receptor"] <- str_split(this_row$interacting_pair, pattern = "_", simplify = TRUE)[2]
  }
  
  # for p-values
  for (i in 1:nrow(pv2)){
    this_row <- pv2[i,]
    #status <- sum(this_row$receptor_a == "True", this_row$receptor_b == "True")
    #if (status != 1){
      #next
    #} else if (this_row$receptor_a == "True"){
      #pv2[i,"sender"] <- this_row$pop2
      #pv2[i,"receiver"] <- this_row$pop1
      #pv2[i, "ligand"] <- str_split(this_row$interacting_pair, pattern = "_", simplify = TRUE)[2]
      #pv2[i, "receptor"] <- str_split(this_row$interacting_pair, pattern = "_", simplify = TRUE)[1]
    #} else {
      #pv2[i,"sender"] <- this_row$pop1
      #pv2[i,"receiver"] <- this_row$pop2
      #pv2[i, "ligand"] <- str_split(this_row$interacting_pair, pattern = "_", simplify = TRUE)[1]
      #pv2[i, "receptor"] <- str_split(this_row$interacting_pair, pattern = "_", simplify = TRUE)[2]
    #}
    pv2[i,"sender"] <- this_row$pop1
    pv2[i,"receiver"] <- this_row$pop2
    pv2[i, "ligand"] <- str_split(this_row$interacting_pair, pattern = "_", simplify = TRUE)[1]
    pv2[i, "receptor"] <- str_split(this_row$interacting_pair, pattern = "_", simplify = TRUE)[2]
  }
  
  # these interactions, don't have sender and reciever identified
  dropped_interactions <- sm2 %>% filter(sender == ""|receiver == "") %>% dplyr::select(interacting_pair, pop1, pop2, value)

  # pre merge sig_means
  sm_pre <- sm2 %>% filter(sender != "" & !grepl("complex", interacting_pair)) %>%
    filter(receiver != "") %>% 
    mutate(sr = paste(sender, receiver, sep = " -> "), lr = paste(ligand, receptor, sep = " -> ")) %>%
    dplyr::select(interacting_pair, sender, receiver, ligand, receptor, sr, lr, value, rank) %>%
    arrange(sender, receiver) %>% distinct() %>% as.data.frame()
  
  # pre merge pvalues
  pv_pre <- pv2 %>% filter(sender != "" & !grepl("complex", interacting_pair)) %>%
    filter(receiver != "") %>% mutate(sr = paste(sender, receiver, sep = " -> "), 
           lr = paste(ligand, receptor, sep = " -> ")) %>%
    dplyr::select(sender, receiver, ligand, receptor, sr, lr, value) %>%
    arrange(sender, receiver) %>% distinct() %>% as.data.frame()
  
  colnames(pv_pre)[which(colnames(pv_pre) == "value")] <- "pvalue"
  
  # appending pvalues onto significant means
  merged_out <- merge(sm_pre, pv_pre %>% dplyr::select(sr, lr, pvalue), type = "left", by = c("sr", "lr"))
  
  # writing to outs
  outs[["merged_data"]] <- merged_out
  outs[["dropped_interactions"]] <- dropped_interactions

    # if not filtering at all
  if (isTRUE(senders[1] == "all" & receivers[1] == "all" & drop[1] == "none")){
    outs[["filtered"]] <- NULL
  } else {
  
    # filtering parameters
    if(isTRUE(senders[1] == "all")){
      senders <- outs$merged_data$sender %>% unique()
    }  
    if(isTRUE(receivers == "all")){
      receivers <- outs$merged_data$receiver %>% unique()
    }  
    if(isTRUE(drop == "none")){
      drop <- ""
    }
    
    outs[["filtered"]] <- outs$merged_data %>% dplyr::filter(sender %in% senders, receiver %in% receivers)
  }
  return(outs)
}
################################################################################################
#Here is an example of what function 1 does: It takes ~30 seconds to run
#sm <- read.table(paste(cpdb_dir, "/significant_means.txt", sep = ""), check.names = FALSE, header = TRUE, sep = "\t")

# pivot data longer and include ligand/receptor annotations
#library(data.table)
#my_summary <- cpdb_summary(path = cpdb_dir, senders = "all", receivers = "all", drop = "none")

# considering macrophages as senders and CD8 t cells as receivers
#my_senders <- grep("LAMP3", unique(my_seurat@active.ident), value = TRUE)
#my_receivers <- grep("CD8|CD4|Mac_", unique(my_seurat@active.ident), value = TRUE)

#small_summary <- cpdb_summary(path = cpdb_dir, senders = my_senders, receivers = 'all', drop = "none")

#Function 2: Takes the output from function 1 and plots the data. You can choose which interactions you want to show and which senders/receivers as well.
#########################################################################################################
cpdb_dotplot <- function(cpdb_list, filtered = FALSE, senders = "all",
                         receivers = "all", drop = "none", dendro = TRUE){
  
  # selecting data source
  if(isTRUE(filtered)){
    plot_data <- cpdb_list$filtered
    if (is.null(plot_data)){
      return(message(paste("Mising filtered data slot")))
    }
  } else {
    plot_data <- cpdb_list$merged_data
  }
  # filtering
  if (isTRUE(senders == "all" & receivers == "all" & drop == "none")){
    plot_data2 <- plot_data
  } else {
    # filtering parameters
    if(isTRUE(senders == "all")){
      senders <- plot_data$sender %>% unique()
    }
    if(isTRUE(receivers == "all")){
      receivers <- plot_data$receiver %>% unique()
    }
    if(isTRUE(drop == "none")){
      drop <- ""
    }
    plot_data2 <- plot_data[plot_data$sender %in% senders &
                              plot_data$receiver %in% receivers &
                              !plot_data$interacting_pair %in% drop,]
  }
  # arranging by factor
clst <- data.frame(lr= plot_data2$lr, value = plot_data2$value, sr = plot_data2$sr) %>%
  pivot_wider(names_from = lr, values_fill = 0) %>%
  column_to_rownames(var = "sr") %>%
  as.matrix() %>%
  t() %>%
  dist()
temp_clust <- hclust(clst)
plot.new()
p1 <- temp_clust %>%
             as.dendrogram() %>%
             raise.dendrogram(10) %>%
             set("labels_cex", 0.5) %>%
      ggdendrogram(., rotate = TRUE, labels = FALSE) + theme_dendro() + scale_y_reverse()
# color scaling
plot_data2$value <- ifelse(plot_data2$value > 2.5, 2.5, plot_data2$value)
p2 <-  plot_data2 %>%
          mutate(pvalue = replace(pvalue, pvalue == 0, 0.000999),
                 lr = factor(lr, levels = temp_clust$labels[temp_clust$order])) %>%
          ggplot(aes(x = receiver, y = lr)) +
          geom_point(aes(fill = value, size = -log10(pvalue)), shape = 21, color = "black") +
          theme_bw(base_size = 8) +
          theme(axis.text.x = element_text(angle = 90, hjust = 0.95, vjust = 0.2)) +
          scale_fill_gradient2(high = muted("red"), low = "royalblue4", mid = "white",
                               midpoint = 1, breaks = c(0, 1, 2), limits = c(0,2.5)) +
          facet_wrap(.~sender, scales = "free_x", ncol = 500) +
          scale_size(range = c(0.5, 2)) + xlab("") + ylab("")
plot.new()
if (isTRUE(dendro)){
cowplot::plot_grid(p1, NULL, p2, rel_widths = c(0.2, 0, 1), align = "h", scale = c(1.065, 1, 1), axis = "btlr", nrow = 1)
  } else {
    p2
  }
}
#########################################################################################################
#Example dotplots: The plot shows LR interactions on the y axis, the headers (banner up top) represents the sending cell type. 

# uses output of cluster 1
#cpdb_dotplot(my_summary, filtered = FALSE, senders = "all", receivers = "all")

# can also filter as before using the filtered data (filtered = TRUE), or filtering only for the plot. 
#cpdb_dotplot(small_summary, filtered = TRUE)

# saving the file
#ggsave('/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/01242022_HNC_HPV+vsHPV-_PBMCvsTIL/HPV+/LAMP3_DC_senderplot.png', width = 5, height = 7)

#Function 3: Performs a manual ligand receptor expression score by multiplying the sqrt(log(ligand) * log(receptor))/2. It then calculates a background distribution that is shown in the “rando” row/column. It shouldn’t be used for actual p value calculation, but juet to to make sure that the data being pulled from your seurat object matches the cellphone db results. This is the only function that requires the actual seurat object.

cpdb_check <- function(seu, meta, pct = 0.2, ligand = "", receptor = ""){
  
  Idents(seu) <- meta
  # Extracting ligand and receptor log scaled data
  LR <- AverageExpression(seu, features = c(ligand, receptor), assays = "RNA", group.by = meta)$RNA %>%
    as.data.frame() %>%
    t()
  
  # nested loop all senders and all receivers
  outer_list <- list()
  for (sender in rownames(LR)){
    
    inner_list <- list()
    
    for (receiver in rownames(LR)){
      inner_list[[receiver]] <- sqrt(LR[sender, 1] * LR[receiver, 2])/2
    }
    
    outer_list[[sender]] <- do.call(rbind, inner_list)
  } 
  
  # all cells, before subsampling
  bg_distribution <- as.matrix(seu@assays$RNA@data[c(ligand, receptor),]) %>% t()
 
  # lr background scoring with random 500 cells
  score_dist <- vector()
  for (i in 1:500){
    
    lig_score <- bg_distribution[sample(1:nrow(bg_distribution), size = 500, replace = TRUE),1] %>%
      mean()
    
    rec_score <- lig_score <- bg_distribution[sample(1:nrow(bg_distribution),
                                                     size = 500, replace = TRUE),1] %>%
      mean()
    
    lr_score <- sqrt(lig_score*rec_score)/2
    score_dist[i] <- lr_score
  }
  
  # simplifying ligand receptor list
  out_df <- data.frame(sender = "", receiver = "", score = 0)

  for (i in 1:length(outer_list)) {
    this_df <- outer_list[[i]]
    sndr <- names(outer_list)[i]
    
    df <- data.frame(sender = rep(sndr, times = nrow(this_df)), 
                     receiver = rownames(this_df), score = this_df[,1])
    
    out_df <- rbind(out_df, df)
  }
  
  # simplifying datafrmme
  rownames(out_df) <- NULL
  fi_df <- out_df[-1,]
  
  # combining with background df
  plot_df <- rbind(fi_df, data.frame(sender = "rando", receiver = "rando", score = mean(score_dist)))


  for (i in 1:nrow(plot_df)){
    this_score <- plot_df[i, "score"]
    plot_df[i,"pval"] <- (sum(this_score <= score_dist)/500)*2
  }
  
  
  # plotting all senders and receivers
  p1 <- plot_df %>%
          mutate(pval = pval) %>%
          ggplot(aes(x = receiver, y = sender, size = score, fill = score)) +
          geom_point(aes(color = ifelse(pval < 0.05, "red", "black")), shape = 21) +
          scale_fill_continuous(low = "royalblue4", high = "orange") + 
          theme(axis.text.x = element_text(angle = 90, vjust = 0.95, hjust = 0.2)) + 
          scale_color_identity() + ggtitle(paste(ligand, receptor, sep = " -> ")) 
          
  
  df_pct <- data.frame(ligand = NULL, receptor = NULL)
  #ligand/receptor pct by cluster
  for (cluster in unique(seu@meta.data[,meta])){
    dat <- seu@assays$RNA@counts[c(ligand, receptor), seu@meta.data[,meta]==cluster]
    apply(dat, 1, function(x){sum(x > 0)/ncol(dat)})
    
    
    df_pct[cluster,1:2] <- apply(dat, 1, function(x){sum(x > 0)/ncol(dat)})[1:2]
  }
  

  lig_clusters <- rownames(df_pct)[df_pct[,1] >= pct]
  rec_clusters <- rownames(df_pct)[df_pct[,2] >= pct]
  
  p2 <- VlnPlot(seu[,seu@meta.data[,meta] %in% lig_clusters], sort = "decreasing", features = ligand, group.by = meta, pt.size = 0) + coord_flip() + theme_classic(base_size = 12) + theme(axis.title.y = element_blank()) + NoLegend() + ggtitle(paste("Ligand:", ligand, "\n", "clusters with >=", pct, "pct shown"))

  p3 <- VlnPlot(seu[,seu@meta.data[,meta] %in% rec_clusters], sort = "decreasing", features = receptor, group.by = meta, pt.size = 0) + coord_flip() + theme_classic(base_size = 12) + theme(axis.title.y = element_blank()) + NoLegend() + ggtitle(paste("Receptor:", receptor, "\n", "clusters with >=", pct, "pct shown"))
  
  grid.arrange(grobs = list(p1, p2, p3), widths = c(1.5, 0.8, 0.8), layout_matrix = rbind(c(1,1,2), c(1,1,3)))

}
################################################################
################################################################
cpdb_dotplot_two_objs <- function(cpdb_list1, cpdb_list2, out, filtered = FALSE, senders = "all",
                         receivers = "all", drop = "none", width = 2600, height = 1000){
  # testing
  #cpdb_list1 <- hpvn
  #cpdb_list2 <- hpvp
  #filtered <- FALSE
  #senders = grep("mac|mono|mono|nc", cpdb_sum$merged_data$sender, value = TRUE)
  #senders = 'all'
  #receivers = "all"
  #drop = "none"
  
  # selecting data source
  if(isTRUE(filtered)){
    plot_data <- cpdb_list1$filtered
    if (is.null(plot_data)){
      return(message(paste("Mising filtered data slot")))
    }
  } else {
    plot_data <- cpdb_list1$merged_data
  }
  # filtering
  if (isTRUE(senders == "all" & receivers == "all" & drop == "none")){
    plot_data_hpvneg <- plot_data
  } else {
    # filtering parameters
    if(isTRUE(senders == "all")){
      senders <- plot_data$sender %>% unique()
    }
    if(isTRUE(receivers == "all")){
      receivers <- plot_data$receiver %>% unique()
    }
    if(isTRUE(drop == "none")){
      drop <- ""
    }
    plot_data_hpvneg <- plot_data[plot_data$sender %in% senders &
                              plot_data$receiver %in% receivers &
                              !plot_data$interacting_pair %in% drop,]
  }
 
  # selecting data source #2 
  if(isTRUE(filtered)){
    plot_data <- cpdb_list2$filtered
    if (is.null(plot_data)){
      return(message(paste("Mising filtered data slot")))
    }
  } else {
    plot_data <- cpdb_list2$merged_data
  }
  # filtering
  if (isTRUE(senders == "all" & receivers == "all" & drop == "none")){
    plot_data_hpvpos <- plot_data
  } else {
    # filtering parameters
    if(isTRUE(senders == "all")){
      senders <- plot_data$sender %>% unique()
    }
    if(isTRUE(receivers == "all")){
      receivers <- plot_data$receiver %>% unique()
    }
    if(isTRUE(drop == "none")){
      drop <- ""
    }
    plot_data_hpvpos <- plot_data[plot_data$sender %in% senders &
                                    plot_data$receiver %in% receivers &
                                    !plot_data$interacting_pair %in% drop,]
  }
  
  plot_hpvneg_mini <- plot_data_hpvneg[c('sr', 'lr', 'interacting_pair', 'sender', 'receiver', 'ligand', 'receptor')]
  plot_hpvpos_mini <- plot_data_hpvpos[c('sr', 'lr', 'interacting_pair', 'sender', 'receiver', 'ligand', 'receptor')]
  
  only_in_hpvneg <- anti_join(plot_hpvneg_mini, plot_hpvpos_mini)
  only_in_hpvpos <- anti_join(plot_hpvpos_mini, plot_hpvneg_mini)
  
  plot_data_hpvneg <- plot_data_hpvneg[c('sr', 'lr', 'interacting_pair', 'sender', 'receiver', 'ligand', 'receptor', 'value', 'rank', 'pvalue')]
  plot_data_hpvpos <- plot_data_hpvpos[c('sr', 'lr', 'interacting_pair', 'sender', 'receiver', 'ligand', 'receptor', 'value', 'rank', 'pvalue')]
  
  only_in_hpvneg$value <- 0
  only_in_hpvneg$rank <- NA
  only_in_hpvneg$pvalue <- 1
  
  only_in_hpvpos$value <- 0
  only_in_hpvpos$rank <- NA
  only_in_hpvpos$pvalue <- 1
  
  all_hpvneg <- rbind(plot_data_hpvneg, only_in_hpvpos)
  all_hpvpos <- rbind(plot_data_hpvpos, only_in_hpvneg)
  
  all_hpvneg <- all_hpvneg[order(all_hpvneg$sr, all_hpvneg$lr),]
  all_hpvpos <- all_hpvpos[order(all_hpvpos$sr, all_hpvpos$lr),]
  
  
  #####plotting after this##########3
   # arranging by factor
  clst <- data.frame(lr= all_hpvneg$lr, value = all_hpvneg$value, sr = all_hpvneg$sr) %>%
    pivot_wider(names_from = lr, values_fill = 0) %>%
    column_to_rownames(var = "sr") %>% as.matrix() %>% t() %>% dist()
  temp_clust <- hclust(clst)

  all_hpvneg$value <- ifelse(all_hpvneg$value > 2.5, 2.5, all_hpvneg$value)
  p2_hpvneg <-  all_hpvneg %>% mutate(pvalue = replace(pvalue, pvalue == 0, 0.000999),
    lr = factor(lr, levels = temp_clust$labels[temp_clust$order], ordered = TRUE)) %>%
    ggplot(aes(x = receiver, y = lr)) + geom_point(aes(fill = value, size = -log10(pvalue)), shape = 21, color = "black") +
    theme_bw(base_size = 8) + theme(axis.text.x = element_text(angle = 90, hjust = 0.95, vjust = 0.2)) +
    scale_fill_gradient2(high = muted("red"), low = "royalblue4", mid = "white", midpoint = 1, breaks = c(0, 1, 2),
    limits = c(0,2.5)) + facet_wrap(.~sender, scales = "free_x", ncol = 500) + scale_size(range = c(0.5, 2)) +
    xlab("") + ylab("") + ggtitle('HPV- HNC') + theme(plot.title = element_text(hjust = 0.5))

plot.new()
all_hpvpos$value <- ifelse(all_hpvpos$value > 2.5, 2.5, all_hpvpos$value)
p2_hpvpos <-  all_hpvpos %>%
  mutate(pvalue = replace(pvalue, pvalue == 0, 0.000999),
         lr = factor(lr, levels = temp_clust$labels[temp_clust$order], ordered = TRUE)) %>%
  ggplot(aes(x = receiver, y = lr)) +
  geom_point(aes(fill = value, size = -log10(pvalue)), shape = 21, color = "black") +
  theme_bw(base_size = 8) +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.95, vjust = 0.2)) +
  scale_fill_gradient2(high = muted("red"), low = "royalblue4", mid = "white",
                       midpoint = 1,
                       breaks = c(0, 1, 2),
                       limits = c(0,2.5)) +
  facet_wrap(.~sender, scales = "free_x", ncol = 500) +
  scale_size(range = c(0.5, 2)) +
  xlab("") +
  ylab("") + 
  ggtitle('HPV+ HNC') +
  theme(plot.title = element_text(hjust = 0.5))
  p2_hpvpos + p2_hpvneg
  n <- paste(out, 'cpdb_multi_dotplot.png', sep = '')
  png(n, width = width, height = height, res = 300)
  par(mfrow = c(2, 2))
  plot(p2_hpvpos + p2_hpvneg)
  dev.off()
}

#cpdb_check(seu = hnc_hpvpos, meta = 'global.cluster4', pct = 0.2, ligand = 'MIF', receptor = 'CD74')
#cpdb_check(seu = hnc_hpvneg, meta = 'global.cluster4', pct = 0.2, ligand = 'MIF', receptor = 'CD74')
#cpdb_check(seu = my_seurat, meta = "global.cluster4", pct = 0.2, ligand = "MIF", receptor = "CD74")

corrs <- function(SEU, SUM, meta){
  SUM[['merged_data']]$receptor <- gsub(" receptor", "RA", SUM[['merged_data']]$receptor)

  for (i in 1:nrow(SUM[['merged_data']])){
    lig <- SUM[['merged_data']]$ligand[i]
    rec <- SUM[['merged_data']]$receptor[i]
    print(paste(lig, rec, sep = '->'))
    L <- FetchData(object = SEU, vars = c(lig))[,1]
    R <- FetchData(object = SEU, vars = c(rec))[,1] 
    cor <- cor.test(L, R)
    SUM[['merged_data']]$cor.p[i] <- cor$p.value
  }
  
  x <- SUM[['merged_data']][SUM[['merged_data']]$cor.p < 0.01,]
  return(x)
}
```

```{r using cpdb functions HPV+ vs HPV-}
#running the HPV positive sampling
small_summary_hpvpos <- cpdb_summary(path = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/01242022_HNC_HPV+vsHPV-_TIL/HPV+/')

#choosing DC3_LAMP3 as the sender population
small_summary_hpvpos$merged_data <- small_summary_hpvpos$merged_data[small_summary_hpvpos$merged_data$sender == 'DC3_LAMP3',]

#choosing T cells as our receiver populations
small_summary_hpvpos$merged_data <- small_summary_hpvpos$merged_data[small_summary_hpvpos$merged_data$receiver %in% c('CD4', 'CD8', 'Treg'),]

small_summary_hpvpos$merged_data$p_val_adj <- p.adjust(small_summary_hpvpos$merged_data$pvalue, method = c('fdr'))

#mapping the average expression of the gene in the cell type
seu <- hnc

d <- DotPlot(seu[,seu$global.cluster4 == 'DC3_LAMP3'], features = unique(small_summary_hpvpos$merged_data$ligand), group.by = 'global.cluster4')

ligs_keep <- rownames(d$data[d$data$pct.exp > 20,])

d1 <- DotPlot(seu[,seu$global.cluster4 %in% c('CD4', 'CD8', 'Treg'),], features = unique(small_summary_hpvpos$merged_data$receptor), group.by = 'global.cluster4')

recs_keep <- unique(rownames(d1$data[d1$data$pct.exp > 20,]))

small_summary_hpvpos$merged_data <- small_summary_hpvpos$merged_data[small_summary_hpvpos$merged_data$ligand %in% ligs_keep,]

small_summary_hpvpos$merged_data <- small_summary_hpvpos$merged_data[small_summary_hpvpos$merged_data$receptor %in% recs_keep,]

###############################################
#running the hPV negative sampling
small_summary_hpvneg <- cpdb_summary(path = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/01242022_HNC_HPV+vsHPV-_TIL/HPV-/')

#choosing DC3_lAMP3 as the sender population
small_summary_hpvneg$merged_data <- small_summary_hpvneg$merged_data[small_summary_hpvneg$merged_data$sender == 'DC3_LAMP3',]

#choosing T cells as our receiver populations
small_summary_hpvneg$merged_data <- small_summary_hpvneg$merged_data[small_summary_hpvneg$merged_data$receiver %in% c('CD4', 'CD8', 'Treg'),]

small_summary_hpvneg$merged_data$p_val_adj <- p.adjust(small_summary_hpvneg$merged_data$pvalue, method = c('fdr'))

#mapping the average expression of the gene in the cell type
seu <- hnc

d <- DotPlot(seu[,seu$global.cluster4 == 'DC3_LAMP3'], features = unique(small_summary_hpvneg$merged_data$ligand), group.by = 'global.cluster4')

ligs_keep <- rownames(d$data[d$data$pct.exp > 20,])

d1 <- DotPlot(seu[,seu$global.cluster4 %in% c('CD4', 'CD8', 'Treg'),], features = unique(small_summary_hpvneg$merged_data$receptor), group.by = 'global.cluster4')

recs_keep <- unique(rownames(d1$data[d1$data$pct.exp > 20,]))

small_summary_hpvneg$merged_data <- small_summary_hpvneg$merged_data[small_summary_hpvneg$merged_data$ligand %in% ligs_keep,]

small_summary_hpvneg$merged_data <- small_summary_hpvneg$merged_data[small_summary_hpvneg$merged_data$receptor %in% recs_keep,]

################################################################################

hpvp <- corrs(SEU = hnc, SUM = small_summary_hpvpos, meta = 'global.cluster4')
hpvn <- corrs(SEU = hnc, SUM = small_summary_hpvneg, meta = 'global.cluster4')

cpdb_dotplot_two_objs(hpvn, hpvp, filtered = FALSE, out = '/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/')
png(filename="/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/cpdb_LRdotplot_HPV+vsHPV-_>20pctexp_padj", res=300, width = 1000, height = 800) # perhaps width/height as well
par(mfrow= c(1, 2))
print(cpdb_dotplot_two_objs(hpvn, hpvp, filtered = FALSE))
dev.off()
```

```{r VlnPlot of these ligands/receptors}
ligs <- unique(c(unique(small_summary_hpvneg$merged_data$ligand), unique(small_summary_hpvpos$merged_data$ligand)))

recs <- unique(c(unique(small_summary_hpvneg$merged_data$receptor), unique(small_summary_hpvpos$merged_data$receptor)))

hnc_mini <- hnc[,hnc$global.cluster4 %in% c('DC3_LAMP3', 'CD8', 'CD4', 'Treg')]
hnc_mini <- hnc_mini[,hnc_mini$tissue == 'TIL']

p1 <- VlnPlot(hnc_mini, features = recs[19], pt.size = 0, split.by = 'hpv_status', group.by = 'global.cluster4')
p2 <- VlnPlot(hnc_mini, features = recs[20], pt.size = 0, split.by = 'hpv_status', group.by = 'global.cluster4')
p3 <- VlnPlot(hnc_mini, features = recs[21], pt.size = 0, split.by = 'hpv_status', group.by = 'global.cluster4')
p4 <- VlnPlot(hnc_mini, features = recs[22], pt.size = 0, split.by = 'hpv_status', group.by = 'global.cluster4')
p5 <- VlnPlot(hnc_mini, features = recs[23], pt.size = 0, split.by = 'hpv_status', group.by = 'global.cluster4')
p6 <- VlnPlot(hnc_mini, features = recs[24], pt.size = 0, split.by = 'hpv_status', group.by = 'global.cluster4')

cowplot::plot_grid(p1, p2, p3, p4, p5, p6)

```

```{r using cpdb functions lamp3 to Ahmed CD8 T cells}
small_summary_hpvpos <- cpdb_summary(path = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/03282022_HNC_LAMP3DC_RafiAhmedHPVCD8s/HPV+/out/')

#choosing DC3_LAMP3 as the sender population
small_summary_hpvpos$merged_data <- small_summary_hpvpos$merged_data[small_summary_hpvpos$merged_data$sender == 'DC3_LAMP3',]

small_summary_hpvpos$merged_data <- small_summary_hpvpos$merged_data[small_summary_hpvpos$merged_data$receiver %like% c('Ahmed'),]

small_summary_hpvneg <- cpdb_summary(path = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/03282022_HNC_LAMP3DC_RafiAhmedHPVCD8s/HPV-/out/')

small_summary_hpvneg$merged_data <- small_summary_hpvneg$merged_data[small_summary_hpvneg$merged_data$sender == 'DC3_LAMP3',]

small_summary_hpvneg$merged_data <- small_summary_hpvneg$merged_data[small_summary_hpvneg$merged_data$receiver %like% c('Ahmed'),]

#########################################
SEU <- hnc
SUM <- small_summary_hpvpos
meta <- 'global.cluster4'

corrs <- function(SEU, SUM, meta){
  SUM[['merged_data']]$receptor <- gsub(" receptor", "RA", SUM[['merged_data']]$receptor)

  for (i in 1:nrow(SUM[['merged_data']])){
    lig <- SUM[['merged_data']]$ligand[i]
    rec <- SUM[['merged_data']]$receptor[i]
    print(paste(lig, rec, sep = '->'))
    L <- FetchData(object = SEU, vars = c(lig))[,1]
    R <- FetchData(object = SEU, vars = c(rec))[,1] 
    cor <- cor.test(L, R)
    SUM[['merged_data']]$cor.p[i] <- cor$p.value
  }
  
  SUM[['merged_data']] <- SUM[['merged_data']][SUM[['merged_data']]$cor.p < 0.01,]
  return(SUM)
}

hpvp <- corrs(SEU = hnc, SUM = small_summary_hpvpos, meta = 'global.cluster4')
hpvn <- corrs(SEU = hnc, SUM = small_summary_hpvneg, meta = 'global.cluster4')

cpdb_dotplot_two_objs(hpvn, hpvp, filtered = FALSE)
```

```{r pseudotime plotting for genes}
##################### MONOCLE 3 PSEUDOTIME ##################################
library(monocle3)
SEU <- hnc[,hnc$global.cluster4 %in% c('DC3_LAMP3', 'cDC1_CLEC9A', 'cDC2_CD1C', 'cDC2_CD33')]
ROOT <- 'cDC1_CLEC9A'
meta <- 'global.cluster4'

cds <- SeuratWrappers::as.cell_data_set(SEU, group.by = meta)
cds <- preprocess_cds(cds, method = c('PCA'), num_dim = 15, norm_method = c('log'))
cds <- reduce_dimension(cds)
cds <- cluster_cells(cds = cds)
cds <- learn_graph(cds, use_partition = TRUE)

#subsetting so to get our root
root_cells <- colnames(subset(SEU, subset = global.cluster4 == ROOT))

#computing and plotting the pseudotime graph
cds <- order_cells(cds, reduction_method = "UMAP", root_cells = root_cells)

pData(cds)$Pseudotime <- monocle3::pseudotime(cds, reduction_method = "UMAP")

pData(cds)$Pseudotime <- max(pData(cds)$Pseudotime) - pData(cds)$Pseudotime # to reverse

rowData(cds)$gene_name <- rownames(cds)
rowData(cds)$gene_short_name <- rowData(cds)$gene_name

plot_cells(cds = cds, color_cells_by = 'pseudotime', show_trajectory_graph = TRUE)


#################### plot genes in pseudotime ####################
colData(cds)$pseudotime <- pseudotime(cds)
cds = cds[, is.finite(colData(cds)$pseudotime)]

cds_exprs <- SingleCellExperiment::counts(cds)
cds_exprs <- reshape2::melt(round(as.matrix(cds_exprs)))
cds_exprs$value_norm <- cds_exprs$value/size_factors(cds)
colnames(cds_exprs) <- c("f_id", "Cell", 'exp1', "expression")

cds_colData <- as.data.frame(colData(cds))

#cds_exprs$adjusted_expression <- cds_exprs$expression

cds_exprs$feature_label <- cds_exprs$f_id

cds_exprs$f_id <- as.character(cds_exprs$f_id)
cds_exprs$feature_label <- factor(cds_exprs$feature_label)
new_data <- data.frame(pseudotime = colData(cds)$pseudotime)

cds_exprs <- merge(cds_exprs, new_data, by.x = 'Cell', by.y = 0)
library(DescTools)
cds_exprs <- cds_exprs[!cds_exprs$f_id %like any% c('RP5%', 'RP4%', 'RP11%', 'MT-%', 'RP3%', 'RP1%', 'RPS%', 'RPL%'),]

all_genes <- c(ligs, recs)
all_genes <- all_genes[-c(25)]
all_genes <- c(all_genes, 'IL15RA')

library(plyr)

for (p in all_genes){
  gene <- p
  
  t <- cds_exprs[cds_exprs$f_id == gene,]
  test <- aggregate(t$expression, by=list(Category=t$new_bin), FUN=mean)
  test <- merge(t, test, by.x = 'new_bin', by.y = 'Category')
  #ddply(df, .(cut(df$r, 5)), colwise(mean))
  #cds_exprs <- cds_exprs %>% mutate(new_bin = cut(pseudotime, breaks=100))
  q <- ggplot(aes(pseudotime, x), data = test) + geom_point() + scale_y_log10() + ylab("log10 expression (100 bins)") + xlab("pseudotime") + theme_classic() + geom_smooth(method=lm, se=FALSE) + ggtitle(paste(gene, 'expression over Monocle3 pseudotime'), subtitle = lm_eqn(test))
  q
  name <- paste('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/', gene, '_exp_over_Mon3_pseudotime_CilloHNC_DCs.png', sep = '')
  ggsave(q, file = name, width = 10, height = 8)
}
############################################################################3

lm_eqn <- function(df){
  m <- lm(x ~ pseudotime, df);
  a <- as.numeric(format(unname(coef(m)[1]), digits = 2))
  b <- as.numeric(format(unname(coef(m)[2]), digits = 2))
  r2 <- as.numeric(format(summary(m)$r.squared, digits = 3))
  eq <- paste('y=', a, "+", b, 'x, R^2=', r2)
  return(eq)
}

```

```{r Dotplot of markers CXCL9 and CD274}
hnc_myeloid <- hnc[,hnc$global.cluster4 %like% c('Mac_%', '%Mon%', '%DC3%', '%cDC2%', '%cDC1%', '%DC_pDC%')]
hnc_myeloid$global.cluster4[hnc_myeloid$global.cluster4 == 'PBMC_cDC2_CD33'] <- 'cDC2_CD33'
hnc_myeloid$global.cluster4[hnc_myeloid$global.cluster4 == 'TIL_cDC2_CD33'] <- 'cDC2_CD33'
hnc_myeloid$global.cluster4[hnc_myeloid$global.cluster4 == 'Tonsil_cDC2_CD33'] <- 'cDC2_CD33'
hnc_myeloid <- hnc_myeloid[,hnc_myeloid$tissue %like% c('TIL')]

DotPlot(hnc_myeloid, features = c('CXCL9', 'CD274'), col.min = 0, group.by = 'global.cluster4', split.by = 'hpv_status') 

ggsave('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/CXCL9_CD274_expression_Cillo_HNC_dotplot.png', width = 7, height = 10, dpi = 500)

DotPlot(hnc_myeloid, features = c('CXCL9', 'CD274'), col.min = 0, group.by = 'global.cluster4', split.by = 'hpv_status')  + coord_flip() + theme(axis.text.x = element_text(angle = 90, hjust=0.95,vjust=0.2))

ggsave('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/CXCL9_CD274_expression_Cillo_HNC_dotplot_HORIZONTAL.png', width = 10, height = 4, dpi = 500)

##### MOUSE DATA ######## 

moc2_myeloid <-seurat[,seurat$global.cluster2 %like% c('Mac_%', 'Mon_%', 'DC_%')]

DotPlot(moc2_myeloid, features = c('Cxcl9', 'Cd274'), col.min = 0, group.by = 'global.cluster2', split.by = 'tissue')

ggsave('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/Cxcl9_Cd274_expression_MOC2K17KO_dotplot_VERTICAL.png', width = 6, height = 6, dpi = 500)

DotPlot(moc2_myeloid, features = c('Cxcl9', 'Cd274'), col.min = 0, group.by = 'global.cluster2', split.by = 'tissue') + coord_flip() + theme(axis.text.x = element_text(angle = 90, hjust = 0.95, vjust = 0.2))

ggsave('/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/Cxcl9_Cd274_expression_MOC2K17KO_dotplot_HORIZONTAL.png', width = 10, height = 4, dpi = 500)
```

```{r Rafi Ahmed CD8 T cell sigs}
library(DescTools)
library(data.table)

color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3",
"#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", "#AA8282", "#D4B7B7", "#8600BF", "#BA5CE3", "#808000","#AEAE5C", "#1E90FF", "#00BFFF", "#56FF0D", "#FFFF00")

cd8t <- hnc[,hnc$global.cluster == 'CD8']

#clustering the CD8 T cells
cd8t <- RunPCA(cd8t, features = VariableFeatures(object = cd8t))
ElbowPlot(cd8t) #choosing 15 PCs
cd8t <- FindNeighbors(cd8t, dims = 1:15)
cd8t <- FindClusters(cd8t, resolution = 0.4)
DimPlot(cd8t, label = TRUE, label.box = TRUE)

#merging back the similar clusters
cd8t$cd8t_clusters <- cd8t@active.ident

marks <- unique(marks)

#merging back the similar clusters
#cd8t$cd8t_clusters <- cd8t@active.ident

cd8t$cd8t_clusters[cd8t$cd8t_clusters == 10] <- 1
cd8t$cd8t_clusters[cd8t$cd8t_clusters == 4] <- 1
cd8t$cd8t_clusters[cd8t$cd8t_clusters == 8] <- 3
cd8t$cd8t_clusters[cd8t$cd8t_clusters == 9] <- 7
cd8t$cd8t_clusters[cd8t$cd8t_clusters == 11] <- 3

marks <- c()
for (x in sort(unique(cd8t$cd8t_clusters))){
  y <- FindMarkers(cd8t, ident.1 = x, only.pos = TRUE, logfc.threshold = 0.5, min.pct = 0.25, group.by = 'cd8t_clusters')
  
  y <- y[!rownames(y) %like% 'RPL',]
  y <- y[!rownames(y) %like% 'RPS',]
  y <- y[!rownames(y) %like% 'MT-',]
  
  z <- y[order(-y$avg_log2FC, y$p_val_adj),][1:5,] 
  
  marks <- c(marks, rownames(z))
}

DimPlot(cd8t, label = TRUE, label.box = TRUE, group.by = 'cd8t_clusters', cols = color_clusters)
ggsave('/Volumes/hdlab/Projects/HNC_SPORE/Golfinosetal2022/CD8/CD8_Dimplot_noanns.png')
DotPlot(cd8t, features = marks, col.min = 0, group.by = 'cd8t_clusters') + coord_flip()
ggsave('/Volumes/hdlab/Projects/HNC_SPORE/Golfinosetal2022/CD8/CD8_dotplot_noanns.png')

cd8t$cd8t_clusters <- as.character(cd8t$cd8t_clusters)

cd8 <- 

##### IDENTIFYING RAFI AHMED HPV+ SPECIFIC CD8 T CELLS ######

ahmedcd8 <- readRDS('/Volumes/hdlab/Projects/Public_Data/Rafi_Ahmed_HPV+CD8T/all_integrated.rds')

ahmedcd8@meta.data$clusters_AEG <- as.character(ahmedcd8@meta.data$seurat_clusters)
ahmedcd8@meta.data$clusters_AEG[ahmedcd8@meta.data$clusters_AEG == '0'] <- 'AhmedHPV+CD8_TD'
ahmedcd8@meta.data$clusters_AEG[ahmedcd8@meta.data$clusters_AEG == '1'] <- 'AhmedHPV+CD8_Stem'
ahmedcd8@meta.data$clusters_AEG[ahmedcd8@meta.data$clusters_AEG == '2'] <- 'AhmedHPV+CD8_Trans'

DimPlot(ahmedcd8, group.by = 'clusters_AEG')

############# TRANSFER DATA FROM AHMED CD8 TO OUR SEURAT #####################
trans <- FindTransferAnchors(
  ahmedcd8,
  cd8t,
  normalization.method = "LogNormalize",
  recompute.residuals = TRUE,
  reference.assay = NULL,
  reference.neighbors = NULL,
  query.assay = NULL,
  reduction = "pcaproject",
  reference.reduction = NULL,
  project.query = FALSE,
  features = NULL,
  scale = TRUE,
  npcs = 30,
  l2.norm = TRUE,
  dims = 1:30,
  k.anchor = 5,
  k.filter = 200,
  k.score = 30,
  max.features = 200,
  nn.method = "annoy",
  n.trees = 50,
  eps = 0,
  approx.pca = TRUE,
  mapping.score.k = NULL,
  verbose = TRUE
)



cd8t <- TransferData(
  trans,
  'clusters_AEG',
  reference = ahmedcd8,
  query = cd8t,
  weight.reduction = "pcaproject",
  l2.norm = FALSE,
  dims = NULL,
  k.weight = 50,
  sd.weight = 1,
  eps = 0,
  n.trees = 50,
  verbose = TRUE,
  slot = "data",
  prediction.assay = FALSE,
  store.weights = TRUE
)

DimPlot(cd8t, group.by = 'predicted.id')

ggsave('/Volumes/hdlab/Projects/HNC_SPORE/Golfinosetal2022/AhmedCD8_to_CilloCD8.png')

```

```{r Rafi Ahmed B cell sigs}
bs <- hnc[,hnc$global.cluster4 == 'Bcells']

bs <- FindVariableFeatures(bs)
bs <- RunPCA(bs, features = VariableFeatures(object = bs))
ElbowPlot(bs) #choosing 15 PCs
bs <- FindNeighbors(bs, dims = 1:10)
bs <- FindClusters(bs, resolution = 0.4)
DimPlot(bs, label = TRUE, label.box = TRUE)

bs$bs_clusters <- bs@active.ident

marks <- c()
for (x in sort(unique(bs$bs_clusters))){
  y <- FindMarkers(bs, ident.1 = x, only.pos = TRUE, logfc.threshold = 0.25, min.pct = 0.25, group.by = 'bs_clusters')
  
  y <- y[!rownames(y) %like% 'RPL',]
  y <- y[!rownames(y) %like% 'RPS',]
  y <- y[!rownames(y) %like% 'MT-',]
  y <- y[!rownames(y) %like% 'RP5',]
  
  z <- y[order(-y$avg_log2FC, y$p_val_adj),][1:5,] 
  
  marks <- c(marks, rownames(z))
}

marks <- unique(marks)

DotPlot(bs, features = marks, col.min = 0, group.by = 'bs_clusters') + coord_flip()

#merging back the similar clusters
bs$bs_clusters <- bs@active.ident

bs$bs_clusters[bs$bs_clusters == 8] <- 0
bs$bs_clusters[bs$bs_clusters == 7] <- 0
bs$bs_clusters[bs$bs_clusters == 6] <- 0
bs$bs_clusters[bs$bs_clusters == 4] <- 1
bs$bs_clusters[bs$bs_clusters == 3] <- 1

DimPlot(bs, label = TRUE, label.box = TRUE, group.by = 'bs_clusters', cols = color_clusters)

bs$bs_clusters <- as.character(bs$bs_clusters)

############# NOW LOOKING FOR B CELL CLUSTERS FROM AHMED PAPER ################

########### Ahmed B clusters #############

#ABC cluster
VlnPlot(bs, features = c('CCR7', 'ID3', 'GPR183', 'EMP3', 'BANK1', 'ACP5', 'IFITM2', 'CAPG'), group.by = 'bs_clusters', pt.size = 0) #CLUSTER 2?

#ASC cluster
VlnPlot(bs, features = c('MZB1', 'SSR4', 'XBP1', 'HSP90B1', 'HERPUD1', 'SEC11C'), group.by = 'bs_clusters', pt.size = 0) #CLUSTER 5? 

#GCB cluster
VlnPlot(bs, features = c('TCL1A', 'NEIL1', 'MARKSL1', 'GMDS', 'BACH2', 'TMEM123'), group.by = 'bs_clusters', pt.size = 0) #CLUSTER 5?

#TC cluster 
VlnPlot(bs, features = c('HIST1H4C', 'TYMS', 'HMGB2', 'TUBB', 'TUBA1B', 'HIST1H1E', 'HIST1H1C'), group.by = 'bs_clusters', pt.size = 0) #NO OBVIOUS CLUSTER MATCH

```

#Huy's AAI talk 2022

```{r cellphone db dendritic cells to T cells split by hpv status 05/19/2022, warning = FALSE}
############## UP HERE IS SPLIT BY T CELL SUBSETS ###################
#running the combined HPV statuses
small_summary <- cpdb_summary(path = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/05042022_HNC_HPVstatusCombined_DCstoTsubsets_out/')

#running the HPV positive sampling
small_summary_hpvpos <- cpdb_summary(path = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/05042022_HNC_HPV+DCstoTsubsets_out/')

#choosing DC3_LAMP3 as the sender population
small_summary_hpvpos$merged_data <- small_summary_hpvpos$merged_data[small_summary_hpvpos$merged_data$sender %like any% c('DC3_LAMP3', 'cDC2_CD1C', 'cDC1_CLEC9A', 'cDC2_CD33', '%CD4_%', '%CD8_%', '%Treg%'),]

#choosing T cells as our receiver populations
small_summary_hpvpos$merged_data <- small_summary_hpvpos$merged_data[small_summary_hpvpos$merged_data$receiver %like any% c('%CD4_%', '%CD8_%', '%Treg_%', 'DC3_LAMP3', 'cDC2_CD1C', 'cDC2_CD33', 'cDC1_CLEC9A'),]
small_summary_hpvpos$merged_data$p_val_adj <- p.adjust(small_summary_hpvpos$merged_data$pvalue, method = c('fdr'))

saveRDS(small_summary_hpvpos, file = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/05042022_HNC_HPV+DCstoTsubsets_out/small_summary_hpvpos.rds')

#mapping the average expression of the gene in the cell type
#arguments
seu <- hnc
#seu$global.cluster5 <- seu@active.ident
summary <- small_summary_hpvneg
meta <- 'global.cluster5'
out <- '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/05042022_HNC_HPV+DCstoTsubsets_out/'
outname <- 'all_combined.csv'

cpdb_seu_heatmap(seu = hnc, summary = small_summary_hpvpos, meta = 'global.cluster5', out = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/05042022_HNC_HPV+DCstoTsubsets_out/', outname = 'all_hpvpos.csv')

cpdb_seu_heatmap(seu = hnc, summary = small_summary_hpvneg, meta = 'global.cluster5', out = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/05042022_HNC_HPV-DCstoTsubsets_out/', outname = 'all_hpvneg.csv')

cpdb_seu_heatmap(seu = hnc, summary = small_summary, meta = 'global.cluster5', out = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/05042022_HNC_HPVstatusCombined_DCstoTsubsets_out/', outname = 'all_combined.csv')

##### FUNCTION ###### 
cpdb_seu_heatmap <- function(seu, summary, meta, out, outname) {
  all_df <- data.frame()
  for (x in unique(summary$merged_data$sender)){
    df <- summary$merged_data[summary$merged_data$sender == x,]
    d <- DotPlot(seu[,seu[[meta]] == x], features = unique(summary$merged_data$ligand))
    ligs_keep <- rownames(d$data[d$data$pct.exp > 20,])
    df <- df[df$ligand %in% ligs_keep,]
    all_df <- rbind(all_df, df)
  }
  all_df_hpvpos <- unique(all_df)
  
  all_df_hpvpos <- all_df_hpvpos[!substr(all_df_hpvpos$sender, 1, 4) == substr(all_df_hpvpos$receiver, 1, 4),]
  all_df_hpvpos <- all_df_hpvpos[!substr(all_df_hpvpos$sender, 1, 2) == substr(all_df_hpvpos$receiver, 1, 2),]
  #getting rid of any interactions between subsets where the last 2 characters are the same
  all_df_hpvpos <- all_df_hpvpos[!substr(all_df_hpvpos$sender, nchar(all_df_hpvpos$sender) - 2 + 1, nchar(all_df_hpvpos$sender)) == substr(all_df_hpvpos$receiver, nchar(all_df_hpvpos$receiver) - 2 + 1, nchar(all_df_hpvpos$receiver)),]
  
  all_df2 <- data.frame()
  for (y in unique(all_df$receiver)){
    df <- all_df[all_df$receiver == y,]
    d1 <- DotPlot(seu[,seu@active.ident == y], features = unique(all_df$receptor))
    recs_keep <- c(unique(rownames(d1$data[d1$data$pct.exp > 20,])), 'IL15R')
    df <- df[df$receptor %in% recs_keep,]
    all_df2 <- rbind(all_df2, df)
    
  }
  all_df2_hpvpos <- unique(all_df2)
  all_df2_hpvpos <- all_df2_hpvpos[!substr(all_df2_hpvpos$sender, 1, 4) == substr(all_df2_hpvpos$receiver, 1, 4),]
  all_df2_hpvpos <- all_df2_hpvpos[!substr(all_df2_hpvpos$sender, 1, 2) == substr(all_df2_hpvpos$receiver, 1, 2),]
  #getting rid of any interactions between subsets where the last 2 characters are the same
  all_df2_hpvpos <- all_df2_hpvpos[!substr(all_df2_hpvpos$sender, nchar(all_df2_hpvpos$sender) - 2 + 1, nchar(all_df2_hpvpos$sender)) == substr(all_df2_hpvpos$receiver, nchar(all_df2_hpvpos$receiver) - 2 + 1, nchar(all_df2_hpvpos$receiver)),]
  
  all_hpvpos <- unique(rbind(all_df_hpvpos, all_df2_hpvpos))
  
  all_hpvpos <- all_hpvpos[!all_hpvpos$sr %in% c("CD4_1 -> Treg_2", "CD4_2 -> Treg_1", "CD4_3 -> Treg_1", "CD4_3 -> Treg_2", "CD4_4 -> Treg_1", "CD4_4 -> Treg_2", "CD8_1 -> Treg_2", "CD8_2 -> Treg_1", "CD8_3 -> Treg_1", "CD8_4 -> Treg_1", "CD8_4 -> Treg_2", "CD8_5 -> Treg_1", "CD8_5 -> Treg_2", "cDC1_CLEC9A -> DC3_LAMP3", "cDC2_CD1C -> DC3_LAMP3", "DC3_LAMP3 -> cDC1_CLEC9A", "DC3_LAMP3 -> cDC2_CD1C", "Treg_1 -> CD4_2", "Treg_1 -> CD4_3", "Treg_1 -> CD4_4", "Treg_1 -> CD8_2", "Treg_1 -> CD8_3", "Treg_1 -> CD8_4", "Treg_1 -> CD8_5", "Treg_2 -> CD4_1", "Treg_2 -> CD4_3", "Treg_2 -> CD4_4", "Treg_2 -> CD8_1", "Treg_2 -> CD8_3", "Treg_2 -> CD8_4", "Treg_2 -> CD8_5" , "cDC2_CD33 -> DC3_LAMP3", "DC3_LAMP3 -> cDC2_CD33"),]
  
  all_hpvpos$receptor[all_hpvpos$receptor == "TGFbeta receptor1"] <- 'TGFBR1'
  all_hpvpos$lr <- all_hpvpos$lr %>% str_replace_all("TGFbeta receptor1", "TGFBR1")
  all_hpvpos$interacting_pair <- all_hpvpos$interacting_pair %>% str_replace_all("TGFbeta receptor1", "TGFBR1")
  all_hpvpos$receptor[all_hpvpos$receptor == "Type II IFNR"] <- "IFNGR2"
  all_hpvpos$lr <- all_hpvpos$lr %>% str_replace_all("Type II IFNR", "IFNGR2")
  all_hpvpos$interacting_pair <- all_hpvpos$interacting_pair %>% str_replace_all("Type II IFNR", "IFNGR2")
  all_hpvpos$receptor[all_hpvpos$receptor == "TGFbeta receptor2"] <- "TGFBR2"
  all_hpvpos$lr <- all_hpvpos$lr %>% str_replace_all("TGFbeta receptor2", "TGFBR2")
  all_hpvpos$interacting_pair <- all_hpvpos$interacting_pair %>% str_replace_all("TGFbeta receptor2", "TGFBR2")

  for (i in 1:nrow(all_hpvpos)){
    lig <- all_hpvpos$ligand[i]
    rec <- all_hpvpos$receptor[i]
    print(paste(lig, rec, sep = '->'))
    L <- FetchData(object = seu, vars = c(lig))[,1]
    R <- FetchData(object = seu, vars = c(rec))[,1] 
    cor <- cor.test(L, R)
    all_hpvpos$cor.p[i] <- cor$p.value
  }
  
  x <- all_hpvpos[all_hpvpos$cor.p < 0.01,]
  write.csv(x, file = paste(out, outname, sep = ''))

}

################################################
################################################
#running the hPV negative sampling
small_summary_hpvneg <- cpdb_summary(path = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/05042022_HNC_HPV-DCstoTsubsets_out/')

#choosing DC3_lAMP3 as the sender population
small_summary_hpvneg$merged_data <- small_summary_hpvneg$merged_data[small_summary_hpvneg$merged_data$sender %like any% c('DC3_LAMP3', 'cDC2_CD1C', 'cDC1_CLEC9A', 'cDC2_CD33', '%CD4_%', '%CD8_%', '%Treg_%'),]

#choosing T cells as our receiver populations
small_summary_hpvneg$merged_data <- small_summary_hpvneg$merged_data[small_summary_hpvneg$merged_data$receiver %like any% c('%CD4%', '%CD8%', '%Treg%', 'DC3_LAMP3', 'cDC2_CD1C', 'cDC1_CLEC9A', 'cDC2_CD33'),]

small_summary_hpvneg$merged_data$p_val_adj <- p.adjust(small_summary_hpvneg$merged_data$pvalue, method = c('fdr'))

#saveRDS(small_summary_hpvneg, file = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/05042022_HNC_HPV-DCstoTsubsets_out/small_summary_hpvneg.rds')

#mapping the average expression of the gene in the cell type
seu <- hnc

all_df <- data.frame()
for (x in unique(small_summary_hpvneg$merged_data$sender)){
  df <- small_summary_hpvneg$merged_data[small_summary_hpvneg$merged_data$sender == x,]
  print(nrow(df))
  d <- DotPlot(seu[,seu@active.ident == x], features = unique(small_summary_hpvneg$merged_data$ligand))
  ligs_keep <- rownames(d$data[d$data$pct.exp > 20,])
  df <- df[df$ligand %in% ligs_keep,]
  print(nrow(df))
  all_df <- rbind(all_df, df)
}
all_df_hpvneg <- unique(all_df)

all_df_hpvneg <- all_df_hpvneg[!substr(all_df_hpvneg$sender, 1, 4) == substr(all_df_hpvneg$receiver, 1, 4),]
all_df_hpvneg <- all_df_hpvneg[!substr(all_df_hpvneg$sender, 1, 2) == substr(all_df_hpvneg$receiver, 1, 2),]
#getting rid of any interactions between subsets where the last 2 characters are the same
all_df_hpvneg <- all_df_hpvneg[!substr(all_df_hpvneg$sender, nchar(all_df_hpvneg$sender) - 2 + 1, nchar(all_df_hpvneg$sender)) == substr(all_df_hpvneg$receiver, nchar(all_df_hpvneg$receiver) - 2 + 1, nchar(all_df_hpvneg$receiver)),]

all_df2 <- data.frame()
for (y in unique(all_df$receiver)){
  df <- all_df[all_df$receiver == y,]
  print(nrow(df))
  d1 <- DotPlot(seu[,seu@active.ident == y], features = unique(all_df$receptor))
  recs_keep <- c(unique(rownames(d1$data[d1$data$pct.exp > 20,])), 'IL15R')
  df <- df[df$receptor %in% recs_keep,]
  print(nrow(df))
  all_df2 <- rbind(all_df2, df)
  
}
all_df2_hpvneg <- unique(all_df2)

all_df2_hpvneg <- all_df2_hpvneg[!substr(all_df2_hpvneg$sender, 1, 4) == substr(all_df2_hpvneg$receiver, 1, 4),]
all_df2_hpvneg <- all_df2_hpvneg[!substr(all_df2_hpvneg$sender, 1, 2) == substr(all_df2_hpvneg$receiver, 1, 2),]
#getting rid of any interactions between subsets where the last 2 characters are the same
all_df2_hpvneg <- all_df2_hpvneg[!substr(all_df2_hpvneg$sender, nchar(all_df2_hpvneg$sender) - 2 + 1, nchar(all_df2_hpvneg$sender)) == substr(all_df2_hpvneg$receiver, nchar(all_df2_hpvneg$receiver) - 2 + 1, nchar(all_df2_hpvneg$receiver)),]

all_hpvneg <- unique(rbind(all_df_hpvneg, all_df2_hpvneg))

all_hpvneg <- all_hpvneg[!all_hpvneg$sr %in% c("CD4_1 -> Treg_2", "CD4_2 -> Treg_1", "CD4_3 -> Treg_1", "CD4_3 -> Treg_2", "CD4_4 -> Treg_1", "CD4_4 -> Treg_2", "CD8_1 -> Treg_2", "CD8_2 -> Treg_1", "CD8_3 -> Treg_1", "CD8_4 -> Treg_1", "CD8_4 -> Treg_2", "CD8_5 -> Treg_1", "CD8_5 -> Treg_2", "cDC1_CLEC9A -> DC3_LAMP3", "cDC2_CD1C -> DC3_LAMP3", "DC3_LAMP3 -> cDC1_CLEC9A", "DC3_LAMP3 -> cDC2_CD1C", "Treg_1 -> CD4_2", "Treg_1 -> CD4_3", "Treg_1 -> CD4_4", "Treg_1 -> CD8_2", "Treg_1 -> CD8_3", "Treg_1 -> CD8_4", "Treg_1 -> CD8_5", "Treg_2 -> CD4_1", "Treg_2 -> CD4_3", "Treg_2 -> CD4_4", "Treg_2 -> CD8_1", "Treg_2 -> CD8_3", "Treg_2 -> CD8_4", "Treg_2 -> CD8_5" , "cDC2_CD33 -> DC3_LAMP3", "DC3_LAMP3 -> cDC2_CD33"),]

#write.csv(all_hpvneg, file = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/05042022_HNC_HPV-DCstoTsubsets_out/all_hpvneg.csv')

################################################################################

all_df2_hpvpos[['merged_data']] <- all_df2_hpvpos
all_df2_hpvneg[['merged_data']] <- all_df2_hpvneg

hpvp <- corrs(SEU = seu, SUM = all_df2_hpvpos, meta = 'global.cluster5')
hpvn <- corrs(SEU = seu, SUM = all_df2_hpvneg, meta = 'global.cluster5')

cpdb_dotplot_two_objs(hpvn_noCD8, hpvp_noCD8, filtered = FALSE, out = '/Volumes/hdlab/Projects/HNC_SPORE/AACR2022', width = 3200, height = 1800)

write.csv(hpvp, file = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/05042022_HNC_HPV+DCstoTsubsets_out/hpv+_cpdb_DCstoTsubsets.csv')
write.csv(hpvn, file = '/Volumes/hdlab/Projects/HNC_SPORE/CellphoneDB/05042022_HNC_HPV-DCstoTsubsets_out/hpv-cpdb_DCstoTsubsets.csv')

#cpdb_dotplot_two_objs(hpvn, hpvp, filtered = FALSE, out = '/Volumes/hdlab/Projects/HNC_SPORE/AACR2022/', width = 3200, height = 1800)

#sanity check that ligands and receptors are actually expressed in the corresponding cell type
VlnPlot(hnc, features = c('LGALS9', 'LTBR', 'FLT3', 'BTLA', 'CD55', 'FAM3C', 'CD40', 'CD74'), group.by = 'global.cluster4', idents = c('DC3_LAMP3', 'cDC2_CD1C', 'cDC1_CLEC9A'), split.by = 'hpv_status')
VlnPlot(hnc, features = c('CD47', 'CD44', 'LTB', 'FLT3LG', 'TNFRSF14', 'ADGRE5', 'CLEC2D', 'SORL1', 'CD40LG', 'MIF'), group.by = 'global.cluster4', idents = c('Treg','CD4', 'CD8'), split.by = 'hpv_status')

```

```{r VlnPlot for Tfh-like and CD8 progenitor cell subsets}
VlnPlot(hnc, features = c('CXCL13', 'CD40LG', 'CH25H', 'IL21', 'PDCD1'), idents = unique(hnc@active.ident[hnc@active.ident %like% '%CD4%']))
VlnPlot(hnc, features = c('CD20', 'TCF1', 'MS4A1'), idents = unique(hnc@active.ident[hnc@active.ident %like% '%CD8_%']))

#checking if a variation of a gene is in the matrix
#rownames(hnc)[rownames(hnc) %like% '%TCF%']
```
