---
title: "Visium_V2"
author: Athena Golfinos-Owens
output: html_document
date: "2023-05-31"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(Seurat)
library(hdf5r)
library(ggplot2)
library(patchwork)
library(SeuratDisk)
```

# Visium pre-processing

This function loads in the H5 file from your SpaceRanger output, normalizes with SCTransform, runs PCA using the SCT normalized counts using user-input PC values, finds nearest neighbors, finds clusters at resolutions ranging from 0.1 to 3, and runs UMAP. It additionally saves a log file of the parameters used for the pre-processing steps so it is documented and reproducible. 

in_dir: The directory that contains the output directory from Space Ranger. At the very least, you should have the filtered h5 file, the tiff image used in running space ranger, and the "spatial" output directory in this directory. 

h5_name: This is the name of the h5 file within the directory. This name should match the argument listed unless you have changed the name following the Space Ranger run. 

out_dir: Where to save the pre-processing output that will be generated as part of the run

```{r}
in_dir <- "Z:\\Computational_Toxicology\\Athena\\Test_visium\\"
h5_name <- "V1_Mouse_Brain_Sagittal_Anterior_Section_2_filtered_feature_bc_matrix.h5"
out_dir <- "Z:\\Computational_Toxicology\\Athena\\Test_visium\\"

# if using RMarkdown, run this in the console or you will not see the required plots!
seu <- visium_pp(in_dir = "Z:\\Computational_Toxicology\\Athena\\Test_visium\\", h5_name = "V1_Mouse_Brain_Sagittal_Anterior_Section_2_filtered_feature_bc_matrix.h5", out_dir = "Z:\\Computational_Toxicology\\Athena\\Test_visium\\")

visium_pp <- function(in_dir, h5_name = "filtered_feature_bc_matrix.h5", out_dir){
  
  # attach required packages
  require(ggplot2)
  require(Seurat)
  require(patchwork)
  require(ggplot2)
  require(hdf5r)
  
  sink(paste(out_dir, 'preprocessing_console_output.txt', sep = ''), append = T)
  
  cat('\nNEW ANALYSIS\n')
  cat(paste('Date of analysis: ', Sys.Date(), "\n"))
  cat(paste('Input directory: ', in_dir, "\n"))
  cat(paste('H5 file name: ', h5_name, "\n"))
  cat(paste("Output directory: ", out_dir, "\n"))
  cat(paste("Normalization method: SCTransform \n"))
  
  print("Loading in h5 file and low resolution H&E Image")
  
  # load in the data
  seu <- Seurat::Load10X_Spatial(data.dir = in_dir, filename = h5_name)
  
  # visualize the heterogeneity of coverage
  plot1 <- VlnPlot(seu, features = "nCount_Spatial", pt.size = 0.1) + NoLegend()
  plot2 <- SpatialFeaturePlot(seu, features = "nCount_Spatial") + theme(legend.position = "right")
  print(wrap_plots(plot1, plot2))
  
  print("Normalizing data with SCTransform")
  
  # normalize with SCTransform
  seu <- SCTransform(seu, assay = "Spatial", verbose = FALSE)
  
  print("Running PCA using 200 PCs. Please provide user input when prompted to select the number of PCs used for finding nearest neighbors. Choose a # of PCs right after the point where the plot flattens out")
  
  # run PCA using the SCT assay
  seu <- RunPCA(seu, assay = "SCT", verbose = FALSE, npcs = 200)
  print(ElbowPlot(seu, ndims = 200))
  
  # taking user input as to how many principal components to use 
  npcs <- as.integer(readline("Enter the number of PCs to use: "))
  
  cat(paste("Number of PCs chosen by user: ", npcs, "\n"))
  
  # find nearest neighbors
  seu <- FindNeighbors(seu, reduction = "pca", dims = 1:npcs)
  
  # finding clusters ranging from 0.1-3
  for (x in seq.int(0.1, 3, by = 0.1)){
      seu <- FindClusters(seu, resolution = x, verbose = FALSE)
    }
  
  cat("Range of resolutions used for cluster identification: 0.1-3 (by 0.1)")
  
  # run UMAP
  seu <- RunUMAP(seu, reduction = "pca", dims = 1:npcs)
  
  sink()
  
  return(seu)
}
```

# Global  scRNA-seq prep
## Load and save objects
```{r}
seu <- readRDS("Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_seurat.RDS")

#saveRDS(seu, "Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_seurat.RDS")
```


## Generate seurat object from Kimmel spleen data
```{r}
# List of h5 files
h5_files <- list.files("Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/", pattern = "\\.h5$")


# generate one large seurat object that puts the ID into the name

seurat_obj <- NULL
count = 1
# Loop through h5 files
for (h5_file in h5_files) {
  
  print(paste("Starting file ", as.character(h5_file)))
  
  # Read h5 file into a matrix
  seu_temp <- Seurat::Read10X_h5(paste("Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/", h5_file, sep = ''))
  
  if (count == 1){
    seurat_obj <- CreateSeuratObject(seu_temp)
    seurat_obj$orig.ident <- sapply(strsplit(h5_file, "_"), `[`, 2)
  }
  
  if (count > 1){
    seu_temp <- CreateSeuratObject(seu_temp)
    
    seu_temp$orig.ident <- sapply(strsplit(h5_file, "_"), `[`, 2)
    
    # Add the cells from the temp Seurat object to the main Seurat object
    seurat_obj <- merge(seurat_obj, seu_temp)
  }
  
  count = count + 1

}

saveRDS(seu, "Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_seurat.RDS")

```

## Processing scRNA-seq data
```{r}
# run this in the console!
seu <- seu_pp1(seu, out="Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_seurat.RDS", remove_doublets = F, feature_subset = F, batch_id = 'orig.ident', reduction = 'pca', harmony = F, seurat_integration = F, normalize = T, test_npcs = 100, assay = 'RNA')

seu <- seu[,seu$nCount_RNA < 20000]
seu <- seu[,seu$nFeature_RNA < 3000]
```

## Identify clusters
```{r}
# identifying B cell clusters
FeaturePlot(seu, features = c('Cd19', 'Ms4a1')) # clusters 0, 3, 6, 12

# identifying T cell clusters
FeaturePlot(seu, features = c('Cd3e', 'Cd4', 'Cd8a', 'Foxp3'))

# identifying myeloid cell clusters
FeaturePlot(seu, features = c('Cd86', 'Cd68', 'Cxcl9', 'Spp1'))

# identifying dendritic cell clusters
FeaturePlot(seu, features = c('Clec9a','Clec10a', 'Cd1c', 'Lamp3'))

# identifying NK cell cluster
FeaturePlot(seu, features = c('Nkg7', 'Cd3e'))

# identifying cap endo markers
VlnPlot(seu, group.by = 'RNA_snn_res.0.5', features= c('Podxl', 'Pecam1'))

DimPlot(seu, group.by = 'global.cluster', label = T, label.box = T)

#seu$global.cluster <- as.character(seu$RNA_snn_res.0.1)

seu$global.cluster[seu$global.cluster %in% c('0', '3', '6', '12')] <- 'B cells'
seu$global.cluster[seu$global.cluster %in% c('1', '2', '9', '7')] <- 'T/NK cells'
seu$global.cluster[seu$global.cluster %in% c('4')] <- 'Myeloid cells'
seu$global.cluster[seu$RNA_snn_res.0.5 %in% c('13')] <- 'NK cells'
seu$global.cluster[seu$global.cluster %in% c('8')] <- 'Neutrophils'
seu$global.cluster[seu$global.cluster %in% c('5')] <- 'Plasma cells'
seu <- seu[,!seu$global.cluster %in% c('10', '11')]

#removed <- seu[,seu$global.cluster %in% c('10', '11')]
#saveRDS(removed, file = 'Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_removed_populations.RDS')



de_dotplot(seu = seu, metadata = 'global.cluster', out = 'Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_seurat.RDS', n_marks = 10, min_pct = 0.1)

```

# Myeloid cell clustering

## Load and save objects
```{r}
myl <- readRDS("Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_myeloid.RDS")

#saveRDS(myl, "Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_myeloid.RDS")
```

## Object generation
```{r}
myl <- seu[,seu$global.cluster2 == 'Myeloid cells']
```

## Processing scRNA-seq data
```{r}
# run this in the console!
myl <- seu_pp1(myl, out="Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_myeloid", remove_doublets = F, feature_subset = T, batch_id = 'orig.ident', reduction = 'pca', harmony = F, seurat_integration = F, normalize = T, test_npcs = 100, assay = 'RNA')
```

## Identify clusters
```{r}
marker_plotting(seu = myl, group.by = 'RNA_snn_res.0.1', features = c('Cd68', 'Thbs1', 'Clec9a', 'Clec10a', 'Csf1r', 'Cd14', 'Fcgr3', 'Lyz2', 'Cd3e', 'Cd19', 'C1qa', 'C1qb', 'C1qc'))

marks <- de_dotplot(myl, metadata = 'RNA_snn_res.0.1', out = "Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_myeloid")

myl$global.cluster2 <- as.character(myl$RNA_snn_res.0.1)

myl$global.cluster2[myl$global.cluster2 %in% c('0', '1', '4')] <- 'Mon/Macs'
myl$global.cluster2[myl$global.cluster2 %in% c('2', '3')] <- 'Dendritic cells'

DimPlot(myl, group.by = 'global.cluster2')
```
## Transfer myeloid annotations to global
```{r}
# assigning monocyte/macrophage populations
monmac <- colnames(myl)[myl$global.cluster2 == 'Mon/Macs']

seu$global.cluster2[colnames(seu) %in% monmac] <- 'Mon/Mac'

# assigning dendritic cell populations
dcs <- colnames(myl)[myl$global.cluster2 == 'Dendritic cells']

seu$global.cluster2[colnames(seu) %in% dcs] <- 'Dendritic cells'

table(seu$global.cluster2)
```


# T cell scRNA-seq prep

## Load and save objects
```{r}
tcells <- readRDS("Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_tcells.RDS")

#saveRDS(tcells, "Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_tcells.RDS")
```


## Object generation
```{r}
tcells <- seu[,seu$global.cluster == 'T/NK cells']
```

## Processing scRNA-seq data
```{r}
# run this in the console!
tcells <- seu_pp1(tcells, out="Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_tcells.RDS", remove_doublets = F, feature_subset = F, batch_id = 'orig.ident', reduction = 'pca', harmony = F, seurat_integration = F, normalize = T, test_npcs = 100, assay = 'RNA')
```

## Identify clusters
```{r}
DimPlot(tcells, group.by = 'RNA_snn_res.0.2')

# identifying Tregs
marker_plotting(tcells, features = c('Foxp3', 'Il2ra', 'Nt5e', 'Nrp1', 'Cd4'), group.by = 'RNA_snn_res.0.4') # cluster 1

# identifying Tfh cells
marker_plotting(tcells, features = c('Cxcr5', 'Pdcd1', 'Bcl6', 'Icos', 'Sh2d1a', 'Cd40lg'), group.by = 'RNA_snn_res.0.4') # cluster 2

# identifying CD4 T cells
marker_plotting(tcells, features = c('Cd4'), group.by = 'RNA_snn_res.0.4') # cluster 1, 2, 4

# identifying CD8 T cells
marker_plotting(tcells, features = c('Cd8a'), group.by = 'RNA_snn_res.0.4') # cluster 0, 3, 5, 8, 10

# identifying T cells in general
marker_plotting(tcells, features = 'Cd3e', group.by = 'RNA_snn_res.0.4')

# identifying memory T cells
marker_plotting(tcells, features = c('Cd44', 'Sell', 'Il7r', 'Il2rb', 'Klrg1', 'Cd69'), group.by = 'RNA_snn_res.0.4') # no clear population

# identifying Th17 T cells
marker_plotting(tcells, group.by = 'RNA_snn_res.0.4', features = c('Rorc', 'Il17a', 'Il17f', 'Il23r', 'Il22', 'Batf', 'Ccr6', 'Il1r1'))

tcells$global.cluster2 <- as.character(tcells$RNA_snn_res.0.4)


DimPlot(tcells, group.by = 'global.cluster2')

tcells$global.cluster2[tcells$global.cluster2 %in% c('0', '3', '5', '8', '10')] <- 'CD8 T cells'
tcells$global.cluster2[tcells$global.cluster2 %in% c('2', '4')] <- 'CD4 T cells'
tcells$global.cluster2[tcells$global.cluster2 %in% c('1')] <- 'Tregs'
tcells$global.cluster2[tcells$global.cluster2 == c('6')] <- 'Unassigned T:6'
tcells$global.cluster2[tcells$global.cluster2 == c('7')] <- 'Unassigned T:7'
tcells$global.cluster2[tcells$global.cluster2 == c('9')] <- 'Unassigned T:9'




tcell_marks <- de_dotplot(seu = tcells, metadata = 'global.cluster2', out = 'Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_tcells')
```

## Transfer T annotations to global
```{r}
# remove high stress T cell cluster 6 from global.cluster2 from t cells and from seurat object
t_remove <- colnames(tcells)[tcells$global.cluster2 == 'Unassigned T:6']
seu <- seu[,!colnames(seu) %in% t_remove]
tcells <- tcells[,!colnames(tcells) %in% t_remove]

#seu$global.cluster2 <- seu$global.cluster

# assigning Tregs
tregs <- colnames(tcells)[tcells$global.cluster2 == 'Tregs']

seu$global.cluster2[colnames(seu) %in% tregs] <- 'Tregs'


# assigning CD4 T cells
cd4 <- colnames(tcells)[tcells$global.cluster2 == 'CD4 T cells']

seu$global.cluster2[colnames(seu) %in% cd4] <- 'CD4 T cells'

# assigning CD8 T cells
cd8 <- colnames(tcells)[tcells$global.cluster2 == 'CD8 T cells']

seu$global.cluster2[colnames(seu) %in% cd8] <- 'CD8 T cells'

# removing a high mitochondrial cell percentage population
remove <- colnames(tcells)[tcells$global.cluster2 == 'Unassigned T:6'] 
seu <- seu[,!colnames(seu) %in% remove]


DimPlot(seu, group.by = 'global.cluster2')

#seu$global.cluster2[seu$global.cluster2 == 'T cells'] <- 'Unassigned T cells'
```


# B cell scRNA-seq prep

## Load and save objects
```{r}
bcells <- readRDS("Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_bcells.RDS")

#saveRDS(bcells, "Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_bcells.RDS")
```

## Object generation
```{r}
bcells <- seu[,seu$global.cluster == 'B cells']
```


## Processing scRNA-seq data
```{r}
bcells <- seu_pp1(bcells, out="Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_bcells.RDS", remove_doublets = F, feature_subset = F, batch_id = 'orig.ident', reduction = 'pca', harmony = F, seurat_integration = F, normalize = T, test_npcs = 100, assay = 'RNA')
```


## Identify clusters
```{r}

# identifying follicular B cells
marker_plotting(bcells, features = c('Cd45ra', 'Cd19', 'Ighd', 'Cd38', 'Cr2', 'Fcer2a', 'Cd22', 'Pax5'), group.by = 'RNA_snn_res.0.2') 

# identifying marginal zone B cells
marker_plotting(bcells, features = c('Cd1d1', 'Ighm', 'Cd9', 'Cd22', 'Ebf1', 'E2A', 'Slc22a2'), group.by = 'RNA_snn_res.0.2') 

# identifying B1 B cells
marker_plotting(bcells, features = c('Cd5', 'Spn', 'Itgam', 'Sdc1'), group.by = 'RNA_snn_res.0.2') # cluster 5

# identifying transitional/immature B cells
marker_plotting(bcells, features = c('Cd93', 'Cd24a', 'BR2', 'Tnfrsf13b'), group.by = 'RNA_snn_res.0.2')

# germinal center b cells
marker_plotting(bcells, features = c('Fas', 'PNA', 'Tnfrsf13c', 'Bcl6', 'Ebf1'), group.by = 'RNA_snn_res.0.2')

# memory B cells
marker_plotting(tcells, features = c('Cd80', 'Nt5e', 'Pdcd1lg2', 'Cd38', 'Cd84', 'Cd86', 'Pbx1', 'Spib'), group.by = 'RNA_snn_res.0.2')

bcells$global.cluster2[bcells$RNA_snn_res.2.1 == 28] <- 'Myeloid contamination'
bcells$global.cluster2[bcells$RNA_snn_res.2.1 %in% c(24, 27)] <- 'T cell contamination'
```

## Transfer B annotations to global
```{r}
# assigning myeloid cells back to the proper 
my_contam <- colnames(bcells)[bcells$global.cluster2 == 'Myeloid contamination']

seu$global.cluster2[colnames(seu) %in% my_contam] <- 'Myeloid cells'

t_contam <- colnames(bcells)[bcells$global.cluster2 == 'T cell contamination']
seu$global.cluster2[colnames(seu) %in% t_contam] <- 'T cells'

# removing clusters that seem to be T cells
t_not_b <- colnames(bcells)[bcells$RNA_snn_res.3 %in% c(32, 35)]
seu$global.cluster2[colnames(seu) %in% t_not_b] <- 'T cells'

bcells <- bcells[, !bcells$RNA_snn_res.3 %in% c(32, 35)]
```

# Visium prep

## Load and save objects
```{r}
st <- readRDS(file = "Z:\\Computational_Toxicology\\Athena\\SpaceRanger_testruns\\v2_athena_test\\Kidney_out\\kidney_visium_seurat.RDS")

#saveRDS(st, file = "Z:\\Computational_Toxicology\\Athena\\SpaceRanger_testruns\\v2_athena_test\\Kidney_out\\kidney_visium_seurat.RDS")
```

## Pre-processing Visium
```{r}
seu <- visium_pp(in_dir = "Z:\\Computational_Toxicology\\Athena\\SpaceRanger_testruns\\v2_athena_test\\Kidney_out\\outs", h5_name = "filtered_feature_bc_matrix.h5", out_dir = "Z:\\Computational_Toxicology\\Athena\\SpaceRanger_testruns\\v2_athena_test\\Kidney_out\\")
```


# SCDC 
## Load in
```{r}
# single cell object load in
seu <- readRDS("Z:/Computational_Toxicology/Athena/Mouse_scRNAseq/Kimmel_et_al_2019/Spleen/kimmel_seurat.RDS")

st <- readRDS(file = "Z:\\Computational_Toxicology\\Athena\\SpaceRanger_testruns\\v2_athena_test\\Kidney_out\\kidney_visium_seurat.RDS")
```

## SCDC run


```{r}
#sc_ref <- seu
#st_query <- st
#sc_meta <- 'global.cluster2'
#out <- 'C:\\Users\\GOLFIAX\\OneDrive - AbbVie Inc (O365)\\Desktop\\'

st <- scdc_decon(sc_ref = seu, st_query = st, sc_meta = 'global.cluster2', out = 'Z:\\Computational_Toxicology\\Athena\\SpaceRanger_testruns\\v2_athena_test\\outputs\\')

```

# SPOTclean prep 

Two parts of 10x Space Ranger output files are required as input to `SpotClean`: the raw gene-by-spot count matrix, and the slide metadata. In this example, you can download and unzip the [Feature / cell matrix (raw)](https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Adult_Mouse_Brain/V1_Adult_Mouse_Brain_raw_feature_bc_matrix.tar.gz) and [Spatial imaging data](https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Adult_Mouse_Brain/V1_Adult_Mouse_Brain_spatial.tar.gz) from [V1_Adult_Mouse_Brain](https://support.10xgenomics.com/spatial-gene-expression/datasets/1.0.0/V1_Adult_Mouse_Brain). You will get a folder `raw_feature_bc_matrix` containing `barcodes.tsv.gz`, `features.tsv.gz`, `matrix.mtx.gz`, and a folder `spatial` containing `aligned_fiducials.jpg`, `detected_tissue_image.jpg`, `tissue_hires_image.png`, `tissue_lowres_image.png`, `tissue_positions_list.csv`, `scalefactors_json.json`. The former folder contains the raw gene-by-spot count matrix, and the latter contains slide metadata like slide images and spot IDs and positions.

```{r}
library(SpotClean)
spatial_dir <- '/Volumes/hqdinh2/Projects/RawData_FromUWBiotech/HNCVisium_2022-05-05/CK17-5_outs/'
list.files(spatial_dir)

hnc_raw <- read10xRaw("/Volumes/hqdinh2/Projects/RawData_FromUWBiotech/HNCVisium_2022-05-05/CK17-5_outs/")
hnc_slide_info <- read10xSlide(tissue_csv_file=file.path(spatial_dir, "tissue_positions_list.csv"), tissue_img_file = file.path(spatial_dir,
                                       "tissue_lowres_image.png"),
             scale_factor_file = file.path(spatial_dir,
                                       "scalefactors_json.json"))

str(hnc_slide_info)

slide_obj <- createSlide(hnc_raw, hnc_slide_info)
slide_obj
visualizeSlide(slide_obj)
visualizeLabel(slide_obj,"tissue")
metadata(slide_obj)$slide$total_counts <- colSums(hnc_raw)
visualizeHeatmap(slide_obj,"total_counts")
visualizeHeatmap(slide_obj,"LAMP3")
visualizeHeatmap(slide_obj,"CXCL9")

saveRDS(slide_obj, paste(spatial_dir, 'slide_obj.RDS', sep = ''))

decon_ck17_5_seurat <- readRDS(paste(spatial_dir, 'decon_ck17_5_seurat.RDS', sep = ''))

decon_ck17_5_seurat 
names(metadata(decon_ck17_5_seurat ))
summary(metadata(decon_ck17_5_seurat )$contamination_rate)
arcScore(slide_obj)
decon_ck17_5_seurat <- convertToSeurat(decon_ck17_5_seurat ,image_dir = spatial_dir)

saveRDS(slide_obj, paste(spatial_dir, 'decon_ck17_5_seurat.RDS', sep = ''))
```


